{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-03-22T17:21:55.305238Z",
     "start_time": "2019-03-22T17:21:54.814122Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import SearchIO,SeqIO\n",
    "import ete3\n",
    "import os,functools,json\n",
    "\n",
    "ncbi=ete3.ncbi_taxonomy.NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-03-22T17:21:55.654174Z",
     "start_time": "2019-03-22T17:21:55.309412Z"
    }
   },
   "outputs": [],
   "source": [
    "acc_to_gi={} ## accession to gi dict\n",
    "acc_to_taxid={} ## accession to taxid dict\n",
    "\n",
    "for line in open('/Users/evogytis/Documents/skeeters/data/acc2gi.txt','r'):\n",
    "    acc,gi,taxid=line.strip('\\n').split('\\t')\n",
    "    acc_to_gi[acc]=gi\n",
    "    acc_to_taxid[acc]=int(taxid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-03-22T17:42:49.638192Z",
     "start_time": "2019-03-22T17:21:55.656494Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1 \n",
      "\n",
      "2 CMS002_026d_Rb_S149_L004\n",
      "total number of contigs in sample: 596\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Bio/SearchIO/_legacy/NCBIStandalone.py:45: BiopythonWarning: Parsing BLAST plain text output file is not a well supported functionality anymore. Consider generating your BLAST output for parsing as XML or tabular format instead.\n",
      "  BiopythonWarning\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of contigs without blastx hit: 4\n",
      "\n",
      "3 CMS002_045f_Rb_S189_L004\n",
      "total number of contigs in sample: 77\n",
      "number of contigs without blastx hit: 0\n",
      "\n",
      "4 CMS002_050a_Rb_S5_L004\n",
      "total number of contigs in sample: 173\n",
      "number of contigs without blastx hit: 2\n",
      "\n",
      "5 CMS002_016a_Rb_S121_L004\n",
      "total number of contigs in sample: 0\n",
      "number of contigs without blastx hit: 0\n",
      "\n",
      "6 CMS002_0Water2_Rb_S139_L004\n",
      "total number of contigs in sample: 82\n",
      "\n",
      "7 CMS002_045d_Rb_S186_L004\n",
      "total number of contigs in sample: 139\n",
      "number of contigs without blastx hit: 2\n",
      "\n",
      "8 CMS002_049a_Rb_S4_L004\n",
      "total number of contigs in sample: 1175\n",
      "number of contigs without blastx hit: 82\n",
      "\n",
      "9 CMS002_026a_Rb_S146_L004\n",
      "total number of contigs in sample: 123\n",
      "number of contigs without blastx hit: 1\n",
      "\n",
      "10 CMS002_0Water4_Rb_S163_L004\n",
      "total number of contigs in sample: 121\n",
      "\n",
      "11 CMS001_042_Ra_S23\n",
      "total number of contigs in sample: 6697\n",
      "number of contigs without blastx hit: 323\n",
      "\n",
      "12 CMS002_029c_Rb_S161_L004\n",
      "total number of contigs in sample: 5188\n",
      "number of contigs without blastx hit: 199\n",
      "\n",
      "13 CMS001_050_Ra_S23\n",
      "total number of contigs in sample: 7075\n",
      "number of contigs without blastx hit: 322\n",
      "\n",
      "14 CMS001_048_Ra_S5\n",
      "total number of contigs in sample: 2612\n",
      "number of contigs without blastx hit: 97\n",
      "\n",
      "15 CMS002_010a_Rb_S119_L004\n",
      "total number of contigs in sample: 798\n",
      "number of contigs without blastx hit: 22\n",
      "\n",
      "16 CMS001_058_Ra_S9\n",
      "total number of contigs in sample: 7367\n",
      "number of contigs without blastx hit: 342\n",
      "\n",
      "17 CMS001_009_Ra_S13\n",
      "total number of contigs in sample: 20597\n",
      "number of contigs without blastx hit: 727\n",
      "\n",
      "18 CMS002_029b_Rb_S160_L004\n",
      "total number of contigs in sample: 5219\n",
      "number of contigs without blastx hit: 203\n",
      "\n",
      "19 CMS001_water5_RNA_A_S12\n",
      "total number of contigs in sample: 0\n",
      "\n",
      "20 CMS001_014_Ra_S5\n",
      "total number of contigs in sample: 11654\n",
      "number of contigs without blastx hit: 1060\n",
      "\n",
      "21 CMS001_022_Ra_S6\n",
      "total number of contigs in sample: 3112\n",
      "number of contigs without blastx hit: 104\n",
      "\n",
      "22 CMS002_028b_Rb_S155_L004\n",
      "total number of contigs in sample: 6203\n",
      "number of contigs without blastx hit: 223\n",
      "\n",
      "23 CMS002_017a_Rb_S122_L004\n",
      "total number of contigs in sample: 184\n",
      "number of contigs without blastx hit: 9\n",
      "\n",
      "24 CMS001_039_Ra_S9\n",
      "total number of contigs in sample: 4422\n",
      "number of contigs without blastx hit: 186\n",
      "\n",
      "25 CMS001_015_Ra_S13\n",
      "total number of contigs in sample: 2476\n",
      "number of contigs without blastx hit: 79\n",
      "\n",
      "26 CMS002_027a_Rb_S152_L004\n",
      "total number of contigs in sample: 355\n",
      "number of contigs without blastx hit: 8\n",
      "\n",
      "27 CMS002_034a_Rb_S168_L004\n",
      "total number of contigs in sample: 3620\n",
      "number of contigs without blastx hit: 156\n",
      "\n",
      "28 CMS002_001a_Rb_S116_L004\n",
      "total number of contigs in sample: 822\n",
      "number of contigs without blastx hit: 30\n",
      "\n",
      "29 CMS001_019_Ra_S14\n",
      "total number of contigs in sample: 2042\n",
      "number of contigs without blastx hit: 62\n",
      "\n",
      "30 CMS002_036a_Rb_S170_L004\n",
      "total number of contigs in sample: 7825\n",
      "number of contigs without blastx hit: 377\n",
      "\n",
      "31 CMS002_032a_Rb_S166_L004\n",
      "total number of contigs in sample: 14408\n",
      "number of contigs without blastx hit: 987\n",
      "\n",
      "32 CMS002_0Water5_Rb_S175_L004\n",
      "total number of contigs in sample: 101\n",
      "\n",
      "33 CMS002_007a_Rb_S118_L004\n",
      "total number of contigs in sample: 37488\n",
      "number of contigs without blastx hit: 4454\n",
      "\n",
      "34 CMS002_047g_Rb_S200_L004\n",
      "total number of contigs in sample: 2902\n",
      "number of contigs without blastx hit: 110\n",
      "\n",
      "35 CMS001_002_Ra_S1\n",
      "total number of contigs in sample: 4929\n",
      "number of contigs without blastx hit: 242\n",
      "\n",
      "36 CMS002_029a_Rb_S159_L004\n",
      "total number of contigs in sample: 11348\n",
      "number of contigs without blastx hit: 576\n",
      "\n",
      "37 CMS002_028e_Rb_S158_L004\n",
      "total number of contigs in sample: 8602\n",
      "number of contigs without blastx hit: 442\n",
      "\n",
      "38 CMS002_018b_Rb_S129_L004\n",
      "total number of contigs in sample: 400\n",
      "number of contigs without blastx hit: 16\n",
      "\n",
      "39 CMS001_007_Ra_S12\n",
      "total number of contigs in sample: 4009\n",
      "number of contigs without blastx hit: 76\n",
      "\n",
      "40 CMS002_025e_Rb_S144_L004\n",
      "total number of contigs in sample: 273\n",
      "number of contigs without blastx hit: 4\n",
      "\n",
      "41 CMS002_0Water7_Rb_S199_L004\n",
      "total number of contigs in sample: 106\n",
      "\n",
      "42 CMS002_026c_Rb_S148_L004\n",
      "total number of contigs in sample: 520\n",
      "number of contigs without blastx hit: 3\n",
      "\n",
      "43 CMS001_041_Ra_S10\n",
      "total number of contigs in sample: 1069\n",
      "number of contigs without blastx hit: 46\n",
      "\n",
      "44 CMS001_026_Ra_S18\n",
      "total number of contigs in sample: 21264\n",
      "number of contigs without blastx hit: 2162\n",
      "\n",
      "45 CMS001_057_Ra_S11\n",
      "total number of contigs in sample: 6213\n",
      "number of contigs without blastx hit: 247\n",
      "\n",
      "46 CMS002_045a_Rb_S183_L004\n",
      "total number of contigs in sample: 214\n",
      "number of contigs without blastx hit: 3\n",
      "\n",
      "47 CMS002_022a_Rb_S137_L004\n",
      "total number of contigs in sample: 1124\n",
      "number of contigs without blastx hit: 47\n",
      "\n",
      "48 CMS001_044_Ra_S25\n",
      "total number of contigs in sample: 610\n",
      "number of contigs without blastx hit: 31\n",
      "\n",
      "49 CMS002_047j_Rb_S3_L004\n",
      "total number of contigs in sample: 5708\n",
      "number of contigs without blastx hit: 258\n",
      "\n",
      "50 CMS002_0Water1_Rb_S127_L004\n",
      "total number of contigs in sample: 74\n",
      "\n",
      "51 CMS002_039a_Rb_S173_L004\n",
      "total number of contigs in sample: 12653\n",
      "number of contigs without blastx hit: 715\n",
      "\n",
      "52 CMS002_051a_Rb_S6_L004\n",
      "total number of contigs in sample: 881\n",
      "number of contigs without blastx hit: 29\n",
      "\n",
      "53 CMS002_017e_Rb_S126_L004\n",
      "total number of contigs in sample: 185\n",
      "number of contigs without blastx hit: 15\n",
      "\n",
      "54 CMS001_water1_S11\n",
      "total number of contigs in sample: 0\n",
      "\n",
      "55 CMS002_023a_Rb_S138_L004\n",
      "total number of contigs in sample: 514\n",
      "number of contigs without blastx hit: 31\n",
      "\n",
      "56 CMS002_025c_Rb_S142_L004\n",
      "total number of contigs in sample: 204\n",
      "number of contigs without blastx hit: 6\n",
      "\n",
      "57 CMS002_021a_Rb_S136_L004\n",
      "total number of contigs in sample: 49881\n",
      "number of contigs without blastx hit: 7242\n",
      "\n",
      "58 CMS002_042a_Rb_S177_L004\n",
      "total number of contigs in sample: 2955\n",
      "number of contigs without blastx hit: 133\n",
      "\n",
      "59 CMS001_034_Ra_S19\n",
      "total number of contigs in sample: 3655\n",
      "number of contigs without blastx hit: 157\n",
      "\n",
      "60 CMS002_056a_Rb_S9_L004\n",
      "total number of contigs in sample: 10844\n",
      "number of contigs without blastx hit: 633\n",
      "\n",
      "61 CMS001_060_Ra_S12\n",
      "total number of contigs in sample: 29705\n",
      "number of contigs without blastx hit: 3455\n",
      "\n",
      "62 CMS002_041a_Rb_S176_L004\n",
      "total number of contigs in sample: 7446\n",
      "number of contigs without blastx hit: 451\n",
      "\n",
      "63 CMS002_044e_Rb_S182_L004\n",
      "total number of contigs in sample: 171\n",
      "number of contigs without blastx hit: 5\n",
      "\n",
      "64 CMS001_011_Ra_S4\n",
      "total number of contigs in sample: 6193\n",
      "number of contigs without blastx hit: 225\n",
      "\n",
      "65 CMS001_028_Ra_S17\n",
      "total number of contigs in sample: 4865\n",
      "number of contigs without blastx hit: 145\n",
      "\n",
      "66 CMS001_001_Ra_S1\n",
      "total number of contigs in sample: 11896\n",
      "number of contigs without blastx hit: 677\n",
      "\n",
      "67 CMS002_025a_Rb_S140_L004\n",
      "total number of contigs in sample: 18\n",
      "number of contigs without blastx hit: 0\n",
      "\n",
      "68 CMS001_059_Ra_S10\n",
      "total number of contigs in sample: 1311\n",
      "number of contigs without blastx hit: 63\n",
      "\n",
      "69 CMS002_044a_Rb_S178_L004\n",
      "total number of contigs in sample: 272\n",
      "number of contigs without blastx hit: 9\n",
      "\n",
      "70 CMS001_008_Ra_S3\n",
      "total number of contigs in sample: 13396\n",
      "number of contigs without blastx hit: 1292\n",
      "\n",
      "71 CMS002_047b_Rb_S194_L004\n",
      "total number of contigs in sample: 2522\n",
      "number of contigs without blastx hit: 84\n",
      "\n",
      "72 CMS001_013_Ra_S5\n",
      "total number of contigs in sample: 7071\n",
      "number of contigs without blastx hit: 313\n",
      "\n",
      "73 CMS002_025d_Rb_S143_L004\n",
      "total number of contigs in sample: 0\n",
      "number of contigs without blastx hit: 0\n",
      "\n",
      "74 CMS001_036_Ra_S20\n",
      "total number of contigs in sample: 15097\n",
      "number of contigs without blastx hit: 1085\n",
      "\n",
      "75 CMS002_040a_Rb_S174_L004\n",
      "total number of contigs in sample: 6610\n",
      "number of contigs without blastx hit: 286\n",
      "\n",
      "76 CMS002_047c_Rb_S195_L004\n",
      "total number of contigs in sample: 7295\n",
      "number of contigs without blastx hit: 326\n",
      "\n",
      "77 CMS001_025_Ra_S7\n",
      "total number of contigs in sample: 16379\n",
      "number of contigs without blastx hit: 1419\n",
      "\n",
      "78 CMS002_0Water3_Rb_S151_L004\n",
      "total number of contigs in sample: 112\n",
      "\n",
      "79 CMS001_004_Ra_S2\n",
      "total number of contigs in sample: 836\n",
      "number of contigs without blastx hit: 26\n",
      "\n",
      "80 CMS002_029e_Rb_S164_L004\n",
      "total number of contigs in sample: 1909\n",
      "number of contigs without blastx hit: 80\n",
      "\n",
      "81 CMS002_044d_Rb_S181_L004\n",
      "total number of contigs in sample: 304\n",
      "number of contigs without blastx hit: 8\n",
      "\n",
      "82 CMS001_021_Ra_S16\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of contigs in sample: 12284\n",
      "number of contigs without blastx hit: 607\n",
      "\n",
      "83 CMS001_053_Ra_S8\n",
      "total number of contigs in sample: 2587\n",
      "number of contigs without blastx hit: 90\n",
      "\n",
      "84 CMS002_020d_Rb_S134_L004\n",
      "total number of contigs in sample: 10292\n",
      "number of contigs without blastx hit: 832\n",
      "\n",
      "85 CMS001_030_Ra_S7\n",
      "total number of contigs in sample: 1004\n",
      "number of contigs without blastx hit: 49\n",
      "\n",
      "86 CMS001_016_Ra_S6\n",
      "total number of contigs in sample: 21018\n",
      "number of contigs without blastx hit: 2437\n",
      "\n",
      "87 CMS002_047d_Rb_S196_L004\n",
      "total number of contigs in sample: 7764\n",
      "number of contigs without blastx hit: 369\n",
      "\n",
      "88 CMS001_043_Ra_S24\n",
      "total number of contigs in sample: 5072\n",
      "number of contigs without blastx hit: 278\n",
      "\n",
      "89 CMS001_water3_Qiagen_S26\n",
      "total number of contigs in sample: 6169\n",
      "\n",
      "90 CMS001_020_Ra_S15\n",
      "total number of contigs in sample: 6417\n",
      "number of contigs without blastx hit: 291\n",
      "\n",
      "91 CMS002_047i_Rb_S2_L004\n",
      "total number of contigs in sample: 5774\n",
      "number of contigs without blastx hit: 246\n",
      "\n",
      "92 CMS001_051_Ra_S8\n",
      "total number of contigs in sample: 8676\n",
      "number of contigs without blastx hit: 399\n",
      "\n",
      "93 CMS002_044b_Rb_S179_L004\n",
      "total number of contigs in sample: 62\n",
      "number of contigs without blastx hit: 3\n",
      "\n",
      "94 CMS002_025b_Rb_S141_L004\n",
      "total number of contigs in sample: 40\n",
      "number of contigs without blastx hit: 1\n",
      "\n",
      "95 CMS001_032_Ra_S7\n",
      "total number of contigs in sample: 2669\n",
      "number of contigs without blastx hit: 191\n",
      "\n",
      "96 CMS002_017d_Rb_S125_L004\n",
      "total number of contigs in sample: 212\n",
      "number of contigs without blastx hit: 7\n",
      "\n",
      "97 CMS002_038a_Rb_S172_L004\n",
      "total number of contigs in sample: 10828\n",
      "number of contigs without blastx hit: 567\n",
      "\n",
      "98 CMS002_047e_Rb_S197_L004\n",
      "total number of contigs in sample: 5777\n",
      "number of contigs without blastx hit: 233\n",
      "\n",
      "99 CMS001_006_Ra_S5\n",
      "total number of contigs in sample: 7740\n",
      "number of contigs without blastx hit: 390\n",
      "\n",
      "100 CMS002_020e_Rb_S135_L004\n",
      "total number of contigs in sample: 18006\n",
      "number of contigs without blastx hit: 1942\n",
      "\n",
      "101 CMS002_0Water6_Rb_S187_L004\n",
      "total number of contigs in sample: 38\n",
      "\n",
      "102 CMS002_0Water8_Rb_S11_L004\n",
      "total number of contigs in sample: 0\n",
      "\n",
      "103 CMS001_040_Ra_S21\n",
      "total number of contigs in sample: 2053\n",
      "number of contigs without blastx hit: 76\n",
      "\n",
      "104 CMS002_054a_Rb_S8_L004\n",
      "total number of contigs in sample: 375\n",
      "number of contigs without blastx hit: 19\n",
      "\n",
      "105 CMS001_031_Ra_S19\n",
      "total number of contigs in sample: 4721\n",
      "number of contigs without blastx hit: 208\n",
      "\n",
      "106 CMS001_023_Ra_S17\n",
      "total number of contigs in sample: 3713\n",
      "number of contigs without blastx hit: 106\n",
      "\n",
      "107 CMS001_010_Ra_S1\n",
      "total number of contigs in sample: 2854\n",
      "number of contigs without blastx hit: 101\n",
      "\n",
      "108 CMS002_017c_Rb_S124_L004\n",
      "total number of contigs in sample: 234\n",
      "number of contigs without blastx hit: 18\n",
      "\n",
      "109 CMS002_053a_Rb_S7_L004\n",
      "total number of contigs in sample: 41366\n",
      "number of contigs without blastx hit: 2443\n",
      "\n",
      "110 CMS001_047_Ra_S4\n",
      "total number of contigs in sample: 2936\n",
      "number of contigs without blastx hit: 114\n",
      "\n",
      "111 CMS002_047f_Rb_S198_L004\n",
      "total number of contigs in sample: 5212\n",
      "number of contigs without blastx hit: 204\n",
      "\n",
      "112 CMS002_044c_Rb_S180_L004\n",
      "total number of contigs in sample: 166\n",
      "number of contigs without blastx hit: 5\n",
      "\n",
      "113 CMS001_045_Ra_S2\n",
      "total number of contigs in sample: 4270\n",
      "number of contigs without blastx hit: 177\n",
      "\n",
      "114 CMS001_055_Ra_S9\n",
      "total number of contigs in sample: 2423\n",
      "number of contigs without blastx hit: 117\n",
      "\n",
      "115 CMS001_018_Ra_S14\n",
      "total number of contigs in sample: 2598\n",
      "number of contigs without blastx hit: 60\n",
      "\n",
      "116 CMS002_047h_Rb_S1_L004\n",
      "total number of contigs in sample: 5498\n",
      "number of contigs without blastx hit: 229\n",
      "\n",
      "117 CMS002_018a_Rb_S128_L004\n",
      "total number of contigs in sample: 37\n",
      "number of contigs without blastx hit: 5\n",
      "\n",
      "118 CMS001_005_Ra_S3\n",
      "total number of contigs in sample: 23596\n",
      "number of contigs without blastx hit: 2559\n",
      "\n",
      "119 CMS002_028c_Rb_S156_L004\n",
      "total number of contigs in sample: 5929\n",
      "number of contigs without blastx hit: 226\n",
      "\n",
      "120 CMS002_025f_Rb_S145_L004\n",
      "total number of contigs in sample: 0\n",
      "number of contigs without blastx hit: 0\n",
      "\n",
      "121 CMS002_047a_Rb_S193_L004\n",
      "total number of contigs in sample: 6003\n",
      "number of contigs without blastx hit: 218\n",
      "\n",
      "122 CMS001_049_Ra_S6\n",
      "total number of contigs in sample: 2574\n",
      "number of contigs without blastx hit: 76\n",
      "\n",
      "123 CMS001_056_Ra_S10\n",
      "total number of contigs in sample: 2143\n",
      "number of contigs without blastx hit: 63\n",
      "\n",
      "124 CMS001_052_Ra_S7\n",
      "total number of contigs in sample: 2881\n",
      "number of contigs without blastx hit: 121\n",
      "\n",
      "125 CMS002_026e_Rb_S150_L004\n",
      "total number of contigs in sample: 314\n",
      "number of contigs without blastx hit: 1\n",
      "\n",
      "126 CMS002_045g_Rb_S190_L004\n",
      "total number of contigs in sample: 96\n",
      "number of contigs without blastx hit: 1\n",
      "\n",
      "127 CMS002_020a_Rb_S131_L004\n",
      "total number of contigs in sample: 45574\n",
      "number of contigs without blastx hit: 6158\n",
      "\n",
      "128 CMS001_water2_S24\n",
      "total number of contigs in sample: 196\n",
      "\n",
      "129 CMS002_027b_Rb_S153_L004\n",
      "total number of contigs in sample: 1389\n",
      "number of contigs without blastx hit: 34\n",
      "\n",
      "130 CMS001_033_Ra_S8\n",
      "total number of contigs in sample: 6935\n",
      "number of contigs without blastx hit: 346\n",
      "\n",
      "131 CMS002_057a_Rb_S10_L004\n",
      "total number of contigs in sample: 623\n",
      "number of contigs without blastx hit: 24\n",
      "\n",
      "132 CMS002_046b_Rb_S192_L004\n",
      "total number of contigs in sample: 1119\n",
      "number of contigs without blastx hit: 40\n",
      "\n",
      "133 CMS002_013a_Rb_S120_L004\n",
      "total number of contigs in sample: 1716\n",
      "number of contigs without blastx hit: 85\n",
      "\n",
      "134 CMS001_017_Ra_S6\n",
      "total number of contigs in sample: 7223\n",
      "number of contigs without blastx hit: 361\n",
      "\n",
      "135 CMS001_035_Ra_S20\n",
      "total number of contigs in sample: 6126\n",
      "number of contigs without blastx hit: 178\n",
      "\n",
      "136 CMS001_038_Ra_S22\n",
      "total number of contigs in sample: 4575\n",
      "number of contigs without blastx hit: 197\n",
      "\n",
      "137 CMS001_027_Ra_S16\n",
      "total number of contigs in sample: 4251\n",
      "number of contigs without blastx hit: 104\n",
      "\n",
      "138 CMS002_045b_Rb_S184_L004\n",
      "total number of contigs in sample: 261\n",
      "number of contigs without blastx hit: 7\n",
      "\n",
      "139 CMS001_046_Ra_S3\n",
      "total number of contigs in sample: 2156\n",
      "number of contigs without blastx hit: 74\n",
      "\n",
      "140 CMS002_020c_Rb_S133_L004\n",
      "total number of contigs in sample: 8437\n",
      "number of contigs without blastx hit: 767\n",
      "\n",
      "141 CMS002_031a_Rb_S165_L004\n",
      "total number of contigs in sample: 8136\n",
      "number of contigs without blastx hit: 466\n",
      "\n",
      "142 CMS001_029_Ra_S18\n",
      "total number of contigs in sample: 1154\n",
      "number of contigs without blastx hit: 17\n",
      "\n",
      "143 CMS002_046a_Rb_S191_L004\n",
      "total number of contigs in sample: 963\n",
      "number of contigs without blastx hit: 27\n",
      "\n",
      "144 CMS001_054_Ra_S11\n",
      "total number of contigs in sample: 573\n",
      "number of contigs without blastx hit: 17\n",
      "\n",
      "145 CMS002_017b_Rb_S123_L004\n",
      "total number of contigs in sample: 267\n",
      "number of contigs without blastx hit: 8\n",
      "\n",
      "146 CMS002_028a_Rb_S154_L004\n",
      "total number of contigs in sample: 5396\n",
      "number of contigs without blastx hit: 198\n",
      "\n",
      "147 CMS002_045e_Rb_S188_L004\n",
      "total number of contigs in sample: 226\n",
      "number of contigs without blastx hit: 1\n",
      "\n",
      "148 CMS002_035a_Rb_S169_L004\n",
      "total number of contigs in sample: 14201\n",
      "number of contigs without blastx hit: 874\n",
      "\n",
      "149 CMS002_020b_Rb_S132_L004\n",
      "total number of contigs in sample: 4937\n",
      "number of contigs without blastx hit: 544\n",
      "\n",
      "150 CMS001_003_Ra_S2\n",
      "total number of contigs in sample: 17208\n",
      "number of contigs without blastx hit: 1884\n",
      "\n",
      "151 CMS002_019a_Rb_S130_L004\n",
      "total number of contigs in sample: 352\n",
      "number of contigs without blastx hit: 26\n",
      "\n",
      "152 CMS002_026b_Rb_S147_L004\n",
      "total number of contigs in sample: 161\n",
      "number of contigs without blastx hit: 0\n",
      "\n",
      "153 CMS002_045c_Rb_S185_L004\n",
      "total number of contigs in sample: 102\n",
      "number of contigs without blastx hit: 1\n",
      "\n",
      "154 CMS002_037a_Rb_S171_L004\n",
      "total number of contigs in sample: 8566\n",
      "number of contigs without blastx hit: 407\n",
      "\n",
      "155 CMS001_024_Ra_S15\n",
      "total number of contigs in sample: 3674\n",
      "number of contigs without blastx hit: 63\n",
      "\n",
      "156 CMS002_033a_Rb_S167_L004\n",
      "total number of contigs in sample: 14822\n",
      "number of contigs without blastx hit: 804\n",
      "\n",
      "157 CMS002_028d_Rb_S157_L004\n",
      "total number of contigs in sample: 4634\n",
      "number of contigs without blastx hit: 187\n",
      "\n",
      "158 CMS001_037_Ra_S21\n",
      "total number of contigs in sample: 5980\n",
      "number of contigs without blastx hit: 321\n",
      "\n",
      "159 CMS002_004a_Rb_S117_L004\n",
      "total number of contigs in sample: 642\n",
      "number of contigs without blastx hit: 20\n",
      "\n",
      "160 CMS002_029d_Rb_S162_L004\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of contigs in sample: 6400\n",
      "number of contigs without blastx hit: 250\n",
      "\n",
      "161 CMS001_water4_Zymo_S27\n",
      "total number of contigs in sample: 449\n"
     ]
    }
   ],
   "source": [
    "J=json.load(open('/Users/evogytis/Documents/skeeters/treemap/displayed_taxa.json','r')) ## load designated treemap branches\n",
    "branches={b['taxid']:b for b in J} ## flat list of branches indexed by taxid\n",
    "branches['no_hit']={'taxonomy':'no blast hit','taxid':'no_hit'}\n",
    "\n",
    "local_path='/Users/evogytis/Dropbox/Jupyter_notebooks/Biohub/California_mosquitoes/s3_bucket/'\n",
    "\n",
    "sampleCount=1\n",
    "for subdir, dirs, files in os.walk(local_path): ## iterate through samples\n",
    "    sample=os.path.basename(subdir) ## sample name is folder name\n",
    "    \n",
    "    search={}\n",
    "    print('\\n%d %s'%(sampleCount,sample))\n",
    "    for fname in files: ## iterate through files in subfolders\n",
    "        fpath=os.path.join(subdir, fname)\n",
    "        if 'blast' in fname: ## only interested in blast outputs\n",
    "            if 'gsnap' in fname:\n",
    "                blast_type='nt'\n",
    "            elif 'rapsearch' in fname:\n",
    "                blast_type='aa'\n",
    "            \n",
    "            blast_search=SearchIO.parse(fpath,format='blast-tab') ## read blast output\n",
    "            search[blast_type]=SearchIO.to_dict(blast_search) ## turn to dict\n",
    "        \n",
    "        elif 'fasta' in fname: ## fasta contains contigs\n",
    "            contigs=SeqIO.parse(fpath,format='fasta') ## parse fasta contigs\n",
    "            seqs=SeqIO.to_dict(contigs) ## convert to dict\n",
    "            print('total number of contigs in sample: %s'%(len(seqs)))\n",
    "            \n",
    "    if len(search)==2 and contigs and 'water' not in sample.lower(): ## once done with sample, if there's nt and nr blast results, contigs are loaded and they're not controls - proceed\n",
    "        for searchType in search: ## iterate over nt or nr blast results\n",
    "            no_hits=len(seqs)-len(search[searchType]) ## no hits is (total number of contigs - blast hits)\n",
    "            \n",
    "            if searchType=='aa':\n",
    "                if 'attrs' not in branches['no_hit']: ## no-hit branch doesn't have attributes yet\n",
    "                    branches['no_hit']['attrs']={}\n",
    "                branches['no_hit']['attrs'][sample]={'contig_count':no_hits} ## assign sample contig_count\n",
    "                print('number of contigs without blastx hit: %d'%(no_hits))\n",
    "            \n",
    "            for contig in search[searchType]: ## iterate through contigs blasted\n",
    "                blast_hit={} ## keep a record of hits\n",
    "                bitscores={} ## keep a record of bitscores\n",
    "                percent_id={} ## keep a record of percent identities\n",
    "                for hit in search[searchType][contig]: ## iterate over hits to contig\n",
    "                    hit_id=hit.id\n",
    "                    \n",
    "                    ## want to keep hits where best HSP is <1e-2 and longest match is >200 nt\n",
    "                    if min([h.evalue for h in hit.hsps])<1e-2 and max([len(f) for f in hit.fragments])>200:\n",
    "                        blast_hit[hit_id]=hit\n",
    "                        bitscores[hit_id]=min([k.bitscore for k in hit.hsps]) ## keep track of worst HSP bitscore\n",
    "                        percent_id[hit_id]=min([k.ident_pct for k in hit.hsps]) ## keep track of worst HSP percent identity\n",
    "                \n",
    "                cutoff=0.1 ## within 10% of top hit\n",
    "                filteredHits=filter(lambda k: max(bitscores.values())*(1.0-cutoff)<=bitscores[k], blast_hit) ## filter hits down to those whose bitscore within 10% of the best bitscore\n",
    "                hitLineages=[ncbi.get_lineage(acc_to_taxid[k]) for k in filteredHits] ## get lineage of every hit\n",
    "                hitLineages=list(filter(lambda x: 7157 not in x, hitLineages)) ## filter out lineages with mosquito in its path\n",
    "                \n",
    "                if searchType=='aa': ## if nr\n",
    "                    hitLineages=list(filter(lambda x: 10239 in x, hitLineages)) ## keep viruses (nr)\n",
    "                elif searchType=='nt': ## if nt\n",
    "                    hitLineages=list(filter(lambda x: 10239 not in x, hitLineages)) ## ignore viruses (nt)\n",
    "                \n",
    "                if len(list(hitLineages))>0: ## hits exist\n",
    "                    ranksPresent=functools.reduce(set.intersection,map(set,hitLineages)) ## reduce lineages seen down to those that are shared by all hits\n",
    "                    \n",
    "                    lineages=[ncbi.get_lineage(lin) for lin in ranksPresent] ## get full lineage of each taxonomic rank that's shared across all hits\n",
    "                    mrca_path=[lin for lin in lineages if len(lin)==max(map(len,lineages))][-1] ## longest lineage path is the most recent common ancestor\n",
    "                    \n",
    "                    for rank in mrca_path[::-1]: ## invert ancestor path, so ranks iterated from most recent to oldest\n",
    "                        if rank in branches: ## rank is present amongst those designated for treemap\n",
    "                            branch=branches[rank] ## fetch branch from treemap\n",
    "                            \n",
    "                            if 'attrs' not in branch:\n",
    "                                branch['attrs']={} ## branch doesn't have attrs\n",
    "                                \n",
    "                            if sample not in branch['attrs']:\n",
    "                                branch['attrs'][sample]={} ## branch hasn't seen sample\n",
    "                                \n",
    "                            if 'contig_count' not in branch['attrs'][sample]:\n",
    "                                branch['attrs'][sample]['contig_count']=0 ## branch hasn't seen contig_count in sample\n",
    "                            branch['attrs'][sample]['contig_count']+=1 ## add to contig_count\n",
    "                            break ## no need to iterate further\n",
    "    sampleCount+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-03-22T17:42:49.873035Z",
     "start_time": "2019-03-22T17:42:49.722959Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "taxa_tree={'children':[branches['no_hit']]} ## tree structure (+ add no-hit branch from the beginning)\n",
    "\n",
    "J=sorted(J,key=lambda k: len(ncbi.get_lineage(k['taxid'])))\n",
    "\n",
    "for b in branches: ## iterate through flat list of branches\n",
    "    if 'attrs' in branches[b]: ## branch has been annotated before\n",
    "        branches[b]['attrs']['contig_count']=sum([branches[b]['attrs'][c]['contig_count'] for c in branches[b]['attrs'] if 'CMS' in c]) ## compute sum of sample contig counts\n",
    "    else:\n",
    "        branches[b]['attrs']={'contig_count':0} ## nothing landed on the branch\n",
    "\n",
    "for taxon in J: ## iterate over every taxon in treemap\n",
    "    lineage=ncbi.get_lineage(taxon['taxid']) ## get its lineage\n",
    "    \n",
    "    if len(lineage)==1 and taxon not in taxa_tree['children']: ## if root\n",
    "        taxa_tree['children'].append(taxon) ## add root to tree\n",
    "        \n",
    "    for lin in lineage[::-1][1:]: ## iterate through lineage, starting from most recent, ignore first entry (self)\n",
    "        \n",
    "        if lin in branches: ## rank present amongst branches\n",
    "            parent=branches[lin] ## grab parent\n",
    "            \n",
    "            if 'children' not in parent: ## if parent doesn't have children yet - add the attribute\n",
    "                parent['children']=[]\n",
    "                \n",
    "            if taxon not in parent['children']: ## branch wasn't assigned to its parent yet\n",
    "                parent['children'].append(taxon) ## add child to parent\n",
    "            \n",
    "            break\n",
    "\n",
    "json.dump(taxa_tree,open('/Users/evogytis/Documents/skeeters/treemap/skeeters.json','w'),indent=1) ## write json out to repo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-03-22T17:42:50.958670Z",
     "start_time": "2019-03-22T17:42:49.877848Z"
    }
   },
   "outputs": [],
   "source": [
    "## then run python3 server locally in the folder where treemap.html is located:\n",
    "## python3 -m http.server 4000\n",
    "## python -m SimpleHTTPServer 4000"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
