{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Treemap\n",
    "\n",
    "----------------\n",
    "\n",
    "This Jupyter notebook walks through the steps of creating a treemap that can be used to display hierarchical information. As it is now it is designed to show how reads from a metagenomic study are distributed across the different branches of living organisms.\n",
    "\n",
    "*written by Gytis Dudas, 2019*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: process and prepare archived NCBI taxonomy file (optional)\n",
    "\n",
    "NCBI updated virus taxonomy over the course of this project based on ICTV proposals, but this change has been incomplete _e.g._ most sequences under `ssRNA negative-strand viruses` (descriptive category) have been moved to `Negarnaviricota` (genealogical category) but >100 accessions that are distinctly negative sense RNA viruses still remain under the old category following the evacuation. For the sake of consistency we have decided to go with the older version of taxonomy. \n",
    "\n",
    "This cell downloads the required version of taxonomy, unzips it and tar+gzips it, which is the only format `ete3` will take."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-27T10:11:08.415215Z",
     "start_time": "2019-11-27T10:11:08.398139Z"
    }
   },
   "outputs": [],
   "source": [
    "# %%bash\n",
    "\n",
    "# store_folder=/Users/evogytis/Downloads\n",
    "# cd $store_folder\n",
    "\n",
    "# tax_db=\"taxdmp_2019-01-01\" ## latest taxonomy release that still contains original virus taxonomy\n",
    "# tax_db=\"taxdmp_2019-11-01\" ## latest taxonomy release\n",
    "# curl -O ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_archive/$tax_db.zip ## download taxonomy file\n",
    "\n",
    "# rm -rf $store_folder/$tax_db ## remove existing category\n",
    "# mkdir $store_folder/$tax_db\n",
    "\n",
    "# unzip -o $store_folder/$tax_db.zip -d $store_folder/$tax_db\n",
    "\n",
    "# cd $store_folder/$tax_db; tar -czvf $store_folder/$tax_db.tar.gz *[dmp,prt,txt]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: load libraries, update ete3's taxonomy to file provided\n",
    "\n",
    "This cell loads three native libraries (`os`, `json`, and `glob`) and `ete3`. `ete3` is required to place all BLAST hits in the treemap via the `get_lineage` command. Updating the taxonomy database takes ~3 minutes.\n",
    "\n",
    "Also loaded is a CD-HIT file of clustered contigs, and file mapping numbers of reads to each contig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-27T10:11:12.482612Z",
     "start_time": "2019-11-27T10:11:08.419615Z"
    }
   },
   "outputs": [],
   "source": [
    "import ete3\n",
    "import os,json,glob\n",
    "\n",
    "# taxonomy_path='/Users/evogytis/Downloads/taxdmp_2019-01-01.tar.gz'\n",
    "taxonomy_path='/Users/evogytis/Downloads/taxdmp_2019-11-01.tar.gz'\n",
    "base_path='/Users/evogytis/Documents/manuscripts/skeeters/data' ## point to where the data folder to the repo is locally\n",
    "\n",
    "\n",
    "ncbi=ete3.ncbi_taxonomy.NCBITaxa()\n",
    "# ncbi.update_taxonomy_database(taxdump_file=taxonomy_path) ## uncomment to update ete3's taxonomy\n",
    "\n",
    "#########\n",
    "## Load clusters of homologous contigs >500 nt\n",
    "#########\n",
    "from collections import defaultdict, namedtuple\n",
    "Member = namedtuple('Member', ['contig', 'length', 'percent_id', 'percent_id_sign', 'sample', 'coverage'])\n",
    "\n",
    "def parse_cdhit_row(row):\n",
    "    if '*' in row:\n",
    "        index, length, name, percent_id = row.split()\n",
    "        percent_id_sign, percent_id = '0', 100\n",
    "    else:\n",
    "        index, length, name, _, percent_id = row.split()\n",
    "    length = int(length.strip(',nt'))\n",
    "    name = name.strip('>').strip('.')\n",
    "    sample, contig = name.split('~')\n",
    "    coverage = float(contig.split('_')[-1])\n",
    "    \n",
    "    if percent_id != 100:\n",
    "        percent_id_sign, percent_id = percent_id.strip('%').split('/')\n",
    "        percent_id = float(percent_id)\n",
    "    return Member(contig=contig, sample=sample, length=length,\n",
    "                  percent_id=percent_id, percent_id_sign=percent_id_sign, coverage=coverage)\n",
    "\n",
    "clusters = defaultdict(list)\n",
    "with open(base_path+'/500_contigs_cluster.clstr', 'r') as file:\n",
    "    for line in file:\n",
    "        if line.startswith('>Cluster'):\n",
    "            cluster_id = line.split()[-1]\n",
    "        else:\n",
    "            member = parse_cdhit_row(line)\n",
    "            if 'water' in member.sample.lower():\n",
    "                continue\n",
    "            clusters[cluster_id].append(member)\n",
    "            \n",
    "contig2cluster={}\n",
    "for cluster_id in clusters:\n",
    "    for m in clusters[cluster_id]:\n",
    "        contig2cluster[m.contig]=cluster_id\n",
    "\n",
    "#########\n",
    "## Load read counts of each contig\n",
    "#########\n",
    "contig_reads={} ## will contain a mapping of sample: contig ID: number of reads\n",
    "for reads_file in glob.glob(os.path.join(base_path,'s3/contig_quality/*/contig_stats_all.tsv')): ## iterate over sample LCA files\n",
    "    fsample=os.path.basename(os.path.dirname(reads_file))\n",
    "    for line in open(reads_file,'r'): ## iterate over lines\n",
    "        l=line.strip('\\n').split('\\t')\n",
    "        if l[0]=='sample':\n",
    "            header={x:i for i,x in enumerate(l)} ## create header dict\n",
    "        else:\n",
    "            sample=l[header['sample']]\n",
    "            contig_name=l[header['contig_name']] ## get contig name\n",
    "            read_count=int(float(l[header['read_count']])) ## get read count\n",
    "            \n",
    "            assert sample==fsample ## sanity check that sample name in file is same as folder name\n",
    "            if sample not in contig_reads:\n",
    "                contig_reads[sample]={}\n",
    "            \n",
    "            contig_reads[sample][contig_name]=read_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-02T09:34:31.678394Z",
     "start_time": "2019-05-02T09:34:31.335093Z"
    }
   },
   "source": [
    "## Step 3: load pre-defined taxa that will be displayed in the treemap\n",
    "-----\n",
    "\n",
    "The following cell loads a file (`displayed_taxa_reads.json`) that looks like this:\n",
    "\n",
    "```\n",
    "[{\"taxid\": 1,\n",
    "\"taxonomy\": \"root\"},\n",
    "{\"taxid\": 10239,\n",
    "\"taxonomy\": \"Viruses\"},\n",
    "{\"taxid\": 131567,\n",
    "\"taxonomy\": \"cellular organisms\"},\n",
    "{\"taxid\": 2157,\n",
    "\"taxonomy\": \"Archaea\"},\n",
    "{\"taxid\": 2759,\n",
    "\"taxonomy\": \"Eukaryota\"},\n",
    "{\"taxid\": 2,\n",
    "\"taxonomy\": \"Bacteria\"}]\n",
    "```\n",
    "\n",
    "It is loaded as a flat list of branches, which get annotated later (here with contig and read counts) and in the last steps built into a nested tree data structure (_i.e._ here it would be `(Root(Viruses,CellularOrganisms(Eukaryota,Bacteria,Archaea)));`).\n",
    "\n",
    "By default the produced treemap will only contain the taxids listed in this file and anything higher _e.g._ _Metazoa_ would be traversed back to a taxid that does exist in the file which would be _Eukaryota_.\n",
    "\n",
    "At the end of the cell additional branches are added to the tree. These are high-order taxonomic lineages that are currently not listed by NCBI (_e.g._ the Narna-Levi supergroup that links _Narnaviridae_ and _Leviviridae_), additionally linked into an even higher order structure based on strandedness (_i.e._ Baltimore class) which can be paraphyletic but is done here to introduce more structure into the treemap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-27T10:11:12.517997Z",
     "start_time": "2019-11-27T10:11:12.484892Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current taxids that exist as branches: dict_keys([1, 10239, 131567, 4751, 2759, 2, 953, 1236, 1437010, 8782, 9126, 5654, 1286322, 5690, 28211, 1206794, 33213, 203691, 6029, 451864, 6231, 6656, 33090, 5794, 33154, 7711, 5820, 91347, 135619, 2497569, 11157, 11308, 1980410, 1980417, 1980418, 1980416, 1299308, 11270, 2501985, 186766, 11989, 2560063, 2169577, 119163, 11012, 11050, 39738, 12283, 10880, 11006, 249310, 464095, 699189, 232795, 249184, 675071, 'uncurated'])\n",
      "\n",
      "family Tymoviridae (Hepe-Virga supergroup)\n",
      "family Virgaviridae (Hepe-Virga supergroup)\n",
      "family Botourmiaviridae (Narna-Levi supergroup)\n",
      "family Narnaviridae (Narna-Levi supergroup)\n",
      "family Leviviridae (Narna-Levi supergroup)\n",
      "family Luteoviridae (Luteo-Sobemo supergroup)\n",
      "family Solemoviridae (Luteo-Sobemo supergroup)\n",
      "family Tombusviridae (Tombus-Noda supergroup)\n",
      "family Nodaviridae (Tombus-Noda supergroup)\n",
      "family Totiviridae (Toti-Chryso supergroup)\n",
      "family Chrysoviridae (Toti-Chryso supergroup)\n",
      "supergroup family Flaviviridae\n",
      "supergroup family Reoviridae\n",
      "supergroup family Partitiviridae\n",
      "supergroup family Picornavirales\n"
     ]
    }
   ],
   "source": [
    "J=json.load(open(os.path.join(base_path,'../treemap/displayed_taxa_reads.json'),'r')) ## load designated treemap branches\n",
    "branches={b['taxid']:b for b in J} ## flat list of branches indexed by taxid\n",
    "\n",
    "remove_branches=[] ## empty list that will contain taxids whose lineage cannot be recovered by ete3 (because of out-dated taxonomy)\n",
    "for taxid in branches: ## iterate over branches loaded so far\n",
    "    try:\n",
    "        ncbi.get_lineage(taxid) ## attempt to get lineage\n",
    "    except ValueError: ## attempt failed\n",
    "        remove_branches.append(taxid) ## remember taxid for removal later\n",
    "        print('taxid %s not in taxdump.tar.gz file loaded earlier, it will be excluded'%(taxid))\n",
    "        \n",
    "for taxid in remove_branches:\n",
    "    branches.pop(taxid) ## remove taxids that failed\n",
    "        \n",
    "branches['uncurated']={'taxonomy':'uncurated','taxid':'uncurated','attrs':{'colour': '#E7E7E6'}} ## also create a branch that could contain no hits (i.e. total number of queries minus number of queries that hit something)\n",
    "\n",
    "for b in branches: ## assign default colour to branches\n",
    "    branches[b]['attrs']={'colour':'#E7E7E6'} ## default colour is slategrey, but later cell assigns a colour based on descent from Bacteria, Eukaryotes or Viruses\n",
    "    \n",
    "print('current taxids that exist as branches: %s\\n'%(branches.keys()))\n",
    "\n",
    "###########\n",
    "## This code will add additional treemap compartments that group viral families into higher-order structures which don't exist in official NCBI taxonomy\n",
    "###########\n",
    "fam_taxids={249184: 'Tymoviridae',675071: 'Virgaviridae', \n",
    "            2560063: 'Botourmiaviridae', 186766: 'Narnaviridae', 11989: 'Leviviridae',\n",
    "            119163: 'Luteoviridae', 2169577: 'Solemoviridae', \n",
    "            39738: 'Tombusviridae', 12283: 'Nodaviridae', \n",
    "            11006: 'Totiviridae', 249310: 'Chrysoviridae', \n",
    "            11050: 'Flaviviridae', 10880: 'Reoviridae', \n",
    "            11012: 'Partitiviridae', 464095: 'Picornavirales'} ## link taxids to family names\n",
    "\n",
    "## there are 12 supergroups, but Partitis, Reos and Flavis are their own supergroups, whereas Picornas, Mononegs, Orthomyxos, and Bunyas exist as supergroup taxids on NCBI\n",
    "supergroups={'Tymoviridae': 'Hepe-Virga',\n",
    "             'Virgaviridae': 'Hepe-Virga',\n",
    "             'Botourmiaviridae': 'Narna-Levi', \n",
    "             'Narnaviridae': 'Narna-Levi', \n",
    "             'Leviviridae': 'Narna-Levi',\n",
    "             'Luteoviridae': 'Luteo-Sobemo', \n",
    "             'Solemoviridae': 'Luteo-Sobemo', \n",
    "             'Tombusviridae': 'Tombus-Noda', \n",
    "             'Nodaviridae': 'Tombus-Noda', \n",
    "             'Totiviridae': 'Toti-Chryso', \n",
    "             'Chrysoviridae': 'Toti-Chryso'}\n",
    "\n",
    "## adds an extra layer to treemap which is strandedness\n",
    "strands={10880: 'dsRNA viruses', ## Reo\n",
    "         'Toti-Chryso': 'dsRNA viruses', \n",
    "         11012: 'dsRNA viruses', ## Partiti\n",
    "         'Luteo-Sobemo': '(+)ssRNA viruses', \n",
    "         'Tombus-Noda': '(+)ssRNA viruses', \n",
    "         'Narna-Levi': '(+)ssRNA viruses', \n",
    "         11050: '(+)ssRNA viruses', ## Flavi\n",
    "         'Hepe-Virga': '(+)ssRNA viruses', \n",
    "         464095: '(+)ssRNA viruses'} ## Picorna\n",
    "\n",
    "high_order_lineage={} ## will contain new high-order taxid lineages (adds supergroup classification)\n",
    "\n",
    "for fam in fam_taxids: ## iterate over family taxids\n",
    "    if fam_taxids[fam] in supergroups: ## family is part of supergroup\n",
    "        print('family %s (%s supergroup)'%(fam_taxids[fam],supergroups[fam_taxids[fam]]))\n",
    "        sg=supergroups[fam_taxids[fam]] ## get supergroup of family\n",
    "        st=strands[sg] ## get strand of supergroup\n",
    "        high_order_lineage[fam]=[1,10239,st,sg,fam] ## assign new family lineage with supergroup and strand\n",
    "    else:\n",
    "        sg=fam ## family is its own supergroup\n",
    "        print('supergroup family %s'%(fam_taxids[fam]))\n",
    "        st=strands[sg] ## get strand of supergroup\n",
    "    \n",
    "    if st not in branches: ## create new branches\n",
    "        branches[st]={'attrs':{'colour': '#E7E7E6'},'taxid': st, 'taxonomy': st}\n",
    "    if sg not in branches:\n",
    "        branches[sg]={'attrs':{'colour': '#E7E7E6'},'taxid': sg, 'taxonomy': sg}\n",
    "    \n",
    "    high_order_lineage[sg]=[1,10239,st,sg] ## assign strand+supergroup lineage\n",
    "    high_order_lineage[st]=[1,10239,st] ## assign strand lineage\n",
    "    \n",
    "high_order_lineage['uncurated']=[1,10239,'uncurated'] ## uncurated contig reads will sit under Viruses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4: annotation of branches\n",
    "------\n",
    "\n",
    "In this cell we add the information about how many reads (and contigs) belong to each taxonomic compartment. The `summarise` option allows you to choose between relying on the treemap compartments specified in the `displayed_taxa_reads.json` file or creating a new treemap compartment for every new taxid seen in the BLAST results summary file.\n",
    "\n",
    "If `backbone` is chosen every taxid in BLAST output summary is traversed back to the root and all reads get assigned to the first treemap compartment available, _e.g._ if BLAST hit was _Metazoa_ all the reads mapping to the contig would be assigned to _Eukaryota_ which is the highest treemap compartment available.\n",
    "\n",
    "If `all` is chosen a contig that's been assigned to _Metazoa_ will create a new treemap compartment with all the reads from the contig assigned to it. Future contigs assigned to _Metazoa_ will contribute their reads to this compartment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-27T10:11:15.808284Z",
     "start_time": "2019-11-27T10:11:12.521724Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "previously described virus Culex pipiens-associated Tunisia virus with official taxid 2079148 will be redirected as child of 1527522\n",
      "previously described virus Hubei chryso-like virus 1 with official taxid 1922855 will be redirected as child of 2587491\n",
      "previously described virus Culex bunya-like virus with official taxid 2304497 will be redirected as child of 39718\n",
      "previously described virus Guadeloupe Culex tymo-like virus with official taxid 2607736 will be redirected as child of 330383\n",
      "previously described virus Guadeloupe mosquito virus with official taxid 2607735 will be redirected as child of 2600328\n",
      "previously described virus Wenzhou sobemo-like virus 4 with official taxid 1923660 will be redirected as child of 502177\n",
      "previously described virus Wuhan insect virus 33 with official taxid 1923736 will be redirected as child of 336635\n",
      "previously described virus Hubei virga-like virus 2 with official taxid 1923335 will be redirected as child of 1527522\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 197514 was translated into 2555385\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "#########\n",
    "## Choose between summarising read counts onto a \"backbone\", i.e. using pre-defined taxids from displayed_taxa_reads.json file\n",
    "## OR\n",
    "## \"all\" which adds every taxid encountered in BLAST output to the treemap.\n",
    "#########\n",
    "summarise='backbone'\n",
    "# summarise='all'\n",
    "\n",
    "hexapoda_lineage=ncbi.get_lineage(6960) ## get the taxids of Hexapoda lineage\n",
    "\n",
    "###############\n",
    "## Load read counts and taxid assignments for each distinct virus lineage from virus.json\n",
    "###############\n",
    "in_json=open(base_path+'/darkmatter/virus.json','r')\n",
    "virus=json.load(in_json)\n",
    "\n",
    "redirect={} ## will redirect some viruses from the taxids they were submitted under (usually under \"unclassified Viruses\") to something more informative\n",
    "curated_contigs=[] ## will keep track of which contigs were viral and curated (and so already included in the treemap)\n",
    "\n",
    "for pol_group in virus: ## iterate over polymerase groups\n",
    "    taxid=virus[pol_group]['taxid'] ## get assigned taxid to virus\n",
    "    subm_taxid=virus[pol_group]['submission_taxid'] ## get the taxid under which the virus will be submitted (may belong to a group sitting under \"unclassified Viruses\")\n",
    "    \n",
    "    name=virus[pol_group]['provisional_name'] if 'provisional_name' in virus[pol_group] else virus[pol_group]['name'] ## get the official or provisional name of the virus (human-readable)\n",
    "    \n",
    "    if taxid!=subm_taxid: ## assigned taxid does not match submission taxid (previously described viruses under incorrect taxid in the database)\n",
    "        print('previously described virus %s with official taxid %s will be redirected as child of %s'%(name,subm_taxid,taxid))\n",
    "        redirect[subm_taxid]=taxid ## add taxid for redirection later\n",
    "        vir_id=subm_taxid ## virus ID is the incorrect taxid\n",
    "    elif 'provisional_name' in virus[pol_group]: ## new virus, still needs redirect\n",
    "        redirect[name]=taxid ## since new viruses don't exist under their own taxid it will need to be created later\n",
    "        vir_id=name\n",
    "    else: ## taxid matches assigned taxid, not a new virus and sits at the appropriate taxid\n",
    "        vir_id=taxid\n",
    "    \n",
    "    branches[vir_id]={'attrs':{'colour': '#E7E7E6'}} ## create a new branch for treemap\n",
    "    branches[vir_id]['taxonomy']=name ## give it a name\n",
    "    branches[vir_id]['taxid']=vir_id ## give whatever was closest to a taxid (provisional name for new viruses)\n",
    "    \n",
    "    for seg in virus[pol_group]['segments']: ## iterate over segments\n",
    "        seg_clusters=virus[pol_group]['segments'][seg]['clusters'] ## get clusters comprising the segment\n",
    "        for clust in seg_clusters: ## iterate over clusters\n",
    "            for member in clusters[str(clust)]: ## iterate over contigs comprising the cluster\n",
    "                contig_name=member.contig ## get contig name\n",
    "                sample=member.sample ## get sample name\n",
    "                curated_contigs.append(contig_name) ## remember that contig has been assigned to viruses\n",
    "                \n",
    "                reads=contig_reads[sample][contig_name] ## get read counts\n",
    "                \n",
    "                if sample not in branches[vir_id]['attrs']:\n",
    "                    branches[vir_id]['attrs'][sample]={'read_count': 0, 'contig_count': 0}\n",
    "                    \n",
    "                branches[vir_id]['attrs'][sample]['read_count']+=reads ## add to reads for virus\n",
    "                branches[vir_id]['attrs'][sample]['contig_count']+=1 ## increment contig number\n",
    "\n",
    "###############\n",
    "## Load taxid assignments for non-curated taxids\n",
    "###############\n",
    "uncurated_reads=0\n",
    "\n",
    "for lca_file in glob.glob(os.path.join(base_path,'s3/contig_quality/*/contig_stats_lca.tsv')): ## iterate over sample LCA files\n",
    "    sample=os.path.basename(os.path.dirname(lca_file)) ## folder name is sample name\n",
    "    \n",
    "    for line in open(lca_file,'r'): ## iterate over contig IDs\n",
    "        l=line.strip('\\n').split('\\t')\n",
    "        if l[0]=='sample':\n",
    "            header={x:i for i,x in enumerate(l)} ## create header dict\n",
    "        else:\n",
    "            assert sample==l[header['sample']],'Sample in file does not match the location of file.'\n",
    "            contig_name=l[header['contig_name']] ## get contig name\n",
    "            \n",
    "            if l[header['taxid']]!='': ## taxid is not empty, so contig has a blast hit\n",
    "                taxid=int(float(l[header['taxid']])) ## get taxid\n",
    "                reads=contig_reads[sample][contig_name] ## get read count for contig\n",
    "                lineage=ncbi.get_lineage(taxid) ## get lineage of taxid\n",
    "\n",
    "                proceed=False ## assume can't proceed\n",
    "                if taxid not in hexapoda_lineage and 10239 not in ncbi.get_lineage(taxid) and contig_name not in curated_contigs: ## non-hexapod, non-virus, and not curated\n",
    "                    proceed=True ## go ahead if contig has not been curated or assigned to viruses or hexapods\n",
    "\n",
    "                if summarise=='all': ## adding all encountered taxids to treemap\n",
    "                    if taxid not in branches: ## taxid has not been seen before, create a new branch\n",
    "                        branches[taxid]={'attrs':{'colour': '#E7E7E6'}} ## create empty branch\n",
    "                        branches[taxid]['taxid']=taxid ## annotate taxid\n",
    "                        branches[taxid]['taxonomy']=ncbi.get_taxid_translator([taxid])[taxid] ## annotate taxonomy\n",
    "\n",
    "                for rank in lineage[::-1]: ## iterate over lineage of taxid in reverse order (youngest -> oldest)\n",
    "                    if proceed==True and rank in branches: ## given green light before and rank is present amongst those designated for treemap\n",
    "                        branch=branches[rank] ## fetch branch from treemap\n",
    "                        if sample not in branch['attrs']:\n",
    "                            branch['attrs'][sample]={'contig_count':0,'read_count':0}\n",
    "\n",
    "                        branch['attrs'][sample]['read_count']+=reads ## add to read count\n",
    "                        branch['attrs'][sample]['contig_count']+=1 ## add to contig_count\n",
    "                        break ## rank of lineage found in treemap so no need to iterate further, break loop\n",
    "\n",
    "                if 10239 in ncbi.get_lineage(taxid) and contig_name not in curated_contigs: ## couldn't proceed, but taxid is viral and uncurated\n",
    "                    if sample not in branches['uncurated']['attrs']:\n",
    "                        branches['uncurated']['attrs'][sample]={'read_count': 0, 'contig_count': 0}\n",
    "                    uncurated_reads+=reads ## increment reads for uncurated viral contigs\n",
    "                    branches['uncurated']['attrs'][sample]['read_count']+=reads ## increment read count\n",
    "                    branches['uncurated']['attrs'][sample]['contig_count']+=1 ## increment contig count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 5: build tree structure, output to file\n",
    "-------\n",
    "\n",
    "All the branches have been annotated up to this stage and now it's time to build the tree data structure. We start by summarising the contig and read counts across samples. We then specify colours for particular compartment and their descendants (_e.g._ -ssRNA viruses obviously need to be red-ish). The procedure involves fetching branch A, retrieving its lineage and looking for taxids along its lineage which exist as a treemap compartment (let's call it B). Treemap compartment A gets assigned as a child of treemap compartment B.\n",
    "\n",
    "The resulting tree structure is saved to file as a JSON and can be inspected using javascript code included in the repository by running a local server with the command and opening `treemap.html`:\n",
    "```\n",
    "python3 -m http.server 4000\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-27T10:11:15.946957Z",
     "start_time": "2019-11-27T10:11:15.818214Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n"
     ]
    }
   ],
   "source": [
    "taxa_tree={'children':[],'attrs':{'colour':'#E7E7E6'}} ## tree structure (+ add no-hit branch from the beginning)\n",
    "\n",
    "sorted_taxids=branches\n",
    "\n",
    "for b in sorted_taxids: ## iterate through flat list of branches\n",
    "    reads=[branches[b]['attrs'][c]['read_count'] for c in branches[b]['attrs'] if 'CMS' in c]\n",
    "    contigs=[branches[b]['attrs'][c]['contig_count'] for c in branches[b]['attrs'] if 'CMS' in c]\n",
    "    branches[b]['attrs']['read_count']=sum(reads)\n",
    "    branches[b]['attrs']['contig_count']=sum(contigs) ## compute sum of contig counts across samples  \n",
    "    branches[b]['attrs']['sample_count']=len(reads)\n",
    "\n",
    "## Colours assigned to a treemap compartment (all descendants inherit parental colour unless they have their own assigned colour)\n",
    "lineage_colours={'dsRNA viruses': 'grey', ## dsRNA\n",
    "                 '(+)ssRNA viruses': 'grey',  ## (+)ssRNA\n",
    "                 2497569: 'grey', ## (-)ssRNA\n",
    "                 10239: 'grey', ## viruses\n",
    "                 11157: '#FF7D48', ## Mononegas\n",
    "                 1980410: '#AC6569', ## Bunyas\n",
    "                 11308: '#C43A3F', ## Orthomyxos\n",
    "                 11050: '#0074C4', ## Flavi \n",
    "                 464095: '#42C754', ## Picorna\n",
    "                 11012: '#965DAE', ## Partiti\n",
    "                 10880: '#866E8B', ## Reo\n",
    "                 'Hepe-Virga': '#AFD1FF', ## Hepe-Virga\n",
    "                 'Narna-Levi': '#6967BC', ## Narna-Levi\n",
    "                 'Luteo-Sobemo': '#E6C930', ## Luteo-Sobemo\n",
    "                 'Toti-Chryso': '#D8A5CE', ## Toti-Chryso\n",
    "                 'Tombus-Noda': '#618F80', ## Tombus-Noda\n",
    "                 'uncurated': '#ACADAE', ## uncurated viral contigs\n",
    "                 2: '#867A5F', ## Bacteria\n",
    "                 953: '#86665F', ## Wolbachia\n",
    "                 2759: '#E5E4E2', ## Eukaryota\n",
    "                 5794:'#5E716A', ## apicomplexa\n",
    "                 7711: '#B2BEB5', ## chordata\n",
    "                 8782: '#6082B6', ## Aves\n",
    "                 9347: '#658EA9', ## Eutheria\n",
    "                 4751: '#808080', ## Fungi\n",
    "                 5654: '#708090' ## Trypanosomatidae\n",
    "                }\n",
    "\n",
    "for taxid in sorted_taxids: ## iterate over every taxon in treemap\n",
    "    taxon=branches[taxid] ## fetch branch\n",
    "\n",
    "    if taxid in redirect: ## taxid has been redirected because the official taxid is imprecise\n",
    "        lineage=ncbi.get_lineage(redirect[taxid])+[taxid]\n",
    "    elif taxid in high_order_lineage: ## taxid is being adjusted at a higher taxonomic level\n",
    "        lineage=high_order_lineage[taxid]\n",
    "    else:\n",
    "        lineage=ncbi.get_lineage(taxid) ## taxid is correct as assigned\n",
    "\n",
    "    for r,rank in enumerate(lineage): ## iterate over lineage recovered\n",
    "        if rank in high_order_lineage: ## lineage contains high-order readjustment\n",
    "            lineage=high_order_lineage[rank]+lineage[r+1:] ## insert new lineage\n",
    "\n",
    "    for rank in lineage: ## iterate over (potentially) new lineage\n",
    "        if rank in branches and rank in lineage_colours: ## lineage exists and has colour assignment\n",
    "            taxon['attrs']['colour']=lineage_colours[rank] ## assign colour\n",
    "\n",
    "    if len(lineage)==1 and taxon not in taxa_tree['children']: ## if root\n",
    "        taxa_tree['children'].append(taxon) ## add root to tree\n",
    "\n",
    "    for lin in lineage[::-1][1:]: ## iterate through lineage, starting from most recent, ignore first entry (self)\n",
    "        if lin in branches: ## rank present amongst branches\n",
    "            parent=branches[lin] ## grab parent\n",
    "\n",
    "            if 'children' not in parent: ## if parent doesn't have children yet - add the attribute\n",
    "                parent['children']=[]\n",
    "\n",
    "            if taxon not in parent['children']: ## branch wasn't assigned to its parent yet\n",
    "                parent['children'].append(taxon) ## add child to parent\n",
    "\n",
    "            assert taxon['taxid']!=parent['taxid'], 'parent is child %s %s, lineage: %s'%(taxon['taxid'],taxon['taxonomy'],lineage)\n",
    "            break ## break for loop, parent has been identified\n",
    "    \n",
    "json.dump(taxa_tree,open(os.path.join(base_path,'../treemap/skeeters.json'),'w'),indent=1,sort_keys=True) ## write json out to repo\n",
    "\n",
    "print('Done!')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
