---
title: "Bloodmeal Pathogens"
output: html_notebook
---

Can we find pathogens that are associated (and perhaps even carried by) bloodmeals?

```{r}
library(taxizedb)
library(tidyverse)
library(purrr)
library(ggplot2)
library(taxize)
library(usethis)
```

```{r}
setwd('~/src/skeeters') 
```



```{r}
report = read_csv('data/mosquito_reports.csv')
```

```{r}
contig_df = read_tsv('~/src/skeeters/data/contig_quality_df.tsv')
```

```{r}
contig_df = contig_df %>% mutate(bloodfed = str_detect(sample, 'CMS001'), water = str_detect(sample, 'ater'))
```

```{r}
bloodfed_df = contig_df %>% filter(kingdom == "Chordata" & !(Genus == "Homo" | Genus == "Mus")) %>% group_by(sample) %>% 
  filter(pmatch_gsnap > 0.9) %>% 
  summarize(total_length = sum(qlength), bloodfed = first(bloodfed), water = first(water), n_contigs = n())

bloodfed_df %>% ggplot(aes(x = total_length, color = bloodfed)) + geom_freqpoly() + scale_x_log10()

bloodfed_df %>% ggplot(aes(x = n_contigs, color = bloodfed)) + geom_freqpoly() + scale_x_log10()

bloodfed_df %>% ggplot(aes(x = n_contigs, y = total_length, color = bloodfed, shape = water)) + geom_point() + scale_x_log10() + scale_y_log10()
```

Let's go with n_contigs > 2.

```{r}
bloodfed_samples = bloodfed_df %>% filter(n_contigs > 2, sample != "CMS001_water3_Qiagen_S26") %>% pull(sample)
bloodfed_contigs = contig_df %>% filter(kingdom == "Chordata" & !(Genus == "Homo" | Genus == "Mus")) %>%
  filter(sample %in% bloodfed_samples) %>% 
  filter(pmatch_gsnap > 0.9) %>% group_by(sample, Division) %>%
  summarize(total_length = sum(qlength), n_contigs = n())
bloodfed_contigs
```

We see expected divisions, which we can dive into a bit more.

```{r}
division_totals = bloodfed_contigs %>% mutate(long_enough = total_length > 500) %>% group_by(Division) %>%
  summarize(n_division = sum(long_enough)) %>% filter(n_division > 0)
division_totals
```

Primates appear with short matches, plus a bunch in water3, which we will drop for further analyses.

```{r}
contig_df %>% filter(Division=='primates' & kingdom == "Chordata" & !(Genus == "Homo" | Genus == "Mus")) %>% filter(sample %in% bloodfed_samples) %>%
  filter(pmatch_gsnap > 0.9) %>% select(sample, pmatch_gsnap, qlength, CommonName)
```

Rodents: many short contigs to Norway rat, a praprie vole at lower pID, prarie deer mouse to 011. We may not need to go all the way up.

```{r}
contig_df %>% filter(Division=='rodents' & kingdom == "Chordata" & !(Genus == "Homo" | Genus == "Mus")) %>% filter(sample %in% bloodfed_samples) %>%
  filter(pmatch_gsnap > 0.9) %>% select(sample, pmatch_gsnap, qlength, CommonName)
```

Carnivores: racoon, : many short contigs to Norway rat, a prarie vole at lower pID, prarie deer mouse to 011.

004: raccoon (likely)
031: raccoon (maybe)
030: racoon
028: ?!
019: Felinae family (certainly)

```{r}
contig_df %>% filter(Division=='carnivores' & kingdom == "Chordata" & !(Genus == "Homo" | Genus == "Mus")) %>% filter(sample %in% bloodfed_samples) %>%
  filter(pmatch_gsnap > 0.9) %>% select(sample, pmatch_gsnap, qlength, CommonName)
```

Ungulates: 

Odocoileus is deer
Bos: cow/ox
Capra: also goat
Ovis: goat

There are some tricky ones. 060 has long contigs to Odocoileus virginianus (texanus) and to Bos mutus. These only unite at the Order level, which is high.

```{r}
contig_df %>% filter(Division=='even-toed ungulates' & kingdom == "Chordata" & !(Genus == "Homo" | Genus == "Mus")) %>% filter(sample %in% bloodfed_samples) %>%
  filter(pmatch_gsnap > 0.9) %>% select(sample, pmatch_gsnap, qlength, scientific_name) %>% filter(pmatch_gsnap > 0.98) %>% 
  group_by(sample, scientific_name) %>% summarize(len = sum(qlength))
```


# Association Testing

We want a permutation test for species A being enriched in samples with bloodmeal of division D.

This is a hypergeometric test. If there are m white balls (ungulates) and n black balls (non-ungulates) and we draw k samples (anaplasma) what are the chances we draw at least q white balls (ungulates with anaplasma)?

```{r}
phyper(14 - 1, 18, 42 - 18, 14, lower.tail = FALSE)
```

We will use reads, since they are more sensitive (if perhaps less specific).

```{r}
report %>% filter(name == 'Anaplasma', NT_r > 10) %>% select(sample_name, NT_rpm, NT_r) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(Division) %>% 
  filter(total_length > 500) %>% count()

contig_df %>% filter(Genus == "Anaplasma") %>% filter(pmatch_gsnap > 0.85) %>% group_by(sample) %>% 
  summarize(qlen = sum(qlength)) %>% select(sample, qlen) %>% left_join(bloodfed_contigs) %>% group_by(Division) %>% filter(total_length > 500) %>% count()

report %>% filter(name == 'Anaplasma', NT_r > 10) %>% select(sample_name, NT_rpm, NT_r, NT_percentidentity) %>% 
  mutate(sample = sample_name) %>% 
      full_join(contig_df %>% filter(str_detect(scientific_name, "naplasma")) %>% filter(pmatch_gsnap > 0.85) %>% group_by(sample) %>% 
        summarize(qlen = sum(qlength), pmatch = mean(pmatch_gsnap)) %>% select(sample, qlen, pmatch)) %>% 
  replace_na(list(qlen=0, NT_r=0)) %>%
  ggplot(aes(NT_r, qlen, color=NT_percentidentity)) + geom_point()
```

Okay, it appears that the contig_df is missing CMS001_012_Ra_S4 entirely.

```{r}
report %>% filter(name == 'Anaplasma', NT_r > 10) %>% select(sample_name, NT_rpm, NT_r, NT_percentidentity) %>% 
  mutate(sample = sample_name) %>% 
      full_join(contig_df %>% filter(str_detect(scientific_name, "naplasma")) %>% filter(pmatch_gsnap > 0.85) %>% group_by(sample) %>% 
        summarize(qlen = sum(qlength), pmatch = mean(pmatch_gsnap)) %>% select(sample, qlen, pmatch)) %>% 
  replace_na(list(qlen=0, NT_r=0)) %>% filter(qlen < 1)
```

CMS001_060_Ra_S12 appears to have had the contig kicked up to the class level, Alphaproteobacteria. When I run BLAST by hand, the top hits are all Anaplasma, with to ppercent id 99.49. The first hit in another species is much lower at 91%. This is apparently ribosomal RNA. 

```{r}
contig_df %>% filter(sample == "CMS001_060_Ra_S12") %>% filter(str_detect(qseqid, "NODE_2258_length_777_cov_1.060000") | str_detect(qseqid, "NODE_3274"))
```

```{r}
length(contig_df %>% pull(sample) %>% unique())
length(report %>% pull(sample_name) %>% unique())
setdiff(report %>% pull(sample_name) %>% unique(), contig_df %>% pull(sample) %>% unique())
```


What are the samples with many reads that don't assemble?
The %id in NT-read alignment is very high. If that's also on the whole query, this is trutworthy.
What is NT_alignmentlength in the report?

```{r}
report %>% filter(name == 'Anaplasma', NT_r > 150)
```


```{r}
genus_totals = report %>% filter(category_name == "Bacteria", NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name) %>% 
  mutate(sample = sample_name) %>% group_by(name) %>% summarize(n_genus = n())


report %>% filter(category_name == "Bacteria", NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(name, Division) %>% 
  filter(total_length > 500) %>% summarize(n_genus_and_division = n()) %>%
  left_join(division_totals) %>% left_join(genus_totals) %>% ungroup() %>%
  mutate(p = phyper(n_genus_and_division - 1, n_division, 42 - n_division, n_genus, lower.tail = FALSE)) %>%
  arrange(p)
```

It appears that the Anaplasma-ungulate relationship is way the strongest. In contrast, there are no significant results for viruses.

```{r}
genus_totals = report %>% filter(category_name == 'Viruses' & NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name) %>% 
  mutate(sample = sample_name) %>% group_by(name) %>% summarize(n_genus = n())

report %>% filter(category_name == 'Viruses' & NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name, category_name) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(name, Division) %>% 
  filter(total_length > 500) %>% summarize(n_genus_and_division = n()) %>%
  left_join(division_totals) %>% left_join(genus_totals) %>% ungroup() %>%
  mutate(p = phyper(n_genus_and_division - 1, n_division, 42 - n_division, n_genus, lower.tail = FALSE)) %>%
  arrange(p)
```

Eukaryota are harder b/c, well, birds match to birds. Need to rule out the chordates, but that's not in report data yet.

```{r}
genus_totals = report %>% filter(category_name == 'Eukaryota' & NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name) %>% 
  mutate(sample = sample_name) %>% group_by(name) %>% summarize(n_genus = n())

report %>% filter(category_name == 'Eukaryota' & NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name, category_name) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(name, Division) %>% 
  filter(total_length > 500) %>% summarize(n_genus_and_division = n(), mean_rpm = mean(NT_rpm)) %>%
  left_join(division_totals) %>% left_join(genus_totals) %>% ungroup() %>%
  mutate(p = phyper(n_genus_and_division - 1, n_division, 42 - n_division, n_genus, lower.tail = FALSE)) %>%
  arrange(p)
```


```{r}
genus_totals = report %>% filter(category_name == 'Eukaryota' & NT_r > 10, tax_level == 2, str_detect(name, "Trypano")) %>% 
  select(sample_name, NT_rpm, NT_r, name) %>% 
  mutate(sample = sample_name) %>% group_by(name) %>% summarize(n_genus = n())

report %>% filter(category_name == 'Eukaryota' & NT_r > 10, tax_level == 2, str_detect(name, "Trypano")) %>% 
  select(sample_name, NT_rpm, NT_r, name, category_name) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(name, Division) %>% 
  filter(total_length > 500) %>% summarize(n_genus_and_division = n(), mean_rpm = mean(NT_rpm)) %>%
  left_join(division_totals) %>% left_join(genus_totals) %>% ungroup() %>%
  mutate(p = phyper(n_genus_and_division - 1, n_division, 42 - n_division, n_genus, lower.tail = FALSE)) %>%
  arrange(p)
```

Loa loa.

Summary: anaplasma statistical, anectodes of mobuck, trypanosomes, malaria, loa loa.


```{r}
genus_totals = contig_df %>% filter(kingdom != "Chordata" & rank == 'genus') %>% filter(pmatch_gsnap > 0.85) %>% 
  group_by(scientific_name, sample) %>% summarize(total_length = sum(qlength)) %>% ungroup() %>%
  group_by(scientific_name) %>% summarize(n_genus = n())


contig_df %>% filter(kingdom != "Chordata" & rank == 'genus') %>% filter(pmatch_gsnap > 0.85) %>%
  
  TKTKTK: here we need to filter on things appearing at all in these bloodmeal samples so as to get adequate background.
  
  group_by(scientific_name) 

select(scientific_name, sample) %>% 
  left_join(bloodfed_contigs) 
bloodfed_contigs
bloodfed_contigs
%>% group_by(Genus, Division) %>% 
  filter(total_length > 500) %>% summarize(n_genus_and_division = n()) 

%>%
  left_join(division_totals) %>% left_join(genus_totals) %>% ungroup() %>%
  mutate(p = phyper(n_genus_and_division - 1, n_division, 42 - n_division, n_genus, lower.tail = FALSE)) %>%
  arrange(p)

report %>% filter(category_name == 'Eukaryota' & NT_r > 10, tax_level == 2) %>% 
  select(sample_name, NT_rpm, NT_r, name, category_name) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(name, Division) %>% 
  filter(total_length > 500) %>% summarize(n_genus_and_division = n(), mean_rpm = mean(NT_rpm)) %>%
  left_join(division_totals) %>% left_join(genus_totals) %>% ungroup() %>%
  mutate(p = phyper(n_genus_and_division - 1, n_division, 42 - n_division, n_genus, lower.tail = FALSE)) %>%
  arrange(p)
```


```{r}
report %>% pull(category_name) %>% unique()
```


```{r}
test <- function(overlap, n_division, n_genus){binom.test(x, n, p, alternative="two-sided")}
binom.test(14, 18, 14/n_bloodfed_samples, "greater")
```


```{r}
contig_df %>% filter(kingdom == "Bacteria") %>% mutate(bird = (sample %in% bird_samples)) %>%
  filter(pmatch_gsnap > 0.9) %>% summarize(total_length = sum(qlength), bloodfed = first(bloodfed), n_contigs = n())
```

```{r}
bacteria = contig_df %>% filter(kingdom == "Bacteria") %>% group_by(sample, Genus) %>% 
  filter(pmatch_gsnap > 0.85) %>% 
  summarize(total_length = sum(qlength), bloodfed = first(bloodfed), n_contigs = n()) %>%
  left_join(bloodfed_table, fill = 0)
bacteria %>% filter(Genus == 'Anaplasma')
names(bacteria) %<>% stringr::str_replace_all("\\s","_") %>% tolower
bacteria %>% filter(birds > 0) 
```

Plasmodium, Trypanosom

```{r}
report %>% filter(name == 'Plasmodium', NT_r > 10) %>% select(sample_name, NT_rpm, NT_r) %>% 
  mutate(sample = sample_name) %>% left_join(bloodfed_contigs) %>% group_by(Division) %>% count()
report %>% filter(name == 'Plasmodium', NT_r > 10) %>% select(sample_name, NT_rpm, NT_r) %>% 
  mutate(sample = sample_name) %>% count()

report %>% filter(str_detect(name, 'Plasmodium'), NT_r > 10) %>% select(sample_name, NT_r, name)
```

```{r}
n = length(bloodfed_samples)
binom.test(, n = n, p = 0.5, alternative = "greater")
```


```{r}
bloodfed_df %>% filter()
```



```{r}
contig_df %>% filter(kingdom == "Chordata") %>% group_by(sample, Division) %>% 
  filter(pmatch_gsnap > 0.9) %>% summarize(total_length = sum(qlength))
```

# Report formatting

Confirm that genus counts in report is rollup of species counts.

```{r}
species_rolled_up = report %>% filter(tax_level == 1) %>% group_by(genus_taxid) %>% summarize(n = sum(NT_r))
genus_rolled_up = report %>% filter(tax_level == 2) %>% group_by(genus_taxid) %>% summarize(n1 = sum(NT_r))

species_rolled_up %>% left_join(genus_rolled_up) %>% filter(n != n1)
```