---
title: "Generate figures"
author: "Lucy M. Li"
date: "12/17/2018"
output: 
  html_document:
    fig_caption: yes
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
library(parallel)
library(dplyr)
library(tidyr)
library(readr)
library(magrittr)
library(ggplot2)
library(grid)
library(gridExtra)
library(broom)
library(sp)
library(rgdal)
library(maptools)
library(OpenStreetMap)
library(ggmap)
library(DiagrammeR)
library(reutils)
library(XML)
# Attention!
# Google has recently changed its API requirements, and ggmap users are now required to provide an API key and enable billing. ggmap itself is outdated on CRAN; we hope to have the new version up on CRAN soon, but until then, here is the workaround:
# 
# if(!requireNamespace("devtools")) install.packages("devtools")
# devtools::install_github("dkahle/ggmap", ref = "tidyup")
# When you load ggmap, you can set your API key with register_google() (see ?register_google for details), but donâ€™t forget to enable the Maps Static API and geocoding API in the Google Cloud Console and enable billing. Also, go to 'Quotas' on Google Cloud and increase the 'Unsigned requests (if URL signing secret is defined) per day' limit
# See #51 on dkhale/ggmap github page for an extended discussion on details.
# register_google(key="AIzaSyAPP_DRq_N6XZyO7r2ad_-eQw36I_wFWio")
# Requires inputting billing information to increase limit to more than 1 request per day
# For now, use 'stamen' maps instead

```

```{r directories}
data_dir <- "../../../data"
```

```{r scripts}
source("generate_figures_functions.R")
```

# 1. Demographics

```{r load_data}
species_calls <- read.csv(file.path(data_dir, "sample_genus_and_species.csv"),
                          stringsAsFactors = FALSE) %>%
  rename(sample_name=X)
metadata <- read.csv(file.path(data_dir, "project-mosquito_sample-table_cleaned.csv"),
                     stringsAsFactors = FALSE) %>%
  left_join(., species_calls, "sample_name")
merged_metadata <- read.csv(file.path(data_dir, "CMS001_CMS002_MergedAnnotations_181218.csv"))
```

```{r get-shapefiles}
list_of_counties <- c("Alameda", "Placer", "San Diego", "Coachella Valley", "West Valley")
# https://data.ca.gov/dataset/ca-geographic-boundaries
shapefile_names <- c(paste0("CA_", c("State", "Counties", "Places")), "WestValley")
ca_shp <- file.path(data_dir, "shapefiles", shapefile_names) %>%
  lapply(., readOGR, stringsAsFactors=FALSE) %>%
  lapply(., spTransform, CRS("+proj=longlat +datum=WGS84")) %>%
  setNames(., gsub("CA_", "", shapefile_names))
coachellavalley_names <- ca_places_shp@data %>% 
  select(INTPTLON, INTPTLAT) %>%
  apply(1, as.numeric) %>% 
  apply(2, calc_eucl_dist, c(-116.0825, 33.51680)) %>%
  order(.) %>% 
  slice(ca_places_shp@data, .) %>%
  slice(., 1:(which(NAME=="Coachella")-1)) %>%
  `[[`("NAME")
ca_shp$Areas <- ca_shp$Counties %>%
  `[`(., .$NAME %in% c("Alameda", "Placer", "San Diego"), )
ca_shp$Areas@plotOrder <- 1:5
ca_shp$Areas@data %<>% rbind(., .[1:2, ])
places_data <- slice(ca_shp$Places@data, match(c("Ontario", "Coachella"), NAME)) %>%
  mutate(NAME=c("West Valley", "Coachella Valley"))
common_columns <- names(ca_shp$Areas@data) %>% `[`(., . %in% names(places_data))
ca_shp$Areas@data[4:5, common_columns] <- places_data[, common_columns]
ca_shp$Areas@data[4:5, c("COUNTYFP", "COUNTYNS")] <- places_data[, c("PLACEFP", "PLACENS")]
ca_shp$Areas@data %<>% 
  mutate(colours=c("#6b2500", "#de2d6a", "#565fd5", "#ff6f53", "#22b14e"),
         order=c(5, 2, 1, 3, 4))
gpclibPermit() # https://stackoverflow.com/questions/30790036/error-istruegpclibpermitstatus-is-not-true
ca_shp$Areas@polygons %<>% 
  c(list(unionSpatialPolygons(ca_shp$WestValley, IDs=rep(1, length(ca_shp$WestValley@polygons))),
         unionSpatialPolygons(ca_shp$Places[ca_shp$Places$NAME %in% coachellavalley_names, ], rep(1, length(coachellavalley_names)))) %>%
      lapply(., spTransform, CRS("+proj=longlat +datum=WGS84")) %>%
      lapply(., slot, "polygons") %>% 
      lapply(., `[[`, 1)
    )
for (i in seq(ca_shp$Areas@polygons)) {
  slot(slot(ca_shp$Areas, "polygons")[[i]], "ID") <- as.character(i)
}
ca_shp$Areas@data %<>% mutate(., id=sapply(ca_shp$Areas@polygons, function (x) x@ID))
zoom_options <- get_zoom_options (long=ca_shp$State@bbox[1, ], lat=ca_shp$State@bbox[2, ])
map <- openmap(c(zoom_options$lat_range[2], zoom_options$lon_range[1]),
               c(zoom_options$lat_range[1], zoom_options$lon_range[2]), 
               zoom = zoom_options$zoom, type = "osm", mergeTiles = TRUE) %>%
  openproj(CRS("+proj=longlat +datum=WGS84"))
base_cali_map <- autoplot(map) +
  geom_polygon(data=ca_shp$State, aes(x=long, y=lat, group=group), colour = "black", fill = NA) +
  geom_polygon(data=ca_shp$Areas, aes(x=long, y=lat, group=group, fill=factor(id, levels=with(ca_shp$Areas@data, id[order(order)]))), colour = "black", alpha=.75) +
  scale_fill_manual("Collection site", labels=with(ca_shp$Areas@data, NAME[order(order)]), values=with(ca_shp$Areas@data, colours[order(order)])) +
  theme(axis.title=element_blank(), axis.text=element_blank(), axis.ticks=element_blank(),
        legend.position="top", legend.direction="vertical",
        plot.margin = unit(c(1,1,1,1), "mm"))


cms_all_data <- filter(metadata, !is.na(lat), grepl("^CMS", sample_name))
cms_all_data[is.na(cms_all_data$genus), c("genus", "species", "corrected.genus", "corrected.species")] <- 
  data.frame(genus="Culex", species=c("erythrothorax", "tarsalis"),
             corrected.genus="Culex", corrected.species=c("erythrothorax", "tarsalis"),
             stringsAsFactors = FALSE)
cms_all_data$location_name <- apply(cms_all_data[, c("long", "lat")], 1, function (x) {
  lapply(ca_shp$Areas@data[, c("INTPTLON", "INTPTLAT")], as.numeric) %>% 
    data.frame() %>%
    apply(., 1, calc_eucl_dist, x) %>%
    which.min() %>%
    `[`(ca_shp$Areas@data$NAME, .)
})

polygon_plots <- lapply(with(ca_shp$Areas@data, id[order(order)]), function (x) {
  plot_data <- filter(tidy(ca_shp$Areas), id==x)
  width_height <- plot_data[, c("long", "lat")]%>%lapply(range)%>%sapply(diff)
  place_name <- filter(ca_shp$Areas@data, id==x)$NAME
  sites <- filter(cms_all_data, location_name==place_name) %>% mutate(lat_lon=paste0(lat, "_", long)) %>% `[[`("lat_lon") %>% table()
  n <- sum(sites)
  create_map(filter(cms_all_data, location_name==place_name), polygon=plot_data, polygon_col=filter(ca_shp$Areas@data, id==x)$colours) +
    ggtitle(paste0(filter(ca_shp$Areas@data, id==x)$NAME, "\nNumber of sites: ", length(sites), "     n: ", n)) +
    theme(plot.title=element_text(hjust=0.5))
})
grouped_polygons <- split(polygon_plots, c(1, 1, 2, 2, 2)) %>%
  mapply(function (x, y) {
    arrangeGrob(grobs=x, nrow=1)
  }, ., split(seq(polygons), c(1, 1, 2, 2, 2)), SIMPLIFY=FALSE) %>%
  arrangeGrob(grobs=., ncol=1)

full_demo_map <- arrangeGrob(grobs=list(base_cali_map, grouped_polygons), ncol=2, widths=c(3, 5))
ggsave("01_demographics/demographic_map.pdf", full_demo_map, width=18, height=8)
```


```{r demographics-tables}
species_by_location <- cms_all_data %>%
  group_by(., location_name, corrected.genus, corrected.species) %>%
  summarize(., n=n())
species_by_location_plot <- 
  ggplot(species_by_location, 
       aes(x=paste0(corrected.genus, "\n", corrected.species), 
           y=factor(location_name, levels=with(ca_shp$Areas@data, NAME[order(order, decreasing = TRUE)])), 
           fill=n)) +
  theme_light() +
  geom_tile() +
  facet_grid(.~corrected.genus, scales="free_x") +
  xlab("Species") +
  ylab("Location") +
  scale_fill_continuous("Number of\nSamples")
ggsave("01_demographics/species_by_location.pdf", species_by_location_plot, width=12, height=4)

species_by_location %>%
  mutate(species_name=paste(corrected.genus, corrected.species)) %>%
  spread(., key=species_name, value=n, fill=0) %>%
  `[`(., , -2:-3) %>%
  write_csv(., "01_demographics/demographics.csv")
```

# 2. mNGS analysis

```{r mNGS}
# mNGS analysis pipeline
# Placeholder diagram for now
DiagrammeR::mermaid("
graph TD
  id1[raw reads]--filtering--> reads 
  reads --assembly --> id2[contigs] 
", height=200)
```


# 3. Microbiota overview

```{r}
contig_hits <- read.csv(paste0(data_dir, "/blast_topresults_df_sub.csv"))
acc_to_lineage <- rjson::fromJSON(file=paste0(data_dir, "/acc_to_lineage.json"))
```

```{r contig_hits_summary}
contig_hits$lineage <- acc_to_lineage[contig_hits$sseqid]
taxcat <- list(bacteria=2,
               viruses=10239,
               fungi=4751,
               plants=33090,
               chordata=7711)
kingdom_summary <- contig_hits %>%
  split(., .$sample) %>%
  lapply(., function (x) {
    lapply(taxcat, function (id) {
      filtered_x <- filter(x, mapply(`%in%`, id, lineage))
      if (nrow(filtered_x)==0) return (NA)
      filtered_x$lineage %>% sapply(tail, 1) %>% table() %>% sort()
    })
  })
kingdom_numbers <- lapply(kingdom_summary, function (x) {
  lapply(x, is.na) %>% lapply(`!`) %>% sapply(sum) %>% t() %>% as.data.frame()
}) %>%
  do.call(what=rbind, .) %>%
  mutate(., sample=rownames(.)) %>%
  mutate(sample_name=slice(merged_metadata, match(kingdom_numbers$sample, NewIDseqName))$OldIDseqName) %>%
  left_join(., cms_all_data, by="sample_name") %>%
  select(-sample_name) %>%
  reshape2::melt(measure.vars=names(taxcat), variable.name="groups", value.name="nspecies")
```

```{r kingdom_numbers_plot}
kingdom_numbers_plot <- ggplot(kingdom_numbers %>% filter(!is.na(location_name))) + theme_bw() +
  geom_boxplot(aes(x=location_name, y=nspecies)) + 
  facet_wrap(~groups, ncol=1, scales="free_y")
ggsave("03_microbiota_overview/number_species_by_location.pdf", kingdom_numbers_plot, width=4.5, height=7)
```

# 4. Bloodmeal

```{r}
bloodmeal_summary <- lapply(kingdom_summary, `[[`, "chordata") %>%
  lapply(names) %>%
  lapply(esummary, db="taxonomy") %>%
  lapply(content, "parsed")
bloodmeal_summary_df <- bloodmeal_summary %>%
  `[`(!sapply(., is.null)) %>%
  `[`(sapply(., length)>0) %>%
  mapply(data.frame, sample=names(.), ., SIMPLIFY=FALSE) %>%
  do.call(what=rbind, .) %>%
  mutate(., species_name=paste(Genus, Species))
bloodmeal_species <- bloodmeal_summary_df %>%
  filter(., !duplicated(species_name)) %>%
  select(Division, CommonName, species_name, Genus, Species) %>%
  mutate(CommonName=gsub("NA", NA, CommonName)) %>%
  mutate(CommonName=coalesce(CommonName, species_name))
bloodmeal_species_melt <- bloodmeal_summary_df %>%
  split(., .$sample) %>%
  lapply(., function (x) {
    bloodmeal_species$species_name %in% x$species_name
  }) %>%
  do.call(what=cbind, .) %>%
  data.frame(bloodmeal_species, .) %>% 
  reshape2::melt(., id.vars=names(bloodmeal_species), variable.name="sample", value.name="present") %>%
  left_join(., select(kingdom_numbers, sample, location_name), "sample")
```

```{r}
bloodmeal_species_plot <- ggplot(bloodmeal_species_melt) + theme_bw() +
  geom_tile(aes(x=gsub(" ", "\n", CommonName), y=sample, fill=present)) +
  facet_grid(location_name~gsub(" ", "\n", Division), scales="free", space="free") +
  xlab("Species") + ylab("Sample") +
  scale_fill_manual("Present", values=c("white", "steelblue4")) +
  theme(strip.text.y=element_text(angle=0),
        axis.text.x=element_text(angle=0))
ggsave("04_bloodmeal/bloodmeal_species.pdf", bloodmeal_species_plot,
       width=24, height=16)
```

# 5. Vignettes



