{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#This is a notebook AK used to document how demographic data was pulled from from CMS001_CMS002_MergedAnnotations_190325.csv file to generate Table S1 for CMS manuscript"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Set up python tools for pandas processing, read in data file, and check with head."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "AllMosData = pd.read_csv('CMS001_CMS002_MergedAnnotations_190325.csv', delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>OldIDseqName</th>\n",
       "      <th>czbiohub-mosquito_sequences_id</th>\n",
       "      <th>raw_sequence_run_directory</th>\n",
       "      <th>host</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>nucleotide_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>...</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>sample_unit</th>\n",
       "      <th>known_organism</th>\n",
       "      <th>detection_method</th>\n",
       "      <th>library_prep</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>rna_dna_input_ng</th>\n",
       "      <th>library_prep_batch</th>\n",
       "      <th>extraction_batch</th>\n",
       "      <th>Notes:</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>CMS_001_RNA_A_S1</td>\n",
       "      <td>CMS_001_RNA_A_S1</td>\n",
       "      <td>czbiohub-seqbot/fastqs/180323_NB501961_0089_AH...</td>\n",
       "      <td>mosquito</td>\n",
       "      <td>whole</td>\n",
       "      <td>RNA</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>...</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEBNext Ultra II Directional RNA Library Prep Kit</td>\n",
       "      <td>NextSeq</td>\n",
       "      <td>27.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>CMS_002_RNA_A_S1</td>\n",
       "      <td>CMS_002_RNA_A_S1</td>\n",
       "      <td>czbiohub-seqbot/fastqs/180412_NB501961_0097_AH...</td>\n",
       "      <td>mosquito</td>\n",
       "      <td>whole</td>\n",
       "      <td>RNA</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>...</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEBNext Ultra II Directional RNA Library Prep Kit</td>\n",
       "      <td>NextSeq</td>\n",
       "      <td>59.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>CMS_003_RNA_A_S2</td>\n",
       "      <td>CMS_003_RNA_A_S2</td>\n",
       "      <td>czbiohub-seqbot/fastqs/180323_NB501961_0089_AH...</td>\n",
       "      <td>mosquito</td>\n",
       "      <td>whole</td>\n",
       "      <td>RNA</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>...</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEBNext Ultra II Directional RNA Library Prep Kit</td>\n",
       "      <td>NextSeq</td>\n",
       "      <td>88.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>CMS_004_RNA_A_S2</td>\n",
       "      <td>CMS_004_RNA_A_S2</td>\n",
       "      <td>czbiohub-seqbot/fastqs/180412_NB501961_0097_AH...</td>\n",
       "      <td>mosquito</td>\n",
       "      <td>whole</td>\n",
       "      <td>RNA</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>...</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEBNext Ultra II Directional RNA Library Prep Kit</td>\n",
       "      <td>NextSeq</td>\n",
       "      <td>88.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>CMS_005_RNA_A_S3</td>\n",
       "      <td>CMS_005_RNA_A_S3</td>\n",
       "      <td>czbiohub-seqbot/fastqs/180323_NB501961_0089_AH...</td>\n",
       "      <td>mosquito</td>\n",
       "      <td>whole</td>\n",
       "      <td>RNA</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>...</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEBNext Ultra II Directional RNA Library Prep Kit</td>\n",
       "      <td>NextSeq</td>\n",
       "      <td>240.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       NewIDseqName      OldIDseqName czbiohub-mosquito_sequences_id  \\\n",
       "0  CMS001_001_Ra_S1  CMS_001_RNA_A_S1               CMS_001_RNA_A_S1   \n",
       "1  CMS001_002_Ra_S1  CMS_002_RNA_A_S1               CMS_002_RNA_A_S1   \n",
       "2  CMS001_003_Ra_S2  CMS_003_RNA_A_S2               CMS_003_RNA_A_S2   \n",
       "3  CMS001_004_Ra_S2  CMS_004_RNA_A_S2               CMS_004_RNA_A_S2   \n",
       "4  CMS001_005_Ra_S3  CMS_005_RNA_A_S3               CMS_005_RNA_A_S3   \n",
       "\n",
       "                          raw_sequence_run_directory      host sample_type  \\\n",
       "0  czbiohub-seqbot/fastqs/180323_NB501961_0089_AH...  mosquito       whole   \n",
       "1  czbiohub-seqbot/fastqs/180412_NB501961_0097_AH...  mosquito       whole   \n",
       "2  czbiohub-seqbot/fastqs/180323_NB501961_0089_AH...  mosquito       whole   \n",
       "3  czbiohub-seqbot/fastqs/180412_NB501961_0097_AH...  mosquito       whole   \n",
       "4  czbiohub-seqbot/fastqs/180323_NB501961_0089_AH...  mosquito       whole   \n",
       "\n",
       "  nucleotide_type       collection_date  collection_lat  collection_long  \\\n",
       "0             RNA  2017-10-03T00:00:00Z        37.55697       -122.07938   \n",
       "1             RNA  2017-09-21T00:00:00Z        37.55697       -122.07938   \n",
       "2             RNA  2017-09-19T00:00:00Z        37.55697       -122.07938   \n",
       "3             RNA  2017-09-12T00:00:00Z        37.71520       -122.19431   \n",
       "4             RNA  2017-09-29T00:00:00Z        37.78886       -122.29903   \n",
       "\n",
       "   ...    blood_fed sample_unit known_organism detection_method  \\\n",
       "0  ...    blood fed           1            NaN              NaN   \n",
       "1  ...    blood fed           1            NaN              NaN   \n",
       "2  ...    blood fed           1            NaN              NaN   \n",
       "3  ...    blood fed           1            NaN              NaN   \n",
       "4  ...    blood fed           1            NaN              NaN   \n",
       "\n",
       "                                        library_prep  sequencer  \\\n",
       "0  NEBNext Ultra II Directional RNA Library Prep Kit   NextSeq    \n",
       "1  NEBNext Ultra II Directional RNA Library Prep Kit   NextSeq    \n",
       "2  NEBNext Ultra II Directional RNA Library Prep Kit   NextSeq    \n",
       "3  NEBNext Ultra II Directional RNA Library Prep Kit   NextSeq    \n",
       "4  NEBNext Ultra II Directional RNA Library Prep Kit   NextSeq    \n",
       "\n",
       "  rna_dna_input_ng library_prep_batch extraction_batch Notes:  \n",
       "0             27.3                NaN              NaN    NaN  \n",
       "1             59.8                NaN              NaN    NaN  \n",
       "2             88.9                NaN              NaN    NaN  \n",
       "3             88.4                NaN              NaN    NaN  \n",
       "4            240.7                NaN              NaN    NaN  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllMosData.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Print column names to more easily view all column headers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NewIDseqName' 'OldIDseqName' 'czbiohub-mosquito_sequences_id'\n",
      " 'raw_sequence_run_directory' 'host' 'sample_type' 'nucleotide_type'\n",
      " 'collection_date' 'collection_lat' 'collection_long' 'Habitat'\n",
      " 'Population' 'collected_by' 'preservation_method' 'sex' 'life_stage'\n",
      " 'visual_genus' 'visual_species' 'compute_genus' 'compute_species'\n",
      " 'blood_fed' 'sample_unit' 'known_organism' 'detection_method'\n",
      " 'library_prep' 'sequencer' 'rna_dna_input_ng' 'library_prep_batch'\n",
      " 'extraction_batch' 'Notes:']\n"
     ]
    }
   ],
   "source": [
    "print(AllMosData.columns.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Select subset of data relevant for Table S1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>visual_genus</th>\n",
       "      <th>visual_species</th>\n",
       "      <th>compute_genus</th>\n",
       "      <th>compute_species</th>\n",
       "      <th>sex</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.715200</td>\n",
       "      <td>-122.194310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.788860</td>\n",
       "      <td>-122.299030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CMS001_006_Ra_S5</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CMS001_007_Ra_S12</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CMS001_008_Ra_S3</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CMS001_010_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CMS001_011_Ra_S4</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>dorsalis</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>dorsalis</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CMS001_013_Ra_S5</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CMS001_014_Ra_S5</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CMS001_015_Ra_S13</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>inornata</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>inornata</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-27T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CMS001_016_Ra_S6</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CMS001_017_Ra_S6</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CMS001_018_Ra_S14</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-27T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CMS001_019_Ra_S14</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-04T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CMS001_020_Ra_S15</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CMS001_021_Ra_S16</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CMS001_022_Ra_S6</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-27T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CMS001_023_Ra_S17</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CMS001_024_Ra_S15</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-10-06T00:00:00Z</td>\n",
       "      <td>37.581910</td>\n",
       "      <td>-122.048380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CMS001_025_Ra_S7</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.556970</td>\n",
       "      <td>-122.079380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CMS001_026_Ra_S18</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.509530</td>\n",
       "      <td>-121.931130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CMS001_027_Ra_S16</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.866500</td>\n",
       "      <td>-122.310510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CMS001_028_Ra_S17</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.715200</td>\n",
       "      <td>-122.194310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CMS001_029_Ra_S18</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-13T00:00:00Z</td>\n",
       "      <td>37.684730</td>\n",
       "      <td>-121.913410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CMS001_030_Ra_S7</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>2017-09-10T00:00:00Z</td>\n",
       "      <td>37.715200</td>\n",
       "      <td>-122.194310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>CMS002_045e_Rb_S188_L004</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-10-24T00:00:00Z</td>\n",
       "      <td>34.151881</td>\n",
       "      <td>-117.662950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>CMS002_045f_Rb_S189_L004</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-10-24T00:00:00Z</td>\n",
       "      <td>34.151881</td>\n",
       "      <td>-117.662950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>CMS002_045g_Rb_S190_L004</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-10-24T00:00:00Z</td>\n",
       "      <td>34.151881</td>\n",
       "      <td>-117.662950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>CMS002_046a_Rb_S191_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.972500</td>\n",
       "      <td>-117.714444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>CMS002_046b_Rb_S192_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.972500</td>\n",
       "      <td>-117.714444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>CMS002_047a_Rb_S193_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>CMS002_047b_Rb_S194_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>CMS002_047c_Rb_S195_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>CMS002_047d_Rb_S196_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>CMS002_047e_Rb_S197_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CMS002_047f_Rb_S198_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>CMS002_047g_Rb_S200_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>CMS002_047h_Rb_S1_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>CMS002_047i_Rb_S2_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>CMS002_047j_Rb_S3_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-29T00:00:00Z</td>\n",
       "      <td>33.937780</td>\n",
       "      <td>-117.633056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>CMS002_049a_Rb_S4_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-21T00:00:00Z</td>\n",
       "      <td>38.739939</td>\n",
       "      <td>-121.305636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>CMS002_050a_Rb_S5_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-21T00:00:00Z</td>\n",
       "      <td>38.824306</td>\n",
       "      <td>-121.257067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>CMS002_051a_Rb_S6_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-21T00:00:00Z</td>\n",
       "      <td>38.897572</td>\n",
       "      <td>-121.311747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>CMS002_053a_Rb_S7_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-21T00:00:00Z</td>\n",
       "      <td>38.836781</td>\n",
       "      <td>-121.391190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>CMS002_054a_Rb_S8_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-21T00:00:00Z</td>\n",
       "      <td>38.772222</td>\n",
       "      <td>-121.469500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>CMS002_056a_Rb_S9_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-28T00:00:00Z</td>\n",
       "      <td>38.787211</td>\n",
       "      <td>-121.372895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>CMS002_057a_Rb_S10_L004</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unfed</td>\n",
       "      <td>2017-11-28T00:00:00Z</td>\n",
       "      <td>38.787211</td>\n",
       "      <td>-121.372895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>CMS002_Water1_Rb_S127_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>CMS002_Water2_Rb_S139_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>CMS002_Water3_Rb_S151_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>CMS002_Water4_Rb_S163_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>CMS002_Water5_Rb_S175_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>CMS002_Water6_Rb_S187_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>CMS002_Water7_Rb_S199_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>CMS002_Water8_Rb_S11_L004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   NewIDseqName visual_genus    visual_species compute_genus  \\\n",
       "0              CMS001_001_Ra_S1        Culex     erythrothorax         Culex   \n",
       "1              CMS001_002_Ra_S1     Culiseta         particeps      Culiseta   \n",
       "2              CMS001_003_Ra_S2        Culex           pipiens         Culex   \n",
       "3              CMS001_004_Ra_S2     Culiseta          incidens      Culiseta   \n",
       "4              CMS001_005_Ra_S3        Culex     erythrothorax         Culex   \n",
       "5              CMS001_006_Ra_S5        Culex     erythrothorax         Culex   \n",
       "6             CMS001_007_Ra_S12     Culiseta         particeps      Culiseta   \n",
       "7              CMS001_008_Ra_S3     Culiseta         particeps      Culiseta   \n",
       "8             CMS001_009_Ra_S13        Culex     erythrothorax         Culex   \n",
       "9              CMS001_010_Ra_S1        Culex     erythrothorax         Culex   \n",
       "10             CMS001_011_Ra_S4        Aedes          dorsalis         Aedes   \n",
       "11             CMS001_012_Ra_S4        Culex     erythrothorax         Culex   \n",
       "12             CMS001_013_Ra_S5     Culiseta          incidens      Culiseta   \n",
       "13             CMS001_014_Ra_S5        Culex     erythrothorax         Culex   \n",
       "14            CMS001_015_Ra_S13     Culiseta          inornata      Culiseta   \n",
       "15             CMS001_016_Ra_S6        Culex     erythrothorax         Culex   \n",
       "16             CMS001_017_Ra_S6        Culex     erythrothorax         Culex   \n",
       "17            CMS001_018_Ra_S14        Culex     erythrothorax         Culex   \n",
       "18            CMS001_019_Ra_S14        Culex     erythrothorax         Culex   \n",
       "19            CMS001_020_Ra_S15        Culex     erythrothorax         Culex   \n",
       "20            CMS001_021_Ra_S16        Culex     erythrothorax         Culex   \n",
       "21             CMS001_022_Ra_S6        Culex     erythrothorax         Culex   \n",
       "22            CMS001_023_Ra_S17        Culex     erythrothorax         Culex   \n",
       "23            CMS001_024_Ra_S15        Culex          tarsalis      Culiseta   \n",
       "24             CMS001_025_Ra_S7        Culex          tarsalis      Culiseta   \n",
       "25            CMS001_026_Ra_S18        Culex          tarsalis         Culex   \n",
       "26            CMS001_027_Ra_S16        Culex          tarsalis         Culex   \n",
       "27            CMS001_028_Ra_S17        Culex           pipiens         Culex   \n",
       "28            CMS001_029_Ra_S18        Culex           pipiens         Culex   \n",
       "29             CMS001_030_Ra_S7        Culex          tarsalis         Culex   \n",
       "..                          ...          ...               ...           ...   \n",
       "131    CMS002_045e_Rb_S188_L004        Aedes        albopictus         Aedes   \n",
       "132    CMS002_045f_Rb_S189_L004        Aedes        albopictus         Aedes   \n",
       "133    CMS002_045g_Rb_S190_L004        Aedes        albopictus         Aedes   \n",
       "134    CMS002_046a_Rb_S191_L004        Culex  quinquefasciatus         Culex   \n",
       "135    CMS002_046b_Rb_S192_L004        Culex  quinquefasciatus         Culex   \n",
       "136    CMS002_047a_Rb_S193_L004        Culex     erythrothorax         Culex   \n",
       "137    CMS002_047b_Rb_S194_L004        Culex     erythrothorax         Culex   \n",
       "138    CMS002_047c_Rb_S195_L004        Culex     erythrothorax         Culex   \n",
       "139    CMS002_047d_Rb_S196_L004        Culex     erythrothorax         Culex   \n",
       "140    CMS002_047e_Rb_S197_L004        Culex     erythrothorax         Culex   \n",
       "141    CMS002_047f_Rb_S198_L004        Culex     erythrothorax         Culex   \n",
       "142    CMS002_047g_Rb_S200_L004        Culex     erythrothorax         Culex   \n",
       "143      CMS002_047h_Rb_S1_L004        Culex     erythrothorax         Culex   \n",
       "144      CMS002_047i_Rb_S2_L004        Culex     erythrothorax         Culex   \n",
       "145      CMS002_047j_Rb_S3_L004        Culex     erythrothorax         Culex   \n",
       "146      CMS002_049a_Rb_S4_L004        Culex           pipiens         Culex   \n",
       "147      CMS002_050a_Rb_S5_L004        Culex           pipiens         Culex   \n",
       "148      CMS002_051a_Rb_S6_L004        Culex           pipiens         Culex   \n",
       "149      CMS002_053a_Rb_S7_L004        Culex           pipiens         Culex   \n",
       "150      CMS002_054a_Rb_S8_L004        Culex           pipiens         Culex   \n",
       "151      CMS002_056a_Rb_S9_L004        Culex          tarsalis         Culex   \n",
       "152     CMS002_057a_Rb_S10_L004        Culex           pipiens         Culex   \n",
       "153  CMS002_Water1_Rb_S127_L004          NaN               NaN           NaN   \n",
       "154  CMS002_Water2_Rb_S139_L004          NaN               NaN           NaN   \n",
       "155  CMS002_Water3_Rb_S151_L004          NaN               NaN           NaN   \n",
       "156  CMS002_Water4_Rb_S163_L004          NaN               NaN           NaN   \n",
       "157  CMS002_Water5_Rb_S175_L004          NaN               NaN           NaN   \n",
       "158  CMS002_Water6_Rb_S187_L004          NaN               NaN           NaN   \n",
       "159  CMS002_Water7_Rb_S199_L004          NaN               NaN           NaN   \n",
       "160   CMS002_Water8_Rb_S11_L004          NaN               NaN           NaN   \n",
       "\n",
       "      compute_species     sex  blood_fed       collection_date  \\\n",
       "0       erythrothorax  female  blood fed  2017-10-03T00:00:00Z   \n",
       "1           particeps  female  blood fed  2017-09-21T00:00:00Z   \n",
       "2             pipiens  female  blood fed  2017-09-19T00:00:00Z   \n",
       "3            incidens  female  blood fed  2017-09-12T00:00:00Z   \n",
       "4       erythrothorax  female  blood fed  2017-09-29T00:00:00Z   \n",
       "5       erythrothorax  female  blood fed  2017-09-19T00:00:00Z   \n",
       "6           particeps  female  blood fed  2017-09-19T00:00:00Z   \n",
       "7           particeps  female  blood fed  2017-09-19T00:00:00Z   \n",
       "8       erythrothorax  female  blood fed  2017-09-19T00:00:00Z   \n",
       "9       erythrothorax  female  blood fed  2017-09-19T00:00:00Z   \n",
       "10           dorsalis  female  blood fed  2017-09-19T00:00:00Z   \n",
       "11      erythrothorax  female  blood fed  2017-09-19T00:00:00Z   \n",
       "12           incidens  female  blood fed  2017-09-19T00:00:00Z   \n",
       "13      erythrothorax  female  blood fed  2017-10-06T00:00:00Z   \n",
       "14           inornata  female  blood fed  2017-09-27T00:00:00Z   \n",
       "15      erythrothorax  female  blood fed  2017-10-06T00:00:00Z   \n",
       "16      erythrothorax  female  blood fed  2017-10-06T00:00:00Z   \n",
       "17      erythrothorax  female  blood fed  2017-09-27T00:00:00Z   \n",
       "18      erythrothorax  female  blood fed  2017-10-04T00:00:00Z   \n",
       "19      erythrothorax  female  blood fed  2017-10-06T00:00:00Z   \n",
       "20      erythrothorax  female  blood fed  2017-10-06T00:00:00Z   \n",
       "21      erythrothorax  female  blood fed  2017-09-27T00:00:00Z   \n",
       "22      erythrothorax  female  blood fed  2017-10-06T00:00:00Z   \n",
       "23          particeps  female  blood fed  2017-10-06T00:00:00Z   \n",
       "24           incidens  female  blood fed  2017-09-21T00:00:00Z   \n",
       "25           tarsalis  female  blood fed  2017-09-19T00:00:00Z   \n",
       "26           tarsalis  female  blood fed  2017-09-12T00:00:00Z   \n",
       "27            pipiens  female  blood fed  2017-09-12T00:00:00Z   \n",
       "28            pipiens  female  blood fed  2017-09-13T00:00:00Z   \n",
       "29           tarsalis  female  blood fed  2017-09-10T00:00:00Z   \n",
       "..                ...     ...        ...                   ...   \n",
       "131        albopictus     NaN      unfed  2017-10-24T00:00:00Z   \n",
       "132        albopictus     NaN      unfed  2017-10-24T00:00:00Z   \n",
       "133        albopictus     NaN      unfed  2017-10-24T00:00:00Z   \n",
       "134  quinquefasciatus     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "135  quinquefasciatus     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "136     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "137     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "138     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "139     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "140     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "141     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "142     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "143     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "144     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "145     erythrothorax     NaN      unfed  2017-11-29T00:00:00Z   \n",
       "146           pipiens     NaN      unfed  2017-11-21T00:00:00Z   \n",
       "147           pipiens     NaN      unfed  2017-11-21T00:00:00Z   \n",
       "148           pipiens     NaN      unfed  2017-11-21T00:00:00Z   \n",
       "149           pipiens     NaN      unfed  2017-11-21T00:00:00Z   \n",
       "150           pipiens     NaN      unfed  2017-11-21T00:00:00Z   \n",
       "151          tarsalis     NaN      unfed  2017-11-28T00:00:00Z   \n",
       "152           pipiens     NaN      unfed  2017-11-28T00:00:00Z   \n",
       "153               NaN     NaN        NaN                   NaN   \n",
       "154               NaN     NaN        NaN                   NaN   \n",
       "155               NaN     NaN        NaN                   NaN   \n",
       "156               NaN     NaN        NaN                   NaN   \n",
       "157               NaN     NaN        NaN                   NaN   \n",
       "158               NaN     NaN        NaN                   NaN   \n",
       "159               NaN     NaN        NaN                   NaN   \n",
       "160               NaN     NaN        NaN                   NaN   \n",
       "\n",
       "     collection_lat  collection_long  \n",
       "0         37.556970      -122.079380  \n",
       "1         37.556970      -122.079380  \n",
       "2         37.556970      -122.079380  \n",
       "3         37.715200      -122.194310  \n",
       "4         37.788860      -122.299030  \n",
       "5         37.556970      -122.079380  \n",
       "6         37.556970      -122.079380  \n",
       "7         37.556970      -122.079380  \n",
       "8         37.556970      -122.079380  \n",
       "9         37.556970      -122.079380  \n",
       "10        37.556970      -122.079380  \n",
       "11        37.556970      -122.079380  \n",
       "12        37.556970      -122.079380  \n",
       "13        37.581910      -122.048380  \n",
       "14        37.556970      -122.079380  \n",
       "15        37.581910      -122.048460  \n",
       "16        37.581910      -122.048380  \n",
       "17        37.556970      -122.079380  \n",
       "18        37.556970      -122.079380  \n",
       "19        37.581910      -122.048460  \n",
       "20        37.581910      -122.048380  \n",
       "21        37.556970      -122.079380  \n",
       "22        37.581910      -122.048380  \n",
       "23        37.581910      -122.048380  \n",
       "24        37.556970      -122.079380  \n",
       "25        37.509530      -121.931130  \n",
       "26        37.866500      -122.310510  \n",
       "27        37.715200      -122.194310  \n",
       "28        37.684730      -121.913410  \n",
       "29        37.715200      -122.194310  \n",
       "..              ...              ...  \n",
       "131       34.151881      -117.662950  \n",
       "132       34.151881      -117.662950  \n",
       "133       34.151881      -117.662950  \n",
       "134       33.972500      -117.714444  \n",
       "135       33.972500      -117.714444  \n",
       "136       33.937780      -117.633056  \n",
       "137       33.937780      -117.633056  \n",
       "138       33.937780      -117.633056  \n",
       "139       33.937780      -117.633056  \n",
       "140       33.937780      -117.633056  \n",
       "141       33.937780      -117.633056  \n",
       "142       33.937780      -117.633056  \n",
       "143       33.937780      -117.633056  \n",
       "144       33.937780      -117.633056  \n",
       "145       33.937780      -117.633056  \n",
       "146       38.739939      -121.305636  \n",
       "147       38.824306      -121.257067  \n",
       "148       38.897572      -121.311747  \n",
       "149       38.836781      -121.391190  \n",
       "150       38.772222      -121.469500  \n",
       "151       38.787211      -121.372895  \n",
       "152       38.787211      -121.372895  \n",
       "153             NaN              NaN  \n",
       "154             NaN              NaN  \n",
       "155             NaN              NaN  \n",
       "156             NaN              NaN  \n",
       "157             NaN              NaN  \n",
       "158             NaN              NaN  \n",
       "159             NaN              NaN  \n",
       "160             NaN              NaN  \n",
       "\n",
       "[161 rows x 10 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllMosData[['NewIDseqName','visual_genus','visual_species','compute_genus','compute_species','sex','blood_fed','collection_date','collection_lat','collection_long']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Surprising number of NaN for sex field. Use count to investigate. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>OldIDseqName</th>\n",
       "      <th>czbiohub-mosquito_sequences_id</th>\n",
       "      <th>raw_sequence_run_directory</th>\n",
       "      <th>host</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>nucleotide_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_unit</th>\n",
       "      <th>known_organism</th>\n",
       "      <th>detection_method</th>\n",
       "      <th>library_prep</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>rna_dna_input_ng</th>\n",
       "      <th>library_prep_batch</th>\n",
       "      <th>extraction_batch</th>\n",
       "      <th>Notes:</th>\n",
       "      <th>visual_equals_compute</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sex</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>female</th>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>22</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>male</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        NewIDseqName  OldIDseqName  czbiohub-mosquito_sequences_id  \\\n",
       "sex                                                                  \n",
       "female            91            91                              91   \n",
       "male               4             4                               4   \n",
       "\n",
       "        raw_sequence_run_directory  host  sample_type  nucleotide_type  \\\n",
       "sex                                                                      \n",
       "female                          91    91           91               91   \n",
       "male                             4     4            4                4   \n",
       "\n",
       "        collection_date  collection_lat  collection_long  \\\n",
       "sex                                                        \n",
       "female               89              89               89   \n",
       "male                  4               4                4   \n",
       "\n",
       "                ...            sample_unit  known_organism  detection_method  \\\n",
       "sex             ...                                                            \n",
       "female          ...                     91               0                 0   \n",
       "male            ...                      4               0                 0   \n",
       "\n",
       "        library_prep  sequencer  rna_dna_input_ng  library_prep_batch  \\\n",
       "sex                                                                     \n",
       "female            91         91                91                   0   \n",
       "male               4          4                 3                   0   \n",
       "\n",
       "        extraction_batch  Notes:  visual_equals_compute  \n",
       "sex                                                      \n",
       "female                31      22                     91  \n",
       "male                   4       4                      4  \n",
       "\n",
       "[2 rows x 30 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllMosData.groupby('sex').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#These numbers match \"female\" and \"male\" grep I did on cli on source data file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Next, checked on discrepancies between computed species calls and visual calls."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "AllMosData['visual_equals_compute'] = AllMosData['visual_species'] == AllMosData['compute_species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>OldIDseqName</th>\n",
       "      <th>czbiohub-mosquito_sequences_id</th>\n",
       "      <th>raw_sequence_run_directory</th>\n",
       "      <th>host</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>nucleotide_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>...</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>sample_unit</th>\n",
       "      <th>known_organism</th>\n",
       "      <th>detection_method</th>\n",
       "      <th>library_prep</th>\n",
       "      <th>sequencer</th>\n",
       "      <th>rna_dna_input_ng</th>\n",
       "      <th>library_prep_batch</th>\n",
       "      <th>extraction_batch</th>\n",
       "      <th>Notes:</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>visual_equals_compute</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>143</td>\n",
       "      <td>...</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>135</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       NewIDseqName  OldIDseqName  \\\n",
       "visual_equals_compute                               \n",
       "False                            17            17   \n",
       "True                            144           144   \n",
       "\n",
       "                       czbiohub-mosquito_sequences_id  \\\n",
       "visual_equals_compute                                   \n",
       "False                                              17   \n",
       "True                                              144   \n",
       "\n",
       "                       raw_sequence_run_directory  host  sample_type  \\\n",
       "visual_equals_compute                                                  \n",
       "False                                          17     5           12   \n",
       "True                                          144   144          144   \n",
       "\n",
       "                       nucleotide_type  collection_date  collection_lat  \\\n",
       "visual_equals_compute                                                     \n",
       "False                               17                3               3   \n",
       "True                               144              143             143   \n",
       "\n",
       "                       collection_long   ...    blood_fed  sample_unit  \\\n",
       "visual_equals_compute                    ...                             \n",
       "False                                3   ...            4           17   \n",
       "True                               143   ...          144          144   \n",
       "\n",
       "                       known_organism  detection_method  library_prep  \\\n",
       "visual_equals_compute                                                   \n",
       "False                               0                 0            17   \n",
       "True                                0                 0           144   \n",
       "\n",
       "                       sequencer  rna_dna_input_ng  library_prep_batch  \\\n",
       "visual_equals_compute                                                    \n",
       "False                         17                 4                   0   \n",
       "True                         144               135                   0   \n",
       "\n",
       "                       extraction_batch  Notes:  \n",
       "visual_equals_compute                            \n",
       "False                                 8       0  \n",
       "True                                 88      75  \n",
       "\n",
       "[2 rows x 30 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllMosData.groupby('visual_equals_compute').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Quick print all column names again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NewIDseqName' 'OldIDseqName' 'czbiohub-mosquito_sequences_id'\n",
      " 'raw_sequence_run_directory' 'host' 'sample_type' 'nucleotide_type'\n",
      " 'collection_date' 'collection_lat' 'collection_long' 'Habitat'\n",
      " 'Population' 'collected_by' 'preservation_method' 'sex' 'life_stage'\n",
      " 'visual_genus' 'visual_species' 'compute_genus' 'compute_species'\n",
      " 'blood_fed' 'sample_unit' 'known_organism' 'detection_method'\n",
      " 'library_prep' 'sequencer' 'rna_dna_input_ng' 'library_prep_batch'\n",
      " 'extraction_batch' 'Notes:' 'visual_equals_compute']\n"
     ]
    }
   ],
   "source": [
    "print(AllMosData.columns.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Sort dataframe on 'visual_equals_compute' columns and head to get a look at the 17 discordant samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>visual_equals_compute</th>\n",
       "      <th>visual_genus</th>\n",
       "      <th>visual_species</th>\n",
       "      <th>compute_genus</th>\n",
       "      <th>compute_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>CMS002_Water8_Rb_S11_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>CMS001_water3_Qiagen_S26</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CMS001_025_Ra_S7</td>\n",
       "      <td>False</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CMS001_024_Ra_S15</td>\n",
       "      <td>False</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CMS001_045_Ra_S2</td>\n",
       "      <td>False</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>CMS001_water2_S24</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>CMS001_water1_S11</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CMS001_058_Ra_S9</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>CMS001_water4_Zymo_S27</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>CMS001_water5_RNA_A_S12</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>CMS002_Water6_Rb_S187_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>CMS002_Water5_Rb_S175_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>CMS002_Water4_Rb_S163_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>CMS002_Water3_Rb_S151_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>CMS002_Water2_Rb_S139_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>CMS002_Water1_Rb_S127_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>CMS002_Water7_Rb_S199_L004</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>CMS002_033a_Rb_S167_L004</td>\n",
       "      <td>True</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>CMS002_028e_Rb_S158_L004</td>\n",
       "      <td>True</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>CMS002_029a_Rb_S159_L004</td>\n",
       "      <td>True</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   NewIDseqName  visual_equals_compute visual_genus  \\\n",
       "160   CMS002_Water8_Rb_S11_L004                  False          NaN   \n",
       "62     CMS001_water3_Qiagen_S26                  False          NaN   \n",
       "24             CMS001_025_Ra_S7                  False        Culex   \n",
       "23            CMS001_024_Ra_S15                  False        Culex   \n",
       "44             CMS001_045_Ra_S2                  False        Culex   \n",
       "61            CMS001_water2_S24                  False          NaN   \n",
       "60            CMS001_water1_S11                  False          NaN   \n",
       "57             CMS001_058_Ra_S9                  False          NaN   \n",
       "63       CMS001_water4_Zymo_S27                  False          NaN   \n",
       "64      CMS001_water5_RNA_A_S12                  False          NaN   \n",
       "158  CMS002_Water6_Rb_S187_L004                  False          NaN   \n",
       "157  CMS002_Water5_Rb_S175_L004                  False          NaN   \n",
       "156  CMS002_Water4_Rb_S163_L004                  False          NaN   \n",
       "155  CMS002_Water3_Rb_S151_L004                  False          NaN   \n",
       "154  CMS002_Water2_Rb_S139_L004                  False          NaN   \n",
       "153  CMS002_Water1_Rb_S127_L004                  False          NaN   \n",
       "159  CMS002_Water7_Rb_S199_L004                  False          NaN   \n",
       "112    CMS002_033a_Rb_S167_L004                   True        Culex   \n",
       "104    CMS002_028e_Rb_S158_L004                   True        Culex   \n",
       "105    CMS002_029a_Rb_S159_L004                   True        Culex   \n",
       "\n",
       "    visual_species compute_genus compute_species  \n",
       "160            NaN           NaN             NaN  \n",
       "62             NaN           NaN             NaN  \n",
       "24        tarsalis      Culiseta        incidens  \n",
       "23        tarsalis      Culiseta       particeps  \n",
       "44        tarsalis         Culex   erythrothorax  \n",
       "61             NaN           NaN             NaN  \n",
       "60             NaN           NaN             NaN  \n",
       "57             NaN         Culex   erythrothorax  \n",
       "63             NaN           NaN             NaN  \n",
       "64             NaN           NaN             NaN  \n",
       "158            NaN           NaN             NaN  \n",
       "157            NaN           NaN             NaN  \n",
       "156            NaN           NaN             NaN  \n",
       "155            NaN           NaN             NaN  \n",
       "154            NaN           NaN             NaN  \n",
       "153            NaN           NaN             NaN  \n",
       "159            NaN           NaN             NaN  \n",
       "112       tarsalis         Culex        tarsalis  \n",
       "104  erythrothorax         Culex   erythrothorax  \n",
       "105       tarsalis         Culex        tarsalis  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllMosData.sort_values(\"visual_equals_compute\")[['NewIDseqName', 'visual_equals_compute','visual_genus','visual_species','compute_genus','compute_species']].head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Note: The discordant samples correspond primarily to water controls because they have no calls for either visual or compute fields. There are also 4 mosquito specimens where computed and visual calls are different: CMS001_025, CMS001_024, CMS001_045, and CMS001_058. Note that CMS001_058 was also discordant, but this was because no visual call was provided for this specimen. The other mosquito specimens with a discordant call were likely visual mis-calls or mis-annotations. Based on this, \n",
    "\n",
    "##Action items: 1) I'm going to use compute_genus and compute_species calls for TableS1 genus and species calls. 2) Review and confirm overall kmer results & specific discordant results with JB. 3) Should probably highlight these 4 samples with discordant visual and compute calls. 4) Develop and run directed analysis of a transcript (or a set of transcripts) to provide conventional ground truth data to validate [Reviewers will ask for this]. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NewIDseqName' 'OldIDseqName' 'czbiohub-mosquito_sequences_id'\n",
      " 'raw_sequence_run_directory' 'host' 'sample_type' 'nucleotide_type'\n",
      " 'collection_date' 'collection_lat' 'collection_long' 'Habitat'\n",
      " 'Population' 'collected_by' 'preservation_method' 'sex' 'life_stage'\n",
      " 'visual_genus' 'visual_species' 'compute_genus' 'compute_species'\n",
      " 'blood_fed' 'sample_unit' 'known_organism' 'detection_method'\n",
      " 'library_prep' 'sequencer' 'rna_dna_input_ng' 'library_prep_batch'\n",
      " 'extraction_batch' 'Notes:' 'visual_equals_compute']\n"
     ]
    }
   ],
   "source": [
    "print(AllMosData.columns.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Generating new dataframe \"TableS1data\" with subset of relevant data for TableS1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TableS1data = AllMosData.loc[:,('NewIDseqName','compute_genus','compute_species','sex','blood_fed','collected_by','collection_date','collection_lat','collection_long','Habitat')].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>compute_genus</th>\n",
       "      <th>compute_species</th>\n",
       "      <th>sex</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NewIDseqName compute_genus compute_species     sex  blood_fed  \\\n",
       "0  CMS001_001_Ra_S1         Culex   erythrothorax  female  blood fed   \n",
       "1  CMS001_002_Ra_S1      Culiseta       particeps  female  blood fed   \n",
       "2  CMS001_003_Ra_S2         Culex         pipiens  female  blood fed   \n",
       "3  CMS001_004_Ra_S2      Culiseta        incidens  female  blood fed   \n",
       "4  CMS001_005_Ra_S3         Culex   erythrothorax  female  blood fed   \n",
       "\n",
       "  collected_by       collection_date  collection_lat  collection_long  \\\n",
       "0         ALCO  2017-10-03T00:00:00Z        37.55697       -122.07938   \n",
       "1         ALCO  2017-09-21T00:00:00Z        37.55697       -122.07938   \n",
       "2         ALCO  2017-09-19T00:00:00Z        37.55697       -122.07938   \n",
       "3         ALCO  2017-09-12T00:00:00Z        37.71520       -122.19431   \n",
       "4         ALCO  2017-09-29T00:00:00Z        37.78886       -122.29903   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "1             Marsh / Wildlife Refuge / Park  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "4                                Residential  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>compute_genus</th>\n",
       "      <th>compute_species</th>\n",
       "      <th>sex</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NewIDseqName compute_genus compute_species     sex  blood_fed  \\\n",
       "0  CMS001_001_Ra_S1         Culex   erythrothorax  female  blood fed   \n",
       "4  CMS001_005_Ra_S3         Culex   erythrothorax  female  blood fed   \n",
       "2  CMS001_003_Ra_S2         Culex         pipiens  female  blood fed   \n",
       "3  CMS001_004_Ra_S2      Culiseta        incidens  female  blood fed   \n",
       "1  CMS001_002_Ra_S1      Culiseta       particeps  female  blood fed   \n",
       "\n",
       "  collected_by       collection_date  collection_lat  collection_long  \\\n",
       "0         ALCO  2017-10-03T00:00:00Z        37.55697       -122.07938   \n",
       "4         ALCO  2017-09-29T00:00:00Z        37.78886       -122.29903   \n",
       "2         ALCO  2017-09-19T00:00:00Z        37.55697       -122.07938   \n",
       "3         ALCO  2017-09-12T00:00:00Z        37.71520       -122.19431   \n",
       "1         ALCO  2017-09-21T00:00:00Z        37.55697       -122.07938   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "4                                Residential  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "1             Marsh / Wildlife Refuge / Park  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data.sort_values(by = ['collected_by','compute_genus','compute_species'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Strip unnecessary suffix string from 'NewIDseqName' values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>compute_genus</th>\n",
       "      <th>compute_species</th>\n",
       "      <th>sex</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  NewIDseqName compute_genus compute_species     sex  blood_fed collected_by  \\\n",
       "0   CMS001_001         Culex   erythrothorax  female  blood fed         ALCO   \n",
       "1   CMS001_002      Culiseta       particeps  female  blood fed         ALCO   \n",
       "2   CMS001_003         Culex         pipiens  female  blood fed         ALCO   \n",
       "3   CMS001_004      Culiseta        incidens  female  blood fed         ALCO   \n",
       "4   CMS001_005         Culex   erythrothorax  female  blood fed         ALCO   \n",
       "\n",
       "        collection_date  collection_lat  collection_long  \\\n",
       "0  2017-10-03T00:00:00Z        37.55697       -122.07938   \n",
       "1  2017-09-21T00:00:00Z        37.55697       -122.07938   \n",
       "2  2017-09-19T00:00:00Z        37.55697       -122.07938   \n",
       "3  2017-09-12T00:00:00Z        37.71520       -122.19431   \n",
       "4  2017-09-29T00:00:00Z        37.78886       -122.29903   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "1             Marsh / Wildlife Refuge / Park  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "4                                Residential  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data.replace(r'\\_Ra.*', '', regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Strip unnecessary time suffix string from 'collection_date' values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NewIDseqName</th>\n",
       "      <th>compute_genus</th>\n",
       "      <th>compute_species</th>\n",
       "      <th>sex</th>\n",
       "      <th>blood_fed</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>collection_lat</th>\n",
       "      <th>collection_long</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       NewIDseqName compute_genus compute_species     sex  blood_fed  \\\n",
       "0  CMS001_001_Ra_S1         Culex   erythrothorax  female  blood fed   \n",
       "1  CMS001_002_Ra_S1      Culiseta       particeps  female  blood fed   \n",
       "2  CMS001_003_Ra_S2         Culex         pipiens  female  blood fed   \n",
       "3  CMS001_004_Ra_S2      Culiseta        incidens  female  blood fed   \n",
       "4  CMS001_005_Ra_S3         Culex   erythrothorax  female  blood fed   \n",
       "\n",
       "  collected_by collection_date  collection_lat  collection_long  \\\n",
       "0         ALCO      2017-10-03        37.55697       -122.07938   \n",
       "1         ALCO      2017-09-21        37.55697       -122.07938   \n",
       "2         ALCO      2017-09-19        37.55697       -122.07938   \n",
       "3         ALCO      2017-09-12        37.71520       -122.19431   \n",
       "4         ALCO      2017-09-29        37.78886       -122.29903   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "1             Marsh / Wildlife Refuge / Park  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "4                                Residential  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data.replace(r'T00.*', '', regex=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Tidying up dataframe: renaming columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TableS1data.rename(columns={'NewIDseqName':'Specimen ID','compute_genus':'Genus','compute_species':'Species','sex':'Sex','blood_fed':'Status','collected_by':'Collection Site','collection_date':'Collection Date','collection_lat':'Latitude','collection_long':'Longitude'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specimen ID</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Status</th>\n",
       "      <th>Collection Site</th>\n",
       "      <th>Collection Date</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Specimen ID     Genus        Species     Sex     Status  \\\n",
       "0  CMS001_001_Ra_S1     Culex  erythrothorax  female  blood fed   \n",
       "1  CMS001_002_Ra_S1  Culiseta      particeps  female  blood fed   \n",
       "2  CMS001_003_Ra_S2     Culex        pipiens  female  blood fed   \n",
       "3  CMS001_004_Ra_S2  Culiseta       incidens  female  blood fed   \n",
       "4  CMS001_005_Ra_S3     Culex  erythrothorax  female  blood fed   \n",
       "\n",
       "  Collection Site       Collection Date  Latitude  Longitude  \\\n",
       "0            ALCO  2017-10-03T00:00:00Z  37.55697 -122.07938   \n",
       "1            ALCO  2017-09-21T00:00:00Z  37.55697 -122.07938   \n",
       "2            ALCO  2017-09-19T00:00:00Z  37.55697 -122.07938   \n",
       "3            ALCO  2017-09-12T00:00:00Z  37.71520 -122.19431   \n",
       "4            ALCO  2017-09-29T00:00:00Z  37.78886 -122.29903   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "1             Marsh / Wildlife Refuge / Park  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "4                                Residential  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specimen ID</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Status</th>\n",
       "      <th>Collection Site</th>\n",
       "      <th>Collection Date</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002_Ra_S1</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19T00:00:00Z</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12T00:00:00Z</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29T00:00:00Z</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Specimen ID     Genus        Species     Sex     Status  \\\n",
       "0  CMS001_001_Ra_S1     Culex  erythrothorax  female  blood fed   \n",
       "1  CMS001_002_Ra_S1  Culiseta      particeps  female  blood fed   \n",
       "2  CMS001_003_Ra_S2     Culex        pipiens  female  blood fed   \n",
       "3  CMS001_004_Ra_S2  Culiseta       incidens  female  blood fed   \n",
       "4  CMS001_005_Ra_S3     Culex  erythrothorax  female  blood fed   \n",
       "\n",
       "  Collection Site       Collection Date  Latitude  Longitude  \\\n",
       "0            ALCO  2017-10-03T00:00:00Z  37.55697 -122.07938   \n",
       "1            ALCO  2017-09-21T00:00:00Z  37.55697 -122.07938   \n",
       "2            ALCO  2017-09-19T00:00:00Z  37.55697 -122.07938   \n",
       "3            ALCO  2017-09-12T00:00:00Z  37.71520 -122.19431   \n",
       "4            ALCO  2017-09-29T00:00:00Z  37.78886 -122.29903   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "1             Marsh / Wildlife Refuge / Park  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "4                                Residential  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data[['Specimen ID','Genus','Species','Sex','Status','Collection Site','Collection Date','Latitude','Longitude','Habitat']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Playing with table formatting - specifying text alignment in table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col9 {\n",
       "            text-align:  left;\n",
       "        }</style>  \n",
       "<table id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250a\" > \n",
       "<thead>    <tr> \n",
       "        <th class=\"blank level0\" ></th> \n",
       "        <th class=\"col_heading level0 col0\" >Specimen ID</th> \n",
       "        <th class=\"col_heading level0 col1\" >Genus</th> \n",
       "        <th class=\"col_heading level0 col2\" >Species</th> \n",
       "        <th class=\"col_heading level0 col3\" >Sex</th> \n",
       "        <th class=\"col_heading level0 col4\" >Status</th> \n",
       "        <th class=\"col_heading level0 col5\" >Collection Site</th> \n",
       "        <th class=\"col_heading level0 col6\" >Collection Date</th> \n",
       "        <th class=\"col_heading level0 col7\" >Latitude</th> \n",
       "        <th class=\"col_heading level0 col8\" >Longitude</th> \n",
       "        <th class=\"col_heading level0 col9\" >Habitat</th> \n",
       "    </tr></thead> \n",
       "<tbody>    <tr> \n",
       "        <th id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250alevel0_row0\" class=\"row_heading level0 row0\" >0</th> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col0\" class=\"data row0 col0\" >CMS001_001_Ra_S1</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col1\" class=\"data row0 col1\" >Culex</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col2\" class=\"data row0 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col3\" class=\"data row0 col3\" >female</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col4\" class=\"data row0 col4\" >blood fed</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col5\" class=\"data row0 col5\" >ALCO</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col6\" class=\"data row0 col6\" >2017-10-03T00:00:00Z</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col7\" class=\"data row0 col7\" >37.557</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col8\" class=\"data row0 col8\" >-122.079</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow0_col9\" class=\"data row0 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250alevel0_row1\" class=\"row_heading level0 row1\" >1</th> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col0\" class=\"data row1 col0\" >CMS001_002_Ra_S1</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col1\" class=\"data row1 col1\" >Culiseta</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col2\" class=\"data row1 col2\" >particeps</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col3\" class=\"data row1 col3\" >female</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col4\" class=\"data row1 col4\" >blood fed</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col5\" class=\"data row1 col5\" >ALCO</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col6\" class=\"data row1 col6\" >2017-09-21T00:00:00Z</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col7\" class=\"data row1 col7\" >37.557</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col8\" class=\"data row1 col8\" >-122.079</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow1_col9\" class=\"data row1 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250alevel0_row2\" class=\"row_heading level0 row2\" >2</th> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col0\" class=\"data row2 col0\" >CMS001_003_Ra_S2</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col1\" class=\"data row2 col1\" >Culex</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col2\" class=\"data row2 col2\" >pipiens</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col3\" class=\"data row2 col3\" >female</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col4\" class=\"data row2 col4\" >blood fed</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col5\" class=\"data row2 col5\" >ALCO</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col6\" class=\"data row2 col6\" >2017-09-19T00:00:00Z</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col7\" class=\"data row2 col7\" >37.557</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col8\" class=\"data row2 col8\" >-122.079</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow2_col9\" class=\"data row2 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250alevel0_row3\" class=\"row_heading level0 row3\" >3</th> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col0\" class=\"data row3 col0\" >CMS001_004_Ra_S2</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col1\" class=\"data row3 col1\" >Culiseta</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col2\" class=\"data row3 col2\" >incidens</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col3\" class=\"data row3 col3\" >female</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col4\" class=\"data row3 col4\" >blood fed</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col5\" class=\"data row3 col5\" >ALCO</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col6\" class=\"data row3 col6\" >2017-09-12T00:00:00Z</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col7\" class=\"data row3 col7\" >37.7152</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col8\" class=\"data row3 col8\" >-122.194</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow3_col9\" class=\"data row3 col9\" > Industrial (waste water treatment plant)</td> \n",
       "    </tr>    <tr> \n",
       "        <th id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250alevel0_row4\" class=\"row_heading level0 row4\" >4</th> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col0\" class=\"data row4 col0\" >CMS001_005_Ra_S3</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col1\" class=\"data row4 col1\" >Culex</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col2\" class=\"data row4 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col3\" class=\"data row4 col3\" >female</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col4\" class=\"data row4 col4\" >blood fed</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col5\" class=\"data row4 col5\" >ALCO</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col6\" class=\"data row4 col6\" >2017-09-29T00:00:00Z</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col7\" class=\"data row4 col7\" >37.7889</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col8\" class=\"data row4 col8\" >-122.299</td> \n",
       "        <td id=\"T_c15456f0_6bd8_11e9_afa8_48bf6bdf250arow4_col9\" class=\"data row4 col9\" >Residential</td> \n",
       "    </tr></tbody> \n",
       "</table> "
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x11eaf20b8>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data.style.set_properties(**{'text-align': 'left'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Playing with table formatting - specifying column alignment, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Using style to eliminate index, align text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col9 {\n",
       "            text-align:  left;\n",
       "        }</style>  \n",
       "<table id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250a\" > \n",
       "<thead>    <tr> \n",
       "        <th class=\"col_heading level0 col0\" >Specimen ID</th> \n",
       "        <th class=\"col_heading level0 col1\" >Genus</th> \n",
       "        <th class=\"col_heading level0 col2\" >Species</th> \n",
       "        <th class=\"col_heading level0 col3\" >Sex</th> \n",
       "        <th class=\"col_heading level0 col4\" >Status</th> \n",
       "        <th class=\"col_heading level0 col5\" >Collection Site</th> \n",
       "        <th class=\"col_heading level0 col6\" >Collection Date</th> \n",
       "        <th class=\"col_heading level0 col7\" >Latitude</th> \n",
       "        <th class=\"col_heading level0 col8\" >Longitude</th> \n",
       "        <th class=\"col_heading level0 col9\" >Habitat</th> \n",
       "    </tr></thead> \n",
       "<tbody>    <tr> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col0\" class=\"data row0 col0\" >CMS001_001_Ra_S1</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col1\" class=\"data row0 col1\" >Culex</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col2\" class=\"data row0 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col3\" class=\"data row0 col3\" >female</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col4\" class=\"data row0 col4\" >blood fed</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col5\" class=\"data row0 col5\" >ALCO</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col6\" class=\"data row0 col6\" >2017-10-03T00:00:00Z</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col7\" class=\"data row0 col7\" >37.557</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col8\" class=\"data row0 col8\" >-122.079</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow0_col9\" class=\"data row0 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col0\" class=\"data row1 col0\" >CMS001_002_Ra_S1</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col1\" class=\"data row1 col1\" >Culiseta</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col2\" class=\"data row1 col2\" >particeps</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col3\" class=\"data row1 col3\" >female</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col4\" class=\"data row1 col4\" >blood fed</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col5\" class=\"data row1 col5\" >ALCO</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col6\" class=\"data row1 col6\" >2017-09-21T00:00:00Z</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col7\" class=\"data row1 col7\" >37.557</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col8\" class=\"data row1 col8\" >-122.079</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow1_col9\" class=\"data row1 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col0\" class=\"data row2 col0\" >CMS001_003_Ra_S2</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col1\" class=\"data row2 col1\" >Culex</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col2\" class=\"data row2 col2\" >pipiens</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col3\" class=\"data row2 col3\" >female</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col4\" class=\"data row2 col4\" >blood fed</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col5\" class=\"data row2 col5\" >ALCO</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col6\" class=\"data row2 col6\" >2017-09-19T00:00:00Z</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col7\" class=\"data row2 col7\" >37.557</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col8\" class=\"data row2 col8\" >-122.079</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow2_col9\" class=\"data row2 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col0\" class=\"data row3 col0\" >CMS001_004_Ra_S2</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col1\" class=\"data row3 col1\" >Culiseta</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col2\" class=\"data row3 col2\" >incidens</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col3\" class=\"data row3 col3\" >female</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col4\" class=\"data row3 col4\" >blood fed</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col5\" class=\"data row3 col5\" >ALCO</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col6\" class=\"data row3 col6\" >2017-09-12T00:00:00Z</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col7\" class=\"data row3 col7\" >37.7152</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col8\" class=\"data row3 col8\" >-122.194</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow3_col9\" class=\"data row3 col9\" > Industrial (waste water treatment plant)</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col0\" class=\"data row4 col0\" >CMS001_005_Ra_S3</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col1\" class=\"data row4 col1\" >Culex</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col2\" class=\"data row4 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col3\" class=\"data row4 col3\" >female</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col4\" class=\"data row4 col4\" >blood fed</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col5\" class=\"data row4 col5\" >ALCO</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col6\" class=\"data row4 col6\" >2017-09-29T00:00:00Z</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col7\" class=\"data row4 col7\" >37.7889</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col8\" class=\"data row4 col8\" >-122.299</td> \n",
       "        <td id=\"T_c4f58888_6bd8_11e9_942b_48bf6bdf250arow4_col9\" class=\"data row4 col9\" >Residential</td> \n",
       "    </tr></tbody> \n",
       "</table> "
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x11eaf20f0>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1data.style.hide_index().set_properties(**{'text-align': 'left'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#More sophisticated (props to StackOverflow): specifying multiple items - hide index, specify header and cell alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250a th {\n",
       "          text-align: left;\n",
       "    }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col9 {\n",
       "            text-align:  left;\n",
       "        }</style>  \n",
       "<table id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250a\" > \n",
       "<thead>    <tr> \n",
       "        <th class=\"col_heading level0 col0\" >Specimen ID</th> \n",
       "        <th class=\"col_heading level0 col1\" >Genus</th> \n",
       "        <th class=\"col_heading level0 col2\" >Species</th> \n",
       "        <th class=\"col_heading level0 col3\" >Sex</th> \n",
       "        <th class=\"col_heading level0 col4\" >Status</th> \n",
       "        <th class=\"col_heading level0 col5\" >Collection Site</th> \n",
       "        <th class=\"col_heading level0 col6\" >Collection Date</th> \n",
       "        <th class=\"col_heading level0 col7\" >Latitude</th> \n",
       "        <th class=\"col_heading level0 col8\" >Longitude</th> \n",
       "        <th class=\"col_heading level0 col9\" >Habitat</th> \n",
       "    </tr></thead> \n",
       "<tbody>    <tr> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col0\" class=\"data row0 col0\" >CMS001_001_Ra_S1</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col1\" class=\"data row0 col1\" >Culex</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col2\" class=\"data row0 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col3\" class=\"data row0 col3\" >female</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col4\" class=\"data row0 col4\" >blood fed</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col5\" class=\"data row0 col5\" >ALCO</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col6\" class=\"data row0 col6\" >2017-10-03T00:00:00Z</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col7\" class=\"data row0 col7\" >37.557</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col8\" class=\"data row0 col8\" >-122.079</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow0_col9\" class=\"data row0 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col0\" class=\"data row1 col0\" >CMS001_002_Ra_S1</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col1\" class=\"data row1 col1\" >Culiseta</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col2\" class=\"data row1 col2\" >particeps</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col3\" class=\"data row1 col3\" >female</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col4\" class=\"data row1 col4\" >blood fed</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col5\" class=\"data row1 col5\" >ALCO</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col6\" class=\"data row1 col6\" >2017-09-21T00:00:00Z</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col7\" class=\"data row1 col7\" >37.557</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col8\" class=\"data row1 col8\" >-122.079</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow1_col9\" class=\"data row1 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col0\" class=\"data row2 col0\" >CMS001_003_Ra_S2</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col1\" class=\"data row2 col1\" >Culex</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col2\" class=\"data row2 col2\" >pipiens</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col3\" class=\"data row2 col3\" >female</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col4\" class=\"data row2 col4\" >blood fed</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col5\" class=\"data row2 col5\" >ALCO</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col6\" class=\"data row2 col6\" >2017-09-19T00:00:00Z</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col7\" class=\"data row2 col7\" >37.557</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col8\" class=\"data row2 col8\" >-122.079</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow2_col9\" class=\"data row2 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col0\" class=\"data row3 col0\" >CMS001_004_Ra_S2</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col1\" class=\"data row3 col1\" >Culiseta</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col2\" class=\"data row3 col2\" >incidens</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col3\" class=\"data row3 col3\" >female</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col4\" class=\"data row3 col4\" >blood fed</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col5\" class=\"data row3 col5\" >ALCO</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col6\" class=\"data row3 col6\" >2017-09-12T00:00:00Z</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col7\" class=\"data row3 col7\" >37.7152</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col8\" class=\"data row3 col8\" >-122.194</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow3_col9\" class=\"data row3 col9\" > Industrial (waste water treatment plant)</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col0\" class=\"data row4 col0\" >CMS001_005_Ra_S3</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col1\" class=\"data row4 col1\" >Culex</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col2\" class=\"data row4 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col3\" class=\"data row4 col3\" >female</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col4\" class=\"data row4 col4\" >blood fed</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col5\" class=\"data row4 col5\" >ALCO</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col6\" class=\"data row4 col6\" >2017-09-29T00:00:00Z</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col7\" class=\"data row4 col7\" >37.7889</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col8\" class=\"data row4 col8\" >-122.299</td> \n",
       "        <td id=\"T_c96a5846_6bd8_11e9_9dd4_48bf6bdf250arow4_col9\" class=\"data row4 col9\" >Residential</td> \n",
       "    </tr></tbody> \n",
       "</table> "
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x12324be48>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "styler = TableS1data.style\n",
    "#hide the index values\n",
    "styler = styler.hide_index()\n",
    "#left align cell values\n",
    "styler = styler.set_properties(**{'text-align': 'left'})\n",
    "# select the table header with th and set it left align\n",
    "styler.set_table_styles([dict(selector=\"th\", props=[(\"text-align\", \"left\")])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Specimen ID' 'Genus' 'Species' 'Sex' 'Status' 'Collection Site'\n",
      " 'Collection Date' 'Latitude' 'Longitude' 'Habitat']\n"
     ]
    }
   ],
   "source": [
    "print(TableS1data.columns.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Strip unnecessary suffixes values in 'Specimen ID' and 'Collection Date' columns, and store in a new data frame, TableS1datafinal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "TableS1datafinal = TableS1data.replace(r'\\_Ra.*', '', regex=True).replace(r'T00.*', '', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specimen ID</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Status</th>\n",
       "      <th>Collection Site</th>\n",
       "      <th>Collection Date</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Habitat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMS001_001</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-10-03</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMS001_002</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-21</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMS001_003</td>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-19</td>\n",
       "      <td>37.55697</td>\n",
       "      <td>-122.07938</td>\n",
       "      <td>Marsh / Wildlife Refuge / Park</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMS001_004</td>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-12</td>\n",
       "      <td>37.71520</td>\n",
       "      <td>-122.19431</td>\n",
       "      <td>Industrial (waste water treatment plant)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMS001_005</td>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>female</td>\n",
       "      <td>blood fed</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>2017-09-29</td>\n",
       "      <td>37.78886</td>\n",
       "      <td>-122.29903</td>\n",
       "      <td>Residential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Specimen ID     Genus        Species     Sex     Status Collection Site  \\\n",
       "0  CMS001_001     Culex  erythrothorax  female  blood fed            ALCO   \n",
       "1  CMS001_002  Culiseta      particeps  female  blood fed            ALCO   \n",
       "2  CMS001_003     Culex        pipiens  female  blood fed            ALCO   \n",
       "3  CMS001_004  Culiseta       incidens  female  blood fed            ALCO   \n",
       "4  CMS001_005     Culex  erythrothorax  female  blood fed            ALCO   \n",
       "\n",
       "  Collection Date  Latitude  Longitude  \\\n",
       "0      2017-10-03  37.55697 -122.07938   \n",
       "1      2017-09-21  37.55697 -122.07938   \n",
       "2      2017-09-19  37.55697 -122.07938   \n",
       "3      2017-09-12  37.71520 -122.19431   \n",
       "4      2017-09-29  37.78886 -122.29903   \n",
       "\n",
       "                                     Habitat  \n",
       "0             Marsh / Wildlife Refuge / Park  \n",
       "1             Marsh / Wildlife Refuge / Park  \n",
       "2             Marsh / Wildlife Refuge / Park  \n",
       "3   Industrial (waste water treatment plant)  \n",
       "4                                Residential  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1datafinal.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Clean up format of TableS1datafinal data frame - remove index column, align left both column headers and cell values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250a th {\n",
       "          text-align: left;\n",
       "    }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col9 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col0 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col1 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col2 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col3 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col4 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col5 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col6 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col7 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col8 {\n",
       "            text-align:  left;\n",
       "        }    #T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col9 {\n",
       "            text-align:  left;\n",
       "        }</style>  \n",
       "<table id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250a\" > \n",
       "<thead>    <tr> \n",
       "        <th class=\"col_heading level0 col0\" >Specimen ID</th> \n",
       "        <th class=\"col_heading level0 col1\" >Genus</th> \n",
       "        <th class=\"col_heading level0 col2\" >Species</th> \n",
       "        <th class=\"col_heading level0 col3\" >Sex</th> \n",
       "        <th class=\"col_heading level0 col4\" >Status</th> \n",
       "        <th class=\"col_heading level0 col5\" >Collection Site</th> \n",
       "        <th class=\"col_heading level0 col6\" >Collection Date</th> \n",
       "        <th class=\"col_heading level0 col7\" >Latitude</th> \n",
       "        <th class=\"col_heading level0 col8\" >Longitude</th> \n",
       "        <th class=\"col_heading level0 col9\" >Habitat</th> \n",
       "    </tr></thead> \n",
       "<tbody>    <tr> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col0\" class=\"data row0 col0\" >CMS001_001</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col1\" class=\"data row0 col1\" >Culex</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col2\" class=\"data row0 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col3\" class=\"data row0 col3\" >female</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col4\" class=\"data row0 col4\" >blood fed</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col5\" class=\"data row0 col5\" >ALCO</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col6\" class=\"data row0 col6\" >2017-10-03</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col7\" class=\"data row0 col7\" >37.557</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col8\" class=\"data row0 col8\" >-122.079</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow0_col9\" class=\"data row0 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col0\" class=\"data row1 col0\" >CMS001_002</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col1\" class=\"data row1 col1\" >Culiseta</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col2\" class=\"data row1 col2\" >particeps</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col3\" class=\"data row1 col3\" >female</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col4\" class=\"data row1 col4\" >blood fed</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col5\" class=\"data row1 col5\" >ALCO</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col6\" class=\"data row1 col6\" >2017-09-21</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col7\" class=\"data row1 col7\" >37.557</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col8\" class=\"data row1 col8\" >-122.079</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow1_col9\" class=\"data row1 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col0\" class=\"data row2 col0\" >CMS001_003</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col1\" class=\"data row2 col1\" >Culex</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col2\" class=\"data row2 col2\" >pipiens</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col3\" class=\"data row2 col3\" >female</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col4\" class=\"data row2 col4\" >blood fed</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col5\" class=\"data row2 col5\" >ALCO</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col6\" class=\"data row2 col6\" >2017-09-19</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col7\" class=\"data row2 col7\" >37.557</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col8\" class=\"data row2 col8\" >-122.079</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow2_col9\" class=\"data row2 col9\" >Marsh / Wildlife Refuge / Park</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col0\" class=\"data row3 col0\" >CMS001_004</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col1\" class=\"data row3 col1\" >Culiseta</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col2\" class=\"data row3 col2\" >incidens</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col3\" class=\"data row3 col3\" >female</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col4\" class=\"data row3 col4\" >blood fed</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col5\" class=\"data row3 col5\" >ALCO</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col6\" class=\"data row3 col6\" >2017-09-12</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col7\" class=\"data row3 col7\" >37.7152</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col8\" class=\"data row3 col8\" >-122.194</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow3_col9\" class=\"data row3 col9\" > Industrial (waste water treatment plant)</td> \n",
       "    </tr>    <tr> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col0\" class=\"data row4 col0\" >CMS001_005</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col1\" class=\"data row4 col1\" >Culex</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col2\" class=\"data row4 col2\" >erythrothorax</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col3\" class=\"data row4 col3\" >female</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col4\" class=\"data row4 col4\" >blood fed</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col5\" class=\"data row4 col5\" >ALCO</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col6\" class=\"data row4 col6\" >2017-09-29</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col7\" class=\"data row4 col7\" >37.7889</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col8\" class=\"data row4 col8\" >-122.299</td> \n",
       "        <td id=\"T_0f3fc6ec_6bd9_11e9_8027_48bf6bdf250arow4_col9\" class=\"data row4 col9\" >Residential</td> \n",
       "    </tr></tbody> \n",
       "</table> "
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x12327c518>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "styler = TableS1datafinal.style\n",
    "#hide the index values\n",
    "styler = styler.hide_index()\n",
    "#left align cell values\n",
    "styler = styler.set_properties(**{'text-align': 'left'})\n",
    "# select the table header with th and set it left align\n",
    "styler.set_table_styles([dict(selector=\"th\", props=[(\"text-align\", \"left\")])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Specimen ID,Genus,Species,Sex,Status,Collection Site,Collection Date,Latitude,Longitude,Habitat\\nCMS001_001,Culex,erythrothorax,female,blood fed,ALCO,2017-10-03,37.55697,-122.07938,Marsh / Wildlife Refuge / Park\\nCMS001_002,Culiseta,particeps,female,blood fed,ALCO,2017-09-21,37.55697,-122.07938,Marsh / Wildlife Refuge / Park\\nCMS001_003,Culex,pipiens,female,blood fed,ALCO,2017-09-19,37.55697,-122.07938,Marsh / Wildlife Refuge / Park\\nCMS001_004,Culiseta,incidens,female,blood fed,ALCO,2017-09-12,37.7152,-122.19431000000002, Industrial (waste water treatment plant)\\nCMS001_005,Culex,erythrothorax,female,blood fed,ALCO,2017-09-29,37.78886,-122.29903,Residential\\n'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TableS1datafinal.to_csv(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
