{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to download the assembled contigs, blast them against nt and nr, then return those contigs to AWS.\n",
    "\n",
    "1. Set up an m5a.12xlarge instance, with keys and ports for jupyter from the `czbiohub-miniconda` AMI. Storage is at `/mnt/data`.\n",
    "\n",
    "`aegea launch --iam-role S3fromEC2 --ami-tags Name=czbiohub-jupyter -t m5a.12xlarge  batson-blast`\n",
    "\n",
    "Setup jupyter on the remote\n",
    "\n",
    "`aegea ssh batson@batson-blast`\n",
    "`tmux`\n",
    "`jupyter notebook`\n",
    "\n",
    "Port forwarding from laptop\n",
    "\n",
    "`aegea ssh batson@batson-blast -NL localhost:8899:localhost:8888`\n",
    "\n",
    "2. Download contigs\n",
    "\n",
    "`mkdir /mnt/data/contigs`\n",
    "`aws s3 sync s3://czbiohub-mosquito/contigs/ /mnt/data/contigs --exclude \"*\" --include \"*.fasta\" --dryrun`\n",
    "\n",
    "3. Install requirements\n",
    "\n",
    "conda install -c bioconda -c conda-forge blast\n",
    "\n",
    "mkdir /mnt/data/blast\n",
    "cd /mnt/data/blast\n",
    "update_blastdb.pl --decompress nt nr taxdb\n",
    "\n",
    "4. Run BLAST\n",
    "\n",
    "Loop over all contigs, and run for each sample\n",
    "\n",
    "`BLASTDB=/mnt/data/blast blastn -db nt -num_threads 48 -query /mnt/data/contigs/{SAMPLE}/contigs.fasta -outfmt 7 -out /mnt/data/contigs/{SAMPLE}/blastn_nt.m9 -evalue 1e-1`\n",
    "\n",
    "`BLASTDB=/mnt/data/blast blastx -db nr -num_threads 48 -query /mnt/data/contigs/{SAMPLE}/contigs.fasta -outfmt 7 -out /mnt/data/contigs/{SAMPLE}/blastx_nr.m9 -evalue 1e-1`\n",
    "\n",
    "5. Upload samples\n",
    "\n",
    "`aws s3 sync /mnt/data/contigs/ s3://czbiohub-mosquito/contigs/ --exclude \"*\" --include \"*.m9\" --dryrun`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/mnt/data/contigs’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "!mkdir /mnt/data/contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "160\r\n"
     ]
    }
   ],
   "source": [
    "!aws s3 sync s3://czbiohub-mosquito/contigs/ /mnt/data/contigs --exclude \"*\" --include \"*.fasta\" --dryrun | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://czbiohub-mosquito/contigs/CMS001_004_Ra_S2/contigs.fasta to ../../../../mnt/data/contigs/CMS001_004_Ra_S2/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_007_Ra_S12/contigs.fasta to ../../../../mnt/data/contigs/CMS001_007_Ra_S12/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_002_Ra_S1/contigs.fasta to ../../../../mnt/data/contigs/CMS001_002_Ra_S1/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_010_Ra_S1/contigs.fasta to ../../../../mnt/data/contigs/CMS001_010_Ra_S1/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_006_Ra_S5/contigs.fasta to ../../../../mnt/data/contigs/CMS001_006_Ra_S5/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_001_Ra_S1/contigs.fasta to ../../../../mnt/data/contigs/CMS001_001_Ra_S1/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_008_Ra_S3/contigs.fasta to ../../../../mnt/data/contigs/CMS001_008_Ra_S3/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_011_Ra_S4/contigs.fasta to ../../../../mnt/data/contigs/CMS001_011_Ra_S4/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_015_Ra_S13/contigs.fasta to ../../../../mnt/data/contigs/CMS001_015_Ra_S13/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_003_Ra_S2/contigs.fasta to ../../../../mnt/data/contigs/CMS001_003_Ra_S2/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_018_Ra_S14/contigs.fasta to ../../../../mnt/data/contigs/CMS001_018_Ra_S14/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_013_Ra_S5/contigs.fasta to ../../../../mnt/data/contigs/CMS001_013_Ra_S5/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_019_Ra_S14/contigs.fasta to ../../../../mnt/data/contigs/CMS001_019_Ra_S14/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_009_Ra_S13/contigs.fasta to ../../../../mnt/data/contigs/CMS001_009_Ra_S13/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_005_Ra_S3/contigs.fasta to ../../../../mnt/data/contigs/CMS001_005_Ra_S3/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_014_Ra_S5/contigs.fasta to ../../../../mnt/data/contigs/CMS001_014_Ra_S5/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_016_Ra_S6/contigs.fasta to ../../../../mnt/data/contigs/CMS001_016_Ra_S6/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_022_Ra_S6/contigs.fasta to ../../../../mnt/data/contigs/CMS001_022_Ra_S6/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_024_Ra_S15/contigs.fasta to ../../../../mnt/data/contigs/CMS001_024_Ra_S15/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_023_Ra_S17/contigs.fasta to ../../../../mnt/data/contigs/CMS001_023_Ra_S17/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_017_Ra_S6/contigs.fasta to ../../../../mnt/data/contigs/CMS001_017_Ra_S6/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_030_Ra_S7/contigs.fasta to ../../../../mnt/data/contigs/CMS001_030_Ra_S7/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_028_Ra_S17/contigs.fasta to ../../../../mnt/data/contigs/CMS001_028_Ra_S17/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_027_Ra_S16/contigs.fasta to ../../../../mnt/data/contigs/CMS001_027_Ra_S16/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_029_Ra_S18/contigs.fasta to ../../../../mnt/data/contigs/CMS001_029_Ra_S18/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_021_Ra_S16/contigs.fasta to ../../../../mnt/data/contigs/CMS001_021_Ra_S16/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_031_Ra_S19/contigs.fasta to ../../../../mnt/data/contigs/CMS001_031_Ra_S19/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_032_Ra_S7/contigs.fasta to ../../../../mnt/data/contigs/CMS001_032_Ra_S7/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_025_Ra_S7/contigs.fasta to ../../../../mnt/data/contigs/CMS001_025_Ra_S7/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_020_Ra_S15/contigs.fasta to ../../../../mnt/data/contigs/CMS001_020_Ra_S15/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_026_Ra_S18/contigs.fasta to ../../../../mnt/data/contigs/CMS001_026_Ra_S18/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_034_Ra_S19/contigs.fasta to ../../../../mnt/data/contigs/CMS001_034_Ra_S19/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_039_Ra_S9/contigs.fasta to ../../../../mnt/data/contigs/CMS001_039_Ra_S9/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_043_Ra_S24/contigs.fasta to ../../../../mnt/data/contigs/CMS001_043_Ra_S24/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_041_Ra_S10/contigs.fasta to ../../../../mnt/data/contigs/CMS001_041_Ra_S10/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_040_Ra_S21/contigs.fasta to ../../../../mnt/data/contigs/CMS001_040_Ra_S21/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_038_Ra_S22/contigs.fasta to ../../../../mnt/data/contigs/CMS001_038_Ra_S22/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_036_Ra_S20/contigs.fasta to ../../../../mnt/data/contigs/CMS001_036_Ra_S20/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_035_Ra_S20/contigs.fasta to ../../../../mnt/data/contigs/CMS001_035_Ra_S20/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_033_Ra_S8/contigs.fasta to ../../../../mnt/data/contigs/CMS001_033_Ra_S8/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_045_Ra_S2/contigs.fasta to ../../../../mnt/data/contigs/CMS001_045_Ra_S2/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_044_Ra_S25/contigs.fasta to ../../../../mnt/data/contigs/CMS001_044_Ra_S25/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_047_Ra_S4/contigs.fasta to ../../../../mnt/data/contigs/CMS001_047_Ra_S4/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_048_Ra_S5/contigs.fasta to ../../../../mnt/data/contigs/CMS001_048_Ra_S5/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_046_Ra_S3/contigs.fasta to ../../../../mnt/data/contigs/CMS001_046_Ra_S3/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_042_Ra_S23/contigs.fasta to ../../../../mnt/data/contigs/CMS001_042_Ra_S23/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_049_Ra_S6/contigs.fasta to ../../../../mnt/data/contigs/CMS001_049_Ra_S6/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_037_Ra_S21/contigs.fasta to ../../../../mnt/data/contigs/CMS001_037_Ra_S21/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_053_Ra_S8/contigs.fasta to ../../../../mnt/data/contigs/CMS001_053_Ra_S8/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_052_Ra_S7/contigs.fasta to ../../../../mnt/data/contigs/CMS001_052_Ra_S7/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_055_Ra_S9/contigs.fasta to ../../../../mnt/data/contigs/CMS001_055_Ra_S9/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_059_Ra_S10/contigs.fasta to ../../../../mnt/data/contigs/CMS001_059_Ra_S10/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_water5_RNA_A_S12/contigs.fasta to ../../../../mnt/data/contigs/CMS001_water5_RNA_A_S12/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_056_Ra_S10/contigs.fasta to ../../../../mnt/data/contigs/CMS001_056_Ra_S10/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_054_Ra_S11/contigs.fasta to ../../../../mnt/data/contigs/CMS001_054_Ra_S11/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_water1_S11/contigs.fasta to ../../../../mnt/data/contigs/CMS001_water1_S11/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_water4_Zymo_S27/contigs.fasta to ../../../../mnt/data/contigs/CMS001_water4_Zymo_S27/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_004a_Rb_S117_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_004a_Rb_S117_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_050_Ra_S23/contigs.fasta to ../../../../mnt/data/contigs/CMS001_050_Ra_S23/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_water3_Qiagen_S26/contigs.fasta to ../../../../mnt/data/contigs/CMS001_water3_Qiagen_S26/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_water2_S24/contigs.fasta to ../../../../mnt/data/contigs/CMS001_water2_S24/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_057_Ra_S11/contigs.fasta to ../../../../mnt/data/contigs/CMS001_057_Ra_S11/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_058_Ra_S9/contigs.fasta to ../../../../mnt/data/contigs/CMS001_058_Ra_S9/contigs.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://czbiohub-mosquito/contigs/CMS002_013a_Rb_S120_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_013a_Rb_S120_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_016a_Rb_S121_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_016a_Rb_S121_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_060_Ra_S12/contigs.fasta to ../../../../mnt/data/contigs/CMS001_060_Ra_S12/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS001_051_Ra_S8/contigs.fasta to ../../../../mnt/data/contigs/CMS001_051_Ra_S8/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_017b_Rb_S123_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_017b_Rb_S123_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_017c_Rb_S124_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_017c_Rb_S124_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_018a_Rb_S128_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_018a_Rb_S128_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_010a_Rb_S119_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_010a_Rb_S119_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_017e_Rb_S126_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_017e_Rb_S126_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_001a_Rb_S116_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_001a_Rb_S116_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_017a_Rb_S122_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_017a_Rb_S122_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_020b_Rb_S132_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_020b_Rb_S132_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_017d_Rb_S125_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_017d_Rb_S125_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_022a_Rb_S137_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_022a_Rb_S137_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_019a_Rb_S130_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_019a_Rb_S130_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_007a_Rb_S118_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_007a_Rb_S118_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_025a_Rb_S140_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_025a_Rb_S140_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_020a_Rb_S131_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_020a_Rb_S131_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_025c_Rb_S142_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_025c_Rb_S142_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_018b_Rb_S129_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_018b_Rb_S129_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_026a_Rb_S146_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_026a_Rb_S146_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_025d_Rb_S143_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_025d_Rb_S143_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_025e_Rb_S144_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_025e_Rb_S144_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_025f_Rb_S145_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_025f_Rb_S145_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_025b_Rb_S141_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_025b_Rb_S141_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_023a_Rb_S138_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_023a_Rb_S138_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_020c_Rb_S133_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_020c_Rb_S133_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_026d_Rb_S149_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_026d_Rb_S149_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_026e_Rb_S150_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_026e_Rb_S150_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_020e_Rb_S135_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_020e_Rb_S135_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_020d_Rb_S134_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_020d_Rb_S134_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_026c_Rb_S148_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_026c_Rb_S148_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_027b_Rb_S153_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_027b_Rb_S153_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_026b_Rb_S147_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_026b_Rb_S147_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_027a_Rb_S152_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_027a_Rb_S152_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_028a_Rb_S154_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_028a_Rb_S154_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_021a_Rb_S136_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_021a_Rb_S136_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_028c_Rb_S156_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_028c_Rb_S156_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_028d_Rb_S157_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_028d_Rb_S157_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_029c_Rb_S161_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_029c_Rb_S161_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_029b_Rb_S160_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_029b_Rb_S160_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_028b_Rb_S155_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_028b_Rb_S155_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_029a_Rb_S159_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_029a_Rb_S159_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_034a_Rb_S168_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_034a_Rb_S168_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_029d_Rb_S162_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_029d_Rb_S162_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_033a_Rb_S167_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_033a_Rb_S167_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_029e_Rb_S164_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_029e_Rb_S164_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_028e_Rb_S158_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_028e_Rb_S158_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_031a_Rb_S165_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_031a_Rb_S165_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_044c_Rb_S180_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_044c_Rb_S180_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_041a_Rb_S176_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_041a_Rb_S176_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_035a_Rb_S169_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_035a_Rb_S169_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_037a_Rb_S171_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_037a_Rb_S171_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_036a_Rb_S170_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_036a_Rb_S170_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_042a_Rb_S177_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_042a_Rb_S177_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_044b_Rb_S179_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_044b_Rb_S179_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_040a_Rb_S174_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_040a_Rb_S174_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_044d_Rb_S181_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_044d_Rb_S181_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_044a_Rb_S178_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_044a_Rb_S178_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_045g_Rb_S190_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045g_Rb_S190_L004/contigs.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download: s3://czbiohub-mosquito/contigs/CMS002_045f_Rb_S189_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045f_Rb_S189_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_044e_Rb_S182_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_044e_Rb_S182_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_045b_Rb_S184_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045b_Rb_S184_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_045c_Rb_S185_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045c_Rb_S185_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_045a_Rb_S183_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045a_Rb_S183_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_038a_Rb_S172_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_038a_Rb_S172_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_045d_Rb_S186_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045d_Rb_S186_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_045e_Rb_S188_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_045e_Rb_S188_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_039a_Rb_S173_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_039a_Rb_S173_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_046a_Rb_S191_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_046a_Rb_S191_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047b_Rb_S194_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047b_Rb_S194_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_032a_Rb_S166_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_032a_Rb_S166_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047f_Rb_S198_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047f_Rb_S198_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_046b_Rb_S192_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_046b_Rb_S192_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047g_Rb_S200_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047g_Rb_S200_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047h_Rb_S1_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047h_Rb_S1_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047a_Rb_S193_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047a_Rb_S193_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_050a_Rb_S5_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_050a_Rb_S5_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_054a_Rb_S8_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_054a_Rb_S8_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047e_Rb_S197_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047e_Rb_S197_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_049a_Rb_S4_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_049a_Rb_S4_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047c_Rb_S195_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047c_Rb_S195_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047d_Rb_S196_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047d_Rb_S196_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_057a_Rb_S10_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_057a_Rb_S10_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047i_Rb_S2_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047i_Rb_S2_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water8_Rb_S11_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water8_Rb_S11_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water3_Rb_S151_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water3_Rb_S151_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water5_Rb_S175_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water5_Rb_S175_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water7_Rb_S199_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water7_Rb_S199_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water2_Rb_S139_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water2_Rb_S139_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_051a_Rb_S6_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_051a_Rb_S6_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_047j_Rb_S3_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_047j_Rb_S3_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water1_Rb_S127_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water1_Rb_S127_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water6_Rb_S187_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water6_Rb_S187_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_0Water4_Rb_S163_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_0Water4_Rb_S163_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_053a_Rb_S7_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_053a_Rb_S7_L004/contigs.fasta\n",
      "download: s3://czbiohub-mosquito/contigs/CMS002_056a_Rb_S9_L004/contigs.fasta to ../../../../mnt/data/contigs/CMS002_056a_Rb_S9_L004/contigs.fasta\n"
     ]
    }
   ],
   "source": [
    "!aws s3 sync s3://czbiohub-mosquito/contigs/ /mnt/data/contigs --exclude \"*\" --include \"*.fasta\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To download the contigs, we will sync to\n",
    "\n",
    "`s3://czbiohub-mosquito/contigs/SAMPLE/contigs.fasta`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To setup blast db, follow https://czbiohub.atlassian.net/wiki/spaces/DS/pages/903905690/nt+nr+BLAST+etc+on+EC2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = !ls /mnt/data/contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_samples = ['CMS002_045c_Rb_S185_L004']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NT: Beginning sample  CMS001_001_Ra_S1\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    print(\"NT: Beginning sample \", sample)\n",
    "    \n",
    "    !BLASTDB=/mnt/data/blast blastn -db nt -num_threads 48 \\\n",
    "         -query /mnt/data/contigs/{sample}/contigs.fasta -outfmt \"7 std staxid ssciname scomname stitle\" \\\n",
    "         -out /mnt/data/contigs/{sample}/blast_nt.m9 -evalue 1e-1\n",
    "\n",
    "for sample in samples:\n",
    "    print(\"NR: Beginning sample \", sample)\n",
    "\n",
    "    !BLASTDB=/mnt/data/blast blastx -db nr -num_threads 48 \\\n",
    "         -query /mnt/data/contigs/{sample}/contigs.fasta -outfmt \"7 std staxid ssciname scomname stitle\" \\\n",
    "         -out /mnt/data/contigs/{sample}/blast_nr.m9 -evalue 1e-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "1=-outfmt \"$BLAST_OUTFMT\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"7 std staxid ssciname scomname stitle\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\r\n",
      "  blastx [-h] [-help] [-import_search_strategy filename]\r\n",
      "    [-export_search_strategy filename] [-task task_name] [-db database_name]\r\n",
      "    [-dbsize num_letters] [-gilist filename] [-seqidlist filename]\r\n",
      "    [-negative_gilist filename] [-negative_seqidlist filename]\r\n",
      "    [-entrez_query entrez_query] [-db_soft_mask filtering_algorithm]\r\n",
      "    [-db_hard_mask filtering_algorithm] [-subject subject_input_file]\r\n",
      "    [-subject_loc range] [-query input_file] [-out output_file]\r\n",
      "    [-evalue evalue] [-word_size int_value] [-gapopen open_penalty]\r\n",
      "    [-gapextend extend_penalty] [-qcov_hsp_perc float_value]\r\n",
      "    [-max_hsps int_value] [-xdrop_ungap float_value] [-xdrop_gap float_value]\r\n",
      "    [-xdrop_gap_final float_value] [-searchsp int_value]\r\n",
      "    [-sum_stats bool_value] [-max_intron_length length] [-seg SEG_options]\r\n",
      "    [-soft_masking soft_masking] [-matrix matrix_name]\r\n",
      "    [-threshold float_value] [-culling_limit int_value]\r\n",
      "    [-best_hit_overhang float_value] [-best_hit_score_edge float_value]\r\n",
      "    [-window_size int_value] [-ungapped] [-lcase_masking] [-query_loc range]\r\n",
      "    [-strand strand] [-parse_deflines] [-query_gencode int_value]\r\n",
      "    [-outfmt format] [-show_gis] [-num_descriptions int_value]\r\n",
      "    [-num_alignments int_value] [-line_length line_length] [-html]\r\n",
      "    [-max_target_seqs num_sequences] [-num_threads int_value] [-remote]\r\n",
      "    [-comp_based_stats compo] [-use_sw_tback] [-version]\r\n",
      "\r\n",
      "DESCRIPTION\r\n",
      "   Translated Query-Protein Subject BLAST 2.7.1+\r\n",
      "\r\n",
      "OPTIONAL ARGUMENTS\r\n",
      " -h\r\n",
      "   Print USAGE and DESCRIPTION;  ignore all other parameters\r\n",
      " -help\r\n",
      "   Print USAGE, DESCRIPTION and ARGUMENTS; ignore all other parameters\r\n",
      " -version\r\n",
      "   Print version number;  ignore other arguments\r\n",
      "\r\n",
      " *** Input query options\r\n",
      " -query <File_In>\r\n",
      "   Input file name\r\n",
      "   Default = `-'\r\n",
      " -query_loc <String>\r\n",
      "   Location on the query sequence in 1-based offsets (Format: start-stop)\r\n",
      " -strand <String, `both', `minus', `plus'>\r\n",
      "   Query strand(s) to search against database/subject\r\n",
      "   Default = `both'\r\n",
      " -query_gencode <Integer, values between: 1-6, 9-16, 21-25>\r\n",
      "   Genetic code to use to translate query (see user manual for details)\r\n",
      "   Default = `1'\r\n",
      "\r\n",
      " *** General search options\r\n",
      " -task <String, Permissible values: 'blastx' 'blastx-fast' >\r\n",
      "   Task to execute\r\n",
      "   Default = `blastx'\r\n",
      " -db <String>\r\n",
      "   BLAST database name\r\n",
      "    * Incompatible with:  subject, subject_loc\r\n",
      " -out <File_Out>\r\n",
      "   Output file name\r\n",
      "   Default = `-'\r\n",
      " -evalue <Real>\r\n",
      "   Expectation value (E) threshold for saving hits \r\n",
      "   Default = `10'\r\n",
      " -word_size <Integer, >=2>\r\n",
      "   Word size for wordfinder algorithm\r\n",
      " -gapopen <Integer>\r\n",
      "   Cost to open a gap\r\n",
      " -gapextend <Integer>\r\n",
      "   Cost to extend a gap\r\n",
      " -max_intron_length <Integer, >=0>\r\n",
      "   Length of the largest intron allowed in a translated nucleotide sequence\r\n",
      "   when linking multiple distinct alignments\r\n",
      "   Default = `0'\r\n",
      " -matrix <String>\r\n",
      "   Scoring matrix name (normally BLOSUM62)\r\n",
      " -threshold <Real, >=0>\r\n",
      "   Minimum word score such that the word is added to the BLAST lookup table\r\n",
      " -comp_based_stats <String>\r\n",
      "   Use composition-based statistics:\r\n",
      "       D or d: default (equivalent to 2 )\r\n",
      "       0 or F or f: No composition-based statistics\r\n",
      "       1: Composition-based statistics as in NAR 29:2994-3005, 2001\r\n",
      "       2 or T or t : Composition-based score adjustment as in Bioinformatics\r\n",
      "   21:902-911,\r\n",
      "       2005, conditioned on sequence properties\r\n",
      "       3: Composition-based score adjustment as in Bioinformatics 21:902-911,\r\n",
      "       2005, unconditionally\r\n",
      "   Default = `2'\r\n",
      "\r\n",
      " *** BLAST-2-Sequences options\r\n",
      " -subject <File_In>\r\n",
      "   Subject sequence(s) to search\r\n",
      "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\r\n",
      "   negative_seqidlist, db_soft_mask, db_hard_mask\r\n",
      " -subject_loc <String>\r\n",
      "   Location on the subject sequence in 1-based offsets (Format: start-stop)\r\n",
      "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\r\n",
      "   negative_seqidlist, db_soft_mask, db_hard_mask, remote\r\n",
      "\r\n",
      " *** Formatting options\r\n",
      " -outfmt <String>\r\n",
      "   alignment view options:\r\n",
      "     0 = Pairwise,\r\n",
      "     1 = Query-anchored showing identities,\r\n",
      "     2 = Query-anchored no identities,\r\n",
      "     3 = Flat query-anchored showing identities,\r\n",
      "     4 = Flat query-anchored no identities,\r\n",
      "     5 = BLAST XML,\r\n",
      "     6 = Tabular,\r\n",
      "     7 = Tabular with comment lines,\r\n",
      "     8 = Seqalign (Text ASN.1),\r\n",
      "     9 = Seqalign (Binary ASN.1),\r\n",
      "    10 = Comma-separated values,\r\n",
      "    11 = BLAST archive (ASN.1),\r\n",
      "    12 = Seqalign (JSON),\r\n",
      "    13 = Multiple-file BLAST JSON,\r\n",
      "    14 = Multiple-file BLAST XML2,\r\n",
      "    15 = Single-file BLAST JSON,\r\n",
      "    16 = Single-file BLAST XML2,\r\n",
      "    18 = Organism Report\r\n",
      "   \r\n",
      "   Options 6, 7 and 10 can be additionally configured to produce\r\n",
      "   a custom format specified by space delimited format specifiers.\r\n",
      "   The supported format specifiers are:\r\n",
      "   \t    qseqid means Query Seq-id\r\n",
      "   \t       qgi means Query GI\r\n",
      "   \t      qacc means Query accesion\r\n",
      "   \t   qaccver means Query accesion.version\r\n",
      "   \t      qlen means Query sequence length\r\n",
      "   \t    sseqid means Subject Seq-id\r\n",
      "   \t sallseqid means All subject Seq-id(s), separated by a ';'\r\n",
      "   \t       sgi means Subject GI\r\n",
      "   \t    sallgi means All subject GIs\r\n",
      "   \t      sacc means Subject accession\r\n",
      "   \t   saccver means Subject accession.version\r\n",
      "   \t   sallacc means All subject accessions\r\n",
      "   \t      slen means Subject sequence length\r\n",
      "   \t    qstart means Start of alignment in query\r\n",
      "   \t      qend means End of alignment in query\r\n",
      "   \t    sstart means Start of alignment in subject\r\n",
      "   \t      send means End of alignment in subject\r\n",
      "   \t      qseq means Aligned part of query sequence\r\n",
      "   \t      sseq means Aligned part of subject sequence\r\n",
      "   \t    evalue means Expect value\r\n",
      "   \t  bitscore means Bit score\r\n",
      "   \t     score means Raw score\r\n",
      "   \t    length means Alignment length\r\n",
      "   \t    pident means Percentage of identical matches\r\n",
      "   \t    nident means Number of identical matches\r\n",
      "   \t  mismatch means Number of mismatches\r\n",
      "   \t  positive means Number of positive-scoring matches\r\n",
      "   \t   gapopen means Number of gap openings\r\n",
      "   \t      gaps means Total number of gaps\r\n",
      "   \t      ppos means Percentage of positive-scoring matches\r\n",
      "   \t    frames means Query and subject frames separated by a '/'\r\n",
      "   \t    qframe means Query frame\r\n",
      "   \t    sframe means Subject frame\r\n",
      "   \t      btop means Blast traceback operations (BTOP)\r\n",
      "   \t    staxid means Subject Taxonomy ID\r\n",
      "   \t  ssciname means Subject Scientific Name\r\n",
      "   \t  scomname means Subject Common Name\r\n",
      "   \tsblastname means Subject Blast Name\r\n",
      "   \t sskingdom means Subject Super Kingdom\r\n",
      "   \t   staxids means unique Subject Taxonomy ID(s), separated by a ';'\r\n",
      "   \t\t\t (in numerical order)\r\n",
      "   \t sscinames means unique Subject Scientific Name(s), separated by a ';'\r\n",
      "   \t scomnames means unique Subject Common Name(s), separated by a ';'\r\n",
      "   \tsblastnames means unique Subject Blast Name(s), separated by a ';'\r\n",
      "   \t\t\t (in alphabetical order)\r\n",
      "   \tsskingdoms means unique Subject Super Kingdom(s), separated by a ';'\r\n",
      "   \t\t\t (in alphabetical order) \r\n",
      "   \t    stitle means Subject Title\r\n",
      "   \tsalltitles means All Subject Title(s), separated by a '<>'\r\n",
      "   \t   sstrand means Subject Strand\r\n",
      "   \t     qcovs means Query Coverage Per Subject\r\n",
      "   \t   qcovhsp means Query Coverage Per HSP\r\n",
      "   \t    qcovus means Query Coverage Per Unique Subject (blastn only)\r\n",
      "   When not provided, the default value is:\r\n",
      "   'qaccver saccver pident length mismatch gapopen qstart qend sstart send\r\n",
      "   evalue bitscore', which is equivalent to the keyword 'std'\r\n",
      "   Default = `0'\r\n",
      " -show_gis\r\n",
      "   Show NCBI GIs in deflines?\r\n",
      " -num_descriptions <Integer, >=0>\r\n",
      "   Number of database sequences to show one-line descriptions for\r\n",
      "   Not applicable for outfmt > 4\r\n",
      "   Default = `500'\r\n",
      "    * Incompatible with:  max_target_seqs\r\n",
      " -num_alignments <Integer, >=0>\r\n",
      "   Number of database sequences to show alignments for\r\n",
      "   Default = `250'\r\n",
      "    * Incompatible with:  max_target_seqs\r\n",
      " -line_length <Integer, >=1>\r\n",
      "   Line length for formatting alignments\r\n",
      "   Not applicable for outfmt > 4\r\n",
      "   Default = `60'\r\n",
      " -html\r\n",
      "   Produce HTML output?\r\n",
      "\r\n",
      " *** Query filtering options\r\n",
      " -seg <String>\r\n",
      "   Filter query sequence with SEG (Format: 'yes', 'window locut hicut', or\r\n",
      "   'no' to disable)\r\n",
      "   Default = `12 2.2 2.5'\r\n",
      " -soft_masking <Boolean>\r\n",
      "   Apply filtering locations as soft masks\r\n",
      "   Default = `false'\r\n",
      " -lcase_masking\r\n",
      "   Use lower case filtering in query and subject sequence(s)?\r\n",
      "\r\n",
      " *** Restrict search or results\r\n",
      " -gilist <String>\r\n",
      "   Restrict search of database to list of GI's\r\n",
      "    * Incompatible with:  negative_gilist, seqidlist, negative_seqidlist,\r\n",
      "   remote, subject, subject_loc\r\n",
      " -seqidlist <String>\r\n",
      "   Restrict search of database to list of SeqId's\r\n",
      "    * Incompatible with:  gilist, negative_gilist, negative_seqidlist, remote,\r\n",
      "   subject, subject_loc\r\n",
      " -negative_gilist <String>\r\n",
      "   Restrict search of database to everything except the listed GIs\r\n",
      "    * Incompatible with:  gilist, seqidlist, remote, subject, subject_loc\r\n",
      " -negative_seqidlist <String>\r\n",
      "   Restrict search of database to everything except the listed SeqIDs\r\n",
      "    * Incompatible with:  gilist, seqidlist, remote, subject, subject_loc\r\n",
      " -entrez_query <String>\r\n",
      "   Restrict search with the given Entrez query\r\n",
      "    * Requires:  remote\r\n",
      " -db_soft_mask <String>\r\n",
      "   Filtering algorithm ID to apply to the BLAST database as soft masking\r\n",
      "    * Incompatible with:  db_hard_mask, subject, subject_loc\r\n",
      " -db_hard_mask <String>\r\n",
      "   Filtering algorithm ID to apply to the BLAST database as hard masking\r\n",
      "    * Incompatible with:  db_soft_mask, subject, subject_loc\r\n",
      " -qcov_hsp_perc <Real, 0..100>\r\n",
      "   Percent query coverage per hsp\r\n",
      " -max_hsps <Integer, >=1>\r\n",
      "   Set maximum number of HSPs per subject sequence to save for each query\r\n",
      " -culling_limit <Integer, >=0>\r\n",
      "   If the query range of a hit is enveloped by that of at least this many\r\n",
      "   higher-scoring hits, delete the hit\r\n",
      "    * Incompatible with:  best_hit_overhang, best_hit_score_edge\r\n",
      " -best_hit_overhang <Real, (>0 and <0.5)>\r\n",
      "   Best Hit algorithm overhang value (recommended value: 0.1)\r\n",
      "    * Incompatible with:  culling_limit\r\n",
      " -best_hit_score_edge <Real, (>0 and <0.5)>\r\n",
      "   Best Hit algorithm score edge value (recommended value: 0.1)\r\n",
      "    * Incompatible with:  culling_limit\r\n",
      " -max_target_seqs <Integer, >=1>\r\n",
      "   Maximum number of aligned sequences to keep \r\n",
      "   Not applicable for outfmt <= 4\r\n",
      "   Default = `500'\r\n",
      "    * Incompatible with:  num_descriptions, num_alignments\r\n",
      "\r\n",
      " *** Statistical options\r\n",
      " -dbsize <Int8>\r\n",
      "   Effective length of the database \r\n",
      " -searchsp <Int8, >=0>\r\n",
      "   Effective length of the search space\r\n",
      " -sum_stats <Boolean>\r\n",
      "   Use sum statistics\r\n",
      "\r\n",
      " *** Search strategy options\r\n",
      " -import_search_strategy <File_In>\r\n",
      "   Search strategy to use\r\n",
      "    * Incompatible with:  export_search_strategy\r\n",
      " -export_search_strategy <File_Out>\r\n",
      "   File name to record the search strategy used\r\n",
      "    * Incompatible with:  import_search_strategy\r\n",
      "\r\n",
      " *** Extension options\r\n",
      " -xdrop_ungap <Real>\r\n",
      "   X-dropoff value (in bits) for ungapped extensions\r\n",
      " -xdrop_gap <Real>\r\n",
      "   X-dropoff value (in bits) for preliminary gapped extensions\r\n",
      " -xdrop_gap_final <Real>\r\n",
      "   X-dropoff value (in bits) for final gapped alignment\r\n",
      " -window_size <Integer, >=0>\r\n",
      "   Multiple hits window size, use 0 to specify 1-hit algorithm\r\n",
      " -ungapped\r\n",
      "   Perform ungapped alignment only?\r\n",
      "\r\n",
      " *** Miscellaneous options\r\n",
      " -parse_deflines\r\n",
      "   Should the query and subject defline(s) be parsed?\r\n",
      " -num_threads <Integer, (>=1 and =<48)>\r\n",
      "   Number of threads (CPUs) to use in the BLAST search\r\n",
      "   Default = `1'\r\n",
      "    * Incompatible with:  remote\r\n",
      " -remote\r\n",
      "   Execute search remotely?\r\n",
      "    * Incompatible with:  gilist, seqidlist, negative_gilist,\r\n",
      "   negative_seqidlist, subject_loc, num_threads\r\n",
      " -use_sw_tback\r\n",
      "   Compute locally optimal Smith-Waterman alignments?\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!blastx -help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!aws s3 sync /mnt/data/contigs/ s3://czbiohub-mosquito/contigs/ --exclude \"*\" --include \"*_nt.m9\" --dryrun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    !cat /mnt/data/contigs/{sample}/contigs.fasta | sed 's/>N/>{samples[0]}~N/g' > /mnt/data/contigs/prefixed/{sample}_contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "160\r\n"
     ]
    }
   ],
   "source": [
    "!ls /mnt/data/contigs/prefixed | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat /mnt/data/contigs/prefixed/*.fasta > /mnt/data/contigs/all.fasta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">CMS001_001_Ra_S1~NODE_1_length_10868_cov_37.868316\r\n",
      "GATCTCTTGGTGACTGTTTTGTTGACAATTGCCAAGCCGGCGGTAGCTGTATTGTTACTC\r\n",
      "TATCGCTTCAAATCGTTCTCGTGACGCCACGCGAACAACCATGCCGACGTTTCCGAACCA\r\n",
      "ACGCAAAACCCTGGCATCTGCCAAGCCGTCGAGGACGCCCCGAGACGGTACTAAAGTCGC\r\n",
      "GGAGAATCGCGGGTATTGTTACCTCGCGCTCTTTGAGGCTTTAAACGCGTCGTCTGAGAA\r\n",
      "GAAGTTAGACGTTGCGTCAATCAAGGCTCGTTTGGGAGCCTTTCCCCTCGTGAGACGCGT\r\n",
      "TGTTGGGGAGTTGTACGCTCACGTGACATTTGACTTGTTTGTTCCCTGCGTGCGTAGGGT\r\n",
      "AAGCAATACGATGTTCCACGTGGACGAATGGCGCCCCCCAATGTTGTTCTCTGAGGTACT\r\n",
      "CGCGATGACGATCTTTTCGAGTGCGAGAATTGGTGCGGATGACCGTGCCCACTTGCAGCA\r\n",
      "GCAACAGCTGATCAGGGTGCAGGACTTATGTAAGACTGCAGGCCTGGACCGCAACACCGT\r\n"
     ]
    }
   ],
   "source": [
    "!head /mnt/data/contigs/all.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!/home/ubuntu/plastbinary_linux_20160121/plast -p plastx \\\n",
    "    -i /mnt/data/contigs/all.fasta \\\n",
    "    -d /mnt/data/blast/nr.pal \\\n",
    "    -o /mnt/data/plast_output.tab \\\n",
    "    -e 1e-2 -a 48 -max-hit-per-query 30 -outfmt 1 \\\n",
    "    -bargraph -verbose \\\n",
    "    -max-database-size 200000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plast 2.3.1\r\n",
      "  - Build date: 2016-01-21 13:53:54\r\n",
      "  - OS: Linux-2.6.32-279.el6.x86_64\r\n",
      "  - Compiler: /usr/bin/gcc (4.4.6)\r\n",
      "  - Host CPU: 48 cores available\r\n",
      "\r\n",
      "[*] denotes mandatory argument.\r\n",
      "\t-p [*]:\t Program Name [plastp, tplastn, plastx, tplastx or plastn]\r\n",
      "\t-d [*]:\t Subject database file\r\n",
      "\t-i [*]:\t Query database file\r\n",
      "\t-o :\t PLAST report Output File\r\n",
      "\t-e :\t Expectation value\r\n",
      "\t-n :\t Size of neighbourhood peforming ungapped extension\r\n",
      "\t-s :\t Ungapped threshold trigger a small gapped extension\r\n",
      "\t-g :\t threshold for small gapped extension\r\n",
      "\t-b :\t bandwith for small gapped extension\r\n",
      "\t-a :\t Number of processors to use\r\n",
      "\t-G :\t Cost to open a gap\r\n",
      "\t-E :\t Cost to extend a gap\r\n",
      "\t-xdrop-ungap :\t X dropoff value for Ungapped alignment (in bits) (zero invokes default behavior 20 bits)\r\n",
      "\t-X :\t X dropoff value for gapped alignment (in bits) (zero invokes default behavior)\r\n",
      "\t-Z :\t X dropoff value for final gapped alignment in bits (0.0 invokes default behavior)\r\n",
      "\t-index-threshold :\t Index threshold to calculate the similarity between neighbour\r\n",
      "\t-F :\t Filter query sequence\r\n",
      "\t-M :\t Score matrix (BLOSUM62 or BLOSUM50)\r\n",
      "\t-strand :\t strands for plastn: 'plus', 'minus' or 'both' (default)\r\n",
      "\t-r :\t reward for a nucleotide match (plastn)\r\n",
      "\t-q :\t penalty for a nucleotide mismatch (plastn)\r\n",
      "\t-force-query-order :\t Force queries ordering in output file.\r\n",
      "\t-max-database-size :\t Maximum allowed size (in bytes) for a database. If greater, database is segmented.\r\n",
      "\t-max-hit-per-query :\t Maximum hits per query. 0 value will dump all hits (default)\r\n",
      "\t-max-hsp-per-hit :\t Maximum alignments per hit. 0 value will dump all hits (default)\r\n",
      "\t-outfmt :\t Output format: 1 for tabulated (default), 2 for extended tubulated, 4 for NCBI Blast-like.\r\n",
      "\t-strands-list :\t List of the strands (ex: \"1,2,6\") to be used when using algo using nucleotids databases.\r\n",
      "\t-optim-codon-stop :\t size of the allowed range between the last invalid character and the next stop codon\r\n",
      "\t-factory-dispatcher :\t Factory that creates dispatcher.\r\n",
      "\t-factory-statistics :\t Factory that creates statistics builder.\r\n",
      "\t-factory-indexation :\t Factory that creates indexation builder.\r\n",
      "\t-factory-hit-ungap :\t Factory that creates ungap hits iterator.\r\n",
      "\t-factory-hit-smallgap :\t Factory that creates small gap hits iterator.\r\n",
      "\t-factory-hit-fullgap :\t Factory that creates full gap hits iterator.\r\n",
      "\t-factory-hit-composition :\t Factory that creates composition hits iterator.\r\n",
      "\t-factory-gap-result :\t Factory that creates gap alignments result.\r\n",
      "\t-factory-ungap-result :\t Factory that creates ungap alignments result.\r\n",
      "\t-splitter :\t Factory that creates an alignment splitter. String: 'normal' or 'banded' (default)\r\n",
      "\t-optim-filter-ungap :\t Optimization that filters out through ungap alignments.\r\n",
      "\t-bargraph :\t Display a progress bar during execution.\r\n",
      "\t-bargraph-size :\t Nb of characters of the bargraph.\r\n",
      "\t-progression-file :\t Dump in a file the current execution percentage.\r\n",
      "\t-verbose :\t Display information during algorithm execution.\r\n",
      "\t-full-stats :\t Dump algorithm statistics.\r\n",
      "\t-stats :\t Dump generic statistics.\r\n",
      "\t-stats-fmt :\t Format of statistics: 'raw' (default) or 'xml'\r\n",
      "\t-stats-auto :\t Automatic stats file creation\r\n",
      "\t-alignment-progress :\t Dump in a file the growing number of ungap/ungap alignments during algorithm.\r\n",
      "\t-resources-progress :\t Dump in a file information about resources during algorithm.\r\n",
      "\t-plastrc :\t Pathname of the plast config file.\r\n",
      "\t-xmlfilter :\t Uri of a XML filter file.\r\n",
      "\t-seeds-use-ratio :\t Ratio of seeds to be used.\r\n",
      "\t-seeds-index-filter :\t seeds length to be used for the indexation filter.\r\n",
      "\t-complete-subject-database-stats-file :\t File path to the stats of the complete subject database\r\n",
      "\t-W :\t size of the seeds\r\n",
      "\t-h :\t help\r\n",
      "Citing PLAST: Nguyen VH, Lavenier D. (2009) PLAST: parallel local alignment search tool for database comparison. BMC Bioinformatics, vol 10, no 329.\r\n"
     ]
    }
   ],
   "source": [
    "!/home/ubuntu/plastbinary_linux_20160121/plast -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
