{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Building PSSM profiles from alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-10T10:56:49.167536Z",
     "start_time": "2020-01-10T10:56:19.911949Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ArenaRdRp_PF06317_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             ArenaRdRp_PF06317_full.fasta\n",
      "# output HMM file:                  ArenaRdRp_PF06317_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     ArenaRdRp_PF06317_full    16  2640  1096     1.62  0.591 \n",
      "\n",
      "# CPU time: 2.29u 0.03s 00:00:02.31 Elapsed: 00:00:02.79\n",
      "BirnaRdRp_PF04197_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             BirnaRdRp_PF04197_full.fasta\n",
      "# output HMM file:                  BirnaRdRp_PF04197_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     BirnaRdRp_PF04197_full     8  1078   846     1.04  0.590 \n",
      "\n",
      "# CPU time: 1.58u 0.02s 00:00:01.60 Elapsed: 00:00:01.95\n",
      "BromoTobamoTogaRdRp_PF00978_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             BromoTobamoTogaRdRp_PF00978_full.fasta\n",
      "# output HMM file:                  BromoTobamoTogaRdRp_PF00978_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     BromoTobamoTogaRdRp_PF00978_full   367  1199   311     8.55  0.590 \n",
      "\n",
      "# CPU time: 0.94u 0.01s 00:00:00.95 Elapsed: 00:00:01.75\n",
      "BunyaRdRp_PF04196_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             BunyaRdRp_PF04196_full.fasta\n",
      "# output HMM file:                  BunyaRdRp_PF04196_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     BunyaRdRp_PF04196_full   157  1488   708     8.02  0.590 \n",
      "\n",
      "# CPU time: 1.69u 0.02s 00:00:01.71 Elapsed: 00:00:02.24\n",
      "FlaviRdRp_PF00972_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             FlaviRdRp_PF00972_full.fasta\n",
      "# output HMM file:                  FlaviRdRp_PF00972_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     FlaviRdRp_PF00972_full    15   844   549     1.83  0.590 \n",
      "\n",
      "# CPU time: 1.27u 0.02s 00:00:01.29 Elapsed: 00:00:01.62\n",
      "LuteoTotiRotaRdRp_PF02123_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             LuteoTotiRotaRdRp_PF02123_full.fasta\n",
      "# output HMM file:                  LuteoTotiRotaRdRp_PF02123_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     LuteoTotiRotaRdRp_PF02123_full   200  1340   359    12.25  0.590 \n",
      "\n",
      "# CPU time: 0.94u 0.01s 00:00:00.95 Elapsed: 00:00:01.05\n",
      "MitoRdRp_PF05919_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             MitoRdRp_PF05919_full.fasta\n",
      "# output HMM file:                  MitoRdRp_PF05919_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     MitoRdRp_PF05919_full   141   989   240     4.71  0.590 \n",
      "\n",
      "# CPU time: 0.52u 0.00s 00:00:00.52 Elapsed: 00:00:00.55\n",
      "MononegaRdRp_PF00946_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             MononegaRdRp_PF00946_full.fasta\n",
      "# output HMM file:                  MononegaRdRp_PF00946_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     MononegaRdRp_PF00946_full   356  2584   849     7.43  0.590 \n",
      "\n",
      "# CPU time: 2.44u 0.03s 00:00:02.46 Elapsed: 00:00:02.88\n",
      "OrthomyxoRdRp_PF00602_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             OrthomyxoRdRp_PF00602_full.fasta\n",
      "# output HMM file:                  OrthomyxoRdRp_PF00602_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     OrthomyxoRdRp_PF00602_full    59   878   700     1.96  0.590 \n",
      "\n",
      "# CPU time: 1.62u 0.03s 00:00:01.65 Elapsed: 00:00:02.19\n",
      "Partitiviridae_ICTV_Mafft_211216_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             Partitiviridae_ICTV_Mafft_211216_full.fasta\n",
      "# output HMM file:                  Partitiviridae_ICTV_Mafft_211216_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Partitiviridae_ICTV_Mafft_211216_full    45   907   623     4.11  0.590 \n",
      "\n",
      "# CPU time: 1.48u 0.03s 00:00:01.51 Elapsed: 00:00:02.15\n",
      "PicornaRdRp_PF00680_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             PicornaRdRp_PF00680_full.fasta\n",
      "# output HMM file:                  PicornaRdRp_PF00680_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     PicornaRdRp_PF00680_full   704  1806   398    13.75  0.590 \n",
      "\n",
      "# CPU time: 1.34u 0.03s 00:00:01.37 Elapsed: 00:00:02.17\n",
      "ReoRdRp_PF07925_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             ReoRdRp_PF07925_full.fasta\n",
      "# output HMM file:                  ReoRdRp_PF07925_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     ReoRdRp_PF07925_full    16  1360  1252     1.15  0.590 \n",
      "\n",
      "# CPU time: 2.50u 0.03s 00:00:02.52 Elapsed: 00:00:03.30\n",
      "Reo_alternative_full.fasta\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             Reo_alternative_full.fasta\n",
      "# output HMM file:                  Reo_alternative_pssm.txt\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Reo_alternative_full    50  1792  1229     4.57  0.590 \n",
      "\n",
      "# CPU time: 2.71u 0.04s 00:00:02.75 Elapsed: 00:00:04.08\n",
      "Working...    done.\n",
      "Pressed and indexed 13 HMMs (13 names).\n",
      "Models pressed into binary file:   RdRp_profiles.pssm.txt.h3m\n",
      "SSI index for binary model file:   RdRp_profiles.pssm.txt.h3i\n",
      "Profiles (MSV part) pressed into:  RdRp_profiles.pssm.txt.h3f\n",
      "Profiles (remainder) pressed into: RdRp_profiles.pssm.txt.h3p\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "base_path=/Users/evogytis/Documents/manuscripts/skeeters/data/darkmatter/RdRp_profiles ## where alignments are kept\n",
    "cd $base_path\n",
    "\n",
    "for msa in *full.fasta ## iterate over alignments of each Pfam group\n",
    "    do\n",
    "    echo $msa\n",
    "    out=${msa/_full.fasta/_pssm.txt}\n",
    "    hmmbuild $out $msa ## build PSSM\n",
    "    \n",
    "done;\n",
    "\n",
    "cat *_pssm.txt > RdRp_profiles.pssm.txt ## concatenate Pfam PSSMs into a single file\n",
    "hmmpress -f RdRp_profiles.pssm.txt ## compress for search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract representative contigs of each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-10T10:57:06.934824Z",
     "start_time": "2020-01-10T10:56:49.172183Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3857 failed, probably because longest contig is from water\n",
      "44862 failed, probably because longest contig is from water\n",
      "68543 failed, probably because longest contig is from water\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict, namedtuple\n",
    "import os,re,glob\n",
    "from Bio import SeqIO\n",
    "\n",
    "Member = namedtuple('Member', ['contig', 'length', 'percent_id', 'percent_id_sign', 'sample',\n",
    "                               'coverage', 'uuid', 'reads'])\n",
    "\n",
    "base_path='/Users/evogytis/Documents/manuscripts/skeeters/data/'\n",
    "\n",
    "# Use abundance information from mapping reads back to contigs\n",
    "count_df = pd.read_csv('/Users/evogytis/Downloads/contig_counts.csv')\n",
    "count_df['uuid'] = count_df['sample'] + '~' + count_df['contig']\n",
    "count_lookup = {uuid: counts for uuid, counts in zip(count_df['uuid'], count_df['counts'])}\n",
    "\n",
    "def parse_cdhit_row(row):\n",
    "    if '*' in row:\n",
    "        index, length, name, percent_id = row.split()\n",
    "        percent_id_sign, percent_id = '0', 100\n",
    "    else:\n",
    "        index, length, name, _, percent_id = row.split()\n",
    "    length = int(length.strip(',nt'))\n",
    "    name = name.strip('>').strip('.')\n",
    "    uuid = name\n",
    "    sample, contig = name.split('~')\n",
    "    coverage = float(contig.split('_')[-1])\n",
    "    reads = count_lookup[uuid] if uuid in count_lookup else 0\n",
    "    \n",
    "    if percent_id != 100:\n",
    "        percent_id_sign, percent_id = percent_id.strip('%').split('/')\n",
    "        percent_id = float(percent_id)\n",
    "        \n",
    "    return Member(contig=contig, sample=sample, length=length,\n",
    "                  percent_id=percent_id, percent_id_sign=percent_id_sign, coverage=coverage,\n",
    "                  uuid=uuid, reads=reads)\n",
    "\n",
    "clusters = defaultdict(list)\n",
    "with open(os.path.join(base_path,'500_contigs_cluster.clstr'), 'r') as file:\n",
    "    for line in file:\n",
    "        if line.startswith('>Cluster'):\n",
    "            cluster_id = line.split()[-1]\n",
    "        else:\n",
    "            member = parse_cdhit_row(line)\n",
    "            if 'water' in member.sample.lower():\n",
    "                continue\n",
    "            clusters[cluster_id].append(member)\n",
    "            \n",
    "def get_representative(cluster):\n",
    "    representative = [member for member in cluster if member.percent_id_sign == '0'][0]\n",
    "    return representative\n",
    "\n",
    "fetch={} ## cluster ID of representative contig\n",
    "for clustID in clusters:\n",
    "    try:\n",
    "        representative_contig=get_representative(clusters[clustID])\n",
    "        fetch[representative_contig.contig]=clustID\n",
    "    except:\n",
    "        print(clustID,'failed, probably because longest contig is from water')\n",
    "\n",
    "store_representatives=open(os.path.join(base_path,'darkmatter/representative_contigs.fasta'),'w')\n",
    "\n",
    "for contig_file in glob.glob(os.path.join(base_path,'s3/contigs','*','contigs.fasta')):\n",
    "    sample=os.path.basename(os.path.dirname(contig_file))\n",
    "    for seq in SeqIO.parse(open(contig_file,'r'),format='fasta'):\n",
    "        if seq.id in fetch:\n",
    "            seq_name=seq.id\n",
    "            cluster_id=fetch[seq_name]\n",
    "            seq=str(seq.seq)\n",
    "            \n",
    "            store_representatives.write('>%s|%s|%s\\n%s\\n'%(cluster_id,sample,seq_name,seq))\n",
    "            \n",
    "store_representatives.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-10T11:00:09.646984Z",
     "start_time": "2020-01-10T10:57:06.969505Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Bio/Seq.py:2715: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18340 82003 0.2236503542553321\n"
     ]
    }
   ],
   "source": [
    "longestORF=re.compile('[A-Z]+') ## longest ORF is delimited by *\n",
    "\n",
    "RdRp_motifs=re.compile('GDD|GDN|SDD|ADN') ## RdRp catalytic domain residues\n",
    "\n",
    "repr_contig_path=os.path.join(base_path,'darkmatter/representative_contigs.fasta') ## path to representative contigs\n",
    "RdRp_candidates=open(os.path.join(base_path,'darkmatter/RdRp_candidates.fasta'),'w') ## file where contigs likely to be polymerases will be stored\n",
    "RdRp_candidate_proteins=open(os.path.join(base_path,'darkmatter/RdRp_candidates_prot.fasta'),'w') ## file where proteins likely to be polymerases will be stored\n",
    "\n",
    "c=0\n",
    "t=0\n",
    "\n",
    "for seq in SeqIO.parse(open(repr_contig_path,'r'),format='fasta'): ## iterate over representative contigs\n",
    "    candidateORFs=[]\n",
    "    candidateProt=[]\n",
    "    \n",
    "    for s,sequence in enumerate([seq.seq,seq.seq.complement()[::-1]]): ## fwd+rev\n",
    "        for i in range(3): ## reading frames\n",
    "            \n",
    "            translation=str(sequence[i:].translate()) ## translate direction+frame\n",
    "            \n",
    "            for protein in longestORF.findall(translation): ## find all proteins\n",
    "#                 if RdRp_motifs.search(protein) and len(protein)>200: ## protein has RdRp-like motif and is longer than 200 amino acids\n",
    "                if len(protein)>200:\n",
    "                    candidateORFs.append(sequence) ## add sequence for later\n",
    "                    candidateProt.append(protein)\n",
    "                    \n",
    "    if len(candidateORFs)>0:\n",
    "        c+=1\n",
    "        RdRp_candidates.write('>%s\\n%s\\n'%(seq.id,seq.seq))\n",
    "        for p in range(len(candidateProt)):\n",
    "            RdRp_candidate_proteins.write('>%s|%s\\n%s\\n'%(seq.id,p,candidateProt[p]))\n",
    "    t+=1\n",
    "    \n",
    "RdRp_candidates.close()\n",
    "RdRp_candidate_proteins.close()\n",
    "\n",
    "print(c,t,c/t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-10T11:01:12.801595Z",
     "start_time": "2020-01-10T11:00:09.651126Z"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "profiledb=/Users/evogytis/Documents/manuscripts/skeeters/data/darkmatter/RdRp_profiles/RdRp_profiles.pssm.txt ## PSSMs of RdRps\n",
    "candidates=/Users/evogytis/Documents/manuscripts/skeeters/data/darkmatter/RdRp_candidates_prot.fasta ## proteins to be scanned with PSSMs\n",
    "\n",
    "out=${candidates/.fasta/.pssm.out}\n",
    "\n",
    "hmmscan --noali -o $out $profiledb $candidates ## scan away\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-01-10T11:01:14.170210Z",
     "start_time": "2020-01-10T11:01:12.805413Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1|CMS001_039_Ra_S9|NODE_1_length_18277_cov_48.170824\n",
      "FlaviRdRp_PF00972_full: 5.9e-27\n",
      "\n",
      "2|CMS002_032a_Rb_S166_L004|NODE_1_length_13844_cov_454.512457\n",
      "MononegaRdRp_PF00946_full: 1.4e-216\n",
      "\n",
      "3|CMS001_039_Ra_S9|NODE_2_length_13816_cov_21.969794\n",
      "contig NODE_2_length_13816_cov_21.969794 does not belong to a cluster\n",
      "MononegaRdRp_PF00946_full: 6.9e-217\n",
      "\n",
      "4|CMS002_053a_Rb_S7_L004|NODE_1_length_13637_cov_31.429499\n",
      "MononegaRdRp_PF00946_full: 2e-222\n",
      "\n",
      "5|CMS002_017c_Rb_S124_L004|NODE_1_length_12898_cov_39.149754\n",
      "MononegaRdRp_PF00946_full: 9.1e-163\n",
      "\n",
      "6|CMS001_008_Ra_S3|NODE_1_length_12382_cov_11.752133\n",
      "MononegaRdRp_PF00946_full: 6.8e-155\n",
      "\n",
      "7|CMS002_020c_Rb_S133_L004|NODE_1_length_12368_cov_173.470507\n",
      "MononegaRdRp_PF00946_full: 6.8e-155\n",
      "\n",
      "8|CMS002_035a_Rb_S169_L004|NODE_1_length_11789_cov_864.832650\n",
      "MononegaRdRp_PF00946_full: 1.3e-264\n",
      "\n",
      "9|CMS002_047g_Rb_S200_L004|NODE_1_length_11763_cov_406.947715\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6.4e-77\n",
      "PicornaRdRp_PF00680_full: 1.3e-07\n",
      "\n",
      "11|CMS002_029d_Rb_S162_L004|NODE_1_length_10957_cov_129.744853\n",
      "contig NODE_1_length_10957_cov_129.744853 does not belong to a cluster\n",
      "BromoTobamoTogaRdRp_PF00978_full: 5.9e-77\n",
      "PicornaRdRp_PF00680_full: 1.1e-07\n",
      "\n",
      "12|CMS001_029_Ra_S18|NODE_1_length_10946_cov_79.583034\n",
      "FlaviRdRp_PF00972_full: 8.6e-181\n",
      "PicornaRdRp_PF00680_full: 1.1e-12\n",
      "\n",
      "13|CMS002_047h_Rb_S1_L004|NODE_1_length_10922_cov_58.303550\n",
      "BromoTobamoTogaRdRp_PF00978_full: 4.1e-84\n",
      "\n",
      "15|CMS001_020_Ra_S15|NODE_1_length_10817_cov_49.966667\n",
      "contig NODE_1_length_10817_cov_49.966667 does not belong to a cluster\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6.3e-77\n",
      "PicornaRdRp_PF00680_full: 1.3e-07\n",
      "\n",
      "16|CMS001_032_Ra_S7|NODE_1_length_10763_cov_35.536028\n",
      "contig NODE_1_length_10763_cov_35.536028 does not belong to a cluster\n",
      "FlaviRdRp_PF00972_full: 1.4e-182\n",
      "PicornaRdRp_PF00680_full: 1.1e-11\n",
      "\n",
      "17|CMS001_048_Ra_S5|NODE_1_length_10707_cov_6.003951\n",
      "contig NODE_1_length_10707_cov_6.003951 does not belong to a cluster\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6.3e-77\n",
      "PicornaRdRp_PF00680_full: 1.2e-07\n",
      "\n",
      "18|CMS001_010_Ra_S1|NODE_1_length_10647_cov_8.761022\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6.2e-77\n",
      "PicornaRdRp_PF00680_full: 1.2e-07\n",
      "\n",
      "20|CMS001_018_Ra_S14|NODE_1_length_10371_cov_9.071595\n",
      "contig NODE_1_length_10371_cov_9.071595 does not belong to a cluster\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6e-77\n",
      "PicornaRdRp_PF00680_full: 1.2e-07\n",
      "\n",
      "23|CMS001_002_Ra_S1|NODE_3_length_9996_cov_1810.573546\n",
      "PicornaRdRp_PF00680_full: 9.1e-97\n",
      "\n",
      "24|CMS001_060_Ra_S12|NODE_1_length_9916_cov_285.959955\n",
      "contig NODE_1_length_9916_cov_285.959955 does not belong to a cluster\n",
      "PicornaRdRp_PF00680_full: 5.7e-93\n",
      "not significant LuteoTotiRotaRdRp_PF02123_full 1.1e-05\n",
      "\n",
      "25|CMS002_020c_Rb_S133_L004|NODE_2_length_9884_cov_284.199042\n",
      "PicornaRdRp_PF00680_full: 1e-93\n",
      "LuteoTotiRotaRdRp_PF02123_full: 5.4e-06\n",
      "\n",
      "27|CMS002_038a_Rb_S172_L004|NODE_2_length_9783_cov_30.124253\n",
      "contig NODE_2_length_9783_cov_30.124253 does not belong to a cluster\n",
      "PicornaRdRp_PF00680_full: 1.9e-98\n",
      "\n",
      "28|CMS002_034a_Rb_S168_L004|NODE_1_length_9781_cov_13.772156\n",
      "PicornaRdRp_PF00680_full: 6.1e-97\n",
      "\n",
      "29|CMS001_051_Ra_S8|NODE_1_length_9777_cov_11.106392\n",
      "contig NODE_1_length_9777_cov_11.106392 does not belong to a cluster\n",
      "PicornaRdRp_PF00680_full: 1.5e-98\n",
      "\n",
      "30|CMS002_037a_Rb_S171_L004|NODE_1_length_9379_cov_122.411954\n",
      "contig NODE_1_length_9379_cov_122.411954 does not belong to a cluster\n",
      "MononegaRdRp_PF00946_full: 6e-190\n",
      "\n",
      "32|CMS001_031_Ra_S19|NODE_1_length_9312_cov_34.616892\n",
      "contig NODE_1_length_9312_cov_34.616892 does not belong to a cluster\n",
      "MononegaRdRp_PF00946_full: 8.9e-190\n",
      "\n",
      "39|CMS001_051_Ra_S8|NODE_2_length_7775_cov_34.360223\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.6e-61\n",
      "PicornaRdRp_PF00680_full: 3.6e-14\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 6.3e-05\n",
      "\n",
      "42|CMS001_057_Ra_S11|NODE_2_length_7695_cov_18.576267\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3e-60\n",
      "PicornaRdRp_PF00680_full: 1.4e-11\n",
      "\n",
      "43|CMS002_053a_Rb_S7_L004|NODE_4_length_7692_cov_11.659619\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.7e-61\n",
      "PicornaRdRp_PF00680_full: 1.4e-11\n",
      "\n",
      "44|CMS002_035a_Rb_S169_L004|NODE_2_length_7663_cov_235.389929\n",
      "BunyaRdRp_PF04196_full: 7.1e-87\n",
      "ArenaRdRp_PF06317_full: 3.9e-09\n",
      "OrthomyxoRdRp_PF00602_full: 6.2e-09\n",
      "\n",
      "45|CMS002_054a_Rb_S8_L004|NODE_1_length_7470_cov_515.249290\n",
      "MononegaRdRp_PF00946_full: 9.3e-199\n",
      "\n",
      "49|CMS002_029d_Rb_S162_L004|NODE_3_length_7380_cov_3.942352\n",
      "LuteoTotiRotaRdRp_PF02123_full: 2.2e-62\n",
      "PicornaRdRp_PF00680_full: 2.7e-14\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 0.00014\n",
      "\n",
      "50|CMS002_022a_Rb_S137_L004|NODE_2_length_7363_cov_8.020725\n",
      "contig NODE_2_length_7363_cov_8.020725 does not belong to a cluster\n",
      "BromoTobamoTogaRdRp_PF00978_full: 2.1e-51\n",
      "\n",
      "55|CMS002_053a_Rb_S7_L004|NODE_5_length_7168_cov_39.954872\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.2e-64\n",
      "PicornaRdRp_PF00680_full: 2.5e-09\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 1.4e-05\n",
      "\n",
      "56|CMS002_029c_Rb_S161_L004|NODE_2_length_7155_cov_5.347415\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.3e-68\n",
      "PicornaRdRp_PF00680_full: 1.8e-08\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 2.1e-05\n",
      "\n",
      "57|CMS002_040a_Rb_S174_L004|NODE_2_length_7146_cov_3.261423\n",
      "LuteoTotiRotaRdRp_PF02123_full: 6.5e-68\n",
      "PicornaRdRp_PF00680_full: 1.3e-08\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 2.9e-05\n",
      "\n",
      "59|CMS002_053a_Rb_S7_L004|NODE_6_length_7067_cov_2259.198283\n",
      "PicornaRdRp_PF00680_full: 3.1e-19\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6e-17\n",
      "\n",
      "60|CMS002_044b_Rb_S179_L004|NODE_1_length_7060_cov_2.221681\n",
      "PicornaRdRp_PF00680_full: 5.2e-110\n",
      "Partitiviridae_ICTV_Mafft_211216_full: 2.1e-12\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.1e-06\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 1.5e-05\n",
      "\n",
      "61|CMS001_040_Ra_S21|NODE_1_length_7059_cov_13.082641\n",
      "contig NODE_1_length_7059_cov_13.082641 does not belong to a cluster\n",
      "BunyaRdRp_PF04196_full: 3.9e-211\n",
      "\n",
      "63|CMS001_013_Ra_S5|NODE_1_length_6965_cov_47.988240\n",
      "BunyaRdRp_PF04196_full: 4.4e-147\n",
      "\n",
      "70|CMS002_045b_Rb_S184_L004|NODE_1_length_6634_cov_114.513192\n",
      "BunyaRdRp_PF04196_full: 3e-52\n",
      "\n",
      "76|CMS002_028d_Rb_S157_L004|NODE_2_length_6403_cov_43.049162\n",
      "BunyaRdRp_PF04196_full: 2.2e-51\n",
      "\n",
      "78|CMS001_038_Ra_S22|NODE_2_length_6269_cov_19.304587\n",
      "contig NODE_2_length_6269_cov_19.304587 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 8.8e-34\n",
      "PicornaRdRp_PF00680_full: 8.1e-16\n",
      "Partitiviridae_ICTV_Mafft_211216_full: 1.5e-08\n",
      "not significant MitoRdRp_PF05919_full 0.00011\n",
      "\n",
      "81|CMS002_017a_Rb_S122_L004|NODE_1_length_6223_cov_33.014156\n",
      "LuteoTotiRotaRdRp_PF02123_full: 4.9e-27\n",
      "PicornaRdRp_PF00680_full: 2e-18\n",
      "Partitiviridae_ICTV_Mafft_211216_full: 3.9e-06\n",
      "\n",
      "88|CMS001_022_Ra_S6|NODE_2_length_6047_cov_3.099497\n",
      "BromoTobamoTogaRdRp_PF00978_full: 2e-21\n",
      "PicornaRdRp_PF00680_full: 9.6e-14\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3e-10\n",
      "\n",
      "98|CMS001_060_Ra_S12|NODE_4_length_5737_cov_11.181979\n",
      "contig NODE_4_length_5737_cov_11.181979 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.1e-78\n",
      "PicornaRdRp_PF00680_full: 9.2e-33\n",
      "BromoTobamoTogaRdRp_PF00978_full: 7.1e-09\n",
      "not significant FlaviRdRp_PF00972_full 5.4e-05\n",
      "\n",
      "128|CMS001_001_Ra_S1|NODE_5_length_5116_cov_2209.828736\n",
      "BromoTobamoTogaRdRp_PF00978_full: 2.8e-22\n",
      "PicornaRdRp_PF00680_full: 4.2e-16\n",
      "LuteoTotiRotaRdRp_PF02123_full: 4.8e-08\n",
      "\n",
      "226|CMS001_047_Ra_S4|NODE_2_length_4094_cov_7.237491\n",
      "contig NODE_2_length_4094_cov_7.237491 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 1.8e-08\n",
      "\n",
      "246|CMS001_012_Ra_S4|NODE_4_length_4019_cov_14.444698\n",
      "contig NODE_4_length_4019_cov_14.444698 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 9.3e-06\n",
      "\n",
      "253|CMS001_012_Ra_S4|NODE_5_length_3972_cov_22.541463\n",
      "contig NODE_5_length_3972_cov_22.541463 does not belong to a cluster\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 3e-05\n",
      "\n",
      "296|CMS001_001_Ra_S1|NODE_9_length_3713_cov_129.105611\n",
      "Reo_alternative_full: 0.0\n",
      "PicornaRdRp_PF00680_full: 5.3e-06\n",
      "\n",
      "312|CMS001_025_Ra_S7|NODE_8_length_3666_cov_8.406520\n",
      "BromoTobamoTogaRdRp_PF00978_full: 2.9e-25\n",
      "PicornaRdRp_PF00680_full: 9.5e-24\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.6e-12\n",
      "\n",
      "326|CMS002_022a_Rb_S137_L004|NODE_3_length_3619_cov_89.634670\n",
      "LuteoTotiRotaRdRp_PF02123_full: 4.5e-36\n",
      "\n",
      "354|CMS001_053_Ra_S8|NODE_2_length_3531_cov_10.029820\n",
      "contig NODE_2_length_3531_cov_10.029820 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.4e-37\n",
      "\n",
      "399|CMS001_038_Ra_S22|NODE_4_length_3364_cov_2.638272\n",
      "MononegaRdRp_PF00946_full: 5.3e-28\n",
      "\n",
      "416|CMS001_026_Ra_S18|NODE_13_length_3331_cov_4.147203\n",
      "contig NODE_13_length_3331_cov_4.147203 does not belong to a cluster\n",
      "not significant MitoRdRp_PF05919_full 2.8e-05\n",
      "\n",
      "458|CMS002_056a_Rb_S9_L004|NODE_4_length_3249_cov_3490.915826\n",
      "PicornaRdRp_PF00680_full: 4.3e-43\n",
      "LuteoTotiRotaRdRp_PF02123_full: 2.3e-28\n",
      "\n",
      "484|CMS001_035_Ra_S20|NODE_10_length_3200_cov_65.512648\n",
      "PicornaRdRp_PF00680_full: 2.4e-40\n",
      "LuteoTotiRotaRdRp_PF02123_full: 2.4e-15\n",
      "\n",
      "493|CMS001_044_Ra_S25|NODE_5_length_3184_cov_44.723206\n",
      "MitoRdRp_PF05919_full: 9.9e-18\n",
      "\n",
      "499|CMS002_056a_Rb_S9_L004|NODE_5_length_3180_cov_494.778601\n",
      "contig NODE_5_length_3180_cov_494.778601 does not belong to a cluster\n",
      "PicornaRdRp_PF00680_full: 2.5e-40\n",
      "LuteoTotiRotaRdRp_PF02123_full: 2.4e-15\n",
      "\n",
      "506|CMS002_019a_Rb_S130_L004|NODE_1_length_3167_cov_3289.574434\n",
      "PicornaRdRp_PF00680_full: 4.5e-46\n",
      "LuteoTotiRotaRdRp_PF02123_full: 6.4e-28\n",
      "\n",
      "509|CMS002_047j_Rb_S3_L004|NODE_5_length_3164_cov_625.481697\n",
      "MitoRdRp_PF05919_full: 3.1e-18\n",
      "\n",
      "516|CMS001_040_Ra_S21|NODE_5_length_3145_cov_135.722295\n",
      "contig NODE_5_length_3145_cov_135.722295 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 3.1e-18\n",
      "\n",
      "531|CMS002_020b_Rb_S132_L004|NODE_4_length_3121_cov_1052.131406\n",
      "PicornaRdRp_PF00680_full: 1.1e-41\n",
      "LuteoTotiRotaRdRp_PF02123_full: 9.2e-29\n",
      "not significant MitoRdRp_PF05919_full 9.1e-05\n",
      "\n",
      "561|CMS002_045d_Rb_S186_L004|NODE_2_length_3060_cov_2583.644653\n",
      "PicornaRdRp_PF00680_full: 2.9e-44\n",
      "LuteoTotiRotaRdRp_PF02123_full: 4.7e-28\n",
      "\n",
      "706|CMS002_038a_Rb_S172_L004|NODE_12_length_2836_cov_341.811888\n",
      "PicornaRdRp_PF00680_full: 3.2e-41\n",
      "LuteoTotiRotaRdRp_PF02123_full: 8e-26\n",
      "BromoTobamoTogaRdRp_PF00978_full: 1.5e-06\n",
      "not significant Partitiviridae_ICTV_Mafft_211216_full 0.0001\n",
      "\n",
      "1124|CMS001_036_Ra_S20|NODE_11_length_2476_cov_1.939558\n",
      "contig NODE_11_length_2476_cov_1.939558 does not belong to a cluster\n",
      "MononegaRdRp_PF00946_full: 2.7e-59\n",
      "\n",
      "1160|CMS001_027_Ra_S16|NODE_3_length_2447_cov_21.006751\n",
      "OrthomyxoRdRp_PF00602_full: 1.5e-132\n",
      "BunyaRdRp_PF04196_full: 1.5e-10\n",
      "\n",
      "1168|CMS002_026a_Rb_S146_L004|NODE_2_length_2442_cov_5.123890\n",
      "OrthomyxoRdRp_PF00602_full: 4.1e-129\n",
      "\n",
      "1170|CMS002_029c_Rb_S161_L004|NODE_9_length_2441_cov_36.508037\n",
      "OrthomyxoRdRp_PF00602_full: 4.1e-122\n",
      "BunyaRdRp_PF04196_full: 7.7e-11\n",
      "\n",
      "1181|CMS002_017c_Rb_S124_L004|NODE_5_length_2434_cov_22.260501\n",
      "OrthomyxoRdRp_PF00602_full: 5.8e-122\n",
      "BunyaRdRp_PF04196_full: 1.2e-07\n",
      "\n",
      "1249|CMS001_009_Ra_S13|NODE_70_length_2389_cov_1.563149\n",
      "contig NODE_70_length_2389_cov_1.563149 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 7.9e-19\n",
      "\n",
      "1399|CMS002_053a_Rb_S7_L004|NODE_134_length_2299_cov_1.260126\n",
      "contig NODE_134_length_2299_cov_1.260126 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 1.3e-22\n",
      "\n",
      "1438|CMS002_053a_Rb_S7_L004|NODE_136_length_2278_cov_2.307133\n",
      "contig NODE_136_length_2278_cov_2.307133 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 4e-24\n",
      "\n",
      "1479|CMS002_053a_Rb_S7_L004|NODE_141_length_2261_cov_4.574176\n",
      "contig NODE_141_length_2261_cov_4.574176 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 1e-18\n",
      "\n",
      "1636|CMS002_021a_Rb_S136_L004|NODE_301_length_2191_cov_563.679754\n",
      "Partitiviridae_ICTV_Mafft_211216_full: 7.3e-63\n",
      "PicornaRdRp_PF00680_full: 1.7e-55\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.5e-16\n",
      "BromoTobamoTogaRdRp_PF00978_full: 2.8e-06\n",
      "\n",
      "1715|CMS001_012_Ra_S4|NODE_19_length_2157_cov_2.648077\n",
      "contig NODE_19_length_2157_cov_2.648077 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.9e-54\n",
      "PicornaRdRp_PF00680_full: 6.1e-11\n",
      "BromoTobamoTogaRdRp_PF00978_full: 7.2e-06\n",
      "\n",
      "1780|CMS002_021a_Rb_S136_L004|NODE_323_length_2136_cov_1.284119\n",
      "contig NODE_323_length_2136_cov_1.284119 does not belong to a cluster\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 4.3e-05\n",
      "\n",
      "2118|CMS002_017d_Rb_S125_L004|NODE_7_length_2017_cov_358.241237\n",
      "FlaviRdRp_PF00972_full: 2.3e-154\n",
      "PicornaRdRp_PF00680_full: 3.2e-10\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 3.5e-05\n",
      "\n",
      "2923|CMS001_005_Ra_S3|NODE_94_length_1823_cov_2.475945\n",
      "contig NODE_94_length_1823_cov_2.475945 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.5e-47\n",
      "PicornaRdRp_PF00680_full: 2.6e-11\n",
      "\n",
      "3061|CMS002_053a_Rb_S7_L004|NODE_257_length_1795_cov_3.930733\n",
      "contig NODE_257_length_1795_cov_3.930733 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 1.1e-20\n",
      "\n",
      "3110|CMS002_026d_Rb_S149_L004|NODE_8_length_1787_cov_26.230409\n",
      "Partitiviridae_ICTV_Mafft_211216_full: 2.7e-79\n",
      "PicornaRdRp_PF00680_full: 1.3e-59\n",
      "LuteoTotiRotaRdRp_PF02123_full: 7.6e-16\n",
      "BromoTobamoTogaRdRp_PF00978_full: 6e-08\n",
      "\n",
      "3256|CMS002_020a_Rb_S131_L004|NODE_200_length_1765_cov_198.981043\n",
      "Partitiviridae_ICTV_Mafft_211216_full: 1.4e-63\n",
      "PicornaRdRp_PF00680_full: 2.1e-56\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.1e-16\n",
      "BromoTobamoTogaRdRp_PF00978_full: 2.6e-06\n",
      "\n",
      "3883|CMS002_053a_Rb_S7_L004|NODE_293_length_1668_cov_1.297297\n",
      "contig NODE_293_length_1668_cov_1.297297 does not belong to a cluster\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 0.00011\n",
      "\n",
      "4269|CMS002_053a_Rb_S7_L004|NODE_311_length_1619_cov_2.165370\n",
      "contig NODE_311_length_1619_cov_2.165370 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 3.7e-17\n",
      "\n",
      "4553|CMS001_038_Ra_S22|NODE_25_length_1583_cov_1.702523\n",
      "MononegaRdRp_PF00946_full: 1.4e-106\n",
      "\n",
      "5308|CMS001_009_Ra_S13|NODE_344_length_1503_cov_2.269285\n",
      "contig NODE_344_length_1503_cov_2.269285 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1e-44\n",
      "PicornaRdRp_PF00680_full: 1.3e-11\n",
      "\n",
      "6512|CMS002_021a_Rb_S136_L004|NODE_1168_length_1406_cov_2.164786\n",
      "LuteoTotiRotaRdRp_PF02123_full: 1.5e-42\n",
      "PicornaRdRp_PF00680_full: 4.8e-11\n",
      "\n",
      "6823|CMS001_001_Ra_S1|NODE_81_length_1383_cov_1.592649\n",
      "contig NODE_81_length_1383_cov_1.592649 does not belong to a cluster\n",
      "BunyaRdRp_PF04196_full: 6.8e-15\n",
      "not significant ArenaRdRp_PF06317_full 2e-05\n",
      "\n",
      "9191|CMS002_053a_Rb_S7_L004|NODE_580_length_1250_cov_4.236999\n",
      "contig NODE_580_length_1250_cov_4.236999 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 3.7e-23\n",
      "\n",
      "10146|CMS002_046b_Rb_S192_L004|NODE_38_length_1206_cov_4.790080\n",
      "contig NODE_38_length_1206_cov_4.790080 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 1.2e-18\n",
      "\n",
      "11277|CMS002_017d_Rb_S125_L004|NODE_10_length_1161_cov_248.109779\n",
      "FlaviRdRp_PF00972_full: 3.6e-22\n",
      "\n",
      "12259|CMS001_009_Ra_S13|NODE_835_length_1125_cov_1.704198\n",
      "contig NODE_835_length_1125_cov_1.704198 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 7.8e-17\n",
      "\n",
      "12955|CMS001_025_Ra_S7|NODE_330_length_1102_cov_10.451707\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 0.00012\n",
      "\n",
      "14663|CMS002_020a_Rb_S131_L004|NODE_738_length_1054_cov_1.305015\n",
      "not significant MitoRdRp_PF05919_full 6.9e-05\n",
      "\n",
      "16869|CMS001_060_Ra_S12|NODE_1203_length_1000_cov_4.565547\n",
      "not significant BromoTobamoTogaRdRp_PF00978_full 6.9e-05\n",
      "\n",
      "17639|CMS002_007a_Rb_S118_L004|NODE_2294_length_983_cov_1.416115\n",
      "not significant PicornaRdRp_PF00680_full 7.7e-05\n",
      "\n",
      "19507|CMS002_020a_Rb_S131_L004|NODE_961_length_943_cov_2.577367\n",
      "not significant OrthomyxoRdRp_PF00602_full 2.2e-05\n",
      "\n",
      "19890|CMS001_016_Ra_S6|NODE_503_length_936_cov_1.477299\n",
      "contig NODE_503_length_936_cov_1.477299 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 4.5e-43\n",
      "PicornaRdRp_PF00680_full: 6.8e-11\n",
      "\n",
      "22000|CMS001_060_Ra_S12|NODE_1584_length_898_cov_0.742996\n",
      "contig NODE_1584_length_898_cov_0.742996 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 2e-06\n",
      "\n",
      "22774|CMS001_009_Ra_S13|NODE_1562_length_885_cov_1.720297\n",
      "contig NODE_1562_length_885_cov_1.720297 does not belong to a cluster\n",
      "FlaviRdRp_PF00972_full: 3.8e-06\n",
      "\n",
      "22872|CMS002_020d_Rb_S134_L004|NODE_164_length_884_cov_1.019827\n",
      "contig NODE_164_length_884_cov_1.019827 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 2.7e-66\n",
      "PicornaRdRp_PF00680_full: 1.2e-14\n",
      "\n",
      "23471|CMS002_045a_Rb_S183_L004|NODE_16_length_875_cov_1.358396\n",
      "contig NODE_16_length_875_cov_1.358396 does not belong to a cluster\n",
      "MononegaRdRp_PF00946_full: 4.2e-35\n",
      "\n",
      "28474|CMS002_007a_Rb_S118_L004|NODE_3545_length_805_cov_1.910714\n",
      "contig NODE_3545_length_805_cov_1.910714 does not belong to a cluster\n",
      "not significant ArenaRdRp_PF06317_full 0.074\n",
      "\n",
      "31787|CMS002_023a_Rb_S138_L004|NODE_33_length_768_cov_1.162084\n",
      "contig NODE_33_length_768_cov_1.162084 does not belong to a cluster\n",
      "not significant ArenaRdRp_PF06317_full 0.15\n",
      "\n",
      "32822|CMS002_021a_Rb_S136_L004|NODE_5038_length_757_cov_2.616176\n",
      "contig NODE_5038_length_757_cov_2.616176 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 4e-24\n",
      "\n",
      "33175|CMS002_053a_Rb_S7_L004|NODE_1954_length_754_cov_1.305761\n",
      "contig NODE_1954_length_754_cov_1.305761 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 3.2e-07\n",
      "LuteoTotiRotaRdRp_PF02123_full: 6.8e-07\n",
      "\n",
      "37171|CMS002_007a_Rb_S118_L004|NODE_4548_length_718_cov_1.023401\n",
      "contig NODE_4548_length_718_cov_1.023401 does not belong to a cluster\n",
      "OrthomyxoRdRp_PF00602_full: 1.6e-06\n",
      "\n",
      "47143|CMS002_053a_Rb_S7_L004|NODE_2754_length_647_cov_1.924561\n",
      "contig NODE_2754_length_647_cov_1.924561 does not belong to a cluster\n",
      "MitoRdRp_PF05919_full: 2.2e-08\n",
      "\n",
      "47173|CMS001_014_Ra_S5|NODE_648_length_646_cov_1.527241\n",
      "contig NODE_648_length_646_cov_1.527241 does not belong to a cluster\n",
      "not significant PicornaRdRp_PF00680_full 0.00011\n",
      "\n",
      "47685|CMS001_028_Ra_S17|NODE_190_length_643_cov_0.816254\n",
      "contig NODE_190_length_643_cov_0.816254 does not belong to a cluster\n",
      "not significant BirnaRdRp_PF04197_full 0.011\n",
      "\n",
      "47830|CMS001_022_Ra_S6|NODE_168_length_642_cov_1.302655\n",
      "contig NODE_168_length_642_cov_1.302655 does not belong to a cluster\n",
      "BunyaRdRp_PF04196_full: 4.2e-14\n",
      "\n",
      "49310|CMS001_005_Ra_S3|NODE_2230_length_633_cov_1.062950\n",
      "contig NODE_2230_length_633_cov_1.062950 does not belong to a cluster\n",
      "not significant PicornaRdRp_PF00680_full 0.00016\n",
      "\n",
      "50240|CMS002_020e_Rb_S135_L004|NODE_830_length_628_cov_1.197822\n",
      "contig NODE_830_length_628_cov_1.197822 does not belong to a cluster\n",
      "LuteoTotiRotaRdRp_PF02123_full: 3.4e-41\n",
      "PicornaRdRp_PF00680_full: 4.3e-12\n",
      "\n",
      "99\n",
      "99 hits out of 21177 (no hits: 21078; fraction: 0.004674883127921802)\n"
     ]
    }
   ],
   "source": [
    "from Bio.SearchIO import HmmerIO\n",
    "\n",
    "hitfile=os.path.join(base_path,'darkmatter/RdRp_candidates_prot.pssm.out')\n",
    "hits=HmmerIO.Hmmer3TextParser(open(hitfile,'r'))\n",
    "RdRp_clusters=set()\n",
    "\n",
    "# count=0\n",
    "total=0\n",
    "for query in sorted(hits,key=lambda k: int(k.id.split('|')[0])):\n",
    "    if len(query.hits)>0:\n",
    "        cluster_name,sample,contig_name,_=query.id.split('|')\n",
    "    \n",
    "        \n",
    "        families=[hit.id for hit in query.hits]\n",
    "        print('%s'%('|'.join(query.id.split('|')[:-1])))\n",
    "        \n",
    "        if len(clusters[cluster_name])<2:\n",
    "            print('contig %s does not belong to a cluster'%(contig_name))\n",
    "        \n",
    "        for hit in query.hits:\n",
    "            if hit.evalue<1e-5:\n",
    "                print('%s: %s'%(hit.id,hit.evalue))\n",
    "                if cluster_name in clusters:\n",
    "                    RdRp_clusters.add(cluster_name)\n",
    "            else:\n",
    "                print('not significant',hit.id,hit.evalue)\n",
    "#             print(query.id\n",
    "        print()\n",
    "        \n",
    "    total+=1\n",
    "\n",
    "count=len(RdRp_clusters)\n",
    "print(len(RdRp_clusters))\n",
    "print('%s hits out of %s (no hits: %s; fraction: %s)'%(count,total,total-count,count/total))\n",
    "# 163 hits out of 21177 (no hits: 21014; fraction: 0.0076970297964773105) ## all proteins >200 aa, unfiltered by catalytic motif\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
