{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting a static treemap\n",
    "\n",
    "--------\n",
    "\n",
    "This Jupyter notebook will generate a static version of the treemap. In the first step each parent in the JSON gets all of its children's values (read or contig counts) are summed to the parent. In the second step we're assigning a lineage to each treemap compartment to help later. In the last step we produce a copy of the nested treemap structure that's flat for use later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-27T12:47:45.130820Z",
     "start_time": "2020-02-27T12:47:43.384398Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "138\n",
      "13014814\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "typeface='Helvetica Neue'\n",
    "mpl.rcParams['mathtext.fontset']='custom'\n",
    "mpl.rcParams['font.sans-serif']=typeface\n",
    "mpl.rcParams['mathtext.default']='sf'\n",
    "mpl.rcParams['axes.labelweight']=300\n",
    "mpl.rcParams['font.family']=typeface\n",
    "mpl.rcParams['font.size']=22\n",
    "\n",
    "import numpy as np\n",
    "import json,copy\n",
    "import ete3\n",
    "import squarify\n",
    "\n",
    "ncbi=ete3.ncbi_taxonomy.NCBITaxa()\n",
    "\n",
    "treemap_path='/Users/evogytis/Documents/manuscripts/skeeters/treemap/skeeters.json'\n",
    "J=json.load(open(treemap_path,'r'))\n",
    "\n",
    "def sumValues(node,stat,level=None,order=None):\n",
    "    '''\n",
    "    Sum children's values and assign to parent (currently annotations are branch-specific).\n",
    "    '''\n",
    "    if level==None:\n",
    "        level=0\n",
    "    if order==None:\n",
    "        order=0\n",
    "        \n",
    "    node['attrs']['height']=level ## remember height\n",
    "    order+=1 ## remember order of visitation\n",
    "    \n",
    "    if 'children' in node: ## node with children\n",
    "        level+=1 ## increment level\n",
    "        for child in node['children']: ## iterate over children\n",
    "            if stat not in node['attrs']:\n",
    "                node['attrs'][stat]=0\n",
    "            \n",
    "            order=sumValues(child,stat,level,order) ## traverse first (reaches tips first)\n",
    "            node['attrs'][stat]+=child['attrs'][stat] ## only then add stat\n",
    "            \n",
    "    node['attrs']['order']=order ## assign order of visitation\n",
    "    return order\n",
    "\n",
    "def flatten(node,container=None): ## flat list of JSON branches\n",
    "    \"\"\"\n",
    "    Unroll JSON tree structure into a flat list.\n",
    "    \"\"\"\n",
    "    if container==None:\n",
    "        container=[]\n",
    "        \n",
    "    container.append(node)\n",
    "    if 'children' in node:\n",
    "        for child in node['children']:\n",
    "            flatten(child,container)\n",
    "            \n",
    "    return container\n",
    "\n",
    "def treemapLineages(node,lineage=None):\n",
    "    \"\"\"\n",
    "    Assign lineages to each compartment of the treemap.\n",
    "    \"\"\"\n",
    "    if lineage==None:\n",
    "        lineage=[]\n",
    "    \n",
    "    lin=list(lineage)\n",
    "    \n",
    "    if 'taxid' in node:\n",
    "        node['lineage']=lin+[node['taxid']]\n",
    "    \n",
    "    if 'children' in node:\n",
    "        if 'taxid' in node:\n",
    "            lin.append(node['taxid'])\n",
    "            \n",
    "        for child in node['children']:\n",
    "            lineage=treemapLineages(child,lin)\n",
    "    \n",
    "    return lineage\n",
    "\n",
    "sumValues(J,'read_count') ## sum read counts across children\n",
    "# sumValues(J,'contig_count')\n",
    "            \n",
    "treemapLineages(J) ## assign lineages to each treemap compartment\n",
    "\n",
    "flatJ=flatten(copy.deepcopy(J)) ## get a flat version of the JSON\n",
    "print(len(flatJ)) ## report number of branches in JSON treemap\n",
    "print(flatJ[0]['attrs']['read_count'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting treemap\n",
    "\n",
    "In this cell we define the `size` function which will be used to determine the size of treemap compartments (read counts by default). The parameter `edge_length` by default is computed as the square of the value at the root and passed on as both the `dx` (width) and `dy` (height) of the treemap (_i.e._ treemap is a square)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2020-02-27T12:47:43.349Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3607.6050227262963\n",
      "root (of )\t13014814\n",
      "cellular organisms (of root)\t1869529\n",
      "Bacteria (of cellular organisms)\t680711\n",
      "Spirochaetes (of Bacteria)\t106937\n",
      "Terrabacteria group (of Bacteria)\t48446\n",
      "Gammaproteobacteria (of Bacteria)\t100406\n",
      "Enterobacterales (of Gammaproteobacteria)\t71438\n",
      "Oceanospirillales (of Gammaproteobacteria)\t25934\n",
      "Alphaproteobacteria (of Bacteria)\t301745\n",
      "Wolbachia (of Alphaproteobacteria)\t295760\n",
      "Eukaryota (of cellular organisms)\t850497\n",
      "Viridiplantae (of Eukaryota)\t61664\n",
      "Opisthokonta (of Eukaryota)\t308830\n",
      "Fungi (of Opisthokonta)\t78577\n",
      "Dikarya (of Fungi)\t20896\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:86: RuntimeWarning: divide by zero encountered in double_scalars\n",
      "  for rect in layout(sizes, x, y, dx, dy)\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:38: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  height = covered_area / dx\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:24: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  width = covered_area / dy\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:27: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  rects.append({\"x\": x, \"y\": y, \"dx\": width, \"dy\": size / width})\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:28: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  y += size / width\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:41: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  rects.append({\"x\": x, \"y\": y, \"dx\": size / height, \"dy\": height})\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/squarify/__init__.py:42: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  x += size / height\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Microsporidia (of Fungi)\t56145\n",
      "Bilateria (of Opisthokonta)\t204649\n",
      "Ecdysozoa (of Bilateria)\t36026\n",
      "Chordata (of Bilateria)\t151023\n",
      "Boreoeutheria (of Chordata)\t73151\n",
      "Rodentia (of Boreoeutheria)\t12698\n",
      "Pecora (of Boreoeutheria)\t17306\n",
      "Aves (of Chordata)\t50572\n",
      "Passeriformes (of Aves)\t11824\n",
      "Trypanosomatidae (of Eukaryota)\t449053\n",
      "Leishmaniinae (of Trypanosomatidae)\t249890\n",
      "Trypanosoma (of Trypanosomatidae)\t12810\n",
      "Viruses (of root)\t10930168\n",
      "(-)ss RNA viruses (of Viruses)\t2249955\n",
      "Bunyavirales (of (-)ss RNA viruses)\t1352939\n",
      "Phasmaviridae (of Bunyavirales)\t361786\n",
      "76|Phasma-like (of Phasmaviridae)\t205279\n",
      "Barstukas virus (of Phasmaviridae)\t156507\n",
      "Phenuiviridae (of Bunyavirales)\t437625\n",
      "63|Phenui-like (of Phenuiviridae)\t437625\n",
      "Peribunyaviridae (of Bunyavirales)\t553528\n",
      "Culex bunyavirus 2 (of Peribunyaviridae)\t545058\n",
      "Orthomyxoviridae (of (-)ss RNA viruses)\t207043\n",
      "Quaranjavirus (of Orthomyxoviridae)\t207043\n",
      "Wuhan mosquito virus 6 (of Quaranjavirus)\t113519\n",
      "Guadeloupe mosquito quaranja-like virus 1 (of Quaranjavirus)\t26150\n",
      "Åªsinis virus (of Quaranjavirus)\t61694\n",
      "Mononega-Chu (of (-)ss RNA viruses)\t689973\n",
      "Mononegavirales (of Mononega-Chu)\t493896\n",
      "Rhabdoviridae (of Mononegavirales)\t381308\n",
      "Merida virus (of Rhabdoviridae)\t253014\n",
      "4|Rhabdo-like (of Rhabdoviridae)\t13248\n",
      "30|Rhabdo-like (of Rhabdoviridae)\t21608\n",
      "2|Rhabdo-like (of Rhabdoviridae)\t93438\n",
      "Xinmoviridae (of Mononegavirales)\t112588\n",
      "Aedes anphevirus (of Xinmoviridae)\t10340\n",
      "Gordis virus (of Xinmoviridae)\t102248\n",
      "Chuviridae (of Mononega-Chu)\t196077\n",
      "Culex mosquito virus 4 (of Chuviridae)\t196077\n",
      "uncurated (of Viruses)\t374696\n",
      "(+)ssRNA viruses (of Viruses)\t7360539\n",
      "Flaviviridae (of (+)ssRNA viruses)\t75048\n",
      "Culex flavivirus (of Flaviviridae)\t39399\n",
      "1|Flavi-like (of Flaviviridae)\t30224\n",
      "Picornavirales (of (+)ssRNA viruses)\t701408\n",
      "Iflaviridae (of Picornavirales)\t700791\n",
      "Culex iflavi-like virus 4 (of Iflaviridae)\t602584\n",
      "24|Ifla-like (of Iflaviridae)\t40486\n",
      "25|Ifla-like (of Iflaviridae)\t51675\n",
      "Luteo-Sobemo (of (+)ssRNA viruses)\t4272489\n",
      "Solemoviridae (of Luteo-Sobemo)\t1609134\n",
      "Guadeloupe mosquito virus (of Solemoviridae)\t913066\n",
      "Culex mosquito virus 6 (of Solemoviridae)\t92258\n",
      "Wenzhou sobemo-like virus 4 (of Solemoviridae)\t533167\n",
      "Kellev virus (of Solemoviridae)\t70643\n",
      "Luteoviridae (of Luteo-Sobemo)\t2663355\n",
      "Marma virus (of Luteoviridae)\t2612944\n",
      "Culex-associated Luteo-like virus (of Luteoviridae)\t49491\n",
      "Tombus-Noda (of (+)ssRNA viruses)\t880694\n",
      "Tombusviridae (of Tombus-Noda)\t880694\n",
      "Hubei mosquito virus 4 (of Tombusviridae)\t879164\n",
      "Narna-Levi (of (+)ssRNA viruses)\t920403\n",
      "Narnaviridae (of Narna-Levi)\t919142\n",
      "Culex narnavirus 1 (of Narnaviridae)\t919082\n",
      "Hepe-Virga (of (+)ssRNA viruses)\t510497\n",
      "Virgaviridae (of Hepe-Virga)\t509604\n",
      "Hubei virga-like virus 2 (of Virgaviridae)\t342734\n",
      "Culex pipiens-associated Tunisia virus (of Virgaviridae)\t166870\n",
      "dsRNA viruses (of Viruses)\t944978\n",
      "Partitiviridae (of dsRNA viruses)\t630676\n",
      "3256|Partiti-like (of Partitiviridae)\t10793\n",
      "1636|Partiti-like (of Partitiviridae)\t614765\n",
      "Reoviridae (of dsRNA viruses)\t204502\n",
      "296|Reo-like (of Reoviridae)\t197759\n",
      "Toti-Chryso (of dsRNA viruses)\t109800\n",
      "Totiviridae (of Toti-Chryso)\t85416\n",
      "Aedes aegypti totivirus (of Totiviridae)\t56124\n",
      "42|Toti-like (of Totiviridae)\t12059\n",
      "Chrysoviridae (of Toti-Chryso)\t24384\n",
      "Hubei chryso-like virus 1 (of Chrysoviridae)\t21597\n",
      "100.0 3607.6050227262963\n"
     ]
    }
   ],
   "source": [
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42 ## TrueType fonts for manageable export\n",
    "\n",
    "size=lambda k: k['attrs']['read_count'] ## area will be read counts\n",
    "# size=lambda k: k['attrs']['contig_count'] ## area will be number of contigs\n",
    "\n",
    "edge_length=np.sqrt(size(J['children'][0])) ## assuming\n",
    "print(edge_length)\n",
    "\n",
    "def computeCoordinates(node,x,y,dx,dy):\n",
    "    \"\"\"\n",
    "    Compute treemap compartment coordinates using the squarify library.\n",
    "    \"\"\"\n",
    "    node['attrs']['x']=x ## assign coordinates to compartment\n",
    "    node['attrs']['y']=y\n",
    "    node['attrs']['dx']=dx\n",
    "    node['attrs']['dy']=dy\n",
    "        \n",
    "    if 'children' in node: ## there are children, need to traverse further\n",
    "        children=sorted(node['children'],key=lambda q: size(q)) ## get children sorted by read count\n",
    "        children_values=[size(child) for child in children] ## get read counts for each child\n",
    "        sizes=squarify.normalize_sizes(children_values+[size(node)-sum(children_values)],dx,dy) ## normalize children values\n",
    "        \n",
    "        rects=squarify.padded_squarify(sizes,x,y,dx,dy) ## compute rectangles (padded)\n",
    "#         rects=squarify.squarify(sizes,x,y,dx,dy) ## compute rectangles (unpadded)\n",
    "        for ch,rec in zip(children,rects): ## iterate over children\n",
    "            x=rec['x']\n",
    "            y=rec['y']\n",
    "            dx=rec['dx']\n",
    "            dy=rec['dy']\n",
    "            \n",
    "            computeCoordinates(ch,x,y,dx,dy) ## recurse\n",
    "\n",
    "def plot(ax,node,parent,previous=None,level=None):\n",
    "    \"\"\"\n",
    "    Plot treemap.\n",
    "    \"\"\"\n",
    "    if previous==None:\n",
    "        previous=0.0\n",
    "    if level==None:\n",
    "        level=0\n",
    "    \n",
    "    if 'x' in node['attrs']:\n",
    "        x=node['attrs']['x']\n",
    "        y=node['attrs']['y']\n",
    "        w=node['attrs']['dx']\n",
    "        h=node['attrs']['dy']\n",
    "        \n",
    "        if w==0 or h==0:\n",
    "            print('box %s has zero width or height: %s %s'%(node['taxonomy'],w,h))\n",
    "        \n",
    "        c=node['attrs']['colour']\n",
    "#         lw=max([1,(7-level)])\n",
    "        lw=1\n",
    "        \n",
    "        rect=plt.Rectangle((x,y),w,h,facecolor=c,edgecolor='w',\n",
    "                           alpha=1.0,lw=lw,zorder=level) ## add rectangle\n",
    "        \n",
    "        rotation=0\n",
    "        if h>(w*2):\n",
    "            rotation=90 ## rotate labels if rectangle is taller than wider\n",
    "        \n",
    "        \n",
    "        if 'taxonomy' in node and size(node)>10000.0: ## only add names for compartments big enough\n",
    "            name=node['taxonomy']\n",
    "            print('%s (of %s)\\t%d'%(node['taxonomy'],parent['taxonomy'] if 'taxonomy' in parent else '',size(node)))\n",
    "            if 'children' in node and name!='root':\n",
    "                if level<5: ## when dealing with deep compartments plot label bottom right\n",
    "                    ax.text(x+w-0.0015,y+0.0015,name,ha='right',va='bottom',\n",
    "                        size=22-level,zorder=10000,rotation=rotation)\n",
    "                else: ## otherwise top left\n",
    "                    ax.text(x+0.0015,y+h-0.0015,name,ha='left',va='top',\n",
    "                        size=25-level,zorder=level+10,rotation=rotation)\n",
    "\n",
    "            elif 'children' not in node: ## terminal compartment, text is centered\n",
    "                ax.text(x+w/2,y+h/2,name,ha='center',va='center',\n",
    "                    size=16,zorder=level+10,rotation=rotation)\n",
    "\n",
    "        ax.add_patch(rect)\n",
    "        ax.plot()\n",
    "    \n",
    "    if 'children' in node:\n",
    "        level+=1\n",
    "        for child in node['children']:\n",
    "            plot(ax,child,node,previous,level) ## recurse\n",
    "\n",
    "fig = plt.figure(figsize=(30, 30),facecolor='w') \n",
    "gs = gridspec.GridSpec(1,1,wspace=0.01,hspace=0.01) \n",
    "\n",
    "ax=plt.subplot(gs[0],facecolor='w')\n",
    "\n",
    "computeCoordinates(J,0.0,0.0,edge_length,edge_length) ## compute treemap coordinates, start at (0,0), width and height are edge_length (square treemap)\n",
    "plot(ax,J,None) ## plot treemap\n",
    "\n",
    "if edge_length>3000.0:\n",
    "    unit=100.0 ## legend unit\n",
    "    u='%d reads'%(int(unit)) ## legend label\n",
    "    X=edge_length+100 ## x-axis offset\n",
    "else:\n",
    "    unit=4.0\n",
    "    u='%d contigs'%(int(unit))\n",
    "    X=-7\n",
    "\n",
    "ax.text(X+unit*1.1,unit/2,u,size=40,va='center',ha='left') ## add text explaining what the legend square represents\n",
    "    \n",
    "print(unit,edge_length)\n",
    "\n",
    "legend=plt.Rectangle((X,0.0),unit,unit,lw=2,facecolor='k',edgecolor='none',zorder=100,clip_on=False)\n",
    "ax.add_patch(legend) ## add legend square\n",
    "\n",
    "branchesByLevel=sorted(flatJ[2:], key=lambda k: (-k['attrs']['order'],k['attrs']['height']))\n",
    "colours={}\n",
    "\n",
    "for b in branchesByLevel:\n",
    "    if b['attrs']['colour'] not in colours: ## plot legend squares until colours are exhausted (needs to be iterated over in the correct order)\n",
    "        y=edge_length*0.1+len(colours)*edge_length/25.0 # y coordinate\n",
    "        \n",
    "        ax.add_patch(plt.Rectangle((X,y-unit),unit,unit,lw=2,facecolor=b['attrs']['colour'],edgecolor='none',zorder=1000,clip_on=False)) ## add legend rectangle\n",
    "        \n",
    "        ax.text(X+unit*1.1,y-unit/2.0,b['taxonomy'],size=26,ha='left',va='center') ## add name for compartment\n",
    "        colours[b['attrs']['colour']]=b['taxonomy'] ## remember colour\n",
    "\n",
    "[ax.spines[loc].set_visible(False) for loc in ax.spines]\n",
    "ax.tick_params(size=0,labelsize=0)\n",
    "ax.set_aspect(1)\n",
    "\n",
    "plt.savefig('/Users/evogytis/Documents/manuscripts/skeeters/figures/fig2_treemap.pdf',dpi=300,bbox_inches='tight')\n",
    "plt.savefig('/Users/evogytis/Documents/manuscripts/skeeters/figures/fig2_treemap.png',dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Display read counts for treemap compartments split by sample\n",
    "\n",
    "This cell will collect specified compartments of the treemap and display all the reads under the compartment split by sample, excluding nested compartments included in the plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2020-02-27T12:47:43.352Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def split_reads(node,container=None,exclude=None):\n",
    "    \"\"\"\n",
    "    Traverse subtree from a node, accumulating sample-specific read counts.\n",
    "    \"\"\"\n",
    "    if container==None:\n",
    "        container={}\n",
    "    if exclude==None:\n",
    "        exclude=lambda node: True\n",
    "        \n",
    "    for sample in [attr for attr in node['attrs'] if 'CMS' in attr]:\n",
    "        if sample not in container:\n",
    "            container[sample]=0\n",
    "        container[sample]+=node['attrs'][sample]['read_count'] ## add reads to sample\n",
    "        \n",
    "    if 'children' in node: ## need to recurse\n",
    "        for child in node['children']: ## iterate over children\n",
    "            if exclude(child)==True: ## child not excluded\n",
    "                split_reads(child,container,exclude=exclude) ## recurse\n",
    "            else:\n",
    "                print('%s reads excluded from %s'%(child['taxonomy'],node['taxonomy']))\n",
    "    \n",
    "    return container\n",
    "\n",
    "fig = plt.figure(figsize=(40, 35),facecolor='w') \n",
    "gs = gridspec.GridSpec(1,1,wspace=0.01,hspace=0.01) \n",
    "\n",
    "ax=plt.subplot(gs[0],facecolor='w')\n",
    "\n",
    "ylabels=[]\n",
    "read_counts=[]\n",
    "colours=[]\n",
    "\n",
    "picked=[b for b in flatJ[1:] if b['attrs']['read_count']>1e3 and ('children' not in b or b['taxonomy']=='Bacteria')] ## programmatically get things that have high reads, are terminal (except if it's Bacteria)\n",
    "\n",
    "include=[b['taxid'] for b in picked] ## list of taxids from branches\n",
    "\n",
    "for k in sorted(picked,key=lambda w: w['attrs']['read_count']): ## iterate over picked branches, sorted by read count\n",
    "    taxid_reads_per_sample=split_reads(k,exclude=lambda x: x in k['lineage'] and x not in include) ## do traversals accumulating reads (excluding other picked branches)\n",
    "    read_counts.append(list(filter(lambda w:w!=0,sorted(taxid_reads_per_sample.values(),reverse=True)))) ## convert dict to list, sort by read count\n",
    "\n",
    "    cur_clade=set(k['lineage'])\n",
    "    include_clades={b['taxid']: set(b['lineage']) for b in flatJ[1:] if b['taxid'] in include}\n",
    "    \n",
    "    taxid_label=[]\n",
    "    for t in include_clades: ## iterate over all other treemap compartments designated for display here\n",
    "        if cur_clade.issubset(include_clades[t]): ## if another compartment's lienage is a subset of current one's (it's nested within current one)\n",
    "            if isinstance(t,str):\n",
    "                taxid_label.append(t)\n",
    "            elif isinstance(t,int):\n",
    "                taxid_label.append(ncbi.get_taxid_translator([t])[t]) ## continue adding to taxid label, first element is what's being displayed, every other element will be a compartment that's excluded\n",
    "\n",
    "    if len(taxid_label)==1:\n",
    "        ylabels.append(taxid_label[0])\n",
    "    else:\n",
    "        ylabels.append('%s (excluding %s)'%(taxid_label[0],', '.join(taxid_label[1:]))) ## format label\n",
    "\n",
    "    colours.append(k['attrs']['colour']) ## remember colour\n",
    "\n",
    "for i in range(len(ylabels)): ## iterate over compartments\n",
    "    for j in range(len(read_counts[i])): ## iterate over sample-specific reads\n",
    "        x=read_counts[i][j] ## current sample's read counts\n",
    "        sx=sum(read_counts[i][:j]) ## sum of all other samples up to current one\n",
    "        ax.barh(i,x,left=sx,fc=colours[i],ec='k',lw=1,alpha=1.0,align='center',zorder=10) ## plot stack\n",
    "        \n",
    "ax.set_yticks(range(len(ylabels)))\n",
    "ax.set_yticklabels(ylabels)\n",
    "ax.set_ylim(-0.5,len(ylabels)-0.5)\n",
    "\n",
    "ax.set_xlabel('read count',size=30)\n",
    "\n",
    "customfmt=lambda y,pos: r'$%.2f\\times10^{%d}$'%(y*(10**-int(np.log10(y))),int(np.log10(y))) if y!=0.0 else '0.0'\n",
    "\n",
    "ax.xaxis.set_major_formatter(mpl.ticker.FuncFormatter(customfmt)) ## custom log formatted tick labels\n",
    "ax.tick_params(size=0,labelsize=22)\n",
    "ax.tick_params(axis='x',labelsize=28)\n",
    "\n",
    "ax.grid(axis='x',ls='--',zorder=0)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
