---
title: "Generate figures"
author: "Lucy M. Li"
date: "`r format(Sys.Date(), '%b %d, %Y')`"
output: 
  html_document:
    fig_caption: yes
---


# Notebook setup



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE)
# !diagnostics off
library(parallel)
library(dplyr)
library(tidyr)
library(readr)
library(magrittr)
library(ggplot2)
library(grid)
library(gridExtra)
library(stringr)
library(reutils)
library(rjson)
library(knitr)
# library(broom)
# library(sp)
# library(rgdal)
# library(maptools)
# library(OpenStreetMap)
# library(ggmap)
# library(XML)
# library(xtable)
# library(tibble)
# library(ape)
# library(seqinr)
# library(ggtree)
# library(kmer)
```

```{r directories}
data_dir <- "../../data"
fig_dir <- "../../figures/fig2"
if (!dir.exists(data_dir)) dir.create(data_dir)
if (!dir.exists(fig_dir)) dir.create(fig_dir)
```


```{r scripts}
#source("generate_figures_functions.R")
```

# Metadata

```{r load_data}
metadata <- read.csv(list.files(data_dir, "CMS001_CMS002_Merged", full.names = TRUE), stringsAsFactors = FALSE) %>%
  mutate(collected_by=gsub(" ", "", collected_by)) %>%
  mutate(collection_date=as.Date(collection_date)) %>%
  mutate(compute_genus_species=paste(compute_genus, compute_species)) %>% 
  left_join(., read.csv(file.path(data_dir, "project-mosquito_sample-table.csv")) %>% 
              rename(NewIDseqName=sample_name) %>% 
              select(NewIDseqName, total_reads, nonhost_reads, subsampled_fraction, total_ercc_reads, nucleotide_type)
            )

total_mosquito_nonhost_reads <- metadata %>%
  filter(., !grepl("ater", compute_genus_species)) %>%
  group_by(, compute_genus_species, collected_by) %>%
  summarize(total_mosquitos=n(), total_nonhost=sum(nonhost_reads), mean_nonhost=mean(nonhost_reads))
```



```{r summary_metadata}
summary_metadata <- metadata %>%
  filter(!is.na(collected_by)) %>%
  group_by(collected_by, compute_genus_species) %>%
  summarize(n=n()) %>%
  spread(key=compute_genus_species, value=n, fill=0) %>%
  left_join(group_by(metadata, collected_by) %>% summarize_at(paste0("collection_", c("lat", "long")), mean, na.rm=TRUE), by="collected_by") %>%
  mutate(collected_by_display=c(ALCO="Alameda", PLCR="Placer", WVAL="West Valley", COAV="Coachella Valley", SAND="San Diego")[collected_by])

kable(summary_metadata)
```


# Read in data

For each group, get the taxid of each taxonomic group.

```{r taxonomy}
groups_taxid <- sapply(c("viruses", "eubacteria", "culicidae", "eukaryota", "fungi", "chordata", "nematoda", "apicomplexa"), function (x) {
  esearch(x, db="taxonomy") %>%
    content(., "parsed") %>%
    as.character(.) %>%
    as.numeric(.)
})
groups_taxid
```


Read in blast results.

```{r blast_results, cache=TRUE, warning=FALSE, message=FALSE}
# full_lca_results.txt generated by running visualization.ipynb in the skeeters/analysis/contig_blast_analysis folder
# jupyter nbconvert --to notebook --execute visualization.ipynb
blast_results <- read_tsv(file.path(data_dir, "full_lca_results.txt")) %>% 
  # Calculate the proportion of contig covered by each blast hit
  mutate(align_prop=align_length*sapply(blast_type=="nr", ifelse, 3, 1)/as.numeric(str_split(query, "_", simplify=TRUE)[, 4])) %>%
  left_join(., metadata, by=c("sample"="NewIDseqName"), suffix=c("", ".y")) %>%
  select(-ends_with(".y"))  %>%
  # Add total read counts
  left_join(., total_mosquito_nonhost_reads, by=c("compute_genus_species", "collected_by"))
head(blast_results)
```

For each taxid found in each sample, the proportion of nonhost reads that belong to that taxid (according to nt and nr, separately) is calculated below.

```{r blast_results_summary}
blast_results_summary <- blast_results %>%
  group_by(taxid, sample, blast_type) %>%
  summarize(read_count=sum(read_count), nonhost_reads=first(nonhost_reads),
            total_mosquitos=first(total_mosquitos), sci_name=first(sci_name),
            lineage=first(lineage)) %>%
  ungroup() %>%
  mutate(prop_nonhost=read_count/nonhost_reads)
head(blast_results_summary)
```

# Viruses

```{r viral_reads, message=FALSE, warning=FALSE}
viral_contigs <- blast_results %>%
  filter(grepl(groups_taxid["viruses"] %>% paste0(", ", ., ", "), lineage), blast_type=="nr")
viral_thresholds <- c(align_prop=0.8, identity=0.8)
```

Contigs are assigned to known viruses if they are at least `r scales::percent(viral_thresholds[["identity"]], accuracy=1)` identical across at least `r scales::percent(viral_thresholds[["align_prop"]], accuracy=1)` of the bases to a known virus.


```{r close_viral_hits}
close_viral_hits <- viral_contigs %>%
  filter(., align_prop > viral_thresholds[["align_prop"]]/3, identity > (viral_thresholds[["identity"]]*100)) %>%
  mutate(sample=factor(sample, levels=grep("ater", metadata$NewIDseqName, invert=TRUE, value=TRUE)))
write.table(close_viral_hits, file.path(data_dir, "close_viral_hits.txt"), sep="\t", row.names=FALSE, quote=FALSE)
```

## Per mosquito viral composition

```{r per_mosquito_virus_composition, fig.width=15, fig.height=6}
per_mosquito_virus_df <- close_viral_hits %>%
  filter(., (read_count/nonhost_reads)>median(read_count/nonhost_reads)) %>%
  mutate(sample=factor(sample, levels=grep("ater", metadata$NewIDseqName, value=TRUE, invert=TRUE))) %>%
  mutate(sample=as.character(sample) %>% gsub("CMS", "", .) %>% str_split(., "_") %>% sapply(function (x) paste(x[1:2], collapse="_")))
per_mosquito_virus_plot <- list()
per_mosquito_virus_plot$by_location_species <-
  ggplot(per_mosquito_virus_df) +
  theme_bw() +
  geom_bar(aes(x=sample, y=read_count/nonhost_reads, fill=sci_name), stat="identity") +
  facet_grid(~compute_genus_species+collected_by, space="free_x", scale="free_x") +
  guides(fill=guide_legend(title="Virus species", nrow=5, label.theme=element_text(size=12), keywidth=1, keyheight=1)) +
  scale_y_continuous(labels=scales::percent) +
  theme(axis.text.x=element_text(angle=45, hjust=1), legend.position="top",
        strip.text=element_text(angle=90)) +
  xlab("Sample") + ylab("Nonhost reads")

per_mosquito_virus_plot$plain <- 
  per_mosquito_virus_df %>%
  split(., paste(.$compute_genus_species, .$collected_by)) %>%
  lapply(function (x) {
    all_samples <- metadata %>%
      filter(., collected_by==x$collected_by[1], compute_genus_species==x$compute_genus_species[1]) %>%
      `[[`("NewIDseqName")
    ordered_samples <- group_by(x, sample) %>%
      summarise(sum=sum(read_count)) %>%
      with(sample[order(sum)]) %>%
      match(., x$sample) %>%
      slice(x, .)
  }) %>%
  do.call(what=rbind) %>%
  ggplot() %>%
  `+`(list(theme_bw(),
    geom_bar(aes(x=sample, y=read_count/nonhost_reads, fill=sci_name), stat="identity"),
    guides(fill=guide_legend(title="Virus species", nrow=5, label.theme=element_text(size=12), keywidth=1, keyheight=1)),
    scale_y_continuous(labels=scales::percent),
    theme(axis.text.x=element_text(angle=45, hjust=1), legend.position="top"),
    xlab("Sample"), ylab("Nonhost reads")))
    

# 
# 
# per_mosquito_virus_df <- filter(close_viral_hits, !is.na(sample)) %>%
#   mutate(taxup=lineage %>% lapply(fromJSON) %>% sapply(function (x) max(groups_taxid[["viruses"]], x[length(x)-2]))) %>%
#   group_by(., sample, taxup) %>%
#   summarise(., read_count=sum(read_count), ) %>% 
#   left_join(., select(close_viral_hits, -read_count), by=c("sample")) %>%
#   ungroup() %>%
#   mutate(sample=as.character(sample) %>% gsub("CMS", "", .) %>% str_split(., "_") %>% sapply(function (x) paste(x[1:2], collapse="_")))
# per_mosquito_virus_df <- 
#   esummary(unique(per_mosquito_virus_df$taxup), db="taxonomy") %>% 
#   content("parse") %>%
#   select(Id, ScientificName) %>%
#   mutate(Id=as.numeric(Id)) %>%
#   left_join(per_mosquito_virus_df, ., by=c("taxup"="Id"))
# 
# 
# per_mosquito_virus_plot <- ggplot(per_mosquito_virus_df) %>%
#   `+`(list(theme_bw(),
#            geom_bar(aes(x=sample, y=read_count/nonhost_reads, fill=ScientificName), stat="identity"),
#            facet_grid(~collected_by+compute_genus_species, space="free_x", scale="free_x"),
#            guides(fill=guide_legend(title="Virus species", nrow=3, label.theme=element_text(size=14), keywidth=1, keyheight=1)),
#            scale_y_continuous(labels=scales::percent),
#            theme(axis.text.x=element_text(angle=45, hjust=1), legend.position="top"),
#            xlab("Sample"),
#            ylab("Nonhost reads")
#           )
#      )
ggsave(file.path(fig_dir, "per_mosquito_virus.pdf"), per_mosquito_virus_plot$plain,
       width=15, height=6)
ggsave(file.path(fig_dir, "per_mosquito_virus_by_loc_spp.pdf"), per_mosquito_virus_plot$by_location_species,
       width=18, height=6)

per_mosquito_virus_plot$plain
```

## Viral prevalence vs. abundance

For each virus taxid, I calculated the prevalence of that virus within a single mosquito species at a collection site.

```{r viral_prevalence}
close_viral_prev <- close_viral_hits %>%
  group_by(collected_by, compute_genus_species, taxid) %>%
  summarise(num_mosquitos=n_distinct(sample), read_prop=mean(read_count/nonhost_reads)) %>%
  left_join(., close_viral_hits[, c("collected_by", "compute_genus_species", "taxid", "sci_name", "total_mosquitos")], by=c("collected_by", "compute_genus_species", "taxid"))
```

For sites where at least 10 mosquitos of a given species were collected, the proportion of mosquitos containing a particular virus was compared to the mean abundance of that virus across that species of mosquito at that site.

```{r prev_abundance, fig.width=4, fig.height=3}
viral_prev_abundance_plot <- ggplot(filter(close_viral_prev, total_mosquitos>=10), aes(x=read_prop, y=num_mosquitos/total_mosquitos)) + 
  theme_bw() +
  geom_point() +
  scale_x_log10(labels=scales::percent) +
  scale_y_continuous(labels=scales::percent) +
  xlab("Bulk abundance\n(proportion of nonhost reads)") +
  ylab("Prevalence\n(proportion of mosquito)")
ggsave(file.path(fig_dir, "viral_prev_abundance.pdf"), width=4, height=3)
viral_prev_abundance_plot
```


## Save output

```{r output}
save.image(file="generate_figures.RData")
```

