{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from ete3 import NCBITaxa\n",
    "import boto3\n",
    "import tempfile\n",
    "import subprocess\n",
    "import os\n",
    "import io\n",
    "import re\n",
    "import time\n",
    "import json\n",
    "import pdb\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from lca_functions import *\n",
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [],
   "source": [
    "identity_qcov_threshold = 0.9\n",
    "min_total_contig_len = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load metadata\n",
    "metadata = pd.read_csv(\"../../data/metadata/CMS001_CMS002_MergedAnnotations.csv\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read counts data\n",
    "all_read_counts = pd.read_csv(\"s3://czbiohub-mosquito/contig_quality_concat/contig_counts.csv\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load data\n",
    "contig_stats_lca = pd.read_csv(\"s3://czbiohub-mosquito/contig_quality_concat/contig_stats_lca.tsv\", sep=\"\\t\", header=0)\n",
    "contig_stats_lca = contig_stats_lca.assign(identity_qcov=(contig_stats_lca[\"identity\"]/100*contig_stats_lca[\"align_length\"]/contig_stats_lca[\"contig_length\"]).apply(lambda x: min(x, 1)))\n",
    "contig_stats_lca = contig_stats_lca[~(contig_stats_lca[\"hexapoda\"] | contig_stats_lca[\"taxid\"].isnull())]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "contig_stats_lca = pd.merge(pd.merge(contig_stats_lca, metadata, how=\"left\", left_on=\"sample\", right_on=\"NewIDseqName\"), \n",
    "                            all_read_counts.groupby(\"sample\").sum().reset_index().rename(columns={\"counts\":\"total_nonhost\"}),\n",
    "                           how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [],
   "source": [
    "contig_stats_lca[\"read_prop\"] = contig_stats_lca[\"read_count\"]/contig_stats_lca[\"total_nonhost\"]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rows_taxid (df, taxid, taxid_colname=\"taxid\", identity_qcov_cutoff=None):\n",
    "    if (not isinstance(df, pd.DataFrame)):\n",
    "        return (taxid in ncbi.get_lineage(df))\n",
    "    outdf = df[df[taxid_colname].apply(get_rows_taxid, taxid=taxid)]\n",
    "    if identity_qcov_cutoff is not None:\n",
    "        outdf = outdf[outdf[\"identity_qcov\"]>=identity_qcov_cutoff]\n",
    "    return (outdf)\n",
    "\n",
    "def filter_using_water_by_criterion (df, colname, split_by_taxid=True):\n",
    "    if (len(df)==0):\n",
    "        return (df)\n",
    "    if split_by_taxid:\n",
    "        water_counts = df[df[\"sample\"].apply(lambda x: \"ater\" in x)].groupby([\"taxid\", \"sample\"])[colname].sum().reset_index().groupby(\"taxid\").max()\n",
    "        df_groupby = df.groupby([\"taxid\", \"sample\"])\n",
    "        sample_sums = df_groupby[colname].sum().reset_index()\n",
    "        for i, x in enumerate(water_counts[colname]):\n",
    "            sample_sums[\"tokeep\"] = sample_sums.loc[sample_sums[\"taxid\"]==water_counts.index[i], colname] > x\n",
    "            df = pd.merge(df, sample_sums.drop(columns=colname), how=\"left\")\n",
    "            df = df[df[\"tokeep\"]!=False].drop(columns=\"tokeep\")\n",
    "    else:\n",
    "        metrics = df.groupby(\"sample\")[colname].sum().sort_values()\n",
    "        selected_samples = metrics.index[([i for i, x in enumerate(metrics.index) if 'ater' in x][-1]+1):]\n",
    "        df = df[df[\"sample\"].isin(selected_samples)]\n",
    "    return (df)\n",
    "    \n",
    "def use_water_filter (df, by_read_count=True, by_length=True, split_by_taxid=True):\n",
    "    if (not df[\"sample\"].apply(lambda x: \"ater\" in x).any()):\n",
    "        return (df)\n",
    "    if split_by_taxid:\n",
    "        if (by_read_count):\n",
    "            df = filter_using_water_by_criterion(df, \"read_count\", split_by_taxid)\n",
    "        if (by_length):\n",
    "            df = filter_using_water_by_criterion(df, \"contig_length\", split_by_taxid)\n",
    "    else:\n",
    "        if (by_read_count):\n",
    "            filter_using_water_by_criterion(df, \"read_count\", split_by_taxid)\n",
    "        if (by_length):\n",
    "            filter_using_water_by_criterion(df, \"contig_length\", split_by_taxid)\n",
    "    return (df)\n",
    "\n",
    "def filter_by_criterion (df, colname, minthreshold, bysample=True):\n",
    "    if bysample:\n",
    "        sums = df.groupby([\"sample\", \"taxid\"])[colname].sum().reset_index()\n",
    "        sums[\"tokeep\"] = sums[colname] >= minthreshold\n",
    "        df = pd.merge(df, sums.drop(columns=colname), how=\"left\")\n",
    "        df = df[df[\"tokeep\"]!=False].drop(columns=\"tokeep\")\n",
    "    else:\n",
    "        df = df[df[colname] >= minthreshold]\n",
    "    return (df)\n",
    "\n",
    "\n",
    "def check_if_in_any_taxid(taxid, taxid_list):\n",
    "    if taxid in taxid_list:\n",
    "        return (taxid)\n",
    "    taxids = ncbi.get_lineage(taxid)\n",
    "    check_in = [i for i, x in enumerate(taxids) if x in taxid_list]\n",
    "    if (len(check_in)==0):\n",
    "        return (np.nan)\n",
    "    return (taxids[check_in[0]])\n",
    "\n",
    "\n",
    "\n",
    "def clean_taxids(df, taxids, root_taxid, taxid_colname=\"taxid\"):\n",
    "    if isinstance(taxids[0], str):\n",
    "        taxids = dict(zip([ncbi.get_name_translator([x])[x][0] for x in taxids], taxids))\n",
    "    else:\n",
    "        taxids = ncbi.get_taxid_translator(taxids)\n",
    "    if isinstance(root_taxid, str):\n",
    "        root_taxid_number = ncbi.get_name_translator([root_taxid])[root_taxid][0]\n",
    "        root_taxid = {root_taxid_number:root_taxid}\n",
    "    else:\n",
    "        root_taxid = ncbi.get_taxid_translator([root_taxid])\n",
    "    df[taxid_colname] = df[taxid_colname].apply(check_if_in_any_taxid, taxid_list=taxids)\n",
    "    df[taxid_colname][df[taxid_colname].isnull()] = list(root_taxid.keys())[0]\n",
    "    taxids.update(root_taxid)\n",
    "    df[\"sci_name\"] = df[taxid_colname].apply(lambda x: taxids[x])\n",
    "    return (df)\n",
    "    \n",
    "\n",
    "def get_summary_table (df, colnames, metric):\n",
    "    df = df.groupby(colnames)[metric].sum().reset_index()\n",
    "    sort_order = colnames+[metric]\n",
    "    sort_order.remove(\"sample\")\n",
    "    return (df.sort_values(by=sort_order, ascending=False))\n",
    "\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 197514.0 was translated into 2555385\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814.0 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ska_genus</th>\n",
       "      <th>ska_species</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>sample</th>\n",
       "      <th>read_prop</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sci_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_046a_Rb_S191_L004</td>\n",
       "      <td>0.133324</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_027a_Rb_S152_L004</td>\n",
       "      <td>0.117018</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_027b_Rb_S153_L004</td>\n",
       "      <td>0.111187</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_046b_Rb_S192_L004</td>\n",
       "      <td>0.109397</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_013a_Rb_S120_L004</td>\n",
       "      <td>0.126148</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_001a_Rb_S116_L004</td>\n",
       "      <td>0.113309</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_004a_Rb_S117_L004</td>\n",
       "      <td>0.109971</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_010a_Rb_S119_L004</td>\n",
       "      <td>0.104038</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_022a_Rb_S137_L004</td>\n",
       "      <td>0.053287</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_023a_Rb_S138_L004</td>\n",
       "      <td>0.015931</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>PLCR</td>\n",
       "      <td>CMS002_050a_Rb_S5_L004</td>\n",
       "      <td>0.147540</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>PLCR</td>\n",
       "      <td>CMS002_054a_Rb_S8_L004</td>\n",
       "      <td>0.059391</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>PLCR</td>\n",
       "      <td>CMS002_051a_Rb_S6_L004</td>\n",
       "      <td>0.039792</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>PLCR</td>\n",
       "      <td>CMS002_057a_Rb_S10_L004</td>\n",
       "      <td>0.038874</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>0.462327</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_054_Ra_S11</td>\n",
       "      <td>0.459820</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_059_Ra_S10</td>\n",
       "      <td>0.423869</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_030_Ra_S7</td>\n",
       "      <td>0.380236</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_041_Ra_S10</td>\n",
       "      <td>0.354086</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_044_Ra_S25</td>\n",
       "      <td>0.095463</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_029_Ra_S18</td>\n",
       "      <td>0.036827</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026e_Rb_S150_L004</td>\n",
       "      <td>0.117545</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026a_Rb_S146_L004</td>\n",
       "      <td>0.107003</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045a_Rb_S183_L004</td>\n",
       "      <td>0.104032</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026d_Rb_S149_L004</td>\n",
       "      <td>0.053729</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026b_Rb_S147_L004</td>\n",
       "      <td>0.050962</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026c_Rb_S148_L004</td>\n",
       "      <td>0.031594</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045g_Rb_S190_L004</td>\n",
       "      <td>0.013202</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045e_Rb_S188_L004</td>\n",
       "      <td>0.006265</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045f_Rb_S189_L004</td>\n",
       "      <td>0.005276</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045d_Rb_S186_L004</td>\n",
       "      <td>0.003834</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045b_Rb_S184_L004</td>\n",
       "      <td>0.001966</td>\n",
       "      <td>952</td>\n",
       "      <td>Wolbachieae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ska_genus       ska_species collected_by                    sample  \\\n",
       "30     Culex  quinquefasciatus        WVAL   CMS002_046a_Rb_S191_L004   \n",
       "28     Culex  quinquefasciatus        WVAL   CMS002_027a_Rb_S152_L004   \n",
       "29     Culex  quinquefasciatus        WVAL   CMS002_027b_Rb_S153_L004   \n",
       "31     Culex  quinquefasciatus        WVAL   CMS002_046b_Rb_S192_L004   \n",
       "25     Culex  quinquefasciatus         SAND  CMS002_013a_Rb_S120_L004   \n",
       "22     Culex  quinquefasciatus         SAND  CMS002_001a_Rb_S116_L004   \n",
       "23     Culex  quinquefasciatus         SAND  CMS002_004a_Rb_S117_L004   \n",
       "24     Culex  quinquefasciatus         SAND  CMS002_010a_Rb_S119_L004   \n",
       "26     Culex  quinquefasciatus         SAND  CMS002_022a_Rb_S137_L004   \n",
       "27     Culex  quinquefasciatus         SAND  CMS002_023a_Rb_S138_L004   \n",
       "18     Culex           pipiens         PLCR    CMS002_050a_Rb_S5_L004   \n",
       "20     Culex           pipiens         PLCR    CMS002_054a_Rb_S8_L004   \n",
       "19     Culex           pipiens         PLCR    CMS002_051a_Rb_S6_L004   \n",
       "21     Culex           pipiens         PLCR   CMS002_057a_Rb_S10_L004   \n",
       "11     Culex           pipiens         ALCO          CMS001_004_Ra_S2   \n",
       "16     Culex           pipiens         ALCO         CMS001_054_Ra_S11   \n",
       "17     Culex           pipiens         ALCO         CMS001_059_Ra_S10   \n",
       "13     Culex           pipiens         ALCO          CMS001_030_Ra_S7   \n",
       "14     Culex           pipiens         ALCO         CMS001_041_Ra_S10   \n",
       "15     Culex           pipiens         ALCO         CMS001_044_Ra_S25   \n",
       "12     Culex           pipiens         ALCO         CMS001_029_Ra_S18   \n",
       "4      Aedes        albopictus        WVAL   CMS002_026e_Rb_S150_L004   \n",
       "0      Aedes        albopictus        WVAL   CMS002_026a_Rb_S146_L004   \n",
       "5      Aedes        albopictus        WVAL   CMS002_045a_Rb_S183_L004   \n",
       "3      Aedes        albopictus        WVAL   CMS002_026d_Rb_S149_L004   \n",
       "1      Aedes        albopictus        WVAL   CMS002_026b_Rb_S147_L004   \n",
       "2      Aedes        albopictus        WVAL   CMS002_026c_Rb_S148_L004   \n",
       "10     Aedes        albopictus        WVAL   CMS002_045g_Rb_S190_L004   \n",
       "8      Aedes        albopictus        WVAL   CMS002_045e_Rb_S188_L004   \n",
       "9      Aedes        albopictus        WVAL   CMS002_045f_Rb_S189_L004   \n",
       "7      Aedes        albopictus        WVAL   CMS002_045d_Rb_S186_L004   \n",
       "6      Aedes        albopictus        WVAL   CMS002_045b_Rb_S184_L004   \n",
       "\n",
       "    read_prop  taxid     sci_name  \n",
       "30   0.133324    952  Wolbachieae  \n",
       "28   0.117018    952  Wolbachieae  \n",
       "29   0.111187    952  Wolbachieae  \n",
       "31   0.109397    952  Wolbachieae  \n",
       "25   0.126148    952  Wolbachieae  \n",
       "22   0.113309    952  Wolbachieae  \n",
       "23   0.109971    952  Wolbachieae  \n",
       "24   0.104038    952  Wolbachieae  \n",
       "26   0.053287    952  Wolbachieae  \n",
       "27   0.015931    952  Wolbachieae  \n",
       "18   0.147540    952  Wolbachieae  \n",
       "20   0.059391    952  Wolbachieae  \n",
       "19   0.039792    952  Wolbachieae  \n",
       "21   0.038874    952  Wolbachieae  \n",
       "11   0.462327    952  Wolbachieae  \n",
       "16   0.459820    952  Wolbachieae  \n",
       "17   0.423869    952  Wolbachieae  \n",
       "13   0.380236    952  Wolbachieae  \n",
       "14   0.354086    952  Wolbachieae  \n",
       "15   0.095463    952  Wolbachieae  \n",
       "12   0.036827    952  Wolbachieae  \n",
       "4    0.117545    952  Wolbachieae  \n",
       "0    0.107003    952  Wolbachieae  \n",
       "5    0.104032    952  Wolbachieae  \n",
       "3    0.053729    952  Wolbachieae  \n",
       "1    0.050962    952  Wolbachieae  \n",
       "2    0.031594    952  Wolbachieae  \n",
       "10   0.013202    952  Wolbachieae  \n",
       "8    0.006265    952  Wolbachieae  \n",
       "9    0.005276    952  Wolbachieae  \n",
       "7    0.003834    952  Wolbachieae  \n",
       "6    0.001966    952  Wolbachieae  "
      ]
     },
     "execution_count": 387,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known wolbachia sequence\n",
    "wolbachia_taxid = 952\n",
    "wolbachia_contigs = get_rows_taxid(contig_stats_lca, taxid=wolbachia_taxid, taxid_colname=\"taxid\", identity_qcov_cutoff=identity_qcov_threshold)\n",
    "# Filter contigs that were less abundant or shorter than contigs found in water belonging to the same taxid\n",
    "wolbachia_contigs = use_water_filter(wolbachia_contigs, by_read_count=True, by_length=True)\n",
    "# Filter taxids for which the total length of contigs in a sample was less than min_total_contig_len (default 1000 bp)\n",
    "wolbachia_contigs = filter_by_criterion(wolbachia_contigs, \"contig_length\", minthreshold=min_total_contig_len, bysample=True)\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "wolbachia_contigs = get_summary_table(wolbachia_contigs, colnames=[\"ska_genus\", \"ska_species\", \"collected_by\", \"sample\"], metric=\"read_prop\")\n",
    "wolbachia_contigs = wolbachia_contigs.assign(taxid=wolbachia_taxid)\n",
    "wolbachia_contigs = wolbachia_contigs.assign(sci_name=ncbi.get_taxid_translator([wolbachia_taxid])[wolbachia_taxid])\n",
    "wolbachia_contigs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 197514.0 was translated into 2555385\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814.0 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:74: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ska_genus</th>\n",
       "      <th>ska_species</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>sample</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>read_prop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_014_Ra_S5</td>\n",
       "      <td>35325.0</td>\n",
       "      <td>dsRNA viruses</td>\n",
       "      <td>0.000483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>35325.0</td>\n",
       "      <td>dsRNA viruses</td>\n",
       "      <td>0.000075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029e_Rb_S164_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.114450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029d_Rb_S162_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.096161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029b_Rb_S160_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.163665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029c_Rb_S161_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.127774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029d_Rb_S162_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.127377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029a_Rb_S159_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.120244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029e_Rb_S164_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.106881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029c_Rb_S161_L004</td>\n",
       "      <td>699189.0</td>\n",
       "      <td>Iflaviridae</td>\n",
       "      <td>0.000334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029d_Rb_S162_L004</td>\n",
       "      <td>186766.0</td>\n",
       "      <td>Narnaviridae</td>\n",
       "      <td>0.175359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029c_Rb_S161_L004</td>\n",
       "      <td>186766.0</td>\n",
       "      <td>Narnaviridae</td>\n",
       "      <td>0.056292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029e_Rb_S164_L004</td>\n",
       "      <td>11308.0</td>\n",
       "      <td>Orthomyxoviridae</td>\n",
       "      <td>0.035058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029c_Rb_S161_L004</td>\n",
       "      <td>11308.0</td>\n",
       "      <td>Orthomyxoviridae</td>\n",
       "      <td>0.033683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029d_Rb_S162_L004</td>\n",
       "      <td>11308.0</td>\n",
       "      <td>Orthomyxoviridae</td>\n",
       "      <td>0.032036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029a_Rb_S159_L004</td>\n",
       "      <td>11006.0</td>\n",
       "      <td>Totiviridae</td>\n",
       "      <td>0.000047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029b_Rb_S160_L004</td>\n",
       "      <td>11006.0</td>\n",
       "      <td>Totiviridae</td>\n",
       "      <td>0.000044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029e_Rb_S164_L004</td>\n",
       "      <td>11006.0</td>\n",
       "      <td>Totiviridae</td>\n",
       "      <td>0.000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>PLCR</td>\n",
       "      <td>CMS002_056a_Rb_S9_L004</td>\n",
       "      <td>12283.0</td>\n",
       "      <td>Nodaviridae</td>\n",
       "      <td>0.095015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>PLCR</td>\n",
       "      <td>CMS002_056a_Rb_S9_L004</td>\n",
       "      <td>11308.0</td>\n",
       "      <td>Orthomyxoviridae</td>\n",
       "      <td>0.078784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_035a_Rb_S169_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.108252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_040a_Rb_S174_L004</td>\n",
       "      <td>1980410.0</td>\n",
       "      <td>Bunyavirales</td>\n",
       "      <td>0.000338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_034a_Rb_S168_L004</td>\n",
       "      <td>699189.0</td>\n",
       "      <td>Iflaviridae</td>\n",
       "      <td>0.018459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_039a_Rb_S173_L004</td>\n",
       "      <td>699189.0</td>\n",
       "      <td>Iflaviridae</td>\n",
       "      <td>0.004733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_042a_Rb_S177_L004</td>\n",
       "      <td>12283.0</td>\n",
       "      <td>Nodaviridae</td>\n",
       "      <td>0.000669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_035a_Rb_S169_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.302652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_031a_Rb_S165_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.127679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_041a_Rb_S176_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.091415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_040a_Rb_S174_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.052688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_031a_Rb_S165_L004</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.000205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_018_Ra_S14</td>\n",
       "      <td>12283.0</td>\n",
       "      <td>Nodaviridae</td>\n",
       "      <td>0.010327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_057_Ra_S11</td>\n",
       "      <td>12283.0</td>\n",
       "      <td>Nodaviridae</td>\n",
       "      <td>0.009620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_015_Ra_S13</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.000427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_053_Ra_S8</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.000193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_035_Ra_S20</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.000079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045e_Rb_S188_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>2.709435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045c_Rb_S185_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.690800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045g_Rb_S190_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.685174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045d_Rb_S186_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.634507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045f_Rb_S189_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.447428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045a_Rb_S183_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.223747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045b_Rb_S184_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.049351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045c_Rb_S185_L004</td>\n",
       "      <td>699189.0</td>\n",
       "      <td>Iflaviridae</td>\n",
       "      <td>0.000081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_045g_Rb_S190_L004</td>\n",
       "      <td>699189.0</td>\n",
       "      <td>Iflaviridae</td>\n",
       "      <td>0.000068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044b_Rb_S179_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.066176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044c_Rb_S180_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.037406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044e_Rb_S182_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.024078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044b_Rb_S179_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.022059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044a_Rb_S178_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.021384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044d_Rb_S181_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.003480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044b_Rb_S179_L004</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.465074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044c_Rb_S180_L004</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.296758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044a_Rb_S178_L004</td>\n",
       "      <td>10239.0</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.159119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017c_Rb_S124_L004</td>\n",
       "      <td>2585030.0</td>\n",
       "      <td>unclassified Riboviria</td>\n",
       "      <td>0.590061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_018a_Rb_S128_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.103608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017c_Rb_S124_L004</td>\n",
       "      <td>11157.0</td>\n",
       "      <td>Mononegavirales</td>\n",
       "      <td>0.065910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017d_Rb_S125_L004</td>\n",
       "      <td>11051.0</td>\n",
       "      <td>Flavivirus</td>\n",
       "      <td>0.116310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017c_Rb_S124_L004</td>\n",
       "      <td>11051.0</td>\n",
       "      <td>Flavivirus</td>\n",
       "      <td>0.002516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017e_Rb_S126_L004</td>\n",
       "      <td>11051.0</td>\n",
       "      <td>Flavivirus</td>\n",
       "      <td>0.001181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017a_Rb_S122_L004</td>\n",
       "      <td>11051.0</td>\n",
       "      <td>Flavivirus</td>\n",
       "      <td>0.000492</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>190 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ska_genus    ska_species collected_by                    sample  \\\n",
       "189  Culiseta      particeps         ALCO          CMS001_014_Ra_S5   \n",
       "188  Culiseta       incidens         ALCO          CMS001_005_Ra_S3   \n",
       "187     Culex       tarsalis        WVAL   CMS002_029e_Rb_S164_L004   \n",
       "183     Culex       tarsalis        WVAL   CMS002_029d_Rb_S162_L004   \n",
       "175     Culex       tarsalis        WVAL   CMS002_029b_Rb_S160_L004   \n",
       "179     Culex       tarsalis        WVAL   CMS002_029c_Rb_S161_L004   \n",
       "182     Culex       tarsalis        WVAL   CMS002_029d_Rb_S162_L004   \n",
       "173     Culex       tarsalis        WVAL   CMS002_029a_Rb_S159_L004   \n",
       "186     Culex       tarsalis        WVAL   CMS002_029e_Rb_S164_L004   \n",
       "178     Culex       tarsalis        WVAL   CMS002_029c_Rb_S161_L004   \n",
       "181     Culex       tarsalis        WVAL   CMS002_029d_Rb_S162_L004   \n",
       "177     Culex       tarsalis        WVAL   CMS002_029c_Rb_S161_L004   \n",
       "185     Culex       tarsalis        WVAL   CMS002_029e_Rb_S164_L004   \n",
       "176     Culex       tarsalis        WVAL   CMS002_029c_Rb_S161_L004   \n",
       "180     Culex       tarsalis        WVAL   CMS002_029d_Rb_S162_L004   \n",
       "172     Culex       tarsalis        WVAL   CMS002_029a_Rb_S159_L004   \n",
       "174     Culex       tarsalis        WVAL   CMS002_029b_Rb_S160_L004   \n",
       "184     Culex       tarsalis        WVAL   CMS002_029e_Rb_S164_L004   \n",
       "171     Culex       tarsalis         PLCR    CMS002_056a_Rb_S9_L004   \n",
       "170     Culex       tarsalis         PLCR    CMS002_056a_Rb_S9_L004   \n",
       "164     Culex       tarsalis         COAV  CMS002_035a_Rb_S169_L004   \n",
       "167     Culex       tarsalis         COAV  CMS002_040a_Rb_S174_L004   \n",
       "162     Culex       tarsalis         COAV  CMS002_034a_Rb_S168_L004   \n",
       "165     Culex       tarsalis         COAV  CMS002_039a_Rb_S173_L004   \n",
       "169     Culex       tarsalis         COAV  CMS002_042a_Rb_S177_L004   \n",
       "163     Culex       tarsalis         COAV  CMS002_035a_Rb_S169_L004   \n",
       "161     Culex       tarsalis         COAV  CMS002_031a_Rb_S165_L004   \n",
       "168     Culex       tarsalis         COAV  CMS002_041a_Rb_S176_L004   \n",
       "166     Culex       tarsalis         COAV  CMS002_040a_Rb_S174_L004   \n",
       "160     Culex       tarsalis         COAV  CMS002_031a_Rb_S165_L004   \n",
       "..        ...            ...          ...                       ...   \n",
       "34      Culex  erythrothorax         ALCO         CMS001_018_Ra_S14   \n",
       "65      Culex  erythrothorax         ALCO         CMS001_057_Ra_S11   \n",
       "30      Culex  erythrothorax         ALCO         CMS001_015_Ra_S13   \n",
       "61      Culex  erythrothorax         ALCO          CMS001_053_Ra_S8   \n",
       "47      Culex  erythrothorax         ALCO         CMS001_035_Ra_S20   \n",
       "21      Aedes     albopictus        WVAL   CMS002_045e_Rb_S188_L004   \n",
       "19      Aedes     albopictus        WVAL   CMS002_045c_Rb_S185_L004   \n",
       "24      Aedes     albopictus        WVAL   CMS002_045g_Rb_S190_L004   \n",
       "20      Aedes     albopictus        WVAL   CMS002_045d_Rb_S186_L004   \n",
       "22      Aedes     albopictus        WVAL   CMS002_045f_Rb_S189_L004   \n",
       "16      Aedes     albopictus        WVAL   CMS002_045a_Rb_S183_L004   \n",
       "17      Aedes     albopictus        WVAL   CMS002_045b_Rb_S184_L004   \n",
       "18      Aedes     albopictus        WVAL   CMS002_045c_Rb_S185_L004   \n",
       "23      Aedes     albopictus        WVAL   CMS002_045g_Rb_S190_L004   \n",
       "11      Aedes        aegypti        WVAL   CMS002_044b_Rb_S179_L004   \n",
       "13      Aedes        aegypti        WVAL   CMS002_044c_Rb_S180_L004   \n",
       "15      Aedes        aegypti        WVAL   CMS002_044e_Rb_S182_L004   \n",
       "10      Aedes        aegypti        WVAL   CMS002_044b_Rb_S179_L004   \n",
       "8       Aedes        aegypti        WVAL   CMS002_044a_Rb_S178_L004   \n",
       "14      Aedes        aegypti        WVAL   CMS002_044d_Rb_S181_L004   \n",
       "9       Aedes        aegypti        WVAL   CMS002_044b_Rb_S179_L004   \n",
       "12      Aedes        aegypti        WVAL   CMS002_044c_Rb_S180_L004   \n",
       "7       Aedes        aegypti        WVAL   CMS002_044a_Rb_S178_L004   \n",
       "3       Aedes        aegypti         SAND  CMS002_017c_Rb_S124_L004   \n",
       "6       Aedes        aegypti         SAND  CMS002_018a_Rb_S128_L004   \n",
       "2       Aedes        aegypti         SAND  CMS002_017c_Rb_S124_L004   \n",
       "4       Aedes        aegypti         SAND  CMS002_017d_Rb_S125_L004   \n",
       "1       Aedes        aegypti         SAND  CMS002_017c_Rb_S124_L004   \n",
       "5       Aedes        aegypti         SAND  CMS002_017e_Rb_S126_L004   \n",
       "0       Aedes        aegypti         SAND  CMS002_017a_Rb_S122_L004   \n",
       "\n",
       "         taxid                sci_name  read_prop  \n",
       "189    35325.0           dsRNA viruses   0.000483  \n",
       "188    35325.0           dsRNA viruses   0.000075  \n",
       "187  2585030.0  unclassified Riboviria   0.114450  \n",
       "183  2585030.0  unclassified Riboviria   0.096161  \n",
       "175  1980410.0            Bunyavirales   0.163665  \n",
       "179  1980410.0            Bunyavirales   0.127774  \n",
       "182  1980410.0            Bunyavirales   0.127377  \n",
       "173  1980410.0            Bunyavirales   0.120244  \n",
       "186  1980410.0            Bunyavirales   0.106881  \n",
       "178   699189.0             Iflaviridae   0.000334  \n",
       "181   186766.0            Narnaviridae   0.175359  \n",
       "177   186766.0            Narnaviridae   0.056292  \n",
       "185    11308.0        Orthomyxoviridae   0.035058  \n",
       "176    11308.0        Orthomyxoviridae   0.033683  \n",
       "180    11308.0        Orthomyxoviridae   0.032036  \n",
       "172    11006.0             Totiviridae   0.000047  \n",
       "174    11006.0             Totiviridae   0.000044  \n",
       "184    11006.0             Totiviridae   0.000036  \n",
       "171    12283.0             Nodaviridae   0.095015  \n",
       "170    11308.0        Orthomyxoviridae   0.078784  \n",
       "164  1980410.0            Bunyavirales   0.108252  \n",
       "167  1980410.0            Bunyavirales   0.000338  \n",
       "162   699189.0             Iflaviridae   0.018459  \n",
       "165   699189.0             Iflaviridae   0.004733  \n",
       "169    12283.0             Nodaviridae   0.000669  \n",
       "163    11157.0         Mononegavirales   0.302652  \n",
       "161    11157.0         Mononegavirales   0.127679  \n",
       "168    11157.0         Mononegavirales   0.091415  \n",
       "166    11157.0         Mononegavirales   0.052688  \n",
       "160    10239.0                 Viruses   0.000205  \n",
       "..         ...                     ...        ...  \n",
       "34     12283.0             Nodaviridae   0.010327  \n",
       "65     12283.0             Nodaviridae   0.009620  \n",
       "30     10239.0                 Viruses   0.000427  \n",
       "61     10239.0                 Viruses   0.000193  \n",
       "47     10239.0                 Viruses   0.000079  \n",
       "21   2585030.0  unclassified Riboviria   2.709435  \n",
       "19   2585030.0  unclassified Riboviria   0.690800  \n",
       "24   2585030.0  unclassified Riboviria   0.685174  \n",
       "20   2585030.0  unclassified Riboviria   0.634507  \n",
       "22   2585030.0  unclassified Riboviria   0.447428  \n",
       "16   2585030.0  unclassified Riboviria   0.223747  \n",
       "17   2585030.0  unclassified Riboviria   0.049351  \n",
       "18    699189.0             Iflaviridae   0.000081  \n",
       "23    699189.0             Iflaviridae   0.000068  \n",
       "11   2585030.0  unclassified Riboviria   0.066176  \n",
       "13   2585030.0  unclassified Riboviria   0.037406  \n",
       "15     11157.0         Mononegavirales   0.024078  \n",
       "10     11157.0         Mononegavirales   0.022059  \n",
       "8      11157.0         Mononegavirales   0.021384  \n",
       "14     11157.0         Mononegavirales   0.003480  \n",
       "9      10239.0                 Viruses   0.465074  \n",
       "12     10239.0                 Viruses   0.296758  \n",
       "7      10239.0                 Viruses   0.159119  \n",
       "3    2585030.0  unclassified Riboviria   0.590061  \n",
       "6      11157.0         Mononegavirales   0.103608  \n",
       "2      11157.0         Mononegavirales   0.065910  \n",
       "4      11051.0              Flavivirus   0.116310  \n",
       "1      11051.0              Flavivirus   0.002516  \n",
       "5      11051.0              Flavivirus   0.001181  \n",
       "0      11051.0              Flavivirus   0.000492  \n",
       "\n",
       "[190 rows x 7 columns]"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known viral sequence\n",
    "viral_contigs = get_rows_taxid(contig_stats_lca, taxid=10239, taxid_colname=\"taxid\", identity_qcov_cutoff=identity_qcov_threshold)\n",
    "# Filter contigs that were less abundant or shorter than contigs found in water belonging to the same taxid\n",
    "viral_contigs = use_water_filter(viral_contigs, by_read_count=True, by_length=True, split_by_taxid=True)\n",
    "# Filter taxids for which the total length of contigs in a sample was less than 500bp\n",
    "viral_contigs = filter_by_criterion(viral_contigs, \"contig_length\", minthreshold=500, bysample=True)\n",
    "# Convert taxids to those of interest\n",
    "viral_groups = [\"unclassified Riboviria\", \"Bunyavirales\", \"Orthomyxoviridae\", \"Mononegavirales\", \n",
    "                \"Nodaviridae\", \"Flavivirus\", \"Narnaviridae\", \"Totiviridae\", \"Iflaviridae\", \"dsRNA viruses\"]\n",
    "viral_contigs = clean_taxids(viral_contigs, taxids=viral_groups, root_taxid=10239)\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "viral_contigs = get_summary_table(viral_contigs, colnames=[\"ska_genus\", \"ska_species\", \"collected_by\", \"sample\", \"taxid\", \"sci_name\"], metric=\"read_prop\")\n",
    "viral_contigs\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:74: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ska_genus</th>\n",
       "      <th>ska_species</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>sample</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>read_prop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.001395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.003614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_008_Ra_S3</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.000719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_025_Ra_S7</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.000111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>inornata</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_016_Ra_S6</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.001187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>incidens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_005_Ra_S3</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>0.001814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>COAV</td>\n",
       "      <td>CMS002_042a_Rb_S177_L004</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.022357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_031_Ra_S19</td>\n",
       "      <td>33554.0</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>0.013463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_042_Ra_S23</td>\n",
       "      <td>33554.0</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>0.001881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_028_Ra_S17</td>\n",
       "      <td>33554.0</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>0.000693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_038_Ra_S22</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.008261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_036_Ra_S20</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.000106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_046_Ra_S3</td>\n",
       "      <td>9895.0</td>\n",
       "      <td>Bovidae</td>\n",
       "      <td>0.001607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_036_Ra_S20</td>\n",
       "      <td>9895.0</td>\n",
       "      <td>Bovidae</td>\n",
       "      <td>0.000740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_036_Ra_S20</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.011232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_038_Ra_S22</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.004398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_046_Ra_S3</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.000379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_027_Ra_S16</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.142716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_032_Ra_S7</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.068273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_043_Ra_S24</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.045940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_004_Ra_S2</td>\n",
       "      <td>33554.0</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>0.017711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_030_Ra_S7</td>\n",
       "      <td>33554.0</td>\n",
       "      <td>Carnivora</td>\n",
       "      <td>0.010530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_029_Ra_S18</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.031271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_041_Ra_S10</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.022645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_029_Ra_S18</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.092740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_059_Ra_S10</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.031076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_040_Ra_S21</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.000836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_054_Ra_S11</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.000607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_048_Ra_S5</td>\n",
       "      <td>337687.0</td>\n",
       "      <td>Muroidea</td>\n",
       "      <td>0.054966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_049_Ra_S6</td>\n",
       "      <td>337687.0</td>\n",
       "      <td>Muroidea</td>\n",
       "      <td>0.031226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_011_Ra_S4</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.043413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_037_Ra_S21</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.034659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_058_Ra_S9</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.029673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_010_Ra_S1</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.011607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_013_Ra_S5</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.010893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_020_Ra_S15</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.009948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_015_Ra_S13</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.008596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_050_Ra_S23</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.005573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_045_Ra_S2</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.005050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_035_Ra_S20</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.001133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_018_Ra_S14</td>\n",
       "      <td>33208.0</td>\n",
       "      <td>Metazoa</td>\n",
       "      <td>0.001113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_020_Ra_S15</td>\n",
       "      <td>9979.0</td>\n",
       "      <td>Leporidae</td>\n",
       "      <td>0.002296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_013_Ra_S5</td>\n",
       "      <td>9979.0</td>\n",
       "      <td>Leporidae</td>\n",
       "      <td>0.000493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_057_Ra_S11</td>\n",
       "      <td>9895.0</td>\n",
       "      <td>Bovidae</td>\n",
       "      <td>0.004716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_052_Ra_S7</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.008892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_022_Ra_S6</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.008596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_010_Ra_S1</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.004264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_053_Ra_S8</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.004140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_015_Ra_S13</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.003310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.001867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_018_Ra_S14</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.001835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_057_Ra_S11</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.001495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_045_Ra_S2</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.001182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_035_Ra_S20</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.000272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_037_Ra_S21</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.055043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_050_Ra_S23</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.021959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_023_Ra_S17</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.017734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_047_Ra_S4</td>\n",
       "      <td>8825.0</td>\n",
       "      <td>Neognathae</td>\n",
       "      <td>0.000553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>dorsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>9895.0</td>\n",
       "      <td>Bovidae</td>\n",
       "      <td>0.005477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>dorsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>9881.0</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.006176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ska_genus    ska_species collected_by                    sample     taxid  \\\n",
       "60  Culiseta      particeps         ALCO          CMS001_003_Ra_S2   33208.0   \n",
       "59  Culiseta      particeps         ALCO          CMS001_003_Ra_S2    9881.0   \n",
       "61  Culiseta      particeps         ALCO          CMS001_008_Ra_S3    9881.0   \n",
       "62  Culiseta      particeps         ALCO          CMS001_025_Ra_S7    9881.0   \n",
       "58  Culiseta       inornata         ALCO          CMS001_016_Ra_S6    9881.0   \n",
       "57  Culiseta       incidens         ALCO          CMS001_005_Ra_S3    9606.0   \n",
       "56     Culex       tarsalis         COAV  CMS002_042a_Rb_S177_L004   33208.0   \n",
       "45     Culex       tarsalis         ALCO         CMS001_031_Ra_S19   33554.0   \n",
       "52     Culex       tarsalis         ALCO         CMS001_042_Ra_S23   33554.0   \n",
       "44     Culex       tarsalis         ALCO         CMS001_028_Ra_S17   33554.0   \n",
       "51     Culex       tarsalis         ALCO         CMS001_038_Ra_S22   33208.0   \n",
       "49     Culex       tarsalis         ALCO         CMS001_036_Ra_S20   33208.0   \n",
       "55     Culex       tarsalis         ALCO          CMS001_046_Ra_S3    9895.0   \n",
       "48     Culex       tarsalis         ALCO         CMS001_036_Ra_S20    9895.0   \n",
       "47     Culex       tarsalis         ALCO         CMS001_036_Ra_S20    9881.0   \n",
       "50     Culex       tarsalis         ALCO         CMS001_038_Ra_S22    9881.0   \n",
       "54     Culex       tarsalis         ALCO          CMS001_046_Ra_S3    9881.0   \n",
       "43     Culex       tarsalis         ALCO         CMS001_027_Ra_S16    8825.0   \n",
       "46     Culex       tarsalis         ALCO          CMS001_032_Ra_S7    8825.0   \n",
       "53     Culex       tarsalis         ALCO         CMS001_043_Ra_S24    8825.0   \n",
       "35     Culex        pipiens         ALCO          CMS001_004_Ra_S2   33554.0   \n",
       "38     Culex        pipiens         ALCO          CMS001_030_Ra_S7   33554.0   \n",
       "37     Culex        pipiens         ALCO         CMS001_029_Ra_S18   33208.0   \n",
       "40     Culex        pipiens         ALCO         CMS001_041_Ra_S10   33208.0   \n",
       "36     Culex        pipiens         ALCO         CMS001_029_Ra_S18    8825.0   \n",
       "42     Culex        pipiens         ALCO         CMS001_059_Ra_S10    8825.0   \n",
       "39     Culex        pipiens         ALCO         CMS001_040_Ra_S21    8825.0   \n",
       "41     Culex        pipiens         ALCO         CMS001_054_Ra_S11    8825.0   \n",
       "25     Culex  erythrothorax         ALCO          CMS001_048_Ra_S5  337687.0   \n",
       "26     Culex  erythrothorax         ALCO          CMS001_049_Ra_S6  337687.0   \n",
       "..       ...            ...          ...                       ...       ...   \n",
       "5      Culex  erythrothorax         ALCO          CMS001_011_Ra_S4   33208.0   \n",
       "21     Culex  erythrothorax         ALCO         CMS001_037_Ra_S21   33208.0   \n",
       "33     Culex  erythrothorax         ALCO          CMS001_058_Ra_S9   33208.0   \n",
       "4      Culex  erythrothorax         ALCO          CMS001_010_Ra_S1   33208.0   \n",
       "8      Culex  erythrothorax         ALCO          CMS001_013_Ra_S5   33208.0   \n",
       "15     Culex  erythrothorax         ALCO         CMS001_020_Ra_S15   33208.0   \n",
       "10     Culex  erythrothorax         ALCO         CMS001_015_Ra_S13   33208.0   \n",
       "28     Culex  erythrothorax         ALCO         CMS001_050_Ra_S23   33208.0   \n",
       "23     Culex  erythrothorax         ALCO          CMS001_045_Ra_S2   33208.0   \n",
       "19     Culex  erythrothorax         ALCO         CMS001_035_Ra_S20   33208.0   \n",
       "12     Culex  erythrothorax         ALCO         CMS001_018_Ra_S14   33208.0   \n",
       "14     Culex  erythrothorax         ALCO         CMS001_020_Ra_S15    9979.0   \n",
       "7      Culex  erythrothorax         ALCO          CMS001_013_Ra_S5    9979.0   \n",
       "32     Culex  erythrothorax         ALCO         CMS001_057_Ra_S11    9895.0   \n",
       "29     Culex  erythrothorax         ALCO          CMS001_052_Ra_S7    9881.0   \n",
       "16     Culex  erythrothorax         ALCO          CMS001_022_Ra_S6    9881.0   \n",
       "3      Culex  erythrothorax         ALCO          CMS001_010_Ra_S1    9881.0   \n",
       "30     Culex  erythrothorax         ALCO          CMS001_053_Ra_S8    9881.0   \n",
       "9      Culex  erythrothorax         ALCO         CMS001_015_Ra_S13    9881.0   \n",
       "2      Culex  erythrothorax         ALCO          CMS001_001_Ra_S1    9881.0   \n",
       "11     Culex  erythrothorax         ALCO         CMS001_018_Ra_S14    9881.0   \n",
       "31     Culex  erythrothorax         ALCO         CMS001_057_Ra_S11    9881.0   \n",
       "22     Culex  erythrothorax         ALCO          CMS001_045_Ra_S2    9881.0   \n",
       "18     Culex  erythrothorax         ALCO         CMS001_035_Ra_S20    9881.0   \n",
       "20     Culex  erythrothorax         ALCO         CMS001_037_Ra_S21    8825.0   \n",
       "27     Culex  erythrothorax         ALCO         CMS001_050_Ra_S23    8825.0   \n",
       "17     Culex  erythrothorax         ALCO         CMS001_023_Ra_S17    8825.0   \n",
       "24     Culex  erythrothorax         ALCO          CMS001_047_Ra_S4    8825.0   \n",
       "1      Aedes       dorsalis         ALCO          CMS001_012_Ra_S4    9895.0   \n",
       "0      Aedes       dorsalis         ALCO          CMS001_012_Ra_S4    9881.0   \n",
       "\n",
       "        sci_name  read_prop  \n",
       "60       Metazoa   0.001395  \n",
       "59  Odocoileinae   0.003614  \n",
       "61  Odocoileinae   0.000719  \n",
       "62  Odocoileinae   0.000111  \n",
       "58  Odocoileinae   0.001187  \n",
       "57  Homo sapiens   0.001814  \n",
       "56       Metazoa   0.022357  \n",
       "45     Carnivora   0.013463  \n",
       "52     Carnivora   0.001881  \n",
       "44     Carnivora   0.000693  \n",
       "51       Metazoa   0.008261  \n",
       "49       Metazoa   0.000106  \n",
       "55       Bovidae   0.001607  \n",
       "48       Bovidae   0.000740  \n",
       "47  Odocoileinae   0.011232  \n",
       "50  Odocoileinae   0.004398  \n",
       "54  Odocoileinae   0.000379  \n",
       "43    Neognathae   0.142716  \n",
       "46    Neognathae   0.068273  \n",
       "53    Neognathae   0.045940  \n",
       "35     Carnivora   0.017711  \n",
       "38     Carnivora   0.010530  \n",
       "37       Metazoa   0.031271  \n",
       "40       Metazoa   0.022645  \n",
       "36    Neognathae   0.092740  \n",
       "42    Neognathae   0.031076  \n",
       "39    Neognathae   0.000836  \n",
       "41    Neognathae   0.000607  \n",
       "25      Muroidea   0.054966  \n",
       "26      Muroidea   0.031226  \n",
       "..           ...        ...  \n",
       "5        Metazoa   0.043413  \n",
       "21       Metazoa   0.034659  \n",
       "33       Metazoa   0.029673  \n",
       "4        Metazoa   0.011607  \n",
       "8        Metazoa   0.010893  \n",
       "15       Metazoa   0.009948  \n",
       "10       Metazoa   0.008596  \n",
       "28       Metazoa   0.005573  \n",
       "23       Metazoa   0.005050  \n",
       "19       Metazoa   0.001133  \n",
       "12       Metazoa   0.001113  \n",
       "14     Leporidae   0.002296  \n",
       "7      Leporidae   0.000493  \n",
       "32       Bovidae   0.004716  \n",
       "29  Odocoileinae   0.008892  \n",
       "16  Odocoileinae   0.008596  \n",
       "3   Odocoileinae   0.004264  \n",
       "30  Odocoileinae   0.004140  \n",
       "9   Odocoileinae   0.003310  \n",
       "2   Odocoileinae   0.001867  \n",
       "11  Odocoileinae   0.001835  \n",
       "31  Odocoileinae   0.001495  \n",
       "22  Odocoileinae   0.001182  \n",
       "18  Odocoileinae   0.000272  \n",
       "20    Neognathae   0.055043  \n",
       "27    Neognathae   0.021959  \n",
       "17    Neognathae   0.017734  \n",
       "24    Neognathae   0.000553  \n",
       "1        Bovidae   0.005477  \n",
       "0   Odocoileinae   0.006176  \n",
       "\n",
       "[63 rows x 7 columns]"
      ]
     },
     "execution_count": 382,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known metazoan sequence\n",
    "metazoan_contigs = contig_stats_lca[(contig_stats_lca[\"taxon_group\"]==\"Metazoa\") & (contig_stats_lca[\"identity_qcov\"]>=identity_qcov_threshold)]\n",
    "# Filter contigs that were less abundant or shorter than contigs found in water belonging to the same taxid\n",
    "metazoan_contigs = use_water_filter(metazoan_contigs, by_read_count=True, by_length=True, split_by_taxid=True)\n",
    "# Filter taxids for which the total length of contigs in a sample was less than min_total_contig_len (default 1000 bp)\n",
    "metazoan_contigs = filter_by_criterion(metazoan_contigs, \"contig_length\", minthreshold=min_total_contig_len, bysample=True)\n",
    "# Convert taxids to those of interest\n",
    "metazoan_contigs = clean_taxids(metazoan_contigs, taxids=[\"Leporidae\", \"Muroidea\", \"Homo sapiens\", \"Carnivora\", \"Odocoileinae\", \"Bovidae\", \"Neognathae\"], root_taxid=\"Metazoa\")\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "metazoan_contigs = get_summary_table(metazoan_contigs, colnames=[\"ska_genus\", \"ska_species\", \"collected_by\", \"sample\", \"taxid\", \"sci_name\"], metric=\"read_prop\")\n",
    "metazoan_contigs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:74: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ska_genus</th>\n",
       "      <th>ska_species</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>sample</th>\n",
       "      <th>taxid</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>read_prop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>33090.0</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>0.000407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.005455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Culiseta</td>\n",
       "      <td>particeps</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_008_Ra_S3</td>\n",
       "      <td>2759.0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>0.000349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_029d_Rb_S162_L004</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.025096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_032_Ra_S7</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.000552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Culex</td>\n",
       "      <td>tarsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_036_Ra_S20</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.000261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Culex</td>\n",
       "      <td>quinquefasciatus</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_004a_Rb_S117_L004</td>\n",
       "      <td>5820.0</td>\n",
       "      <td>Plasmodium</td>\n",
       "      <td>0.010679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_059_Ra_S10</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.021565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_040_Ra_S21</td>\n",
       "      <td>5820.0</td>\n",
       "      <td>Plasmodium</td>\n",
       "      <td>0.003294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Culex</td>\n",
       "      <td>pipiens</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_041_Ra_S10</td>\n",
       "      <td>2759.0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>0.002005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_047b_Rb_S194_L004</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.004210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_047c_Rb_S195_L004</td>\n",
       "      <td>33634.0</td>\n",
       "      <td>Stramenopiles</td>\n",
       "      <td>0.001860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_047a_Rb_S193_L004</td>\n",
       "      <td>33090.0</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>0.000285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_047j_Rb_S3_L004</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.002883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_035_Ra_S20</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.488084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_007_Ra_S12</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.087206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_022_Ra_S6</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.056589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_049_Ra_S6</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.035639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_001_Ra_S1</td>\n",
       "      <td>33682.0</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>0.000093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_022_Ra_S6</td>\n",
       "      <td>5654.0</td>\n",
       "      <td>Trypanosomatidae</td>\n",
       "      <td>0.992131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_035_Ra_S20</td>\n",
       "      <td>5654.0</td>\n",
       "      <td>Trypanosomatidae</td>\n",
       "      <td>0.321557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_018_Ra_S14</td>\n",
       "      <td>5654.0</td>\n",
       "      <td>Trypanosomatidae</td>\n",
       "      <td>0.142531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_049_Ra_S6</td>\n",
       "      <td>5654.0</td>\n",
       "      <td>Trypanosomatidae</td>\n",
       "      <td>0.032690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Culex</td>\n",
       "      <td>erythrothorax</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_022_Ra_S6</td>\n",
       "      <td>2759.0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>0.022414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>dorsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.000284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>dorsalis</td>\n",
       "      <td>ALCO</td>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>2759.0</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>0.002776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026c_Rb_S148_L004</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.035487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026d_Rb_S149_L004</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.025449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>albopictus</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_026b_Rb_S147_L004</td>\n",
       "      <td>4751.0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>0.001018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>WVAL</td>\n",
       "      <td>CMS002_044e_Rb_S182_L004</td>\n",
       "      <td>33090.0</td>\n",
       "      <td>Viridiplantae</td>\n",
       "      <td>0.094262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017b_Rb_S123_L004</td>\n",
       "      <td>33630.0</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>0.001454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017a_Rb_S122_L004</td>\n",
       "      <td>33630.0</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>0.000820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aedes</td>\n",
       "      <td>aegypti</td>\n",
       "      <td>SAND</td>\n",
       "      <td>CMS002_017c_Rb_S124_L004</td>\n",
       "      <td>33630.0</td>\n",
       "      <td>Alveolata</td>\n",
       "      <td>0.000258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ska_genus       ska_species collected_by                    sample  \\\n",
       "32  Culiseta         particeps         ALCO         CMS001_009_Ra_S13   \n",
       "31  Culiseta         particeps         ALCO         CMS001_009_Ra_S13   \n",
       "30  Culiseta         particeps         ALCO          CMS001_008_Ra_S3   \n",
       "29     Culex          tarsalis        WVAL   CMS002_029d_Rb_S162_L004   \n",
       "27     Culex          tarsalis         ALCO          CMS001_032_Ra_S7   \n",
       "28     Culex          tarsalis         ALCO         CMS001_036_Ra_S20   \n",
       "26     Culex  quinquefasciatus         SAND  CMS002_004a_Rb_S117_L004   \n",
       "25     Culex           pipiens         ALCO         CMS001_059_Ra_S10   \n",
       "23     Culex           pipiens         ALCO         CMS001_040_Ra_S21   \n",
       "24     Culex           pipiens         ALCO         CMS001_041_Ra_S10   \n",
       "20     Culex     erythrothorax        WVAL   CMS002_047b_Rb_S194_L004   \n",
       "21     Culex     erythrothorax        WVAL   CMS002_047c_Rb_S195_L004   \n",
       "19     Culex     erythrothorax        WVAL   CMS002_047a_Rb_S193_L004   \n",
       "22     Culex     erythrothorax        WVAL     CMS002_047j_Rb_S3_L004   \n",
       "16     Culex     erythrothorax         ALCO         CMS001_035_Ra_S20   \n",
       "10     Culex     erythrothorax         ALCO         CMS001_007_Ra_S12   \n",
       "14     Culex     erythrothorax         ALCO          CMS001_022_Ra_S6   \n",
       "18     Culex     erythrothorax         ALCO          CMS001_049_Ra_S6   \n",
       "9      Culex     erythrothorax         ALCO          CMS001_001_Ra_S1   \n",
       "13     Culex     erythrothorax         ALCO          CMS001_022_Ra_S6   \n",
       "15     Culex     erythrothorax         ALCO         CMS001_035_Ra_S20   \n",
       "11     Culex     erythrothorax         ALCO         CMS001_018_Ra_S14   \n",
       "17     Culex     erythrothorax         ALCO          CMS001_049_Ra_S6   \n",
       "12     Culex     erythrothorax         ALCO          CMS001_022_Ra_S6   \n",
       "8      Aedes          dorsalis         ALCO          CMS001_012_Ra_S4   \n",
       "7      Aedes          dorsalis         ALCO          CMS001_012_Ra_S4   \n",
       "5      Aedes        albopictus        WVAL   CMS002_026c_Rb_S148_L004   \n",
       "6      Aedes        albopictus        WVAL   CMS002_026d_Rb_S149_L004   \n",
       "4      Aedes        albopictus        WVAL   CMS002_026b_Rb_S147_L004   \n",
       "3      Aedes           aegypti        WVAL   CMS002_044e_Rb_S182_L004   \n",
       "1      Aedes           aegypti         SAND  CMS002_017b_Rb_S123_L004   \n",
       "0      Aedes           aegypti         SAND  CMS002_017a_Rb_S122_L004   \n",
       "2      Aedes           aegypti         SAND  CMS002_017c_Rb_S124_L004   \n",
       "\n",
       "      taxid          sci_name  read_prop  \n",
       "32  33090.0     Viridiplantae   0.000407  \n",
       "31   4751.0             Fungi   0.005455  \n",
       "30   2759.0         Eukaryota   0.000349  \n",
       "29   4751.0             Fungi   0.025096  \n",
       "27  33682.0        Euglenozoa   0.000552  \n",
       "28   4751.0             Fungi   0.000261  \n",
       "26   5820.0        Plasmodium   0.010679  \n",
       "25  33682.0        Euglenozoa   0.021565  \n",
       "23   5820.0        Plasmodium   0.003294  \n",
       "24   2759.0         Eukaryota   0.002005  \n",
       "20  33682.0        Euglenozoa   0.004210  \n",
       "21  33634.0     Stramenopiles   0.001860  \n",
       "19  33090.0     Viridiplantae   0.000285  \n",
       "22   4751.0             Fungi   0.002883  \n",
       "16  33682.0        Euglenozoa   0.488084  \n",
       "10  33682.0        Euglenozoa   0.087206  \n",
       "14  33682.0        Euglenozoa   0.056589  \n",
       "18  33682.0        Euglenozoa   0.035639  \n",
       "9   33682.0        Euglenozoa   0.000093  \n",
       "13   5654.0  Trypanosomatidae   0.992131  \n",
       "15   5654.0  Trypanosomatidae   0.321557  \n",
       "11   5654.0  Trypanosomatidae   0.142531  \n",
       "17   5654.0  Trypanosomatidae   0.032690  \n",
       "12   2759.0         Eukaryota   0.022414  \n",
       "8    4751.0             Fungi   0.000284  \n",
       "7    2759.0         Eukaryota   0.002776  \n",
       "5    4751.0             Fungi   0.035487  \n",
       "6    4751.0             Fungi   0.025449  \n",
       "4    4751.0             Fungi   0.001018  \n",
       "3   33090.0     Viridiplantae   0.094262  \n",
       "1   33630.0         Alveolata   0.001454  \n",
       "0   33630.0         Alveolata   0.000820  \n",
       "2   33630.0         Alveolata   0.000258  "
      ]
     },
     "execution_count": 407,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known eukaryotic sequence\n",
    "eukaryotic_contigs = contig_stats_lca[(contig_stats_lca[\"taxon_group\"]==\"Eukaryota\") & (contig_stats_lca[\"identity_qcov\"]>=identity_qcov_threshold)]\n",
    "# Filter contigs that were less abundant or shorter than contigs found in water belonging to the same taxid\n",
    "eukaryotic_contigs = use_water_filter(eukaryotic_contigs, by_read_count=True, by_length=True, split_by_taxid=True)\n",
    "# Filter taxids for which the total length of contigs in a sample was less than min_total_contig_len (default 1000 bp)\n",
    "eukaryotic_contigs = filter_by_criterion(eukaryotic_contigs, \"contig_length\", minthreshold=min_total_contig_len, bysample=True)\n",
    "# Convert taxids to those of interest\n",
    "eukaryotic_groups = [\"Fungi\", \"Trypanosomatidae\", \"Plasmodium\", \"Stramenopiles\", \"Viridiplantae\", \"Euglenozoa\", \"Alveolata\"]\n",
    "eukaryotic_contigs = clean_taxids(eukaryotic_contigs, taxids=eukaryotic_groups, root_taxid=\"Eukaryota\")\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "eukaryotic_contigs = get_summary_table(eukaryotic_contigs, colnames=[\"ska_genus\", \"ska_species\", \"collected_by\", \"sample\", \"taxid\", \"sci_name\"], metric=\"read_prop\")\n",
    "\n",
    "eukaryotic_contigs\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lucy.li/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=True'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass sort=False\n",
      "\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "all_contigs_df = pd.concat([wolbachia_contigs.assign(group=\"Wolbachia\"), \n",
    "                            viral_contigs.assign(group=\"Virus\"),\n",
    "                            metazoan_contigs.assign(group=\"Metazoa\"), \n",
    "                            eukaryotic_contigs.assign(group=\"Other Eukaryotes\")])\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_contigs_df.to_csv(\"../../figures/fig3/all_contigs_df.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
