---
title: "Generate figures"
author: "Lucy M. Li"
date: "`r format(Sys.Date(), '%b %d, %Y')`"
output: 
  html_document:
    fig_caption: yes
---


# Notebook setup

Run this notebook after running the generate_figures.ipynb jupyter notebook.

```{r setup, include=FALSE}
# List of packages to load
pkgs <- c("dplyr", "tidyr", "readr", "magrittr", "ggplot2", "grid", "gridExtra", "stringr", "knitr", "scales", "gtable", "rgdal", "broom", "svglite")
# Install any packages not yet installed
if (any(!(pkgs %in% rownames(installed.packages())))) {
  invisible(sapply(pkgs[!(pkgs %in% rownames(installed.packages()))], install.packages))
}
# Load the libraries
invisible(sapply(pkgs, library, character.only=TRUE))
# R Markdown options
knitr::opts_chunk$set(echo=TRUE)
```

```{r directories}
data_dir <- "../../data/metadata"
fig_dir <- "../../figures/fig3"
if (!dir.exists(data_dir)) dir.create(data_dir)
if (!dir.exists(fig_dir)) dir.create(fig_dir)
# helvetica neue - font from other figures.
```


```{r scripts}
source("generate_figures_functions.R")
```

# Metadata

```{r load_data, message=FALSE}
metadata_file_path <- list.files(data_dir, "CMS001_CMS002_Merged", full.names = TRUE)
metadata <- read.csv(metadata_file_path, stringsAsFactors = FALSE) %>%
  mutate(collected_by=gsub(" ", "", collected_by)) %>%
  mutate(collection_date=as.Date(collection_date)) %>%
  mutate(ska_genus_species=paste(ska_genus, ska_species)) %>%
  left_join(read_csv(file.path(data_dir, "../metadata/idseq_metadata.csv")), by=c("NewIDseqName"="sample"))
```


# Read in data



```{r read_in_data, message=FALSE, warning=FALSE}
all_curated_contigs <- read_tsv(file.path(fig_dir, "all_contigs_df.tsv")) %>%
  mutate(collected_by=factor(collected_by, levels=c("PLCR", "ALCO",  "WVAL", "COAV", "SAND")),
         group=factor(group, levels=c("Virus", "Wolbachia", "Metazoa", "Other Eukaryotes")))
all_curated_contigs <- all_curated_contigs %>% 
  filter(group=="Virus") %>% 
  group_by(sample) %>% 
  summarize(read_prop_virus=sum(read_prop)) %>%
  left_join(all_curated_contigs, .) %>%
  ungroup() %>%
  arrange(group, ska_genus, ska_species, collected_by, desc(read_prop_virus)) %>%
  mutate(sample_short=str_split(as.character(sample), "_", simplify=FALSE) %>% 
           lapply(head, 2) %>% 
           sapply(paste, collapse="_") %>%
           gsub("CMS00", "", .)) %>% 
  mutate(sample_short=factor(sample_short, levels=unique(sample_short))) %>%
  left_join(metadata%>%select(NewIDseqName, nonhost_reads), by=c("sample"="NewIDseqName")) %>%
  mutate(sci_name=gsub("Virus", "Other viruses", sci_name))


all_curated_contigs[is.na(all_curated_contigs$ska_species), paste0("ska_", c("genus", "species"))] <-
  filter(all_curated_contigs, is.na(ska_species)) %>%
  `[[`("sample") %>% match(metadata$NewIDseqName) %>%
  slice(metadata, .) %>%
  select(visual_genus, visual_species) %>%
  rename(ska_genus=visual_genus, ska_species=visual_species)
```

```{r get_subset_data_for_figures}
curated_contigs_raw <- filter(all_curated_contigs, group %in% c("Virus", "Wolbachia", "Other Eukaryotes") | family=="Onchocercidae") %>%
  mutate(group=factor(gsub("Metazoa", "Other Eukaryotes", group))) %>%
  mutate(group=factor(droplevels(group), c("Virus", "Wolbachia", "Other Eukaryotes"))) %>%
  mutate(sample=factor(sample), sample_short=factor(sample_short)) %>%
  mutate(baltimore_group=coalesce(baltimore_group, as.character(group))) %>%
  mutate(family=coalesce(family, as.character(group))) %>%
  mutate(genome_description=coalesce(genome_description, as.character(group))) %>%
  filter(family!="Other eukaryotes")


subset_contigs <- curated_contigs_raw %>%
  group_by(sample, family) %>%
  summarize(read_prop=sum(read_prop)) %>%
  group_by(family) %>%
  summarize(read_prop=max(read_prop)) %>%
  filter(read_prop>=0.01) %>%
  `[[`("family")

curated_contigs <- curated_contigs_raw %>%
  filter(., family %in% subset_contigs)

curated_contigs %<>%
  group_by(sample) %>%
  summarize(read_prop=sum(read_prop)) %>%
  filter(read_prop>=0.01) %>%
  with(curated_contigs$sample %in% sample) %>%
  filter(curated_contigs, .)

curated_contigs %<>%
  mutate(ska_species=apply(., 1, function (x) ifelse(x["ska_genus"]=="Culiseta", "", x["ska_species"]))) %>%
  arrange(ska_genus, ska_species, collected_by, desc(read_prop_virus)) %>%
  mutate(sample=factor(sample, levels=as.character(sample) %>% unique())) %>%
  mutate(sample_id=as.numeric(sample))  


```


# Plot virus and wolbachia results


```{r plot_color_palette, message=FALSE}
color_palette <- file.path(fig_dir, "fig3_color_palette.tsv") %>%
  read_tsv() %>%
  with(setNames(color, key))
```




```{r plot_data, message=FALSE}
base_plot_background_df <- curated_contigs %>% 
  filter(!duplicated(sample)) %>% 
  select(sample, ska_species, collected_by) %>% 
  list(.) %>% 
  mapply(mutate, ., group=as.list(levels(curated_contigs$group)), SIMPLIFY=FALSE) %>%
  do.call(what=rbind) %>%
  mutate(group=factor(group, levels=unique(group))) %>%
  mutate(fill=c("white", "gray80")[factor(paste(ska_species, collected_by), levels=unique(paste(ska_species, collected_by))) %>% as.numeric() %>% `%%`(2) %>% `+`(1)])

base_plot_addon <- list(scale_y_continuous(labels=scales::percent, breaks=c(0.5), minor_breaks=c(0.25, 0.75), limits=c(0, 1), expand=c(0, 0)),
                        ylab("% non-host reads"),
                        theme(legend.position="none",
                              strip.background.y=element_blank(),
                              strip.text.y=element_blank(),
                              text=element_text(size=18),
                              axis.text.x=element_blank(),
                              axis.title.x=element_blank(),
                              axis.title.y=element_text(vjust=0.5, margin=margin(t=0, r=1.1, b=0, l=0, unit="lines")),
                              axis.line=element_line(size=.2),
                              plot.margin=unit(c(0, 0, 0, 0), "lines")))

base_plot <- 
  ggplot(curated_contigs) +
  theme_bw() + 
  geom_bar(aes(x=sample, y=read_prop, fill=family), stat="identity", position="stack", color="white", size=.2) +
  geom_text(data=curated_contigs %>% filter(!duplicated(group)) %>% mutate(group_label=gsub("Other Eukaryotes", "eukaryotes", group) %>% gsub("Virus", "viruses", .)), 
            aes(label=group_label), x=1, y=max(curated_contigs$read_prop)*0.9, hjust=0, lineheight=.75) +
  facet_grid(group~., switch="both") +
  scale_fill_manual("", breaks=names(color_palette), values=color_palette, labels=names(color_palette)) +
  # xlab("sample ID") +
  base_plot_addon + 
  theme(panel.grid.major.y=element_line(linetype=2, size=0.5, color="gray80"),
        panel.grid.minor.y=element_line(linetype=2, size=0.25, color="gray90"))

plot_grob <- ggplotGrob(base_plot)
plot_grob$heights[filter(plot_grob$layout, grepl("panel-1-1", name))$t] <- unit(2.5, "null")
```

```{r species_label_grob}
species_label_df <- curated_contigs %>%
  group_by(ska_genus, ska_species) %>% 
  summarize(xmin=min(sample_id), xmax=max(sample_id)) %>% 
  ungroup() %>% 
  mutate(xmin=xmin-0.5, xmax=xmax+0.5) %>%
  mutate(species_label=paste0(substr(ska_genus, 1, 2), ". ", ska_species) %>% 
           gsub("Cu. $", "Culiseta spp.", .) %>% 
           gsub("fasciatus", ".", ., fixed=TRUE) %>% 
           gsub("Ae. dorsalis", "*", .) %>%
           gsub("Cu. ", "Cx. ", ., fixed=TRUE))

element_addon <- list(scale_x_continuous(expand = c(0,0)),
                      theme_classic(),
                      theme(legend.position="none",
                            text=element_text(size=18),
                            axis.line=element_line(size=.2),
                            axis.line.y=element_line(color="white"),
                            axis.text.y=element_text(color="white"),
                            axis.ticks.y=element_line(color="white"),
                            axis.title.y=element_text(vjust=0.5, color="white", margin=margin(t=0, r=1.1, b=0, l=0, unit="lines")),
                            axis.line.x=element_blank(),
                            axis.text.x=element_blank(),
                            axis.ticks.x=element_blank(),
                            axis.title.x=element_blank(),
                            plot.margin=unit(c(0, 0, 0, 0), "lines"))
                      )

species_label_grob <- ggplot(species_label_df) +
  geom_rect(aes(xmin=xmin, xmax=xmax, fill=species_label, ymin=0, ymax=1), color="white", fill="gray") +
  geom_text(aes(x=(xmax+xmin)/2, y=0.5, label=species_label), fontface="italic", size=4.5) +
  base_plot_addon + element_addon

```

```{r location_labels_grob}
location_label_df <- curated_contigs %>%
  group_by(ska_genus, ska_species, collected_by) %>% 
  summarize(xmin=min(sample_id), xmax=max(sample_id)) %>% 
  ungroup() %>%
  mutate(xmin=xmin-0.5, xmax=xmax+0.5) 

location_label_grob <- ggplot(location_label_df) +
  geom_rect(aes(xmin=xmin, xmax=xmax, fill=collected_by, ymin=0, ymax=1)) +
  scale_fill_manual("", breaks=names(color_palette), values=color_palette, labels=names(color_palette)) +
  base_plot_addon + element_addon
  

```


```{r presence_absence_grob}
curated_contigs_low_abundance <- curated_contigs_raw %>%
  group_by(sample, group) %>% 
  summarize(sum=sum(read_prop)) %>% 
  filter(sum<0.01)

low_abundance_df <- 
  filter(curated_contigs_raw, sample %in% curated_contigs_low_abundance$sample, group!="Virus") %>% 
  select(sample, family) %>% 
  distinct() %>%
  mutate(present=TRUE) %>%
  rename(low_abundance_family=family) %>%
  left_join(curated_contigs_raw, ., by="sample") %>%
  filter(!is.na(present), group!="Virus") %>%
  mutate(ypos=as.numeric(as.factor(low_abundance_family))/(length(unique(low_abundance_family))+1)) %>%
  mutate(sample=factor(sample, levels=levels(curated_contigs$sample)))

low_abundance_plot_ylab <- filter(low_abundance_df, !duplicated(ypos)) %>% 
  arrange(low_abundance_family) %>% 
  select(ypos, low_abundance_family)

low_abundance_grob <- location_label_grob + 
  geom_point(data=low_abundance_df, aes(x=as.numeric(sample), y=ypos, shape=low_abundance_family)) +
  geom_text(data=low_abundance_plot_ylab, aes(x=length(levels(low_abundance_df$sample))-20, y=ypos, label=low_abundance_family), hjust=0) +
  scale_fill_manual("", breaks=names(color_palette), values=rep("white", length(color_palette)), labels=names(color_palette)) +
  theme(legend.position="none")

```


```{r collection_site_map}
collected_by_dict <- c("ALCO"="Alameda", "COAV"="Coachella Valley", "PLCR"="Placer", 
                       "SAND"="San Diego", "WVAL"="West Valley")

ca_shapefile <- readOGR(dsn=file.path(data_dir, "../shapefiles/CA_State")) %>% spTransform(CRS("+proj=longlat +datum=WGS84"))
ca_shapefile_tidy <- broom::tidy(ca_shapefile)
selected_ca_shapefile_group <- ca_shapefile_tidy %>% 
  group_by(group) %>% 
  summarize(sum=diff(range(lat))+diff(range(long))) %>%
  slice(which.max(sum)) %>% 
  `[[`("group")
ca_shapefile_tidy <- filter(ca_shapefile_tidy, group==selected_ca_shapefile_group)
map_dot_df <- metadata %>% group_by(collected_by) %>% summarize(lat=median(collection_lat, na.rm=TRUE), long=median(collection_long, na.rm=TRUE))
map_text_df <- map_dot_df %>%
  mutate(lat=lat+c(0, 1.8, 1, -.9, -.9), long=long+c(2, -1.7, -.8, 1.5, -2)) %>%
  mutate(label=collected_by_dict[collected_by])
ca_map_plot <- ggplot(ca_shapefile_tidy, aes(x=long, y=lat, group=group)) +
  geom_polygon(color = "black", size = 0.2, fill = "white") +
  coord_equal() +
  theme_classic() +
  geom_point(data=map_dot_df, aes(x=long, y=lat, fill=collected_by, group=NA), cex=6, stroke=0.25, color="black", shape=21, alpha=1) +
  geom_text(data=map_text_df, aes(x=long, y=lat, label=label, group=NA)) +
  geom_segment(data=filter(map_dot_df, collected_by=="COAV") %>% mutate(xend=long, yend=filter(map_text_df, collected_by=="COAV")$lat-.4, lat=lat+.4), 
               aes(x=long, y=lat, xend=xend, yend=yend, group=NA)) +
  scale_fill_manual(breaks=names(color_palette), values=color_palette) +
  xlab("Collection Sites") +
  theme(legend.position="none", 
        axis.line=element_blank(), 
        axis.ticks=element_blank(), 
        axis.title.y=element_blank(), 
        axis.text=element_blank(),
        axis.title.x=element_text(hjust=0, face="bold", size=14),
        plot.margin=unit(c(0, 0, 0, .5), units="lines"))
```

```{r plot_legends}
# Extract legends as separate plots
legends <- split(curated_contigs, curated_contigs$baltimore_group) %>%
  lapply(function (dataframe) {
    legend_nrow <- 2
    #if (dataframe$genome_description[1]=="Other Eukaryotes") legend_nrow <- 3
    selected_colours <- names(color_palette) %in% as.character(dataframe$family)
    subpalette <- color_palette[selected_colours]
    sublabels <- names(color_palette)[selected_colours]
    ggplot(dataframe, aes(x=sample_short, y=read_prop, fill=family)) +
      theme_bw() +
      geom_bar(stat="identity", position="stack") +
      scale_fill_manual(gsub("Other E", "e", dataframe$genome_description[1]), breaks=sublabels, labels=sublabels, values=subpalette) +
      guides(fill=guide_legend(nrow=legend_nrow, byrow=TRUE, title.position="top")) +
      theme(text=element_text(size=14), 
            plot.margin=unit(c(0, 1, 0, 0), units="lines"),
            legend.position="top",
            legend.spacing.y=unit(0.4, "lines"),
            legend.spacing.x=unit(0.3, "lines"),
            legend.key.size=unit(0.9, "lines"),
            legend.title=element_text(margin=margin(0, 0, 0.65, 0, unit="lines"), face="bold"))
  }) %>%
  #c(., list(collection_site=collection_site_plot)) %>%
  lapply(function (plot_x) {
    plot_gtable <- ggplot_gtable(ggplot_build(plot_x))
    leg <- which(sapply(plot_gtable$grobs, function(x) x$name) == "guide-box")
    return (plot_gtable$grobs[[leg]])
  })
```

```{r generate_plot}
legends_grob <- arrangeGrob(rectGrob(gp=gpar(col="white")), arrangeGrob(do.call(what=gtable_cbind, legends[c("V", "IV", "III", "Other Eukaryotes")]), nrow=1), widths=c(1, 9))
collection_site_grob <- arrangeGrob(rectGrob(gp=gpar(col="white")), ca_map_plot, ncol=1, heights=c(1.75, 1))
main_plot_grob <- arrangeGrob(legends_grob, species_label_grob, plot_grob, location_label_grob, low_abundance_grob, ncol=1, heights=c(7, 3, 30, 2, 4))
full_plot <- arrangeGrob(main_plot_grob, collection_site_grob, nrow=1, widths=c(7, 1))
```


```{r save_plot}
lapply(c("pdf", "svg"), function (suffix) {
  ggsave(file.path(fig_dir, paste0("fig3.", suffix)), full_plot, width=ifelse(suffix=="pdf", 20, 20*1.3), height=ifelse(suffix=="pdf", 10, 10*1.3), units="in")
})

```

```{r save-data}
fig3_data_output_viruses <- base_plot$data %>%
  filter(group=="Virus") %>%
  mutate(mosquito_species=paste(ska_genus, ska_species)) %>%
  mutate(collected_by=collected_by_dict[as.character(collected_by)]) %>%
  mutate(sample=paste0("CMS00", sample_short)) %>%
  select(sample, mosquito_species, collected_by, sci_name, taxid, family, baltimore_group, genome_description, read_prop, nonhost_reads) %>%
  write_tsv(file.path(fig_dir, "fig3_data_viruses.tsv"))

fig3_data_output_wolbachia <- base_plot$data %>%
  filter(group=="Wolbachia") %>%
  mutate(mosquito_species=paste(ska_genus, ska_species)) %>%
  mutate(collected_by=collected_by_dict[as.character(collected_by)]) %>%
  mutate(sample=paste0("CMS00", sample_short)) %>%
  select(sample, mosquito_species, collected_by, read_prop, nonhost_reads) %>%
  write_tsv(file.path(fig_dir, "fig3_data_wolbachia.tsv"))

fig3_data_output_eukaryotes <- base_plot$data %>%
    filter(group=="Other Eukaryotes") %>%
    mutate(mosquito_species=paste(ska_genus, ska_species)) %>%
    mutate(collected_by=collected_by_dict[as.character(collected_by)]) %>%
    mutate(sample=paste0("CMS00", sample_short)) %>%
    mutate(taxonomic_group=family) %>%
    select(sample, mosquito_species, collected_by, sci_name, taxid, taxonomic_group, read_prop, nonhost_reads) %>%
  write_tsv(file.path(fig_dir, "fig3_data_eukaryota.tsv"))

```


# Plot prevalence vs abundance of viruses

```{r prev_abundance}
curated_virus_df <- filter(curated_contigs, group=="Virus") %>%
  left_join(., select(metadata, NewIDseqName, collection_lat, collection_long), by=c("sample"="NewIDseqName"))
viral_bulk_abundance <- curated_virus_df %>%
  group_by(ska_genus, ska_species, collection_lat, collection_long, sci_name) %>%
  summarize(bulk_abundance=mean(read_prop), infected_mosquitos=n()) %>%
  ungroup()
viral_bulk_abundance <- curated_virus_df %>%
  group_by(ska_genus, ska_species, collection_lat, collection_long) %>%
  summarize(num_mosquitos=n_distinct(sample_short)) %>%
  ungroup() %>%
  left_join(viral_bulk_abundance, .) %>% 
  mutate(prevalence=infected_mosquitos/num_mosquitos)
```

```{r prev_abundance_vignettes}
select_viruses_prev_abundance <- 
  filter(viral_bulk_abundance, num_mosquitos>=10) %>%
  filter(bulk_abundance>=0.08) %>%
  slice(c(which.min(prevalence), which.max(prevalence))) %>%
  list(.) %>%
  mapply(function (i, df) {
    read_prop_df <- data.frame(mosquito_id=1:df$num_mosquitos[i], read_prop=0)
    read_prop_vector <- left_join(df[i, ], curated_virus_df)$read_prop %>% sort(decreasing=TRUE)
    read_prop_df[seq(read_prop_vector), "read_prop"] <- read_prop_vector
    read_prop_df <- left_join(df[i, ] %>% select(-sci_name), curated_virus_df) %>%
      slice(order(read_prop, decreasing=TRUE)) %>%
      `[[`("sample_short") %>%
      unique() %>%
      mutate(read_prop_df, sample_short=.) %>%
      mutate(sample_short=factor(as.character(sample_short), levels=unique(as.character(sample_short))))
    cbind(read_prop_df, df[i, ])
  }, 1:nrow(.[[1]]), ., SIMPLIFY=FALSE) %>%
  do.call(what=rbind, .)
```

```{r prev_abundance_plot}
prev_abundance_plot_df <- filter(viral_bulk_abundance, num_mosquitos>=10) %>%
  left_join(., select_viruses_prev_abundance) %>% 
  mutate(highlight=sci_name)
prev_abundance_plot_df$highlight[is.na(prev_abundance_plot_df$mosquito_id)] <- "NA"
prev_abundance_plot <- 
  ggplot(prev_abundance_plot_df) +
  theme_bw() +
  geom_point(aes(x=bulk_abundance, y=prevalence, color=highlight)) +
  geom_label(data=filter(select_viruses_prev_abundance, !duplicated(paste(ska_genus, ska_species, collection_lat, collection_long, sci_name))) %>%
                           mutate(bulk_abundance=bulk_abundance-c(0.01, 0.09), prevalence=prevalence+c(0.07, -0.05)),
             aes(x=bulk_abundance, y=prevalence, label=sci_name, fill=sci_name),
             hjust=0.5, alpha=.75, size=3) +
  scale_x_log10(labels=scales::percent) +
  scale_y_continuous(labels=scales::percent) +
  scale_color_manual(values=c("skyblue2", "indianred1", "black")) +
  scale_fill_manual(values=c("skyblue2", "indianred1")) +
  xlab("bulk abundance") + 
  ylab("prevalence") +
  ggtitle("A)") +
  theme(legend.position="none", text=element_text(size=12))
```


```{r prev_abundance_vignettes_plot}
select_virus_prev_abund_plot <- 
  ggplot(select_viruses_prev_abundance, aes(x=sample_short, y=read_prop)) +
  theme_bw() +
  geom_bar(stat="identity") +
  geom_hline(aes(yintercept=bulk_abundance), color="darkmagenta") +
  facet_grid(.~sci_name, scales="free_x", space="free") +
  scale_y_continuous(labels=scales::percent) +
  xlab("Sample") + ylab("% non-host reads") +
  #ggtitle("B)") +
  theme(axis.text.x=element_blank(), axis.title.x=element_blank())
```

```{r prev_abundance_combined_plot}
if (!(dir.exists(file.path(fig_dir, "../figS4")))) dir.create(file.path(fig_dir, "../figS4"))
lapply(c("pdf", "svg"), function (suffix) {
  ggsave(file.path(fig_dir, paste0("../figS4/virus_prev_abundance.", suffix)),
         select_virus_prev_abund_plot,
         width=ifelse(suffix=="pdf", 6, 6*1.3), height=ifelse(suffix=="pdf", 3, 3*1.3), units="in")
})

```


```{r save-data-supp}
select_virus_prev_abund_plot$data %>%
  mutate(mosquito_species=paste(ska_genus, ska_species)) %>%
  mutate(collection_long=round(collection_long, 2), collection_lat=round(collection_lat, 2)) %>%
  mutate(sample=paste0("CMS00", sample_short)) %>%
  select(-ska_genus, -ska_species, -sample_short) %>%
  pivot_wider(id_cols=c(mosquito_species, collection_lat, collection_long, sci_name, bulk_abundance, num_mosquitos, prevalence), values_from=read_prop, names_from=mosquito_id, names_prefix="mosquito_") %>%
  write_tsv(file.path(fig_dir, "../figS4/virus_prev_abundance_data.tsv"))
```

# Wolbachia vs virus

```{r wolbachia_virus}
wolbachia_virus_df <- group_by(fig3_data_output_viruses, sample) %>% 
  summarize(read_prop=sum(read_prop), nonhost_reads=first(nonhost_reads)) %>% 
  left_join(select(fig3_data_output_wolbachia, sample, read_prop, nonhost_reads), by=c("sample", "nonhost_reads"), suffix=c("_virus", "_wolbachia")) %>% 
  replace_na(list(read_prop_virus=0, read_prop_wolbachia=0)) %>% 
  mutate(reads_virus=read_prop_virus*nonhost_reads, 
         reads_wolbachia=read_prop_wolbachia*nonhost_reads) %>%
  left_join(select(metadata, NewIDseqName, total_reads, compression_ratio) %>% mutate(NewIDseqName=strsplit(NewIDseqName, "_")%>%lapply(head, 2)%>%sapply(paste, collapse="_")), by=c("sample"="NewIDseqName")) %>%
  mutate(absolute_reads_virus=reads_virus*compression_ratio/total_reads,
         absolute_reads_wolbachia=reads_wolbachia*compression_ratio/total_reads)
t.test(log10(filter(wolbachia_virus_df, absolute_reads_wolbachia!=0)$absolute_reads_virus),
       log10(filter(wolbachia_virus_df, absolute_reads_wolbachia==0)$absolute_reads_virus))

```
```{r}
ggplot(wolbachia_virus_df, aes(x=absolute_reads_wolbachia, y=absolute_reads_virus)) +
  geom_point() +
  scale_y_log10() +
  scale_x_log10()
```


```{r}
ggplot(wolbachia_virus_df, aes(x=as.character(absolute_reads_wolbachia!=0), y=absolute_reads_virus)) +
  geom_boxplot() +
  scale_y_log10()
#   scale_x_log10()

```

# Save output

```{r output}
save.image(file="generate_figures.RData")
```

