{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import pandas as pd\n",
    "from ete3 import NCBITaxa\n",
    "import subprocess\n",
    "import itertools\n",
    "import os\n",
    "import s3fs\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sqlalchemy import create_engine\n",
    "from lca_functions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "s3 = boto3.resource('s3')\n",
    "client = boto3.client('s3')\n",
    "bucket_name = \"czbiohub-mosquito\"\n",
    "bucket = s3.Bucket(bucket_name)\n",
    "contig_folders = [x[\"Prefix\"] for x in client.list_objects(Bucket=bucket_name, Prefix=\"contigs/\", Delimiter=\"/\")[\"CommonPrefixes\"]]\n",
    "contig_quality_folders = [x[\"Prefix\"] for x in client.list_objects(Bucket=bucket_name, Prefix=\"contig_quality/\", Delimiter=\"/\")[\"CommonPrefixes\"]]\n",
    "\n",
    "ncbi = NCBITaxa()\n",
    "# ncbi.update_taxonomy_database()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### read counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_count_files = [client.list_objects(Bucket=bucket_name, Prefix=x+\"bowtie_csp_counts_1000.txt\") \\\n",
    "                   for x in contig_quality_folders]\n",
    "read_count_files = [\"s3://\"+bucket_name+\"/\"+x[\"Prefix\"] for x in read_count_files if \"Contents\" in x.keys()]\n",
    "read_counts_csp_1000 = pd.concat([pd.read_csv(x, sep=\"\\t\", header=None, names=[\"query\", \"read_count\"]).\\\n",
    "                                  assign(sample=os.path.split(os.path.split(x)[0])[1]) for x in read_count_files])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5577723823263779"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(read_counts_csp_1000[\"read_count\"]>2)/len(read_counts_csp_1000.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "DataFrame constructor not properly called!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-1d7e325334c8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m pd.DataFrame(np.histogram(read_counts_csp_1000[\"read_count\"], \n\u001b[0;32m----> 2\u001b[0;31m                           bins=list(range(21))+[50, 100, 1000, 10000])).transpose()\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    420\u001b[0m                                          dtype=values.dtype, copy=False)\n\u001b[1;32m    421\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 422\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'DataFrame constructor not properly called!'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    423\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    424\u001b[0m         \u001b[0mNDFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmgr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfastpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: DataFrame constructor not properly called!"
     ]
    }
   ],
   "source": [
    "pd.DataFrame(np.histogram(read_counts_csp_1000[\"read_count\"], \n",
    "                          bins=list(range(21))+[50, 100, 1000, 10000])).transpose()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_contigs_by_read_count = read_counts_csp_1000[read_counts_csp_1000[\"read_count\"]>2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_lca_analysis (input_file_name, output_dir, bucket_name, ident_cutoff, align_cutoff, bitscore_cutoff, \\\n",
    "                      blast_type, default=False, ncores=8):\n",
    "    # First list the folders (there is a limit of 1000 files output by AWS by default)\n",
    "    list_of_folders = [client.list_objects(Bucket=bucket_name, Prefix=x[\"Prefix\"]+input_file_name) \\\n",
    "                       for x in client.list_objects(Bucket=bucket_name, Prefix=\"contigs/\", Delimiter=\"/\")[\"CommonPrefixes\"]]\n",
    "    list_of_folders = [\"s3://\"+bucket_name+\"/\"+x[\"Prefix\"] for x in list_of_folders if \"Contents\" in x.keys()]\n",
    "    filenames = pd.DataFrame(list_of_folders, columns=[\"blast_\"+blast_type])\n",
    "    output_string = output_dir\n",
    "    if not default:\n",
    "        output_string += \"/ident\"+str(ident_cutoff)+\"align\"+str(align_cutoff)+\"bitscore\"+str(bitscore_cutoff)\n",
    "    filenames = filenames.assign(filtered_blast=filenames[\"blast_\"+blast_type].str.replace(\"contigs\", output_string).str.replace(\".m9\", \"_filtered.m9\"))\n",
    "    filenames = filenames.assign(lca=filenames[\"filtered_blast\"].str.replace(\"blast_\"+blast_type, \"lca_\"+blast_type).str.replace(\"_filtered\", \"\"))\n",
    "    commands = filenames.apply(lambda x: \"python lca_analysis.py\"+\\\n",
    "                               \" --blast_type \"+blast_type+\\\n",
    "                               \" --fpath \"+x.iloc[0]+\\\n",
    "                               \" --filtered_blast_path \"+x.iloc[1]+\\\n",
    "                               \" --outpath \"+x.iloc[2]+\\\n",
    "                               \" --ident_cutoff \"+str(ident_cutoff)+\\\n",
    "                               \" --align_len_cutoff \"+str(align_cutoff)+\\\n",
    "                               \" --bitscore_cutoff \"+str(bitscore_cutoff), axis=1)\n",
    "    print (commands)\n",
    "    commands_csv_filename = \"lca_\"+blast_type+\"_commands\"\n",
    "    if not default:\n",
    "        commands_csv_filename += \"_ident\"+str(ident_cutoff)+\"align\"+str(align_cutoff)+\"bitscore\"+str(bitscore_cutoff)\n",
    "    commands.to_csv(commands_csv_filename, index=False)\n",
    "    if (len(filenames) < ncores):\n",
    "        ncores = len(filenames)\n",
    "    command_str = \"parallel -a \"+commands_csv_filename+\" -j \"+str(ncores)\n",
    "    process = subprocess.Popen(command_str.split(), stdout=subprocess.PIPE)\n",
    "    output, error = process.communicate()\n",
    "    return (output, error)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_blast_lca (lca_file_name, blast_file_name, outfile, sample_name, blast_type, output_file_name=None):\n",
    "    lca_data = pd.read_csv(lca_file_name, sep=\"\\t\", header=0)\n",
    "    blast_data = pd.read_csv(blast_file_name, sep=\"\\t\", header=0)\n",
    "    blast_data_grouped = blast_data.groupby([\"query\"], as_index=False).\\\n",
    "    agg({'identity':[\"max\"], 'align_length':[\"max\"], 'mismatches':[\"max\"], 'gaps':[\"max\"],\n",
    "         'qstart':[\"min\"], 'qend':[\"max\"], 'sstart':[\"min\"], 'send':[\"max\"],\n",
    "         'bitscore':[\"max\"]})\n",
    "    blast_data_grouped.columns = blast_data_grouped.columns.get_level_values(0)\n",
    "    grouped_df = pd.merge(blast_data_grouped, lca_data, how=\"left\", on=\"query\")\n",
    "    grouped_df.insert(1, \"blast_type\", value=blast_type)\n",
    "    grouped_df.insert(2, \"sample\", value=sample_name)\n",
    "    df_to_s3(grouped_df, outfile)\n",
    "    outfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "aws s3 ls s3://czbiohub-mosquito/plast/ | grep '.m8' | awk 'NF>1{print $NF}' | parallel -j 72 python create_blast_nr.py --fpath s3://czbiohub-mosquito/plast/{}\n",
    "aws s3 ls s3://lucymli/skeeters/ | grep 'CMS00' | awk 'NF>1{print $NF}' | parallel aws s3 sync s3://lucymli/skeeters/{} blast_nr_output/{}\n",
    "head -n 1 $(find blast_nr_output -type f -name '*.m8' | head -n 1) > header_line \n",
    "for x in `ls blast_nr_output`; do \n",
    "    mkdir -p blast_nr_output_full/$x\n",
    "    head -n 1 $(find blast_nr_output -type f -name '*.m8' | head -n 1) > blast_nr_output_full/$x/blast_nr.m9\n",
    "    ls -d $(find blast_nr_output/$x -type f) | xargs -0 -I file cat file > blast_nr_output_full/$x/blast_nr.m9\n",
    "done\n",
    "ls blast_nr_output_full | parallel aws s3 cp blast_nr_output_full/{}/blast_nr.m9 s3://czbiohub-mosquito/contigs/{}/blast_nr.m9\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## Produce a dataframe from every combination of values\n",
    "## From: https://pandas.pydata.org/pandas-docs/stable/user_guide/cookbook.html\n",
    "##\n",
    "def expand_grid(data_dict):\n",
    "    rows = itertools.product(*data_dict.values())\n",
    "    return pd.DataFrame.from_records(rows, columns=data_dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "combinations = pd.concat([\n",
    "    expand_grid({\n",
    "        \"ident_cutoff\":[x*0.1 for x in range(0, 10)], \\\n",
    "        \"align_cutoff\":[x*0.1 for x in range(0, 10)], \\\n",
    "        \"bitscore_cutoff\":[0]\n",
    "    }),\n",
    "    expand_grid({\"ident_cutoff\":[0], \"align_cutoff\":[0], \"bitscore_cutoff\":[x*0.1 for x in range(0, 10)]})\n",
    "], axis=0, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncores = os.cpu_count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### nt hits\n",
    "The nt hits of contigs from each sample are filtered with ident_cutoff=0.9 and align_len_cutoff=0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(b'', None)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_lca_analysis(input_file_name=\"blast_nt.m9\", output_dir=\"contig_quality\", \\\n",
    "                 bucket_name=bucket_name,\\\n",
    "                 ident_cutoff=0.9, align_cutoff=0.9, bitscore_cutoff=0, \\\n",
    "                 blast_type=\"nt\", default=True, ncores=ncores)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sensitivity analysis using different combinations of cutoffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "lca_nt_paths = [\"s3://\"+bucket_name+\"/\"+x[\"Prefix\"]+\"lca_nt.m9\" \\\n",
    " for x in client.list_objects(Bucket=bucket_name, Prefix=\"contig_quality/\", Delimiter=\"/\")[\"CommonPrefixes\"]]\n",
    "blast_nt_paths = [x.replace(\"lca_nt\", \"blast_nt_filtered\") for x in lca_nt_paths]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: CMS001_012_Ra_S4\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(lca_nt_paths)):\n",
    "    sample_name = os.path.basename(os.path.dirname(lca_nt_paths[i]))\n",
    "    outfile = lca_nt_paths[i].replace(\"lca_nt\", \"blast_lca_nt_filtered\")\n",
    "    try:\n",
    "        combine_blast_lca (lca_nt_paths[i], blast_nt_paths[i], outfile, sample_name, \"nt\")\n",
    "    except:\n",
    "        print (\"error: \"+sample_name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combinations.apply(lambda x: \\\n",
    "                   run_lca_analysis(input_file_name=\"blast_nt.m9\", output_dir=\"contig_quality_sensitivity\", \\\n",
    "                                    ident_cutoff=x[\"ident_cutoff\"], align_cutoff=x[\"align_cutoff\"], \\\n",
    "                                    bitscore_cutoff=x[\"bitscore_cutoff\"], \\\n",
    "                                    blast_type=\"nt\", default=False, ncores=ncores), \\\n",
    "                   axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### subset nr hits\n",
    "The subset nr hits of contigs from each sample are filtered with ident_cutoff=0.9 and align_len_cutoff=0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(b'', None)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_lca_analysis(input_file_name=\"blast_subset_nr.m9\", output_dir=\"contig_quality\", \\\n",
    "                 bucket_name=bucket_name,\\\n",
    "                 ident_cutoff=0.9, align_cutoff=0.9, bitscore_cutoff=0, \\\n",
    "                 blast_type=\"nr\", default=True, ncores=ncores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combinations.apply(lambda x: \\\n",
    "                   run_lca_analysis(input_file_name=\"blast_subset_nr.m9\", output_dir=\"contig_quality_sensitivity\", \\\n",
    "                                    ident_cutoff=x[\"ident_cutoff\"], align_cutoff=x[\"align_cutoff\"], \\\n",
    "                                    bitscore_cutoff=x[\"bitscore_cutoff\"], \\\n",
    "                                    blast_type=\"nr\", default=False, ncores=ncores), \\\n",
    "                   axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### nr hits\n",
    "The nr hits of contigs from each sample are filtered with ident_cutoff=0.9 and align_len_cutoff=0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_lca_analysis(input_file_name=\"blast_nr.m9\", output_dir=\"contig_quality\", \\\n",
    "                 bucket_name=bucket_name,\\\n",
    "                 ident_cutoff=0.9, align_cutoff=0.9, bitscore_cutoff=0, \\\n",
    "                 blast_type=\"nr\", default=True, ncores=ncores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "lca_nr_paths = [\"s3://\"+bucket_name+\"/\"+x[\"Prefix\"]+\"lca_nr.m9\" \\\n",
    " for x in client.list_objects(Bucket=bucket_name, Prefix=\"contig_quality/\", Delimiter=\"/\")[\"CommonPrefixes\"]]\n",
    "blast_nr_paths = [x.replace(\"lca_nr\", \"blast_nr_filtered\") for x in lca_nr_paths]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: CMS001_005_Ra_S3\n",
      "error: CMS001_026_Ra_S18\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(lca_nr_paths)):\n",
    "    sample_name = os.path.basename(os.path.dirname(lca_nr_paths[i]))\n",
    "    outfile = lca_nr_paths[i].replace(\"lca_nr\", \"blast_lca_nr_filtered\")\n",
    "    try:\n",
    "        combine_blast_lca (lca_nr_paths[i], blast_nr_paths[i], outfile, sample_name, \"nr\")\n",
    "    except:\n",
    "        print (\"error: \"+sample_name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## play"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "contig_file = pd.read_csv(\"s3://czbiohub-mosquito/contigs/CMS001_003_Ra_S2/blast_nt.m9\", header=None, sep=\"\\t\", comment=\"#\")\n",
    "blast_lca_file = pd.read_csv(\"s3://czbiohub-mosquito/contig_quality/CMS001_003_Ra_S2/blast_lca_nt_filtered.m9\", header=0, sep=\"\\t\", comment=\"#\")\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[False, False, False, False, False, False, True, False, False, False]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[contig_file[contig_file[0]==x][12].isin(ncbi.get_descendant_taxa(\"7157\")).any()\\\n",
    " for x in blast_lca_file[blast_lca_file[\"taxid\"]==33213][\"query\"]]\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "query           NODE_5053_length_360_cov_0.664311\n",
       "blast_type                                     nt\n",
       "sample                           CMS001_003_Ra_S2\n",
       "identity                                   95.238\n",
       "align_length                                   42\n",
       "mismatches                                      2\n",
       "gaps                                            0\n",
       "qstart                                        120\n",
       "qend                                          164\n",
       "sstart                                       2619\n",
       "send                                     20072008\n",
       "bitscore                                     67.6\n",
       "taxid                                       33213\n",
       "Name: 580, dtype: object"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_lca_file[blast_lca_file[\"taxid\"]==33213].iloc[6,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>461492</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LN010170.1</td>\n",
       "      <td>95.238</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>120</td>\n",
       "      <td>161</td>\n",
       "      <td>2619</td>\n",
       "      <td>2578</td>\n",
       "      <td>2.940000e-07</td>\n",
       "      <td>67.6</td>\n",
       "      <td>99802</td>\n",
       "      <td>Spirometra erinaceieuropaei</td>\n",
       "      <td>Spirometra erinaceieuropaei</td>\n",
       "      <td>Spirometra erinaceieuropaei genome assembly S_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461493</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_005057306.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>968</td>\n",
       "      <td>1002</td>\n",
       "      <td>1.060000e-06</td>\n",
       "      <td>65.8</td>\n",
       "      <td>59894</td>\n",
       "      <td>Ficedula albicollis</td>\n",
       "      <td>collared flycatcher</td>\n",
       "      <td>PREDICTED: Ficedula albicollis myelin transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461494</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>CP026249.1</td>\n",
       "      <td>94.737</td>\n",
       "      <td>38</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>164</td>\n",
       "      <td>20071971</td>\n",
       "      <td>20072008</td>\n",
       "      <td>4.920000e-05</td>\n",
       "      <td>60.2</td>\n",
       "      <td>52904</td>\n",
       "      <td>Scophthalmus maximus</td>\n",
       "      <td>turbot</td>\n",
       "      <td>Scophthalmus maximus chromosome 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461495</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_006970601.2</td>\n",
       "      <td>97.143</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>2180</td>\n",
       "      <td>2214</td>\n",
       "      <td>4.920000e-05</td>\n",
       "      <td>60.2</td>\n",
       "      <td>230844</td>\n",
       "      <td>Peromyscus maniculatus bairdii</td>\n",
       "      <td>prairie deer mouse</td>\n",
       "      <td>PREDICTED: Peromyscus maniculatus bairdii upst...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461496</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_015995521.1</td>\n",
       "      <td>97.143</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>3381</td>\n",
       "      <td>3415</td>\n",
       "      <td>4.920000e-05</td>\n",
       "      <td>60.2</td>\n",
       "      <td>230844</td>\n",
       "      <td>Peromyscus maniculatus bairdii</td>\n",
       "      <td>prairie deer mouse</td>\n",
       "      <td>PREDICTED: Peromyscus maniculatus bairdii upst...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461497</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_015995516.1</td>\n",
       "      <td>97.143</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>2201</td>\n",
       "      <td>2235</td>\n",
       "      <td>4.920000e-05</td>\n",
       "      <td>60.2</td>\n",
       "      <td>230844</td>\n",
       "      <td>Peromyscus maniculatus bairdii</td>\n",
       "      <td>prairie deer mouse</td>\n",
       "      <td>PREDICTED: Peromyscus maniculatus bairdii upst...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461498</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_006970600.1</td>\n",
       "      <td>97.143</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>2291</td>\n",
       "      <td>2325</td>\n",
       "      <td>4.920000e-05</td>\n",
       "      <td>60.2</td>\n",
       "      <td>230844</td>\n",
       "      <td>Peromyscus maniculatus bairdii</td>\n",
       "      <td>prairie deer mouse</td>\n",
       "      <td>PREDICTED: Peromyscus maniculatus bairdii upst...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461499</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_026007050.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1996</td>\n",
       "      <td>2028</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>9627</td>\n",
       "      <td>Vulpes vulpes</td>\n",
       "      <td>red fox</td>\n",
       "      <td>PREDICTED: Vulpes vulpes podocan (PODN), trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461500</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_026007049.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1996</td>\n",
       "      <td>2028</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>9627</td>\n",
       "      <td>Vulpes vulpes</td>\n",
       "      <td>red fox</td>\n",
       "      <td>PREDICTED: Vulpes vulpes podocan (PODN), trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461501</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_025427651.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1953</td>\n",
       "      <td>1985</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>286419</td>\n",
       "      <td>Canis lupus dingo</td>\n",
       "      <td>dingo</td>\n",
       "      <td>PREDICTED: Canis lupus dingo podocan (PODN), t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461502</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_025427650.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1953</td>\n",
       "      <td>1985</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>286419</td>\n",
       "      <td>Canis lupus dingo</td>\n",
       "      <td>dingo</td>\n",
       "      <td>PREDICTED: Canis lupus dingo podocan (PODN), t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461503</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_025427649.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1970</td>\n",
       "      <td>2002</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>286419</td>\n",
       "      <td>Canis lupus dingo</td>\n",
       "      <td>dingo</td>\n",
       "      <td>PREDICTED: Canis lupus dingo podocan (PODN), t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461504</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_025427648.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1970</td>\n",
       "      <td>2002</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>286419</td>\n",
       "      <td>Canis lupus dingo</td>\n",
       "      <td>dingo</td>\n",
       "      <td>PREDICTED: Canis lupus dingo podocan (PODN), t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461505</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_025427647.1</td>\n",
       "      <td>96.970</td>\n",
       "      <td>33</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>159</td>\n",
       "      <td>1970</td>\n",
       "      <td>2002</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>286419</td>\n",
       "      <td>Canis lupus dingo</td>\n",
       "      <td>dingo</td>\n",
       "      <td>PREDICTED: Canis lupus dingo podocan (PODN), t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461506</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_023957260.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>162</td>\n",
       "      <td>2887</td>\n",
       "      <td>2916</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>8090</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>Japanese medaka</td>\n",
       "      <td>PREDICTED: Oryzias latipes polycomb protein su...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461507</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>CP020642.1</td>\n",
       "      <td>94.444</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>162</td>\n",
       "      <td>469123</td>\n",
       "      <td>469158</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>8090</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>Japanese medaka</td>\n",
       "      <td>Oryzias latipes strain HSOK chromosome 22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461508</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>CP020628.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>162</td>\n",
       "      <td>1110383</td>\n",
       "      <td>1110354</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>8090</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>Japanese medaka</td>\n",
       "      <td>Oryzias latipes strain HSOK chromosome 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461509</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>CP020786.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>162</td>\n",
       "      <td>814021</td>\n",
       "      <td>813992</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>8090</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>Japanese medaka</td>\n",
       "      <td>Oryzias latipes strain HNI chromosome 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461510</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>CP020672.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>162</td>\n",
       "      <td>1076377</td>\n",
       "      <td>1076348</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>8090</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>Japanese medaka</td>\n",
       "      <td>Oryzias latipes strain Hd-rR chromosome 8 sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461511</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>HF933214.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>162</td>\n",
       "      <td>766820</td>\n",
       "      <td>766791</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>8090</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>Japanese medaka</td>\n",
       "      <td>TPA_asm: Oryzias latipes strain Hd-rR, complet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461512</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_004662178.2</td>\n",
       "      <td>94.444</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>126</td>\n",
       "      <td>161</td>\n",
       "      <td>1815</td>\n",
       "      <td>1850</td>\n",
       "      <td>6.370000e-04</td>\n",
       "      <td>56.5</td>\n",
       "      <td>51337</td>\n",
       "      <td>Jaculus jaculus</td>\n",
       "      <td>lesser Egyptian jerboa</td>\n",
       "      <td>PREDICTED: Jaculus jaculus kinesin family memb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461513</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_026794755.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>986</td>\n",
       "      <td>1020</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>44394</td>\n",
       "      <td>Zonotrichia albicollis</td>\n",
       "      <td>white-throated sparrow</td>\n",
       "      <td>PREDICTED: Zonotrichia albicollis myelin trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461514</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_005489120.3</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>972</td>\n",
       "      <td>1006</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>44394</td>\n",
       "      <td>Zonotrichia albicollis</td>\n",
       "      <td>white-throated sparrow</td>\n",
       "      <td>PREDICTED: Zonotrichia albicollis myelin trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461515</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_026794754.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>1191</td>\n",
       "      <td>1225</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>44394</td>\n",
       "      <td>Zonotrichia albicollis</td>\n",
       "      <td>white-throated sparrow</td>\n",
       "      <td>PREDICTED: Zonotrichia albicollis myelin trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461516</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_014268500.2</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>963</td>\n",
       "      <td>997</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>44394</td>\n",
       "      <td>Zonotrichia albicollis</td>\n",
       "      <td>white-throated sparrow</td>\n",
       "      <td>PREDICTED: Zonotrichia albicollis myelin trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461517</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LT841157.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>650769</td>\n",
       "      <td>650803</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax genome assembly, chromosome: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461518</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_023939681.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>992</td>\n",
       "      <td>1026</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>156563</td>\n",
       "      <td>Cyanistes caeruleus</td>\n",
       "      <td>blue tit</td>\n",
       "      <td>PREDICTED: Cyanistes caeruleus myelin transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461519</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_022610224.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>1648</td>\n",
       "      <td>1682</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>1071383</td>\n",
       "      <td>Kazachstania naganishii CBS 8797</td>\n",
       "      <td>Kazachstania naganishii CBS 8797</td>\n",
       "      <td>Kazachstania naganishii CBS 8797 hypothetical ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461520</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_021696928.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>281</td>\n",
       "      <td>315</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>29088</td>\n",
       "      <td>Neomonachus schauinslandi</td>\n",
       "      <td>Hawaiian monk seal</td>\n",
       "      <td>PREDICTED: Neomonachus schauinslandi DEK proto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461521</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_021696926.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>276</td>\n",
       "      <td>310</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>29088</td>\n",
       "      <td>Neomonachus schauinslandi</td>\n",
       "      <td>Hawaiian monk seal</td>\n",
       "      <td>PREDICTED: Neomonachus schauinslandi DEK proto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461522</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_021477301.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>159</td>\n",
       "      <td>825</td>\n",
       "      <td>853</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>PREDICTED: Danio rerio si:dkey-6e2.2 (si:dkey-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461523</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_017356722.2</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>159</td>\n",
       "      <td>1265</td>\n",
       "      <td>1293</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>PREDICTED: Danio rerio si:dkey-6e2.2 (si:dkey-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461524</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_005166268.4</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>159</td>\n",
       "      <td>976</td>\n",
       "      <td>1004</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>PREDICTED: Danio rerio si:dkey-6e2.2 (si:dkey-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461525</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LT635619.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>630045</td>\n",
       "      <td>630079</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax genome assembly, chromosome: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461526</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LT615263.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>658482</td>\n",
       "      <td>658516</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax genome assembly, chromosome: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461527</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LT615263.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>658506</td>\n",
       "      <td>658540</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax genome assembly, chromosome: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461528</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LT615246.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>648479</td>\n",
       "      <td>648513</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax genome assembly, chromosome: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461529</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>LT615246.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>648503</td>\n",
       "      <td>648537</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax genome assembly, chromosome: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461530</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_018208761.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>155</td>\n",
       "      <td>2770</td>\n",
       "      <td>2798</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>149040</td>\n",
       "      <td>Phialocephala scopiformis</td>\n",
       "      <td>Phialocephala scopiformis</td>\n",
       "      <td>Phialocephala scopiformis hypothetical protein...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461531</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_014873945.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>814</td>\n",
       "      <td>848</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>9172</td>\n",
       "      <td>Sturnus vulgaris</td>\n",
       "      <td>common starling</td>\n",
       "      <td>PREDICTED: Sturnus vulgaris myelin transcripti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461532</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_006727724.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>103</td>\n",
       "      <td>137</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>9713</td>\n",
       "      <td>Leptonychotes weddellii</td>\n",
       "      <td>Weddell seal</td>\n",
       "      <td>PREDICTED: Leptonychotes weddellii DEK oncogen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461533</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_005422164.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>814</td>\n",
       "      <td>848</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>48883</td>\n",
       "      <td>Geospiza fortis</td>\n",
       "      <td>medium ground-finch</td>\n",
       "      <td>PREDICTED: Geospiza fortis myelin transcriptio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461534</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>HE978323.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>161</td>\n",
       "      <td>397681</td>\n",
       "      <td>397653</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>1071383</td>\n",
       "      <td>Kazachstania naganishii CBS 8797</td>\n",
       "      <td>Kazachstania naganishii CBS 8797</td>\n",
       "      <td>Kazachstania naganishii CBS 8797 chromosome 10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461535</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_003329815.2</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>155</td>\n",
       "      <td>727</td>\n",
       "      <td>755</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>418459</td>\n",
       "      <td>Puccinia graminis f. sp. tritici CRL 75-36-700-3</td>\n",
       "      <td>Puccinia graminis f. sp. tritici CRL 75-36-700-3</td>\n",
       "      <td>Puccinia graminis f. sp. tritici CRL 75-36-700...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461536</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_556109.2</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>161</td>\n",
       "      <td>673</td>\n",
       "      <td>701</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>180454</td>\n",
       "      <td>Anopheles gambiae str. PEST</td>\n",
       "      <td>Anopheles gambiae str. PEST</td>\n",
       "      <td>Anopheles gambiae str. PEST AGAP005468-RB (Aga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461537</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_001688613.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>161</td>\n",
       "      <td>673</td>\n",
       "      <td>701</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>180454</td>\n",
       "      <td>Anopheles gambiae str. PEST</td>\n",
       "      <td>Anopheles gambiae str. PEST</td>\n",
       "      <td>Anopheles gambiae str. PEST AGAP005468-RA (Aga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461538</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_001614404.1</td>\n",
       "      <td>94.286</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>161</td>\n",
       "      <td>526</td>\n",
       "      <td>560</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>5855</td>\n",
       "      <td>Plasmodium vivax</td>\n",
       "      <td>malaria parasite P. vivax</td>\n",
       "      <td>Plasmodium vivax hypothetical protein, conserv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461539</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>BX936390.7</td>\n",
       "      <td>100.000</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>159</td>\n",
       "      <td>53677</td>\n",
       "      <td>53649</td>\n",
       "      <td>2.000000e-03</td>\n",
       "      <td>54.7</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>Zebrafish DNA sequence from clone DKEY-6E2 in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461540</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_026816044.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>161</td>\n",
       "      <td>214</td>\n",
       "      <td>241</td>\n",
       "      <td>8.000000e-03</td>\n",
       "      <td>52.8</td>\n",
       "      <td>156304</td>\n",
       "      <td>Ceratina calcarata</td>\n",
       "      <td>Ceratina calcarata</td>\n",
       "      <td>PREDICTED: Ceratina calcarata zinc finger prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461541</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_018029345.2</td>\n",
       "      <td>100.000</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>161</td>\n",
       "      <td>287</td>\n",
       "      <td>314</td>\n",
       "      <td>8.000000e-03</td>\n",
       "      <td>52.8</td>\n",
       "      <td>156304</td>\n",
       "      <td>Ceratina calcarata</td>\n",
       "      <td>Ceratina calcarata</td>\n",
       "      <td>PREDICTED: Ceratina calcarata zinc finger prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461542</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_026816043.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>161</td>\n",
       "      <td>309</td>\n",
       "      <td>336</td>\n",
       "      <td>8.000000e-03</td>\n",
       "      <td>52.8</td>\n",
       "      <td>156304</td>\n",
       "      <td>Ceratina calcarata</td>\n",
       "      <td>Ceratina calcarata</td>\n",
       "      <td>PREDICTED: Ceratina calcarata zinc finger prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461543</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_024269549.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>161</td>\n",
       "      <td>1289</td>\n",
       "      <td>1316</td>\n",
       "      <td>8.000000e-03</td>\n",
       "      <td>52.8</td>\n",
       "      <td>30732</td>\n",
       "      <td>Oryzias melastigma</td>\n",
       "      <td>Indian medaka</td>\n",
       "      <td>PREDICTED: Oryzias melastigma uncharacterized ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461544</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_021578254.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>160</td>\n",
       "      <td>360</td>\n",
       "      <td>333</td>\n",
       "      <td>8.000000e-03</td>\n",
       "      <td>52.8</td>\n",
       "      <td>8022</td>\n",
       "      <td>Oncorhynchus mykiss</td>\n",
       "      <td>rainbow trout</td>\n",
       "      <td>PREDICTED: Oncorhynchus mykiss cGMP-specific 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461545</th>\n",
       "      <td>NODE_5053_length_360_cov_0.664311</td>\n",
       "      <td>XM_020938590.1</td>\n",
       "      <td>94.118</td>\n",
       "      <td>34</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>160</td>\n",
       "      <td>1286</td>\n",
       "      <td>1319</td>\n",
       "      <td>8.000000e-03</td>\n",
       "      <td>52.8</td>\n",
       "      <td>150288</td>\n",
       "      <td>Boleophthalmus pectinirostris</td>\n",
       "      <td>great blue-spotted mudskipper</td>\n",
       "      <td>PREDICTED: Boleophthalmus pectinirostris major...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       0               1        2   3   4   \\\n",
       "461492  NODE_5053_length_360_cov_0.664311      LN010170.1   95.238  42   2   \n",
       "461493  NODE_5053_length_360_cov_0.664311  XM_005057306.1  100.000  35   0   \n",
       "461494  NODE_5053_length_360_cov_0.664311      CP026249.1   94.737  38   2   \n",
       "461495  NODE_5053_length_360_cov_0.664311  XM_006970601.2   97.143  35   1   \n",
       "461496  NODE_5053_length_360_cov_0.664311  XM_015995521.1   97.143  35   1   \n",
       "461497  NODE_5053_length_360_cov_0.664311  XM_015995516.1   97.143  35   1   \n",
       "461498  NODE_5053_length_360_cov_0.664311  XM_006970600.1   97.143  35   1   \n",
       "461499  NODE_5053_length_360_cov_0.664311  XM_026007050.1   96.970  33   1   \n",
       "461500  NODE_5053_length_360_cov_0.664311  XM_026007049.1   96.970  33   1   \n",
       "461501  NODE_5053_length_360_cov_0.664311  XM_025427651.1   96.970  33   1   \n",
       "461502  NODE_5053_length_360_cov_0.664311  XM_025427650.1   96.970  33   1   \n",
       "461503  NODE_5053_length_360_cov_0.664311  XM_025427649.1   96.970  33   1   \n",
       "461504  NODE_5053_length_360_cov_0.664311  XM_025427648.1   96.970  33   1   \n",
       "461505  NODE_5053_length_360_cov_0.664311  XM_025427647.1   96.970  33   1   \n",
       "461506  NODE_5053_length_360_cov_0.664311  XM_023957260.1  100.000  30   0   \n",
       "461507  NODE_5053_length_360_cov_0.664311      CP020642.1   94.444  36   2   \n",
       "461508  NODE_5053_length_360_cov_0.664311      CP020628.1  100.000  30   0   \n",
       "461509  NODE_5053_length_360_cov_0.664311      CP020786.1  100.000  30   0   \n",
       "461510  NODE_5053_length_360_cov_0.664311      CP020672.1  100.000  30   0   \n",
       "461511  NODE_5053_length_360_cov_0.664311      HF933214.1  100.000  30   0   \n",
       "461512  NODE_5053_length_360_cov_0.664311  XM_004662178.2   94.444  36   2   \n",
       "461513  NODE_5053_length_360_cov_0.664311  XM_026794755.1   94.286  35   2   \n",
       "461514  NODE_5053_length_360_cov_0.664311  XM_005489120.3   94.286  35   2   \n",
       "461515  NODE_5053_length_360_cov_0.664311  XM_026794754.1   94.286  35   2   \n",
       "461516  NODE_5053_length_360_cov_0.664311  XM_014268500.2   94.286  35   2   \n",
       "461517  NODE_5053_length_360_cov_0.664311      LT841157.1   94.286  35   2   \n",
       "461518  NODE_5053_length_360_cov_0.664311  XM_023939681.1   94.286  35   2   \n",
       "461519  NODE_5053_length_360_cov_0.664311  XM_022610224.1   94.286  35   2   \n",
       "461520  NODE_5053_length_360_cov_0.664311  XM_021696928.1   94.286  35   2   \n",
       "461521  NODE_5053_length_360_cov_0.664311  XM_021696926.1   94.286  35   2   \n",
       "461522  NODE_5053_length_360_cov_0.664311  XM_021477301.1  100.000  29   0   \n",
       "461523  NODE_5053_length_360_cov_0.664311  XM_017356722.2  100.000  29   0   \n",
       "461524  NODE_5053_length_360_cov_0.664311  XM_005166268.4  100.000  29   0   \n",
       "461525  NODE_5053_length_360_cov_0.664311      LT635619.1   94.286  35   2   \n",
       "461526  NODE_5053_length_360_cov_0.664311      LT615263.1   94.286  35   2   \n",
       "461527  NODE_5053_length_360_cov_0.664311      LT615263.1   94.286  35   2   \n",
       "461528  NODE_5053_length_360_cov_0.664311      LT615246.1   94.286  35   2   \n",
       "461529  NODE_5053_length_360_cov_0.664311      LT615246.1   94.286  35   2   \n",
       "461530  NODE_5053_length_360_cov_0.664311  XM_018208761.1  100.000  29   0   \n",
       "461531  NODE_5053_length_360_cov_0.664311  XM_014873945.1   94.286  35   2   \n",
       "461532  NODE_5053_length_360_cov_0.664311  XM_006727724.1   94.286  35   2   \n",
       "461533  NODE_5053_length_360_cov_0.664311  XM_005422164.1   94.286  35   2   \n",
       "461534  NODE_5053_length_360_cov_0.664311      HE978323.1  100.000  29   0   \n",
       "461535  NODE_5053_length_360_cov_0.664311  XM_003329815.2  100.000  29   0   \n",
       "461536  NODE_5053_length_360_cov_0.664311     XM_556109.2  100.000  29   0   \n",
       "461537  NODE_5053_length_360_cov_0.664311  XM_001688613.1  100.000  29   0   \n",
       "461538  NODE_5053_length_360_cov_0.664311  XM_001614404.1   94.286  35   2   \n",
       "461539  NODE_5053_length_360_cov_0.664311      BX936390.7  100.000  29   0   \n",
       "461540  NODE_5053_length_360_cov_0.664311  XM_026816044.1  100.000  28   0   \n",
       "461541  NODE_5053_length_360_cov_0.664311  XM_018029345.2  100.000  28   0   \n",
       "461542  NODE_5053_length_360_cov_0.664311  XM_026816043.1  100.000  28   0   \n",
       "461543  NODE_5053_length_360_cov_0.664311  XM_024269549.1  100.000  28   0   \n",
       "461544  NODE_5053_length_360_cov_0.664311  XM_021578254.1  100.000  28   0   \n",
       "461545  NODE_5053_length_360_cov_0.664311  XM_020938590.1   94.118  34   2   \n",
       "\n",
       "        5    6    7         8         9             10    11       12  \\\n",
       "461492   0  120  161      2619      2578  2.940000e-07  67.6    99802   \n",
       "461493   0  127  161       968      1002  1.060000e-06  65.8    59894   \n",
       "461494   0  127  164  20071971  20072008  4.920000e-05  60.2    52904   \n",
       "461495   0  127  161      2180      2214  4.920000e-05  60.2   230844   \n",
       "461496   0  127  161      3381      3415  4.920000e-05  60.2   230844   \n",
       "461497   0  127  161      2201      2235  4.920000e-05  60.2   230844   \n",
       "461498   0  127  161      2291      2325  4.920000e-05  60.2   230844   \n",
       "461499   0  127  159      1996      2028  6.370000e-04  56.5     9627   \n",
       "461500   0  127  159      1996      2028  6.370000e-04  56.5     9627   \n",
       "461501   0  127  159      1953      1985  6.370000e-04  56.5   286419   \n",
       "461502   0  127  159      1953      1985  6.370000e-04  56.5   286419   \n",
       "461503   0  127  159      1970      2002  6.370000e-04  56.5   286419   \n",
       "461504   0  127  159      1970      2002  6.370000e-04  56.5   286419   \n",
       "461505   0  127  159      1970      2002  6.370000e-04  56.5   286419   \n",
       "461506   0  133  162      2887      2916  6.370000e-04  56.5     8090   \n",
       "461507   0  127  162    469123    469158  6.370000e-04  56.5     8090   \n",
       "461508   0  133  162   1110383   1110354  6.370000e-04  56.5     8090   \n",
       "461509   0  133  162    814021    813992  6.370000e-04  56.5     8090   \n",
       "461510   0  133  162   1076377   1076348  6.370000e-04  56.5     8090   \n",
       "461511   0  133  162    766820    766791  6.370000e-04  56.5     8090   \n",
       "461512   0  126  161      1815      1850  6.370000e-04  56.5    51337   \n",
       "461513   0  127  161       986      1020  2.000000e-03  54.7    44394   \n",
       "461514   0  127  161       972      1006  2.000000e-03  54.7    44394   \n",
       "461515   0  127  161      1191      1225  2.000000e-03  54.7    44394   \n",
       "461516   0  127  161       963       997  2.000000e-03  54.7    44394   \n",
       "461517   0  127  161    650769    650803  2.000000e-03  54.7     5855   \n",
       "461518   0  127  161       992      1026  2.000000e-03  54.7   156563   \n",
       "461519   0  127  161      1648      1682  2.000000e-03  54.7  1071383   \n",
       "461520   0  127  161       281       315  2.000000e-03  54.7    29088   \n",
       "461521   0  127  161       276       310  2.000000e-03  54.7    29088   \n",
       "461522   0  131  159       825       853  2.000000e-03  54.7     7955   \n",
       "461523   0  131  159      1265      1293  2.000000e-03  54.7     7955   \n",
       "461524   0  131  159       976      1004  2.000000e-03  54.7     7955   \n",
       "461525   0  127  161    630045    630079  2.000000e-03  54.7     5855   \n",
       "461526   0  127  161    658482    658516  2.000000e-03  54.7     5855   \n",
       "461527   0  127  161    658506    658540  2.000000e-03  54.7     5855   \n",
       "461528   0  127  161    648479    648513  2.000000e-03  54.7     5855   \n",
       "461529   0  127  161    648503    648537  2.000000e-03  54.7     5855   \n",
       "461530   0  127  155      2770      2798  2.000000e-03  54.7   149040   \n",
       "461531   0  127  161       814       848  2.000000e-03  54.7     9172   \n",
       "461532   0  127  161       103       137  2.000000e-03  54.7     9713   \n",
       "461533   0  127  161       814       848  2.000000e-03  54.7    48883   \n",
       "461534   0  133  161    397681    397653  2.000000e-03  54.7  1071383   \n",
       "461535   0  127  155       727       755  2.000000e-03  54.7   418459   \n",
       "461536   0  133  161       673       701  2.000000e-03  54.7   180454   \n",
       "461537   0  133  161       673       701  2.000000e-03  54.7   180454   \n",
       "461538   0  127  161       526       560  2.000000e-03  54.7     5855   \n",
       "461539   0  131  159     53677     53649  2.000000e-03  54.7     7955   \n",
       "461540   0  134  161       214       241  8.000000e-03  52.8   156304   \n",
       "461541   0  134  161       287       314  8.000000e-03  52.8   156304   \n",
       "461542   0  134  161       309       336  8.000000e-03  52.8   156304   \n",
       "461543   0  134  161      1289      1316  8.000000e-03  52.8    30732   \n",
       "461544   0  133  160       360       333  8.000000e-03  52.8     8022   \n",
       "461545   0  127  160      1286      1319  8.000000e-03  52.8   150288   \n",
       "\n",
       "                                                      13  \\\n",
       "461492                       Spirometra erinaceieuropaei   \n",
       "461493                               Ficedula albicollis   \n",
       "461494                              Scophthalmus maximus   \n",
       "461495                    Peromyscus maniculatus bairdii   \n",
       "461496                    Peromyscus maniculatus bairdii   \n",
       "461497                    Peromyscus maniculatus bairdii   \n",
       "461498                    Peromyscus maniculatus bairdii   \n",
       "461499                                     Vulpes vulpes   \n",
       "461500                                     Vulpes vulpes   \n",
       "461501                                 Canis lupus dingo   \n",
       "461502                                 Canis lupus dingo   \n",
       "461503                                 Canis lupus dingo   \n",
       "461504                                 Canis lupus dingo   \n",
       "461505                                 Canis lupus dingo   \n",
       "461506                                   Oryzias latipes   \n",
       "461507                                   Oryzias latipes   \n",
       "461508                                   Oryzias latipes   \n",
       "461509                                   Oryzias latipes   \n",
       "461510                                   Oryzias latipes   \n",
       "461511                                   Oryzias latipes   \n",
       "461512                                   Jaculus jaculus   \n",
       "461513                            Zonotrichia albicollis   \n",
       "461514                            Zonotrichia albicollis   \n",
       "461515                            Zonotrichia albicollis   \n",
       "461516                            Zonotrichia albicollis   \n",
       "461517                                  Plasmodium vivax   \n",
       "461518                               Cyanistes caeruleus   \n",
       "461519                  Kazachstania naganishii CBS 8797   \n",
       "461520                         Neomonachus schauinslandi   \n",
       "461521                         Neomonachus schauinslandi   \n",
       "461522                                       Danio rerio   \n",
       "461523                                       Danio rerio   \n",
       "461524                                       Danio rerio   \n",
       "461525                                  Plasmodium vivax   \n",
       "461526                                  Plasmodium vivax   \n",
       "461527                                  Plasmodium vivax   \n",
       "461528                                  Plasmodium vivax   \n",
       "461529                                  Plasmodium vivax   \n",
       "461530                         Phialocephala scopiformis   \n",
       "461531                                  Sturnus vulgaris   \n",
       "461532                           Leptonychotes weddellii   \n",
       "461533                                   Geospiza fortis   \n",
       "461534                  Kazachstania naganishii CBS 8797   \n",
       "461535  Puccinia graminis f. sp. tritici CRL 75-36-700-3   \n",
       "461536                       Anopheles gambiae str. PEST   \n",
       "461537                       Anopheles gambiae str. PEST   \n",
       "461538                                  Plasmodium vivax   \n",
       "461539                                       Danio rerio   \n",
       "461540                                Ceratina calcarata   \n",
       "461541                                Ceratina calcarata   \n",
       "461542                                Ceratina calcarata   \n",
       "461543                                Oryzias melastigma   \n",
       "461544                               Oncorhynchus mykiss   \n",
       "461545                     Boleophthalmus pectinirostris   \n",
       "\n",
       "                                                      14  \\\n",
       "461492                       Spirometra erinaceieuropaei   \n",
       "461493                               collared flycatcher   \n",
       "461494                                            turbot   \n",
       "461495                                prairie deer mouse   \n",
       "461496                                prairie deer mouse   \n",
       "461497                                prairie deer mouse   \n",
       "461498                                prairie deer mouse   \n",
       "461499                                           red fox   \n",
       "461500                                           red fox   \n",
       "461501                                             dingo   \n",
       "461502                                             dingo   \n",
       "461503                                             dingo   \n",
       "461504                                             dingo   \n",
       "461505                                             dingo   \n",
       "461506                                   Japanese medaka   \n",
       "461507                                   Japanese medaka   \n",
       "461508                                   Japanese medaka   \n",
       "461509                                   Japanese medaka   \n",
       "461510                                   Japanese medaka   \n",
       "461511                                   Japanese medaka   \n",
       "461512                            lesser Egyptian jerboa   \n",
       "461513                            white-throated sparrow   \n",
       "461514                            white-throated sparrow   \n",
       "461515                            white-throated sparrow   \n",
       "461516                            white-throated sparrow   \n",
       "461517                         malaria parasite P. vivax   \n",
       "461518                                          blue tit   \n",
       "461519                  Kazachstania naganishii CBS 8797   \n",
       "461520                                Hawaiian monk seal   \n",
       "461521                                Hawaiian monk seal   \n",
       "461522                                         zebrafish   \n",
       "461523                                         zebrafish   \n",
       "461524                                         zebrafish   \n",
       "461525                         malaria parasite P. vivax   \n",
       "461526                         malaria parasite P. vivax   \n",
       "461527                         malaria parasite P. vivax   \n",
       "461528                         malaria parasite P. vivax   \n",
       "461529                         malaria parasite P. vivax   \n",
       "461530                         Phialocephala scopiformis   \n",
       "461531                                   common starling   \n",
       "461532                                      Weddell seal   \n",
       "461533                               medium ground-finch   \n",
       "461534                  Kazachstania naganishii CBS 8797   \n",
       "461535  Puccinia graminis f. sp. tritici CRL 75-36-700-3   \n",
       "461536                       Anopheles gambiae str. PEST   \n",
       "461537                       Anopheles gambiae str. PEST   \n",
       "461538                         malaria parasite P. vivax   \n",
       "461539                                         zebrafish   \n",
       "461540                                Ceratina calcarata   \n",
       "461541                                Ceratina calcarata   \n",
       "461542                                Ceratina calcarata   \n",
       "461543                                     Indian medaka   \n",
       "461544                                     rainbow trout   \n",
       "461545                     great blue-spotted mudskipper   \n",
       "\n",
       "                                                       15  \n",
       "461492  Spirometra erinaceieuropaei genome assembly S_...  \n",
       "461493  PREDICTED: Ficedula albicollis myelin transcri...  \n",
       "461494                  Scophthalmus maximus chromosome 7  \n",
       "461495  PREDICTED: Peromyscus maniculatus bairdii upst...  \n",
       "461496  PREDICTED: Peromyscus maniculatus bairdii upst...  \n",
       "461497  PREDICTED: Peromyscus maniculatus bairdii upst...  \n",
       "461498  PREDICTED: Peromyscus maniculatus bairdii upst...  \n",
       "461499  PREDICTED: Vulpes vulpes podocan (PODN), trans...  \n",
       "461500  PREDICTED: Vulpes vulpes podocan (PODN), trans...  \n",
       "461501  PREDICTED: Canis lupus dingo podocan (PODN), t...  \n",
       "461502  PREDICTED: Canis lupus dingo podocan (PODN), t...  \n",
       "461503  PREDICTED: Canis lupus dingo podocan (PODN), t...  \n",
       "461504  PREDICTED: Canis lupus dingo podocan (PODN), t...  \n",
       "461505  PREDICTED: Canis lupus dingo podocan (PODN), t...  \n",
       "461506  PREDICTED: Oryzias latipes polycomb protein su...  \n",
       "461507          Oryzias latipes strain HSOK chromosome 22  \n",
       "461508           Oryzias latipes strain HSOK chromosome 8  \n",
       "461509            Oryzias latipes strain HNI chromosome 8  \n",
       "461510  Oryzias latipes strain Hd-rR chromosome 8 sequ...  \n",
       "461511  TPA_asm: Oryzias latipes strain Hd-rR, complet...  \n",
       "461512  PREDICTED: Jaculus jaculus kinesin family memb...  \n",
       "461513  PREDICTED: Zonotrichia albicollis myelin trans...  \n",
       "461514  PREDICTED: Zonotrichia albicollis myelin trans...  \n",
       "461515  PREDICTED: Zonotrichia albicollis myelin trans...  \n",
       "461516  PREDICTED: Zonotrichia albicollis myelin trans...  \n",
       "461517    Plasmodium vivax genome assembly, chromosome: 8  \n",
       "461518  PREDICTED: Cyanistes caeruleus myelin transcri...  \n",
       "461519  Kazachstania naganishii CBS 8797 hypothetical ...  \n",
       "461520  PREDICTED: Neomonachus schauinslandi DEK proto...  \n",
       "461521  PREDICTED: Neomonachus schauinslandi DEK proto...  \n",
       "461522  PREDICTED: Danio rerio si:dkey-6e2.2 (si:dkey-...  \n",
       "461523  PREDICTED: Danio rerio si:dkey-6e2.2 (si:dkey-...  \n",
       "461524  PREDICTED: Danio rerio si:dkey-6e2.2 (si:dkey-...  \n",
       "461525    Plasmodium vivax genome assembly, chromosome: 8  \n",
       "461526    Plasmodium vivax genome assembly, chromosome: 8  \n",
       "461527    Plasmodium vivax genome assembly, chromosome: 8  \n",
       "461528    Plasmodium vivax genome assembly, chromosome: 8  \n",
       "461529    Plasmodium vivax genome assembly, chromosome: 8  \n",
       "461530  Phialocephala scopiformis hypothetical protein...  \n",
       "461531  PREDICTED: Sturnus vulgaris myelin transcripti...  \n",
       "461532  PREDICTED: Leptonychotes weddellii DEK oncogen...  \n",
       "461533  PREDICTED: Geospiza fortis myelin transcriptio...  \n",
       "461534  Kazachstania naganishii CBS 8797 chromosome 10...  \n",
       "461535  Puccinia graminis f. sp. tritici CRL 75-36-700...  \n",
       "461536  Anopheles gambiae str. PEST AGAP005468-RB (Aga...  \n",
       "461537  Anopheles gambiae str. PEST AGAP005468-RA (Aga...  \n",
       "461538  Plasmodium vivax hypothetical protein, conserv...  \n",
       "461539  Zebrafish DNA sequence from clone DKEY-6E2 in ...  \n",
       "461540  PREDICTED: Ceratina calcarata zinc finger prot...  \n",
       "461541  PREDICTED: Ceratina calcarata zinc finger prot...  \n",
       "461542  PREDICTED: Ceratina calcarata zinc finger prot...  \n",
       "461543  PREDICTED: Oryzias melastigma uncharacterized ...  \n",
       "461544  PREDICTED: Oncorhynchus mykiss cGMP-specific 3...  \n",
       "461545  PREDICTED: Boleophthalmus pectinirostris major...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_file[contig_file[0]==\"NODE_5053_length_360_cov_0.664311\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = contig_file\n",
    "db=\"nucleotide\"\n",
    "return_taxid_only=True\n",
    "ident_cutoff=0.9\n",
    "align_len_cutoff=0.9\n",
    "bitscore_cutoff=0\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NODE_285_length_1096_cov_0.607458     234514\n",
       "NODE_1136_length_656_cov_0.411054     158521\n",
       "NODE_15469_length_236_cov_0.918239    142173\n",
       "NODE_14382_length_241_cov_0.890244    123947\n",
       "NODE_13150_length_248_cov_0.853801    122442\n",
       "NODE_7415_length_303_cov_1.088496      72063\n",
       "NODE_15289_length_237_cov_0.887500     67027\n",
       "NODE_6460_length_323_cov_1.012195      56131\n",
       "NODE_1369_length_610_cov_0.474672      44797\n",
       "NODE_7634_length_299_cov_0.860360      42498\n",
       "NODE_13451_length_246_cov_0.863905     36120\n",
       "NODE_15422_length_236_cov_0.918239     34179\n",
       "NODE_16432_length_231_cov_0.948052     31562\n",
       "NODE_6654_length_319_cov_0.685950      27936\n",
       "NODE_8220_length_288_cov_22.085308     22213\n",
       "NODE_10099_length_268_cov_0.963351     15842\n",
       "NODE_9043_length_277_cov_0.310000      12949\n",
       "NODE_11445_length_259_cov_0.802198     10864\n",
       "NODE_8440_length_285_cov_0.754808      10035\n",
       "NODE_15589_length_235_cov_0.924051      9619\n",
       "NODE_10308_length_267_cov_0.768421      8183\n",
       "NODE_14865_length_239_cov_0.901235      6172\n",
       "NODE_14885_length_239_cov_0.783951      4445\n",
       "NODE_1654_length_565_cov_3.143443       3437\n",
       "NODE_8951_length_277_cov_1.275000       3405\n",
       "NODE_5806_length_339_cov_0.862595       3315\n",
       "NODE_17126_length_228_cov_0.966887      3050\n",
       "NODE_4658_length_373_cov_0.743243       2697\n",
       "NODE_305_length_1065_cov_2.871457       1951\n",
       "NODE_3760_length_406_cov_1.829787       1947\n",
       "                                       ...  \n",
       "NODE_3619_length_413_cov_0.761905          1\n",
       "NODE_11843_length_256_cov_1.162011         1\n",
       "NODE_13842_length_244_cov_0.874251         1\n",
       "NODE_581_length_836_cov_6.426877           1\n",
       "NODE_9817_length_270_cov_1.512953          1\n",
       "NODE_16668_length_230_cov_0.954248         1\n",
       "NODE_5233_length_355_cov_0.428058          1\n",
       "NODE_6072_length_332_cov_1.349020          1\n",
       "NODE_2734_length_461_cov_0.919271          1\n",
       "NODE_144_length_1380_cov_4.095932          1\n",
       "NODE_12635_length_251_cov_1.321839         1\n",
       "NODE_2454_length_483_cov_0.864532          1\n",
       "NODE_3037_length_442_cov_1.306849          1\n",
       "NODE_6151_length_330_cov_1.675889          1\n",
       "NODE_3316_length_428_cov_0.925926          1\n",
       "NODE_15597_length_235_cov_0.924051         1\n",
       "NODE_3518_length_418_cov_1.005865          1\n",
       "NODE_8537_length_283_cov_0.941748          1\n",
       "NODE_4363_length_384_cov_0.706840          1\n",
       "NODE_7334_length_305_cov_0.741228          1\n",
       "NODE_712_length_783_cov_1.582153           1\n",
       "NODE_60_length_1768_cov_0.777055           1\n",
       "NODE_15901_length_234_cov_0.929936         1\n",
       "NODE_1549_length_582_cov_0.867327          1\n",
       "NODE_16915_length_229_cov_0.960526         1\n",
       "NODE_9517_length_273_cov_0.744898          1\n",
       "NODE_10666_length_264_cov_0.780749         1\n",
       "NODE_14333_length_241_cov_1.085366         1\n",
       "NODE_4328_length_385_cov_0.931818          1\n",
       "NODE_9340_length_274_cov_0.817259          1\n",
       "Name: 0, Length: 831, dtype: int64"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[0].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
