{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from ete3 import NCBITaxa\n",
    "import boto3\n",
    "import tempfile\n",
    "import subprocess\n",
    "import os\n",
    "import io\n",
    "import re\n",
    "import time\n",
    "import json\n",
    "ncbi = NCBITaxa()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('../../figures/fig3/all_contigs_df.tsv', sep='\\t', \n",
    "                dtype={''})\n",
    "df = df[df['group'] == 'Metazoa']\n",
    "df = df[['sample', 'sci_name', 'read_prop', 'taxid']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def taxid2name(taxid):\n",
    "    return ncbi.get_taxid_translator([taxid])[taxid]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rows_taxid (df, taxid, taxid_colname=\"taxid\"):\n",
    "    if (not isinstance(df, pd.DataFrame)):\n",
    "        if isinstance(taxid, str):\n",
    "            taxid = ncbi.get_name_translator([taxid])[taxid][0]\n",
    "        return (taxid in ncbi.get_lineage(df))\n",
    "    outdf = df[df[taxid_colname].apply(get_rows_taxid, taxid=taxid)]\n",
    "\n",
    "    return outdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is a partial order on taxa: a < b if a is an ancestor of b. For a sample with one true bloodmeal, we would like to report a taxon t such that it is related to every other taxon, ie, t < b or b < t for all t reported.\n",
    "\n",
    "put another way, a taxon is admissable if t in lineage(b) or b in lineage(t) for all b."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_least_admissable_taxon(taxa,\n",
    "                               exclude = [], # drop these taxa\n",
    "                               parent=None, # only keep children of the parent\n",
    "                               antiparent=None # only keep taxa not in lineage of parent\n",
    "                              ):\n",
    "    if antiparent:\n",
    "        exclude.extend(ncbi.get_lineage(antiparent))\n",
    "    \n",
    "    taxa = [taxid for taxid in taxa if taxid not in exclude]\n",
    "    lineages = [ncbi.get_lineage(taxid) for taxid in taxa]\n",
    "    \n",
    "    if parent:\n",
    "        lineages = [lineage for lineage in lineages if parent in lineage]\n",
    "        if len(lineages) == 0:\n",
    "            return 0\n",
    "        \n",
    "    if antiparent:\n",
    "        lineages = [lineage for lineage in lineages if antiparent not in lineage]\n",
    "        if len(lineages) == 0:\n",
    "            return 0\n",
    "        \n",
    "    all_taxa = np.unique([taxid for lineage in lineages for taxid in lineage])\n",
    "    non_leaf_taxa = np.unique([taxid for lineage in lineages for taxid in lineage[:-1]])\n",
    "    leaf_taxa = [taxid for taxid in all_taxa if taxid not in non_leaf_taxa]\n",
    "    \n",
    "    leaf_lineages = [ncbi.get_lineage(taxid) for taxid in leaf_taxa]\n",
    "    leaf_common_ancestors = set.intersection(*[set(l) for l in leaf_lineages])\n",
    "    lca = [taxid for taxid in leaf_lineages[0] if taxid in leaf_common_ancestors][-1]\n",
    "    \n",
    "    return lca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "vertebrate_taxid = 7742"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "one_off_exclusions = [314146.0, 9544.0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "least_admissable_taxa = []\n",
    "for sample in df['sample'].unique():\n",
    "    taxid = get_least_admissable_taxon(df[df['sample'] == sample]['taxid'],\n",
    "                                      exclude = one_off_exclusions,\n",
    "                                      parent = vertebrate_taxid)\n",
    "    name = taxid2name(taxid) if taxid else \"NA\"\n",
    "    least_admissable_taxa.append({'sample': sample, 'name': name, 'taxid': taxid})\n",
    "least_admissable_taxa = pd.DataFrame(least_admissable_taxa).sort_values('sample')\n",
    "least_admissable_taxa[['sample', 'taxid', 'name']].to_csv(\n",
    "    '../../figures/fig3/vertebrate_lat.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "least_admissable_taxa = []\n",
    "for sample in df['sample'].unique():\n",
    "    taxid = get_least_admissable_taxon(df[df['sample'] == sample]['taxid'],\n",
    "                                      exclude = one_off_exclusions,\n",
    "                                      antiparent = vertebrate_taxid)\n",
    "    name = taxid2name(taxid) if taxid else \"NA\"\n",
    "    least_admissable_taxa.append({'sample': sample, 'name': name, 'taxid': taxid})\n",
    "least_admissable_taxa = pd.DataFrame(least_admissable_taxa).sort_values('sample')\n",
    "least_admissable_taxa[['sample', 'taxid', 'name']].to_csv(\n",
    "    '../../figures/fig3/non_vertebrate_lat.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manual exclusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_032_exclude = [6231.0, 85066.0,7209.0, 1237907.0, 32561.0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2497    1237907.0\n",
       "2505     175121.0\n",
       "Name: taxid, dtype: float64"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['sample'] == 'CMS001_032_Ra_S7']['taxid'][:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[175121.0, 44394.0, 39616.0, 9126.0, 8930.0, 8825.0, 8782.0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Neognathae'"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxid2name(\n",
    "    get_least_admissable_taxon(\n",
    "        df[df['sample'] == 'CMS001_032_Ra_S7']['taxid'], sample_032_exclude\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1437010.0, 314145.0, 72004.0, 35500.0, 9938.0, 9913.0, 9895.0, 9881.0, 9880.0, 9871.0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Pecora'"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_003_exclude = [6913.0, 74971.0, 336591.0]\n",
    "\n",
    "taxid2name(\n",
    "    get_least_admissable_taxon(\n",
    "        df[df['sample'] == 'CMS001_003_Ra_S2']['taxid'], sample_003_exclude))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[6913.0, 74971.0, 336591.0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Larinioides cornutus'"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxid2name(\n",
    "    get_least_admissable_taxon(\n",
    "        sample_003_exclude))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1437010.0, 337687.0, 79684.0, 39087.0]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Microtus ochrogaster'"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_058_exclude = [637355.0]\n",
    "\n",
    "taxid2name(\n",
    "    get_least_admissable_taxon(\n",
    "        df[df['sample'] == 'CMS001_058_Ra_S9']['taxid'], sample_058_exclude))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Pecora'"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxid2name(get_least_admissable_taxon(df[df['sample'] == 'CMS001_009_Ra_S13']['taxid']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.sort_values(by= ['sample','read_prop']).to_csv('../../figures/fig3/metazoa_sorted.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>read_prop</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2454</th>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>Laurasiatheria</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>314145.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2448</th>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>Boreoeutheria</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>1437010.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2471</th>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>Cervidae</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>9850.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2464</th>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>Bovidae</td>\n",
       "      <td>0.001577</td>\n",
       "      <td>9895.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sample        sci_name  read_prop      taxid\n",
       "2454  CMS001_009_Ra_S13  Laurasiatheria   0.000041   314145.0\n",
       "2448  CMS001_009_Ra_S13   Boreoeutheria   0.000172  1437010.0\n",
       "2471  CMS001_009_Ra_S13        Cervidae   0.000252     9850.0\n",
       "2464  CMS001_009_Ra_S13         Bovidae   0.001577     9895.0"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['sample'] == 'CMS001_009_Ra_S13'].sort_values('read_prop')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sum</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sci_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Zonotrichia albicollis</th>\n",
       "      <td>0.014802</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Passeroidea</th>\n",
       "      <td>0.000353</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cetartiodactyla</th>\n",
       "      <td>0.000485</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cervidae</th>\n",
       "      <td>0.000307</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Odocoileus</th>\n",
       "      <td>0.001616</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amniota</th>\n",
       "      <td>0.075880</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Procyon lotor</th>\n",
       "      <td>0.011936</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Carnivora</th>\n",
       "      <td>0.010041</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Caniformia</th>\n",
       "      <td>0.017999</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Odocoileus hemionus</th>\n",
       "      <td>0.000250</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Odocoileus virginianus texanus</th>\n",
       "      <td>0.000317</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Passeriformes</th>\n",
       "      <td>0.048110</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aves</th>\n",
       "      <td>0.025958</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eutheria</th>\n",
       "      <td>0.004403</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neognathae</th>\n",
       "      <td>0.346778</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bos taurus</th>\n",
       "      <td>0.000833</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Odocoileinae</th>\n",
       "      <td>0.049724</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bovidae</th>\n",
       "      <td>0.051033</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Laurasiatheria</th>\n",
       "      <td>0.043796</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Boreoeutheria</th>\n",
       "      <td>0.081648</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     sum  count\n",
       "sci_name                                       \n",
       "Zonotrichia albicollis          0.014802      3\n",
       "Passeroidea                     0.000353      3\n",
       "Cetartiodactyla                 0.000485      3\n",
       "Cervidae                        0.000307      4\n",
       "Odocoileus                      0.001616      4\n",
       "Amniota                         0.075880      4\n",
       "Procyon lotor                   0.011936      4\n",
       "Carnivora                       0.010041      4\n",
       "Caniformia                      0.017999      5\n",
       "Odocoileus hemionus             0.000250      5\n",
       "Odocoileus virginianus texanus  0.000317      5\n",
       "Passeriformes                   0.048110      5\n",
       "Aves                            0.025958      6\n",
       "Eutheria                        0.004403      6\n",
       "Neognathae                      0.346778      8\n",
       "Bos taurus                      0.000833      8\n",
       "Odocoileinae                    0.049724     16\n",
       "Bovidae                         0.051033     18\n",
       "Laurasiatheria                  0.043796     20\n",
       "Boreoeutheria                   0.081648     27"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = df[['sci_name', 'read_prop']].groupby('sci_name').agg({'read_prop': ['sum','count']})\n",
    "x.columns = x.columns.droplevel()\n",
    "x.sort_values('count').tail(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sci_name</th>\n",
       "      <th>Acarus</th>\n",
       "      <th>Amniota</th>\n",
       "      <th>Anas platyrhynchos</th>\n",
       "      <th>Anatidae</th>\n",
       "      <th>Aphelocoma</th>\n",
       "      <th>Araneidae</th>\n",
       "      <th>Araneomorphae</th>\n",
       "      <th>Archosauria</th>\n",
       "      <th>Arvicolinae</th>\n",
       "      <th>Aves</th>\n",
       "      <th>...</th>\n",
       "      <th>Sturnus vulgaris</th>\n",
       "      <th>Sus scrofa</th>\n",
       "      <th>Taeniopygia guttata</th>\n",
       "      <th>Trachelas tranquillus</th>\n",
       "      <th>Troglodytinae</th>\n",
       "      <th>Ursus arctos horribilis</th>\n",
       "      <th>Zonotrichia</th>\n",
       "      <th>Zonotrichia albicollis</th>\n",
       "      <th>Zonotrichia leucophrys</th>\n",
       "      <th>unclassified Mimetus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CMS001_001_Ra_S1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_003_Ra_S2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_004_Ra_S2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007357</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_005_Ra_S3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_008_Ra_S3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_009_Ra_S13</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_010_Ra_S1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_011_Ra_S4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.023222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_012_Ra_S4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_013_Ra_S5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.002039</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_015_Ra_S13</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_016_Ra_S6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_017_Ra_S6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_018_Ra_S14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_019_Ra_S14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_020_Ra_S15</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_022_Ra_S6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_023_Ra_S17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002649</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_025_Ra_S7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_027_Ra_S16</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000313</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_028_Ra_S17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006363</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_029_Ra_S18</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.019852</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012585</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_030_Ra_S7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_031_Ra_S19</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_032_Ra_S7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000167</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_033_Ra_S8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_035_Ra_S20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_036_Ra_S20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_037_Ra_S21</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.050592</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_038_Ra_S22</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_040_Ra_S21</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_041_Ra_S10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_042_Ra_S23</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_043_Ra_S24</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000692</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011592</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_045_Ra_S2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_046_Ra_S3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_047_Ra_S4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_048_Ra_S5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_049_Ra_S6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_050_Ra_S23</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005586</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_052_Ra_S7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_053_Ra_S8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_054_Ra_S11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_056_Ra_S10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_057_Ra_S11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_058_Ra_S9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS001_059_Ra_S10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001544</td>\n",
       "      <td>0.018930</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS002_025c_Rb_S142_L004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS002_026d_Rb_S149_L004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS002_031a_Rb_S165_L004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS002_042a_Rb_S177_L004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS002_045c_Rb_S185_L004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMS002_056a_Rb_S9_L004</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sci_name                  Acarus   Amniota  Anas platyrhynchos  Anatidae  \\\n",
       "sample                                                                     \n",
       "CMS001_001_Ra_S1             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_003_Ra_S2             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_004_Ra_S2             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_005_Ra_S3             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_008_Ra_S3             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_009_Ra_S13            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_010_Ra_S1             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_011_Ra_S4             0.0  0.023222            0.000000  0.000000   \n",
       "CMS001_012_Ra_S4             0.0  0.000027            0.000000  0.000000   \n",
       "CMS001_013_Ra_S5             0.0  0.002039            0.000000  0.000000   \n",
       "CMS001_015_Ra_S13            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_016_Ra_S6             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_017_Ra_S6             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_018_Ra_S14            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_019_Ra_S14            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_020_Ra_S15            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_022_Ra_S6             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_023_Ra_S17            0.0  0.000000            0.002649  0.000202   \n",
       "CMS001_025_Ra_S7             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_027_Ra_S16            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_028_Ra_S17            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_029_Ra_S18            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_030_Ra_S7             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_031_Ra_S19            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_032_Ra_S7             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_033_Ra_S8             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_035_Ra_S20            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_036_Ra_S20            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_037_Ra_S21            0.0  0.050592            0.000000  0.000000   \n",
       "CMS001_038_Ra_S22            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_040_Ra_S21            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_041_Ra_S10            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_042_Ra_S23            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_043_Ra_S24            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_045_Ra_S2             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_046_Ra_S3             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_047_Ra_S4             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_048_Ra_S5             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_049_Ra_S6             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_050_Ra_S23            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_052_Ra_S7             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_053_Ra_S8             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_054_Ra_S11            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_056_Ra_S10            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_057_Ra_S11            0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_058_Ra_S9             0.0  0.000000            0.000000  0.000000   \n",
       "CMS001_059_Ra_S10            0.0  0.000000            0.000000  0.000000   \n",
       "CMS002_025c_Rb_S142_L004     0.0  0.000000            0.000000  0.000000   \n",
       "CMS002_026d_Rb_S149_L004     0.0  0.000000            0.000000  0.000000   \n",
       "CMS002_031a_Rb_S165_L004     0.0  0.000000            0.000000  0.000000   \n",
       "CMS002_042a_Rb_S177_L004     0.0  0.000000            0.000000  0.000000   \n",
       "CMS002_045c_Rb_S185_L004     0.0  0.000000            0.000000  0.000000   \n",
       "CMS002_056a_Rb_S9_L004       0.0  0.000000            0.000000  0.000000   \n",
       "\n",
       "sci_name                  Aphelocoma  Araneidae  Araneomorphae  Archosauria  \\\n",
       "sample                                                                        \n",
       "CMS001_001_Ra_S1            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_003_Ra_S2            0.000000   0.000143       0.000000     0.000000   \n",
       "CMS001_004_Ra_S2            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_005_Ra_S3            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_008_Ra_S3            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_009_Ra_S13           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_010_Ra_S1            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_011_Ra_S4            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_012_Ra_S4            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_013_Ra_S5            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_015_Ra_S13           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_016_Ra_S6            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_017_Ra_S6            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_018_Ra_S14           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_019_Ra_S14           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_020_Ra_S15           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_022_Ra_S6            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_023_Ra_S17           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_025_Ra_S7            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_027_Ra_S16           0.000313   0.000000       0.000000     0.000000   \n",
       "CMS001_028_Ra_S17           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_029_Ra_S18           0.000000   0.000000       0.000000     0.000064   \n",
       "CMS001_030_Ra_S7            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_031_Ra_S19           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_032_Ra_S7            0.000167   0.000000       0.000000     0.000000   \n",
       "CMS001_033_Ra_S8            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_035_Ra_S20           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_036_Ra_S20           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_037_Ra_S21           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_038_Ra_S22           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_040_Ra_S21           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_041_Ra_S10           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_042_Ra_S23           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_043_Ra_S24           0.000692   0.000000       0.000000     0.000000   \n",
       "CMS001_045_Ra_S2            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_046_Ra_S3            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_047_Ra_S4            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_048_Ra_S5            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_049_Ra_S6            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_050_Ra_S23           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_052_Ra_S7            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_053_Ra_S8            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_054_Ra_S11           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_056_Ra_S10           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_057_Ra_S11           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_058_Ra_S9            0.000000   0.000000       0.000000     0.000000   \n",
       "CMS001_059_Ra_S10           0.000000   0.000000       0.000000     0.000000   \n",
       "CMS002_025c_Rb_S142_L004    0.000000   0.000000       0.000000     0.000000   \n",
       "CMS002_026d_Rb_S149_L004    0.000000   0.000000       0.000025     0.000000   \n",
       "CMS002_031a_Rb_S165_L004    0.000000   0.000000       0.000000     0.000000   \n",
       "CMS002_042a_Rb_S177_L004    0.000000   0.000000       0.000000     0.000000   \n",
       "CMS002_045c_Rb_S185_L004    0.000000   0.000000       0.000000     0.000000   \n",
       "CMS002_056a_Rb_S9_L004      0.000000   0.000000       0.000000     0.000000   \n",
       "\n",
       "sci_name                  Arvicolinae      Aves  ...  Sturnus vulgaris  \\\n",
       "sample                                           ...                     \n",
       "CMS001_001_Ra_S1             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_003_Ra_S2             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_004_Ra_S2             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_005_Ra_S3             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_008_Ra_S3             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_009_Ra_S13            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_010_Ra_S1             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_011_Ra_S4             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_012_Ra_S4             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_013_Ra_S5             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_015_Ra_S13            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_016_Ra_S6             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_017_Ra_S6             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_018_Ra_S14            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_019_Ra_S14            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_020_Ra_S15            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_022_Ra_S6             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_023_Ra_S17            0.000000  0.000015  ...          0.000000   \n",
       "CMS001_025_Ra_S7             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_027_Ra_S16            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_028_Ra_S17            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_029_Ra_S18            0.000000  0.019852  ...          0.000000   \n",
       "CMS001_030_Ra_S7             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_031_Ra_S19            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_032_Ra_S7             0.000000  0.000062  ...          0.000000   \n",
       "CMS001_033_Ra_S8             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_035_Ra_S20            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_036_Ra_S20            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_037_Ra_S21            0.000000  0.000271  ...          0.000000   \n",
       "CMS001_038_Ra_S22            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_040_Ra_S21            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_041_Ra_S10            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_042_Ra_S23            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_043_Ra_S24            0.000000  0.000173  ...          0.011592   \n",
       "CMS001_045_Ra_S2             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_046_Ra_S3             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_047_Ra_S4             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_048_Ra_S5             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_049_Ra_S6             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_050_Ra_S23            0.000000  0.005586  ...          0.000000   \n",
       "CMS001_052_Ra_S7             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_053_Ra_S8             0.000000  0.000000  ...          0.000000   \n",
       "CMS001_054_Ra_S11            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_056_Ra_S10            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_057_Ra_S11            0.000000  0.000000  ...          0.000000   \n",
       "CMS001_058_Ra_S9             0.000229  0.000000  ...          0.000000   \n",
       "CMS001_059_Ra_S10            0.000000  0.000000  ...          0.000000   \n",
       "CMS002_025c_Rb_S142_L004     0.000000  0.000000  ...          0.000000   \n",
       "CMS002_026d_Rb_S149_L004     0.000000  0.000000  ...          0.000000   \n",
       "CMS002_031a_Rb_S165_L004     0.000000  0.000000  ...          0.000000   \n",
       "CMS002_042a_Rb_S177_L004     0.000000  0.000000  ...          0.000000   \n",
       "CMS002_045c_Rb_S185_L004     0.000000  0.000000  ...          0.000000   \n",
       "CMS002_056a_Rb_S9_L004       0.000000  0.000000  ...          0.000000   \n",
       "\n",
       "sci_name                  Sus scrofa  Taeniopygia guttata  \\\n",
       "sample                                                      \n",
       "CMS001_001_Ra_S1             0.00000             0.000000   \n",
       "CMS001_003_Ra_S2             0.00000             0.000000   \n",
       "CMS001_004_Ra_S2             0.00000             0.000000   \n",
       "CMS001_005_Ra_S3             0.00000             0.000000   \n",
       "CMS001_008_Ra_S3             0.00000             0.000000   \n",
       "CMS001_009_Ra_S13            0.00000             0.000000   \n",
       "CMS001_010_Ra_S1             0.00000             0.000000   \n",
       "CMS001_011_Ra_S4             0.00000             0.000000   \n",
       "CMS001_012_Ra_S4             0.00000             0.000000   \n",
       "CMS001_013_Ra_S5             0.00000             0.000000   \n",
       "CMS001_015_Ra_S13            0.00000             0.000000   \n",
       "CMS001_016_Ra_S6             0.00000             0.000000   \n",
       "CMS001_017_Ra_S6             0.00000             0.000000   \n",
       "CMS001_018_Ra_S14            0.00000             0.000000   \n",
       "CMS001_019_Ra_S14            0.00000             0.000000   \n",
       "CMS001_020_Ra_S15            0.00000             0.000000   \n",
       "CMS001_022_Ra_S6             0.00000             0.000000   \n",
       "CMS001_023_Ra_S17            0.00000             0.000000   \n",
       "CMS001_025_Ra_S7             0.00000             0.000000   \n",
       "CMS001_027_Ra_S16            0.00000             0.000000   \n",
       "CMS001_028_Ra_S17            0.00000             0.000000   \n",
       "CMS001_029_Ra_S18            0.00000             0.000086   \n",
       "CMS001_030_Ra_S7             0.00000             0.000000   \n",
       "CMS001_031_Ra_S19            0.00000             0.000000   \n",
       "CMS001_032_Ra_S7             0.00000             0.000000   \n",
       "CMS001_033_Ra_S8             0.00000             0.000000   \n",
       "CMS001_035_Ra_S20            0.00000             0.000000   \n",
       "CMS001_036_Ra_S20            0.00000             0.000000   \n",
       "CMS001_037_Ra_S21            0.00000             0.000000   \n",
       "CMS001_038_Ra_S22            0.00000             0.000000   \n",
       "CMS001_040_Ra_S21            0.00000             0.000000   \n",
       "CMS001_041_Ra_S10            0.00000             0.000000   \n",
       "CMS001_042_Ra_S23            0.00000             0.000000   \n",
       "CMS001_043_Ra_S24            0.00000             0.000000   \n",
       "CMS001_045_Ra_S2             0.00000             0.000000   \n",
       "CMS001_046_Ra_S3             0.00000             0.000000   \n",
       "CMS001_047_Ra_S4             0.00000             0.000000   \n",
       "CMS001_048_Ra_S5             0.00000             0.000000   \n",
       "CMS001_049_Ra_S6             0.00000             0.000000   \n",
       "CMS001_050_Ra_S23            0.00000             0.000000   \n",
       "CMS001_052_Ra_S7             0.00000             0.000000   \n",
       "CMS001_053_Ra_S8             0.00000             0.000000   \n",
       "CMS001_054_Ra_S11            0.00000             0.000000   \n",
       "CMS001_056_Ra_S10            0.00000             0.000000   \n",
       "CMS001_057_Ra_S11            0.00000             0.000000   \n",
       "CMS001_058_Ra_S9             0.00000             0.000000   \n",
       "CMS001_059_Ra_S10            0.00000             0.000000   \n",
       "CMS002_025c_Rb_S142_L004     0.00000             0.000000   \n",
       "CMS002_026d_Rb_S149_L004     0.00000             0.000000   \n",
       "CMS002_031a_Rb_S165_L004     0.00002             0.000000   \n",
       "CMS002_042a_Rb_S177_L004     0.00000             0.000000   \n",
       "CMS002_045c_Rb_S185_L004     0.00000             0.000000   \n",
       "CMS002_056a_Rb_S9_L004       0.00000             0.000000   \n",
       "\n",
       "sci_name                  Trachelas tranquillus  Troglodytinae  \\\n",
       "sample                                                           \n",
       "CMS001_001_Ra_S1                       0.000000       0.000000   \n",
       "CMS001_003_Ra_S2                       0.000000       0.000000   \n",
       "CMS001_004_Ra_S2                       0.000000       0.000000   \n",
       "CMS001_005_Ra_S3                       0.000000       0.000000   \n",
       "CMS001_008_Ra_S3                       0.000000       0.000000   \n",
       "CMS001_009_Ra_S13                      0.000000       0.000000   \n",
       "CMS001_010_Ra_S1                       0.000000       0.000000   \n",
       "CMS001_011_Ra_S4                       0.000000       0.000000   \n",
       "CMS001_012_Ra_S4                       0.000000       0.000000   \n",
       "CMS001_013_Ra_S5                       0.000000       0.000000   \n",
       "CMS001_015_Ra_S13                      0.000000       0.000000   \n",
       "CMS001_016_Ra_S6                       0.000000       0.000000   \n",
       "CMS001_017_Ra_S6                       0.000000       0.000000   \n",
       "CMS001_018_Ra_S14                      0.000000       0.000000   \n",
       "CMS001_019_Ra_S14                      0.000000       0.000000   \n",
       "CMS001_020_Ra_S15                      0.000000       0.000000   \n",
       "CMS001_022_Ra_S6                       0.000000       0.000000   \n",
       "CMS001_023_Ra_S17                      0.000000       0.000000   \n",
       "CMS001_025_Ra_S7                       0.000000       0.000000   \n",
       "CMS001_027_Ra_S16                      0.000000       0.000000   \n",
       "CMS001_028_Ra_S17                      0.000000       0.000000   \n",
       "CMS001_029_Ra_S18                      0.000000       0.000000   \n",
       "CMS001_030_Ra_S7                       0.000000       0.000000   \n",
       "CMS001_031_Ra_S19                      0.000000       0.000000   \n",
       "CMS001_032_Ra_S7                       0.000000       0.000000   \n",
       "CMS001_033_Ra_S8                       0.000000       0.000000   \n",
       "CMS001_035_Ra_S20                      0.000000       0.000000   \n",
       "CMS001_036_Ra_S20                      0.000000       0.000000   \n",
       "CMS001_037_Ra_S21                      0.000000       0.000000   \n",
       "CMS001_038_Ra_S22                      0.000000       0.000000   \n",
       "CMS001_040_Ra_S21                      0.000000       0.000000   \n",
       "CMS001_041_Ra_S10                      0.000000       0.000000   \n",
       "CMS001_042_Ra_S23                      0.000000       0.000000   \n",
       "CMS001_043_Ra_S24                      0.000000       0.000000   \n",
       "CMS001_045_Ra_S2                       0.000000       0.000000   \n",
       "CMS001_046_Ra_S3                       0.000000       0.000000   \n",
       "CMS001_047_Ra_S4                       0.000000       0.000099   \n",
       "CMS001_048_Ra_S5                       0.000000       0.000000   \n",
       "CMS001_049_Ra_S6                       0.000000       0.000000   \n",
       "CMS001_050_Ra_S23                      0.000000       0.000000   \n",
       "CMS001_052_Ra_S7                       0.000000       0.000000   \n",
       "CMS001_053_Ra_S8                       0.000000       0.000000   \n",
       "CMS001_054_Ra_S11                      0.000000       0.000000   \n",
       "CMS001_056_Ra_S10                      0.000000       0.000000   \n",
       "CMS001_057_Ra_S11                      0.000000       0.000000   \n",
       "CMS001_058_Ra_S9                       0.000000       0.000000   \n",
       "CMS001_059_Ra_S10                      0.000000       0.000000   \n",
       "CMS002_025c_Rb_S142_L004               0.000000       0.000000   \n",
       "CMS002_026d_Rb_S149_L004               0.000049       0.000000   \n",
       "CMS002_031a_Rb_S165_L004               0.000000       0.000000   \n",
       "CMS002_042a_Rb_S177_L004               0.000000       0.000000   \n",
       "CMS002_045c_Rb_S185_L004               0.000000       0.000000   \n",
       "CMS002_056a_Rb_S9_L004                 0.000000       0.000000   \n",
       "\n",
       "sci_name                  Ursus arctos horribilis  Zonotrichia  \\\n",
       "sample                                                           \n",
       "CMS001_001_Ra_S1                         0.000000     0.000000   \n",
       "CMS001_003_Ra_S2                         0.000000     0.000000   \n",
       "CMS001_004_Ra_S2                         0.007357     0.000000   \n",
       "CMS001_005_Ra_S3                         0.000000     0.000000   \n",
       "CMS001_008_Ra_S3                         0.000000     0.000000   \n",
       "CMS001_009_Ra_S13                        0.000000     0.000000   \n",
       "CMS001_010_Ra_S1                         0.000000     0.000000   \n",
       "CMS001_011_Ra_S4                         0.000000     0.000000   \n",
       "CMS001_012_Ra_S4                         0.000000     0.000000   \n",
       "CMS001_013_Ra_S5                         0.000000     0.000000   \n",
       "CMS001_015_Ra_S13                        0.000000     0.000000   \n",
       "CMS001_016_Ra_S6                         0.000000     0.000000   \n",
       "CMS001_017_Ra_S6                         0.000000     0.000000   \n",
       "CMS001_018_Ra_S14                        0.000000     0.000000   \n",
       "CMS001_019_Ra_S14                        0.000000     0.000000   \n",
       "CMS001_020_Ra_S15                        0.000000     0.000000   \n",
       "CMS001_022_Ra_S6                         0.000000     0.000000   \n",
       "CMS001_023_Ra_S17                        0.000000     0.000000   \n",
       "CMS001_025_Ra_S7                         0.000000     0.000000   \n",
       "CMS001_027_Ra_S16                        0.000000     0.000000   \n",
       "CMS001_028_Ra_S17                        0.006363     0.000000   \n",
       "CMS001_029_Ra_S18                        0.000000     0.000000   \n",
       "CMS001_030_Ra_S7                         0.000000     0.000000   \n",
       "CMS001_031_Ra_S19                        0.000000     0.000000   \n",
       "CMS001_032_Ra_S7                         0.000000     0.000000   \n",
       "CMS001_033_Ra_S8                         0.000000     0.000000   \n",
       "CMS001_035_Ra_S20                        0.000000     0.000000   \n",
       "CMS001_036_Ra_S20                        0.000000     0.000000   \n",
       "CMS001_037_Ra_S21                        0.000000     0.000000   \n",
       "CMS001_038_Ra_S22                        0.000000     0.000000   \n",
       "CMS001_040_Ra_S21                        0.000000     0.000000   \n",
       "CMS001_041_Ra_S10                        0.000000     0.000000   \n",
       "CMS001_042_Ra_S23                        0.000000     0.000000   \n",
       "CMS001_043_Ra_S24                        0.000000     0.000000   \n",
       "CMS001_045_Ra_S2                         0.000000     0.000000   \n",
       "CMS001_046_Ra_S3                         0.000000     0.000000   \n",
       "CMS001_047_Ra_S4                         0.000000     0.000000   \n",
       "CMS001_048_Ra_S5                         0.000000     0.000000   \n",
       "CMS001_049_Ra_S6                         0.000000     0.000000   \n",
       "CMS001_050_Ra_S23                        0.000000     0.000000   \n",
       "CMS001_052_Ra_S7                         0.000000     0.000000   \n",
       "CMS001_053_Ra_S8                         0.000000     0.000000   \n",
       "CMS001_054_Ra_S11                        0.000000     0.000024   \n",
       "CMS001_056_Ra_S10                        0.000000     0.000000   \n",
       "CMS001_057_Ra_S11                        0.000000     0.000000   \n",
       "CMS001_058_Ra_S9                         0.000000     0.000000   \n",
       "CMS001_059_Ra_S10                        0.000000     0.000000   \n",
       "CMS002_025c_Rb_S142_L004                 0.000000     0.000000   \n",
       "CMS002_026d_Rb_S149_L004                 0.000000     0.000000   \n",
       "CMS002_031a_Rb_S165_L004                 0.000000     0.000000   \n",
       "CMS002_042a_Rb_S177_L004                 0.000000     0.000000   \n",
       "CMS002_045c_Rb_S185_L004                 0.000000     0.000000   \n",
       "CMS002_056a_Rb_S9_L004                   0.000000     0.000000   \n",
       "\n",
       "sci_name                  Zonotrichia albicollis  Zonotrichia leucophrys  \\\n",
       "sample                                                                     \n",
       "CMS001_001_Ra_S1                        0.000000                0.000000   \n",
       "CMS001_003_Ra_S2                        0.000000                0.000000   \n",
       "CMS001_004_Ra_S2                        0.000000                0.000000   \n",
       "CMS001_005_Ra_S3                        0.000000                0.000000   \n",
       "CMS001_008_Ra_S3                        0.000000                0.000000   \n",
       "CMS001_009_Ra_S13                       0.000000                0.000000   \n",
       "CMS001_010_Ra_S1                        0.000000                0.000000   \n",
       "CMS001_011_Ra_S4                        0.000000                0.000000   \n",
       "CMS001_012_Ra_S4                        0.000000                0.000000   \n",
       "CMS001_013_Ra_S5                        0.000000                0.000000   \n",
       "CMS001_015_Ra_S13                       0.000000                0.000000   \n",
       "CMS001_016_Ra_S6                        0.000000                0.000000   \n",
       "CMS001_017_Ra_S6                        0.000000                0.000000   \n",
       "CMS001_018_Ra_S14                       0.000000                0.000000   \n",
       "CMS001_019_Ra_S14                       0.000000                0.000000   \n",
       "CMS001_020_Ra_S15                       0.000000                0.000000   \n",
       "CMS001_022_Ra_S6                        0.000000                0.000000   \n",
       "CMS001_023_Ra_S17                       0.000000                0.000000   \n",
       "CMS001_025_Ra_S7                        0.000000                0.000000   \n",
       "CMS001_027_Ra_S16                       0.000000                0.000000   \n",
       "CMS001_028_Ra_S17                       0.000000                0.000000   \n",
       "CMS001_029_Ra_S18                       0.012585                0.000000   \n",
       "CMS001_030_Ra_S7                        0.000000                0.000000   \n",
       "CMS001_031_Ra_S19                       0.000000                0.000000   \n",
       "CMS001_032_Ra_S7                        0.000673                0.000000   \n",
       "CMS001_033_Ra_S8                        0.000000                0.000000   \n",
       "CMS001_035_Ra_S20                       0.000000                0.000000   \n",
       "CMS001_036_Ra_S20                       0.000000                0.000000   \n",
       "CMS001_037_Ra_S21                       0.000000                0.000000   \n",
       "CMS001_038_Ra_S22                       0.000000                0.000000   \n",
       "CMS001_040_Ra_S21                       0.000000                0.000000   \n",
       "CMS001_041_Ra_S10                       0.000000                0.000000   \n",
       "CMS001_042_Ra_S23                       0.000000                0.000000   \n",
       "CMS001_043_Ra_S24                       0.000000                0.000000   \n",
       "CMS001_045_Ra_S2                        0.000000                0.000000   \n",
       "CMS001_046_Ra_S3                        0.000000                0.000000   \n",
       "CMS001_047_Ra_S4                        0.000000                0.000000   \n",
       "CMS001_048_Ra_S5                        0.000000                0.000000   \n",
       "CMS001_049_Ra_S6                        0.000000                0.000000   \n",
       "CMS001_050_Ra_S23                       0.000000                0.000000   \n",
       "CMS001_052_Ra_S7                        0.000000                0.000000   \n",
       "CMS001_053_Ra_S8                        0.000000                0.000000   \n",
       "CMS001_054_Ra_S11                       0.000000                0.000083   \n",
       "CMS001_056_Ra_S10                       0.000000                0.000000   \n",
       "CMS001_057_Ra_S11                       0.000000                0.000000   \n",
       "CMS001_058_Ra_S9                        0.000000                0.000000   \n",
       "CMS001_059_Ra_S10                       0.001544                0.018930   \n",
       "CMS002_025c_Rb_S142_L004                0.000000                0.000000   \n",
       "CMS002_026d_Rb_S149_L004                0.000000                0.000000   \n",
       "CMS002_031a_Rb_S165_L004                0.000000                0.000000   \n",
       "CMS002_042a_Rb_S177_L004                0.000000                0.000000   \n",
       "CMS002_045c_Rb_S185_L004                0.000000                0.000000   \n",
       "CMS002_056a_Rb_S9_L004                  0.000000                0.000000   \n",
       "\n",
       "sci_name                  unclassified Mimetus  \n",
       "sample                                          \n",
       "CMS001_001_Ra_S1                      0.000000  \n",
       "CMS001_003_Ra_S2                      0.000000  \n",
       "CMS001_004_Ra_S2                      0.000000  \n",
       "CMS001_005_Ra_S3                      0.000000  \n",
       "CMS001_008_Ra_S3                      0.000000  \n",
       "CMS001_009_Ra_S13                     0.000000  \n",
       "CMS001_010_Ra_S1                      0.000000  \n",
       "CMS001_011_Ra_S4                      0.000000  \n",
       "CMS001_012_Ra_S4                      0.000000  \n",
       "CMS001_013_Ra_S5                      0.000000  \n",
       "CMS001_015_Ra_S13                     0.000000  \n",
       "CMS001_016_Ra_S6                      0.000000  \n",
       "CMS001_017_Ra_S6                      0.000000  \n",
       "CMS001_018_Ra_S14                     0.000000  \n",
       "CMS001_019_Ra_S14                     0.000000  \n",
       "CMS001_020_Ra_S15                     0.000000  \n",
       "CMS001_022_Ra_S6                      0.000000  \n",
       "CMS001_023_Ra_S17                     0.000000  \n",
       "CMS001_025_Ra_S7                      0.000000  \n",
       "CMS001_027_Ra_S16                     0.000000  \n",
       "CMS001_028_Ra_S17                     0.000000  \n",
       "CMS001_029_Ra_S18                     0.000000  \n",
       "CMS001_030_Ra_S7                      0.000000  \n",
       "CMS001_031_Ra_S19                     0.000000  \n",
       "CMS001_032_Ra_S7                      0.000000  \n",
       "CMS001_033_Ra_S8                      0.000000  \n",
       "CMS001_035_Ra_S20                     0.000000  \n",
       "CMS001_036_Ra_S20                     0.000000  \n",
       "CMS001_037_Ra_S21                     0.000000  \n",
       "CMS001_038_Ra_S22                     0.000000  \n",
       "CMS001_040_Ra_S21                     0.000000  \n",
       "CMS001_041_Ra_S10                     0.000000  \n",
       "CMS001_042_Ra_S23                     0.000000  \n",
       "CMS001_043_Ra_S24                     0.000000  \n",
       "CMS001_045_Ra_S2                      0.000000  \n",
       "CMS001_046_Ra_S3                      0.000000  \n",
       "CMS001_047_Ra_S4                      0.000000  \n",
       "CMS001_048_Ra_S5                      0.000000  \n",
       "CMS001_049_Ra_S6                      0.000000  \n",
       "CMS001_050_Ra_S23                     0.000000  \n",
       "CMS001_052_Ra_S7                      0.000000  \n",
       "CMS001_053_Ra_S8                      0.000343  \n",
       "CMS001_054_Ra_S11                     0.000000  \n",
       "CMS001_056_Ra_S10                     0.000000  \n",
       "CMS001_057_Ra_S11                     0.000000  \n",
       "CMS001_058_Ra_S9                      0.000000  \n",
       "CMS001_059_Ra_S10                     0.000000  \n",
       "CMS002_025c_Rb_S142_L004              0.000000  \n",
       "CMS002_026d_Rb_S149_L004              0.000000  \n",
       "CMS002_031a_Rb_S165_L004              0.000000  \n",
       "CMS002_042a_Rb_S177_L004              0.000000  \n",
       "CMS002_045c_Rb_S185_L004              0.000000  \n",
       "CMS002_056a_Rb_S9_L004                0.000000  \n",
       "\n",
       "[53 rows x 96 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = df.pivot(index='sample', columns='sci_name', values='read_prop').fillna(0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>read_prop</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2452</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Laurasiatheria</td>\n",
       "      <td>0.003818</td>\n",
       "      <td>314145.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2453</th>\n",
       "      <td>CMS001_008_Ra_S3</td>\n",
       "      <td>Laurasiatheria</td>\n",
       "      <td>0.000337</td>\n",
       "      <td>314145.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2454</th>\n",
       "      <td>CMS001_009_Ra_S13</td>\n",
       "      <td>Laurasiatheria</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>314145.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2455</th>\n",
       "      <td>CMS001_008_Ra_S3</td>\n",
       "      <td>Bubalus bubalis</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>89462.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2457</th>\n",
       "      <td>CMS001_003_Ra_S2</td>\n",
       "      <td>Bos mutus</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>72004.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2701</th>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>Odocoileinae</td>\n",
       "      <td>0.009814</td>\n",
       "      <td>9881.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>Odocoileus virginianus texanus</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>9880.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>Odocoileus hemionus</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>9872.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>CMS001_012_Ra_S4</td>\n",
       "      <td>Odocoileus</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>9871.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2706</th>\n",
       "      <td>CMS002_045c_Rb_S185_L004</td>\n",
       "      <td>Laurasiatheria</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>314145.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        sample                        sci_name  read_prop  \\\n",
       "2452          CMS001_003_Ra_S2                  Laurasiatheria   0.003818   \n",
       "2453          CMS001_008_Ra_S3                  Laurasiatheria   0.000337   \n",
       "2454         CMS001_009_Ra_S13                  Laurasiatheria   0.000041   \n",
       "2455          CMS001_008_Ra_S3                 Bubalus bubalis   0.000271   \n",
       "2457          CMS001_003_Ra_S2                       Bos mutus   0.000079   \n",
       "...                        ...                             ...        ...   \n",
       "2701          CMS001_012_Ra_S4                    Odocoileinae   0.009814   \n",
       "2702          CMS001_012_Ra_S4  Odocoileus virginianus texanus   0.000129   \n",
       "2703          CMS001_012_Ra_S4             Odocoileus hemionus   0.000089   \n",
       "2704          CMS001_012_Ra_S4                      Odocoileus   0.000016   \n",
       "2706  CMS002_045c_Rb_S185_L004                  Laurasiatheria   0.000003   \n",
       "\n",
       "         taxid  \n",
       "2452  314145.0  \n",
       "2453  314145.0  \n",
       "2454  314145.0  \n",
       "2455   89462.0  \n",
       "2457   72004.0  \n",
       "...        ...  \n",
       "2701    9881.0  \n",
       "2702    9880.0  \n",
       "2703    9872.0  \n",
       "2704    9871.0  \n",
       "2706  314145.0  \n",
       "\n",
       "[119 rows x 4 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_rows_taxid(df, 314145.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'gather'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-5d2a96698ecd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_rows_taxid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m314145.0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgather\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/flash/lib/python3.5/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5177\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5178\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5179\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5180\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5181\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'gather'"
     ]
    }
   ],
   "source": [
    "get_rows_taxid(df, 314145.0).gather()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['sample', 'sci_name', 'read_prop', 'taxid']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_viral_family_df (row_x):\n",
    "    segments = row_x[\"segments\"]\n",
    "    df_by_sample = pd.DataFrame([x.split(\"|\") for x in segments[list(segments.keys())[0]][\"contigs\"]])\n",
    "    df_by_sample = df_by_sample.assign(sci_name=row_x[\"family\"], taxid=row_x[\"taxid\"])\n",
    "    df_by_sample = df_by_sample.rename(columns={0:\"sample\", 1:\"contig_name\"})\n",
    "    return (df_by_sample)\n",
    "\n",
    "def get_rows_taxid (df, taxid, taxid_colname=\"taxid\", identity_qcov_cutoff=None):\n",
    "    if (not isinstance(df, pd.DataFrame)):\n",
    "        if isinstance(taxid, str):\n",
    "            taxid = ncbi.get_name_translator([taxid])[taxid][0]\n",
    "        return (taxid in ncbi.get_lineage(df))\n",
    "    outdf = df[df[taxid_colname].apply(get_rows_taxid, taxid=taxid)]\n",
    "    if identity_qcov_cutoff is not None:\n",
    "        outdf = outdf[outdf[\"identity_qcov\"]>=identity_qcov_cutoff]\n",
    "    return (outdf)\n",
    "\n",
    "\n",
    "# def filter_by_criterion (df, colname, minthreshold, bysample=True):\n",
    "#     if bysample:\n",
    "#         sums = df.groupby([\"sample\", \"taxid\"])[colname].sum().reset_index()\n",
    "#         sums[\"tokeep\"] = sums[colname] >= minthreshold\n",
    "#         df = pd.merge(df, sums.drop(columns=colname), how=\"left\")\n",
    "#         df = df[df[\"tokeep\"]!=False].drop(columns=\"tokeep\")\n",
    "#     else:\n",
    "#         df = df[df[colname] >= minthreshold]\n",
    "#     return (df)\n",
    "\n",
    "\n",
    "def check_if_in_any_taxid(taxid, taxid_list):\n",
    "    if taxid in taxid_list:\n",
    "        return (taxid)\n",
    "    taxids = ncbi.get_lineage(taxid)\n",
    "    check_in = [i for i, x in enumerate(taxids) if x in taxid_list]\n",
    "    if (len(check_in)==0):\n",
    "        return (np.nan)\n",
    "    return (taxids[check_in[0]])\n",
    "\n",
    "\n",
    "\n",
    "def clean_taxids(df, taxids, root_taxid, taxid_colname=\"taxid\"):\n",
    "    if isinstance(taxids[0], str):\n",
    "        taxids = dict(zip([ncbi.get_name_translator([x])[x][0] for x in taxids], taxids))\n",
    "    else:\n",
    "        taxids = ncbi.get_taxid_translator(taxids)\n",
    "    if isinstance(root_taxid, str):\n",
    "        root_taxid_number = ncbi.get_name_translator([root_taxid])[root_taxid][0]\n",
    "        root_taxid = {root_taxid_number:root_taxid}\n",
    "    else:\n",
    "        root_taxid = ncbi.get_taxid_translator([root_taxid])\n",
    "    df[taxid_colname] = df[taxid_colname].apply(check_if_in_any_taxid, taxid_list=taxids)\n",
    "    df[taxid_colname][df[taxid_colname].isnull()] = list(root_taxid.keys())[0]\n",
    "    taxids.update(root_taxid)\n",
    "    df[\"sci_name\"] = df[taxid_colname].apply(lambda x: taxids[x])\n",
    "    return (df)\n",
    "    \n",
    "    \n",
    "def get_summary_table (df, colnames, metric):\n",
    "    df = df.groupby(colnames)[metric].sum().reset_index()\n",
    "    if not isinstance(metric, list):\n",
    "        metric = [metric]\n",
    "    sort_order = colnames+metric\n",
    "    sort_order.remove(\"sample\")\n",
    "    return (df.sort_values(by=sort_order, ascending=False))\n",
    "\n",
    "def correct_read_count(df, decontam_df, taxid, taxid_colname=\"taxid\"):\n",
    "    if isinstance(taxid, list):\n",
    "        decontam_table = pd.concat([get_rows_taxid(decontam_df, taxid=x, taxid_colname=taxid_colname) for x in taxid])\n",
    "    else:\n",
    "        decontam_table = get_rows_taxid(decontam_df, taxid=taxid, taxid_colname=taxid_colname)\n",
    "    decontam_table = decontam_table.groupby([\"sample\", \"taxid\"])[\"reads\"].sum().reset_index()\n",
    "    outdf = pd.merge(df, decontam_table)\n",
    "    outdf[\"read_prop\"] = outdf[\"reads\"]/outdf[\"read_count\"]*outdf[\"read_prop\"]\n",
    "    outdf = outdf.drop(columns=[\"read_count\", \"reads\"]).rename(columns={\"reads\":\"read_count\"})\n",
    "    return (outdf)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "identity_qcov_threshold = 0.9\n",
    "metadata_cols = [\"ska_genus\", \"ska_species\", \"collected_by\"]\n",
    "numbers = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read counts data\n",
    "idseq_data = pd.read_csv(\"../../data/project-mosquito_sample-table.csv\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load metadata\n",
    "metadata = pd.read_csv(\"../../data/metadata/CMS001_CMS002_MergedAnnotations.csv\", header=0)\n",
    "metadata = pd.merge(metadata, idseq_data[[\"sample_name\", \"nonhost_reads\", \"total_reads\"]], left_on=\"NewIDseqName\", right_on=\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load LCA data\n",
    "contig_stats_lca = pd.read_csv(\"s3://czbiohub-mosquito/contig_quality_concat/contig_stats_lca.tsv\", sep=\"\\t\", header=0)\n",
    "contig_stats_lca = contig_stats_lca.assign(identity_qcov=(contig_stats_lca[\"identity\"]/100*contig_stats_lca[\"align_length\"]/contig_stats_lca[\"contig_length\"]).apply(lambda x: min(x, 1)))\n",
    "numbers[\"total_num_contigs_with_blast_hits\"] = len(contig_stats_lca)\n",
    "contig_stats_lca = contig_stats_lca[~(contig_stats_lca[\"hexapoda\"])]\n",
    "numbers[\"total_nonhexapoda_contigs\"] = len(contig_stats_lca)\n",
    "contig_stats_lca = pd.merge(contig_stats_lca, metadata, how=\"left\", left_on=\"sample\", right_on=\"NewIDseqName\")\n",
    "contig_stats_lca[\"read_prop\"] = contig_stats_lca[\"read_count\"]/contig_stats_lca[\"nonhost_reads\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curated list of known viruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 197514.0 was translated into 2555385\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814.0 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "# Information about Baltimore classification of virus family groups\n",
    "viral_family_groups = pd.read_csv(\"../../data/virus_family_groups.csv\", header=0)\n",
    "viral_family_groups = viral_family_groups.loc[:, ~viral_family_groups.columns.str.startswith('Unnamed')]\n",
    "# Convert virus json into data frame\n",
    "with open (\"../../data/darkmatter/virus.json\", \"r\") as f:\n",
    "    viral_json = pd.DataFrame(json.load(f)).T\n",
    "viral_contigs = pd.concat(viral_json.apply(get_viral_family_df, axis=1).tolist())\n",
    "numbers[\"num_viral_contigs\"] = len(viral_contigs)\n",
    "# Add read proportions columns\n",
    "viral_contig_stats = get_rows_taxid(contig_stats_lca[[\"sample\", \"read_prop\", \"taxid\", \"contig_name\"]], taxid=10239, taxid_colname=\"taxid\")\n",
    "viral_contigs = pd.merge(viral_contigs, viral_contig_stats).groupby([\"sample\", \"sci_name\", \"taxid\"])[\"read_prop\"].sum().reset_index()\n",
    "viral_contigs = pd.merge(viral_contigs, contig_stats_lca[[\"sample\"]+metadata_cols])\n",
    "viral_contigs = pd.merge(viral_contigs, viral_family_groups, left_on=\"sci_name\", right_on=\"family\").drop(columns=\"family\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curate lists of non-viral contigs with high-confidence hits to NCBI records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load decontam data\n",
    "true_reads = pd.read_csv(\"s3://czbiohub-mosquito/contig_quality_concat/decontam_sample_read_counts.tsv\", sep=\"\\t\", header=0)\n",
    "contam_reads = pd.read_csv(\"s3://czbiohub-mosquito/contig_quality_concat/sample_contamination.tsv\", sep=\"\\t\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deduct from the nonhost reads per sample the number of reads that were removed due to suspected contamination\n",
    "contig_stats_lca[\"nonhost_reads\"] = pd.merge(contig_stats_lca, contam_reads.groupby(\"sample\")[\"reads\"].sum().reset_index()).apply(lambda x: x[\"nonhost_reads\"]-x[\"reads\"], axis=1)\n",
    "contig_stats_lca[\"read_prop\"] = contig_stats_lca[\"read_count\"]/contig_stats_lca[\"nonhost_reads\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 197514.0 was translated into 2555385\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814.0 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n",
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known wolbachia sequence\n",
    "wolbachia_taxid = 952\n",
    "wolbachia_contigs = get_rows_taxid(contig_stats_lca, taxid=wolbachia_taxid, taxid_colname=\"taxid\", identity_qcov_cutoff=identity_qcov_threshold)\n",
    "# Only keep wolbachia groups that were not removed by the decontamination step\n",
    "wolbachia_contigs = wolbachia_contigs[wolbachia_contigs[\"taxid\"].isin(true_reads[\"taxid\"])]\n",
    "numbers[\"total_wolbachia_contigs\"] = len(wolbachia_contigs)\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "wolbachia_contigs = get_summary_table(wolbachia_contigs, colnames=[\"ska_genus\", \"ska_species\", \"collected_by\", \"sample\", \"taxid\"], metric=[\"read_count\", \"read_prop\"])\n",
    "# Correct read counts after decontamination\n",
    "wolbachia_contigs = correct_read_count(wolbachia_contigs, true_reads, wolbachia_taxid, \"taxid\")\n",
    "#wolbachia_contigs = wolbachia_contigs.assign(taxid=wolbachia_taxid)\n",
    "# Create sci_name column to denote that that this table contains Wolbachia samples\n",
    "wolbachia_contigs = wolbachia_contigs.assign(sci_name=ncbi.get_taxid_translator([wolbachia_taxid])[wolbachia_taxid])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known metazoan sequence\n",
    "metazoan_contigs = contig_stats_lca[(contig_stats_lca[\"taxon_group\"]==\"Metazoa\") & (contig_stats_lca[\"identity_qcov\"]>=identity_qcov_threshold)]\n",
    "# Only keep metazoan groups that were not removed by the decontamination step\n",
    "metazoan_contigs = metazoan_contigs[metazoan_contigs[\"taxid\"].isin(true_reads[\"taxid\"])]\n",
    "# Convert taxids to those of interest\n",
    "#metazoan_contigs = clean_taxids(metazoan_contigs, taxids=[\"Leporidae\", \"Muroidea\", \"Homo sapiens\", \"Carnivora\", \"Odocoileinae\", \"Bovidae\", \"Neognathae\"], root_taxid=\"Metazoa\")\n",
    "metazoan_contigs = metazoan_contigs.assign(sci_name=metazoan_contigs[\"taxid\"].apply(lambda x: ncbi.get_taxid_translator([x])[x]))\n",
    "numbers[\"total_metazoan_contigs\"] = len(metazoan_contigs)\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "metazoan_contigs = get_summary_table(metazoan_contigs, colnames=metadata_cols+[\"sample\", \"taxid\", \"sci_name\"], metric=[\"read_count\", \"read_prop\"])\n",
    "# Correct read counts after decontamination\n",
    "metazoan_contigs = correct_read_count(metazoan_contigs, true_reads, taxid=\"Metazoa\", taxid_colname=\"taxid\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ete3/ncbi_taxonomy/ncbiquery.py:240: UserWarning: taxid 257814 was translated into 2608709\n",
      "  warnings.warn(\"taxid %s was translated into %s\" %(taxid, merged_conversion[taxid]))\n"
     ]
    }
   ],
   "source": [
    "# Only keep hits that are almost identical to a known eukaryote sequence\n",
    "eukaryotic_taxids = [\"Fungi\", \"Trypanosomatidae\", \"Plasmodium\", \"Stramenopiles\", \"Viridiplantae\", \"Euglenozoa\", \"Alveolata\"]\n",
    "eukaryotic_contigs = contig_stats_lca[(contig_stats_lca[\"taxon_group\"]==\"Eukaryota\") & (contig_stats_lca[\"identity_qcov\"]>=identity_qcov_threshold)]\n",
    "# Only keep eukaryotic groups that were not removed by the decontamination step\n",
    "eukaryotic_contigs = eukaryotic_contigs[eukaryotic_contigs[\"taxid\"].isin(true_reads[\"taxid\"])]\n",
    "# Convert taxids to those of interest\n",
    "#eukaryotic_contigs = clean_taxids(eukaryotic_contigs, taxids=eukaryotic_taxids, root_taxid=\"Eukaryota\")\n",
    "eukaryotic_contigs = eukaryotic_contigs.assign(sci_name=eukaryotic_contigs[\"taxid\"].apply(lambda x: ncbi.get_taxid_translator([x])[x]))\n",
    "numbers[\"total_eukaryotic_contigs\"] = len(eukaryotic_contigs)\n",
    "# Return a list of species grouped by mosquito species, collection site, sample, and taxid, and sorted by total read count\n",
    "eukaryotic_contigs = get_summary_table(eukaryotic_contigs, colnames=metadata_cols+[\"sample\", \"taxid\", \"sci_name\"], metric=[\"read_count\", \"read_prop\"])\n",
    "# Correct read counts after decontamination\n",
    "eukaryotic_contigs = correct_read_count(eukaryotic_contigs, true_reads, taxid=eukaryotic_taxids, taxid_colname=\"taxid\") \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/josh/anaconda3/envs/flash/lib/python3.5/site-packages/ipykernel_launcher.py:4: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "all_contigs_df = pd.concat([wolbachia_contigs.assign(group=\"Wolbachia\"), \n",
    "                            viral_contigs.assign(group=\"Virus\"),\n",
    "                            metazoan_contigs.assign(group=\"Metazoa\"), \n",
    "                            eukaryotic_contigs.assign(group=\"Other Eukaryotes\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_contigs_df.to_csv(\"../../figures/fig3/all_contigs_df.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "flash",
   "language": "python",
   "name": "flash"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
